<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<HTML>
<HEAD>
	<META HTTP-EQUIV="CONTENT-TYPE" CONTENT="text/html; charset=windows-1252">
	<TITLE>Analysis tools for sequences in ZFIN</TITLE>
	<STYLE>
	<!--
		@page { size: 8.5in 11in; margin-right: 1.25in; margin-top: 1in; margin-bottom: 1in }
		P { margin-bottom: 0.08in; direction: ltr; color: #000000; widows: 2; orphans: 2 }
		P.western { font-family: "Times New Roman", serif; font-size: 12pt; so-language: en-US }
		P.cjk { font-family: "Times New Roman", serif; font-size: 12pt }
		P.ctl { font-family: "Times New Roman", serif; font-size: 12pt; so-language: ar-SA }
	-->
	</STYLE>
</HEAD>
<BODY LANG="en-US" TEXT="#000000" DIR="LTR">
<P CLASS="western" STYLE="margin-bottom: 0in"><FONT SIZE=4><B>Analysis
tools for sequences in ZFIN</B></FONT></P>
<P CLASS="western" STYLE="margin-bottom: 0in"><BR>
</P>
<P CLASS="western" STYLE="margin-bottom: 0in">You are able to
access a variety of sequence analysis tools directly from the
sequence information sections of ZFIN.  The analysis tools available
for each sequence are listed in a pulldown menu in the far right
column of the sequence information section on Gene, Marker and Clone pages.
 When only one analysis tool is available for a sequence, a button
replaces the pulldown menu.</P>
<P CLASS="western" STYLE="margin-bottom: 0in"><BR></P>
<P CLASS="western" STYLE="margin-bottom: 0in">Analysis tools
available for sequences in ZFIN are:</P>
<P CLASS="western" STYLE="margin-bottom: 0in"><B>NCBI BLAST</B> &ndash;
The <B>B</B>asic <B>L</B>ocal <B>A</B>lignment <B>S</B>earch <B>T</B>ool
(BLAST) is used to compare nucleotide or amino acid sequences for
identity and similarity with other sequences.  Selecting this tool
opens a search form for a BLAST analysis at the National Center for
Biotechnology Information (NCBI).  The sequence accession number of
interest is automatically added to the search form for your
convenience.  Just select your search options and click the BLAST!
 button on the NCBI page.</P>
<P CLASS="western" STYLE="margin-bottom: 0in"><B>Ensembl </B>&ndash;
Selecting this tool opens a search form for a BLAST analysis at The
Wellcome Trust Sanger Institute&rsquo;s Ensembl BLASTView resource. 
The sequence accession number of interest is automatically added to
the search form for your convenience.  Just select your search options and click the RUN 
button near the bottom of the Sanger BLASTView page.</P>
<P CLASS="western" STYLE="margin-bottom: 0in"><B>Mega BLAST</B> &ndash;
Mega BLAST is a BLAST algorithm that more efficiently handles longer
nucleotide sequences.  Selecting this tool opens a search form for a
Mega BLAST analysis at the National Center for Biotechnology
Information.  The sequence accession number of interest is
automatically added to the search form for your convenience.  Just select 
your search options and click the BLAST! button on the NCBI page.</P>
<P CLASS="western" STYLE="margin-bottom: 0in"><B>SIB BLAST</B> &ndash;
Selecting this tool runs an amino acid sequence BLAST analysis at
Swiss Institute of Bioinformatics (SIB).   The results are
automatically displayed on the SIB page.</P>
<P CLASS="western" STYLE="margin-bottom: 0in"><B>UCSC BLAT</B> &ndash;
The <B>B</B>last-<B>L</B>ike <B>A</B>lignment <B>T</B>ool (BLAT) is
genome browser that displays the sequence of interest in the context
of other genome features such as known genes from a number of
species, other nucleotide sequences, and gene prediction models. 
Selecting this tool runs a BLAT analysis against the November 2003
assembly at the University of California, Santa Cruz (UCSC) Genome
Bioinformatics Site.  The results are automatically displayed on the
UCSC page.</P>
</BODY>
</HTML>
