<title>ZFMonitor Vol 2(4)</title><head>
</head><body>
<H1><i>THE ZEBRAFISH SCIENCE MONITOR</i></H1>
<B><I>All the Zebrafish News That's Fit to Print</I><BR></B>
Volume 2, Issue 4, October 25, 1993, Eugene, Oregon<BR><BR><hr>
<B>Contents:</B><br>
<A HREF="#1">THE PHYLOGENETIC POSITION OF THE ZEBRAFISH (<i>DANIO RERIO</i>), A MODEL SYSTEM IN DEVELOPMENTAL BIOLOGY: AN INVITATION TO THE COMPARATIVE METHOD</A><p>
<A HREF="#2">EDITORIAL: A MATTER OF POLICY</A><p>
<A HREF="#3">ZEBRAFISH AT THE ISDB MEETING</A><p>
<A HREF="#4">CHARACTERIZATION OF POU GENES IN ZEBRAFISH EMBRYOS</A><p>
<A HREF="#5">A GENOME-WIDE SCREEN FOR MUTATIONS AFFECTING EMBRYONIC DEVELOPMENT IN ZEBRAFISH</A><p>
<A HREF="#6">GENETIC ANALYSIS OF EMBRYONIC PATTERN FORMATION IN <i>DROSOPHILA</i> AND THE ZEBRAFISH, <i>BRACHYDANIO RERIO</i></A><p>
<A HREF="#7">ANALYSIS OF EARLY FISH DEVELOPMENT - THE ROLE OF ACTIVIN-LIKE GENES</A><p>
<A HREF="#8">DNA UPTAKE FROM THE YOLK INTO THE BLASTOMERES IN EARLY DEVELOPMENTAL STAGES OF FISH EMBRYOS</A><p>
<A HREF="#9">GENETICS OF AXIS DEVELOPMENT IN ZEBRAFISH</A><p>
<A HREF="#10">BLASTEMA CELLS OF REGENERATING ZEBRAFISH FINS EXPRESS <i>MSX</i> AND <i>DISTAL-LESS</i> GENES</A><p>
<A HREF="#11">DEVELOPMENT OF IDENTIFIED MOTONEURONS IN EMBRYONIC ZEBRAFISH</A><p>
<A HREF="#12">MUTAGENESIS AND SCREEN FOR EMBRYONIC MUTANTS IN THE ZEBRAFISH, <i>BRACHYDANIO RERIO</i></A><p>
<A HREF="#13">SCALES IN THE ZEBRAFISH, BRACHYDANIO: CELL ORIENTATION AND ANISOTROPIC CELL AGGREGATIONS MAY DETERMINE SCALE ORIENTATION</A><p>
<A HREF="#14">STRUCTURE AND COMPOSITION OF THE ZEBRAFISH EGG CHORION</A><p>
<A HREF="#15">EXPRESSION OF THE <i>XENOPUS</i> HOMEOGENE <i>XHOX3</i> IN ZEBRAFISH EGGS CAUSES A DISRUPTION OF ANTERIOR-POSTERIOR AXIS</A><p>
<A HREF="#16">COMBINATORIAL EXPRESSION OF THREE <i>DISTAL-LESS</i> GENES DURING ZEBRAFISH EMBRYONIC DEVELOPMENT</A><p>
<A HREF="#17">ZEBRAFISH HOMEODOMAIN PROTEIN ISL-1 IS EXPRESSED IN PRIMARY NEURONES PIONEERING AXONAL TRACTS</A><p>
<A HREF="#18">ZEBRAFISH RAR'S AND RXR'S: EVOLUTIONARY CONSERVATION OF STRUCTURE AND FUNCTION</A><p>
<A HREF="#19">MEETING ON ZEBRAFISH DEVELOPMENT AND GENETICS AT COLD SPRING HARBOR LABORATORY</A><p>
<A HREF="#20">ZF ADDRESS UPDATE</A><p>
<A HREF="#21">ZEBRAFISH REFERENCE UPDATE</A>
<p><hr>
<b><H3><A NAME="1">THE PHYLOGENETIC POSITION OF THE ZEBRAFISH (<i>DANIO RERIO</i>), A MODEL SYSTEM IN DEVELOPMENTAL BIOLOGY: AN INVITATION TO THE COMPARATIVE METHOD</A></H3></b>
<p>
By A. Meyer, C.H. Biermann, and G. Ort&iacute;, Department of Ecology &amp; Evolution, State University of New York, Stony Brook, NY 11794-5245, USA (<i>Proc. R. Soc. Lond.</i> B (1993) <b>252</b>:231-236)
	<p>
The zebrafish, <i>Danio</i> (<i>Brachydanio</i>) <i>rerio</i>, has become one of the most widely studied model systems in developmental biology.  We present a DNA-based phylogeny of zebrafish and other species of the genus <i>Danio</i>, and the genera <i>Rasbora, Puntius</i>, and <i>Cyprinus</i>.  Homologous regions of the large (16<i>S</i>) mitochondrial ribosomal RNA gene were amplified by the polymerase chain reaction and directly sequenced.  The phylogeny revealed: (<i>i</i>) the zebrafish, <i>Danio</i> (<i>Brachydanio</i>) <i>rerio</i> is identical in its 16<i>S</i> sequence to its aquarium breeding morph, the leopard danio; (<i>ii</i>) the pearl danio (<i>Danio albolineatus</i>) is more closely related to the zebrafish than the giant danio (<i>Danio aequipinnatus</i>); and (<i>iii</i>) species of the genus <i>Rasbora</i> (<i>hetermorpha, trilineata, elegans, pauciperforata, dorsiocellata</i>) are more closely related to the danios than members of the genus <i>Puntius</i> (<i>tetrazona, conchonius</i>) and <i>Cyprinus</i>, the carp.  All of these species are readily available in the aquarium trade, easily kept and bred in captivity, and amenable to developmental work.  It is hoped that this molecular phylogeny will invite developmental biologists to use the comparative method to ask questions about function (e.g. cellular and genetic aspects) and evolution of zebrafish developmental biology in a phylogenetic context.<b><p>
<H3><A NAME="2">EDITORIAL: A MATTER OF POLICY</A></H3></b>
	<p>
It has been our intention, since <i>The</i> <i>Monitor's </i>inception, that this humble publication serve as a vehicle for the rapid communication of news pertinent to the Zebrafish research community.  We hope that all of you will be eager to send in progress reports of studies that are incomplete or not yet ready to submit to a regular journal. <p>
	<p>
"But, wait a minute"  you argue.  "How can I send in something I hope to publish later? Won't its appearance in <i>The</i> <i>Monitor</i> preclude its later acceptance in a more respectable place, like <i>Nature</i> or <i>Cell</i>?"<p>
	<p>
"Heck, no!"  the editor moans.  "<i>The Monitor</i>, is an 'informal' vehicle dedicated to communicating news, not publishing reviewed manuscripts."<p>
	<p>
All references to information appearing in <i>The</i> <i>Monitor</i> should be made as "personal communications' and then only if you have received explicit permission of the authors.<p>
	<p>
Because all of us have such poor memories, we will include a statement of this policy in each future issue as a reminder of our charter.<b><p>
<H3><A NAME="3">ZEBRAFISH AT THE ISDB MEETING</A></H3></b>
	<p>
Zebrafish research news was reported in a number of presentations at the recent 12th Congress of the International Society of Developmental Biology held in Vienna, Austria, August 8-13, 1993. <p>
This issue of <i>The Monitor</i> includes abstracts from these presentations. 
	<p>
As explained in the editorial on page 2, these abstracts are reproduced here as part of the latest zebrafish news.  They should not be cited.  Reference to information from these abstracts or from any other articles appearing in <i>The Monitor</i>  should be made as personal communications and then only if you have received explicit permission of the authors.<p>
<b><p>
<H3><A NAME="4">CHARACTERIZATION OF POU GENES IN ZEBRAFISH EMBRYOS</A></H3></b><p>
By G. Hauptmann, P. Spaniol, C. Bornmann, and T. Gerster.  Abt. Zellbiologie, Biozentrum der University of Basel, 4056 Basel, SWITZERLAND
	<p>
We have identified by PCR and cDNA cloning five different POU genes expressed during early embryogenesis of zebrafish.  Four of these genes show extended homology to the <i>brn-1</i> class of POU genes previously identified in mammals.	<p>
Northern blot analysis and <i>in situ</i> hybridizations indicate that the expression of these genes begins shortly after gastrulation in the neural tube.  In the 24 h old embryo, various structures in the brain express these genes.  We are currently investigating how the expression patterns of the four <i>brn-1</i>-like genes differ from each other and how much they overlap. Possibly, the four <i>brn-1</i>-like genes play combinatorial roles resulting in differential activation of specific cell fates in the zebrafish brain.	<p>
A fifth gene we have analyzed contains a POU domain that forms the prototype for a novel subclass of these DNA binding proteins.  This particular gene appears to be expressed extremely early in embryogenesis in addition to a large maternal pool of RNA.  Its transcript shows an asymmetric distribution in the prospective neuroectoderm at late gastrula stages. Thereafter, the expression of this gene decreases rapidly and at 24 hours after fertilization, the transcript can only be found in the extreme tip of the tail.  The early expression and the spatial arrangement of the transcript strongly argue that this POU gene is involved in early developmental decisions determining the body plan of the zebrafish.<b><p>
<H3><A NAME="5">A GENOME-WIDE SCREEN FOR MUTATIONS AFFECTING EMBRYONIC DEVELOPMENT IN ZEBRAFISH</A></H3></b><p>
By L. Solnica-Krezel, A.F. Schier, S. Abdelilah, J. Malicki, S. Neuhauss, Z. Rangini, D. Stainier, D. Stemple, F. Zwartkruis, and W. Driever, CVRC, MGH-East 4, 13th St., Bldg. 149, Charlestown, MA 02129 USA
	<p>
Our studies show that mutations affecting specific aspects of vertebrate development can be efficiently identified in zebrafish (<i>Brachydanio rerio</i>) by genetic screens similar to those performed in <i>C. elegans</i> and <i>Drosophila</i>.  We have developed methods for efficient mutagenesis of proliferating germ cells by treating zebrafish males with N-ethyl-N-nitrosourea.  The mutation rates estimated by the specific locus test at four distinct loci vary between 1/400-1/4000. Recessive embryonic lethal or visible mutations are recovered in a two generation screen with an average rate of 3 mutations per line.  Approximately half of the mutations identified so far appear to affect specific regions or processes in developing embryos.  We are focusing on two classes of mutants.  In mutants of the first class formation or differentiation of the notochord and somitic mesoderm are affected. The second class of mutants is characterized by abnormal patterning of the anterior neural tube.
	<p>
<b><p>
<H3><A NAME="6">GENETIC ANALYSIS OF EMBRYONIC PATTERN FORMATION IN <i>DROSOPHILA</i> AND THE ZEBRAFISH, <i>BRACHYDANIO RERIO</i></A></H3></b>
<p>
By C. N&uuml;sslein-Volhard, Max-Planck-Institut f&uuml;r Entwicklungsbiologie, T&uuml;bingen, GERMANY   
	<p>
The systematic searches for mutants affecting embryonic pattern in <i>Drosophila</i> and the subsequent molecular analysis led to the identification of a large number of important regulatory proteins.  Their interactions in several pathways have been established, controlling the progressive subdivisions and differentiation of the initially uniform egg cell.  This approach has led to the fact that <i>Drosophila</i> is now in many respects the best understood system in biology. How much of the knowledge from studying <i>Drosophila</i> can be applied to vertebrate organisms is an open question.  As the body organization of a vertebrate and an insect is very different, a number of pattern-forming mechanisms that do not have a parallel in <i>Drosophila</i> are expected to take place.  Systematic mutational approaches were not yet feasible in a vertebrate and much of our knowledge about controlling genes comes from a surprising degree of homology between <i>Drosophila</i> genes and vertebrate genes.  Homology approaches, however, a <i>priori</i> are restricted to conserved, and exclude novel, structures, and conservation of structure is not always accompanied by conservation of function.  In recent years, a small freshwater fish, the zebrafish, has attracted the interest of experimental embryologists and geneticists.  It has many good qualities of an experimental system, such as transparent eggs and embryos and a speedy development.  We and other laboratories are developing the tools for a genetic analysis in this organism.  It remains to be seen whether this approach will be as fruitful as it has been in <i>Drosophila</i>.<b><p>
<H3><A NAME="7">ANALYSIS OF EARLY FISH DEVELOPMENT - THE ROLE OF ACTIVIN-LIKE GENES</A></H3></b><p>
J. Wittbrodt, M. Rissi, and F. Rosa, Biocenter of the University of Basel, CH-4056 Basel, SWITZERLAND.
	<p>
Cell-cell interactions through diffusible factors of the TGF-&szlig; superfamily are thought to be involved in the specification and modification of cell fate in early vertebrate embryos.  Using medaka (<i>Oryzias latipes</i>) and zebrafish (<i>Brachydanio rerio</i>) as experimental systems, we are analyzing the role of homologous and newly discovered members of the TGF-&szlig; family in fish. Two new members of the DVR subfamily were discovered.  One of these, Zac15, is conserved throughout evolution and differentially expressed during embryonic development.  Whole-mount <i>in situ</i> analysis could first detect the message in the early neurula in the lining of the prospective brain regions.  Later stages show expression in cells derived from all germ layers.  The expression pattern in the neuroectoderm suggests a role of the Zac15 gene in the specification of distinct regions of the CNS.  Using a PCR based approach, we could isolate three genes encoding activin &szlig; chains in medaka and zebrafish.  To interfere with the endogenous activins in transgenic animals, we introduced different types of mutations into a full-length cDNA clone encoding the zebrafish activin &szlig;B.  While one type of mutant allowed the analysis of the contribution of the zygotic expression of activin to axis formation, the other type enabled us to analyze zygotic and maternal contribution of activin protein to mesoderm induction and axis formation <i>in vivo.</i><b><p>
<H3><A NAME="8">DNA UPTAKE FROM THE YOLK INTO THE BLASTOMERES IN EARLY DEVELOPMENTAL STAGES OF FISH EMBRYOS</A></H3></b><p>
By T. Papp, F. Erd&eacute;lyi, A. &Aacute;d&aacute;m, and L. Orb&aacute;n, Institute for Molecular Genetics, Agricultural Biotechnology Center, G&ouml;d&ouml;ll&otilde;, P.O.B. 170, H-2101 HUNGARY
	<p>
Various gene constructs injected into the yolk of fish eggs were transiently expressed in the early embryo.  The expression of the &szlig;-galactosidase gene was first detected in the blastomeres and syncytial layer 5 hours after fertilization.  At 24 hours postfertilization &szlig;-galactosidase staining was observed in the whole embryo including the yolk.  An intensive flow from the yolk towards the blastomeres was revealed in the earliest stages of development of zebrafish and African catfish eggs by injection with FITC-dextran and labeled plasmids.  Our observations suggest that DNA injected into the yolk is transferred to the blastomeres by this flow.  Microinjection into the yolk has some advantages over the conventional microinjection method: lower mortality rate and higher occurrence of transiently expressing embryos.<b><p>
<H3><A NAME="9">GENETICS OF AXIS DEVELOPMENT IN ZEBRAFISH</A></H3></b><p>
By C.B. Kimmel, M.E. Halpern, R.K. Ho, K. Hatta, A.E. Melby, C. Walker, and <b>*</b>B. Trevarrow, Institute of Neuroscience, University of Oregon, Eugene, OR 97403 USA. <b>*</b>Department of Biology, California Institute of Technology, Pasadena, CA 91125 USA
	<p>
We have used zygotic lethal mutations to analyze notochord and floor plate development.  <i>ntl</i> mutants lack a notochord, <i>cyc</i> mutants lack a floor plate, and both tissues are affected in the trunk and tail of <i>flh</i> mutants.  Mosaic analysis and epistatic relationships between <i>cyc</i> and <i>ntl</i> suggest that floor plate development depends on early inductive signaling of ectoderm by dorsal mesoderm. Signaling appears to require <i>flh</i>+ but not <i>ntl</i>+ function in the  trunk.  Neither function is required in the head, and both are required in the tail.<b><p>
<H3><A NAME="10">BLASTEMA CELLS OF REGENERATING ZEBRAFISH FINS EXPRESS <i>MSX</i> AND <i>DISTAL-LESS</i> GENES</A></H3></b>
<p>
By M.-A. Akimenko<I>1</I>, S. Johnson<I>2</I>, M. Ekker<I>1</I>, and M. Westerfield<I>2</I>.  <I>1</I>Loeb Institute and Depts. of Medicine &amp; Anatomy, University of Ottawa, CANADA.  <I>2</I>Institute of Neuroscience, University of Oregon, Eugene, OR 97403-1254 USA
	<p>
Members of the <i>msx</i> and <i>distal-less</i> (<i>dlx</i>) families of homeobox genes are transiently expressed during development of the zebrafish primordial fin fold, which gives rise to the unpaired fins.  Transcripts of some of the genes are also observed in the pectoral fin buds.  However, 3 days after fertilization, <i>msx</i> and <i>dlx</i> expression in fins falls below levels detectable by <i>in situ</i> hybridization.  We have assayed for the presence of <i>msx</i> and <i>dlx</i> transcripts during fin regeneration.  We observe <i>msxB, msxC, msxD</i>, and <i>dlx3</i> transcripts during regeneration of both paired and unpaired fins.  <i>MsxD</i> transcripts are detected in the distal-most layer of cells of the blastema in contrast to <i>msxB</i> and <i>msxC</i> which are expressed in more proximal blastema cells. Amputation performed in different planes along the proximo-distal axis of the fin suggests that <i>msxB</i> expression correlates with the growth rate of the regenerate.  These results suggest a role for <i>msx</i> and <i>dlx</i> genes in the mechanisms of dedifferentiation and redifferentiation necessary for fin regeneration.  Supported by the MRC, the NCIC, and the NIH.<b><p>
<H3><A NAME="11">DEVELOPMENT OF IDENTIFIED MOTONEURONS IN EMBRYONIC ZEBRAFISH</A></H3></b><p>
By J.S. Eisen, Institute of Neuroscience, University of Oregon, Eugene, OR 97403 USA
	<p>
We are interested in how neurons acquire their identities and express cell-specific characteristics, such as shape and synaptic connectivity.  To address these issues, we have examined the development of a set of individually identified motoneurons in the embryonic zebrafish.  This set consists of 3-4 bilaterally symmetric motoneurons that are segmentally iterated along the embryonic axis.  Analysis of mutations affecting specific tissue types suggests that the segmental organization of these motoneurons is influenced by interactions with the paraxial mesoderm.  Transplantation of single motoneurons suggests that their individual identities depend on their spinal cord positions.  Motoneurons in different spinal cord positions seem to have no influence on one another's axonal pathway choices, and proper synaptic connectivity appears to be the result of cell-specific pathway recognition.  Interactions between two motoneurons that form an equivalence pair determine which one of them will survive.<b><p>
<H3><A NAME="12">MUTAGENESIS AND SCREEN FOR EMBRYONIC MUTANTS IN THE ZEBRAFISH, <i>BRACHYDANIO RERIO</A></H3></i></b><p>
By M. Mullins, M. Hammerschmidt, P. Haffter, and C. N&uuml;sslein-Volhard, Max-Planck-Institut f&uuml;r Entwicklungsbiologie, Spemannstrasse 35/III, D-7400 T&uuml;bingen 1, GERMANY
	<p>
We are investigating the embryonic development of the zebrafish through the isolation and characterization of mutations in zygotically-expressed genes involved in embryonic pattern formation, morphogenesis, and differentiation.  We have developed the technology to make practical a large-scale screen. High rates of point mutagenesis of zebrafish spermatogonia were established through the use of the chemical mutagen ethylnitrosourea.  Mutation rates range from 0.1 to 0.3% for an individual locus, similar to those used in genetic screens in <i>Drosophila</i> and <i>C. elegans.</i>	<p>
We performed a small-scale classical F2 diploid screen.  Most of the mutants we isolated have specific defects, but the abnormalities arise late in embryonic development and do not appear to affect embryonic pattern formation.  We have only a small number of mutants displaying early embryonic defects. Our large-scale saturation screen is currently in progress.<b><p>
<H3><A NAME="13">SCALES IN THE ZEBRAFISH, BRACHYDANIO: CELL ORIENTATION AND ANISOTROPIC CELL AGGREGATIONS MAY DETERMINE SCALE ORIENTATION</A></H3></b><p>
<p>
By K. N&uuml;bler-Jung, Biologie Institut I, Alberstr. 21a, D-7800 Freiburg, GERMANY.
	<p>
How do polarized structures (e.g hairs or scales) adopt their specific orientations in animal integuments?	<p>
We find that the ripple patterns on our finger tips, as well as the various orientations of hairs and scales in vertebrates resemble polarity patterns seen in insects.  This may indicate similar orienting mechanisms in insects and vertebrates.	<p>
In the zebrafish, <i>Brachydanio</i>, dermal and epidermal cells together orient the scale anlagen.  Dermal cells (like insect epidermal cells) orient along the dorsoventral body axis and thereby may determine the <b>axis</b> of the scale, while epidermal cells accumulate at the <b>posterior</b> rim of the anlagen.  Interactions between dermis and epidermis thus seem to determine the orientation of scales.<b><p>
<H3><A NAME="14">STRUCTURE AND COMPOSITION OF THE ZEBRAFISH EGG CHORION</A></H3></b>
<p>
By D. Bonsignorio, S. Raisoni, C. Lora Lamia, and F. Cotteli.  Department of Biology, University of Milano, Milano, ITALY
	<p>
The zebrafish egg is completely surrounded by an amorphous thick envelope usually called the chorion.  During oogenesis, this coat is assembled between the egg and the follicle cells.  It is multi-layered and made up by a fibrillar material embedded in an amorphous matrix as in other fish.  In zebrafish, the chorion is made up by at least 4 major layers.  Isolated and purified chorions have been analyzed by SDS-PAGE under reducing and non-reducing conditions.  A reproducible pattern of polipeptides with molecular weights ranging from 30 to 125 KDa was revealed.	<p>
Using various lectins, we have investigated the presence of glycoproteins and we have detected the most representative carbohydrates.<b><p>
<H3><A NAME="15">EXPRESSION OF THE <i>XENOPUS</i> HOMEOGENE <i>XHOX3</i> IN ZEBRAFISH EGGS CAUSES A DISRUPTION OF ANTERIOR-POSTERIOR AXIS</A></H3></b><p>
<p>
By O. Barro<I>1</I>, A. Ruiz i Altaba<I>2</I>, J.S. Joly<I>3</I>, C. Joly<I>1</I>, H. Condamine<I>3</I>, and H. Boulekbache<I>1</I>, <I>1</I>Laboratoire de biologie du d&eacute;veloppement Universit&eacute; Paris 7, 2 place Jussieu, 75251 Paris cedex 05, FRANCE. <I>2</I>Howard Hughes Medical Institute, Columbia University, New York, NY 10032.  <I>3</I>Institut Pasteur, 25 rue du Dr. Roux, 75015 Paris cedex, FRANCE.
	<p>
The ectopic expression in the zebrafish of the <i>Xenopus</i> homeogene <i>XHOX3</i> was carried out by microinjection of synthetic <i>XHOX3</i> mRNA (with an <i>eve</i> homeobox) into uncleaved embryos and resulted in various degrees of anterior-posterior axis disruption.	<p>
The phenotypic variants observed mainly show anomalies in neural tube development, going to microphthalmia to acephali, and could have been classed according to an index of axis deficiency (IAD) (1). A dose-dependent effect of injected <i>XHOX3</i> mRNA was evident; from 5pg to 10pg of mRNA into each egg, we observed a major change from prim 5 stage embryos with a normal phenotype, to an acephalic phenotype. The specificity of exogenous <i>XHOX3</i> protein effects was controlled by the observation of a normal phenotype for the majority (91.8%) of embryos injected with <i>XHOX3</i> transcripts with a deleted HTH motif homeobox. Controls included the injection of lacZ mRNA and the immunostaining of mes-metencephalon, using the 4D9 antibody.	<p>
The nuclear localization of "<i>XHOX3</i>-like" protein was determined using an antibody against the NH2 terminal part of the protein.  The distribution of positive nuclei at 24 h after fertilization was observed in whole-mounts and in sections.  It appeared to be exclusively restricted to posterior mesoderm tissue.	<p>
Our results on the zebrafish embryo are surprisingly similar to those obtained in <i>Xenopus</i> after <i>XHOX3</i>	endogenous gene analysis (2).	<p>
This evidence supports the hypothesis of a high conservation of the mechanism implicated in embryonic development involving the <i>XHOX3</i> gene.
<p>
1) Scharf, S.R. and J.C. Gerhart, <i>Dev. Biol.</i> <b>99</b>:75-87, 1983.<br>
2) Ruiz i Altaba, A. and D.A. Melton, <i>Cell</i> <b>57</b>:317-326, 1989.<b><p>
<H3><A NAME="16">COMBINATORIAL EXPRESSION OF THREE <i>DISTAL-LESS</i> GENES DURING ZEBRAFISH EMBRYONIC DEVELOPMENT</A></H3></b>
<p>
By M. Ekker<I>1</I>, M.-A. Akimenko<I>1</I>, and M. Westerfield<I>2</I>, <I>1</I>Loeb Institute and Departments of Medicine and Anatomy, University of Ottawa, CANADA.  <I>2</I>Institute of Neuroscience, University of Oregon, Eugene, OR 97403-1210 USA
	<p>
We have isolated three zebrafish genes, <i>dlx2</i>, <i>dlx3</i> and <i>dlx4</i>, with homeoboxes related to that of the <i>Drosophila</i> <i>distal-less</i> gene.  Although the overall patterns of expression of the three genes during embryonic development are distinct, we observe remarkable similarities in the expression of two or three <i>dlx</i> genes in specific regions of the embryos.  For example, a subset of cells in the ventral forebrain expresses <i>dlx2</i> and <i>dlx4</i> starting at 16 h after fertilization.  Cells of the olfactory placodes or their precursors express <i>dlx3</i> and <i>dlx4</i>.  Similarly, <i>dlx3</i> and <i>dlx4</i> transcripts are present in the primordial fin fold and in a subset of cells in the otic vesicle.  All three <i>dlx</i> genes are expressed in cells of the visceral arches and their primordia.  <i>dlx2</i> alone is expressed during gastrulation.  Based on these results, we suggest that combinatorial expression of the <i>dlx2</i>, <i>dlx3</i> and <i>dlx4</i> genes is part of a new type of homeobox gene code which could be important in specifying pattern formation or cell fate determination in multiple regions of the embryo.  Supported by the MRC, the NCIC, and the NIH.<b><p>
<H3><A NAME="17">ZEBRAFISH HOMEODOMAIN PROTEIN ISL-1 IS EXPRESSED IN PRIMARY NEURONES PIONEERING AXONAL TRACTS</A></H3></b>
<p>
By V. Korzh, S. Thor, and T. Edlund, Department of Microbiology, University of Ume&aring;, SWEDEN
	<p>
Isl-1 was initially isolated as a factor binding to the insulin gene enhancer.  Isl-1 expression is regulated by the notochord/floor plate inducing signals.  We have shown using anti-rat Isl-1 antibody staining on whole-mounted and sectioned embryos of zebrafish that Isl-1 immunoreactivity is selectively expressed in the primary neurones (primary motor neurons, Rohon-Beard cells, primary interneurons, neurons of the trigeminal ganglia and anterior group of cells) in the spinal cord and brain.  Isl-1 expression is <i>initiated at the very beginning of neurulation (10 hpf)</i>.  Induction of Isl-1 expression in all these cell groups takes place almost simultaneously and before the notochord differentiates from the axial chordamesoderm.  These results suggest that the induction of Isl-1 in the primary neurons is dependent on signaling spreading in the plane of ectoderm.  Isl-1 positive cells will be the <i>first neurons to send axons which will pioneer the major axonal tracts</i>.<b><p>
<H3><A NAME="18">ZEBRAFISH RAR'S AND RXR'S: EVOLUTIONARY CONSERVATION OF STRUCTURE AND FUNCTION</A></H3></b><p>
By M. Petkovich, J. White, B. Jones, and M. Boffa, Cancer Research Labs, Queen's University, Kingston, Ontario, CANADA K7L 3N6.
	<p>
Retinoic acid (RA) is an important signaling molecule in vertebrate pattern formation.  The effects of RA are due largely to regulation of gene transcription, mediated by 2 classes of nuclear receptors - retinoic acid receptors (RAR-a, RAR-&szlig;, RAR-_) and retinoid X receptors (RXR-a, RXR-&szlig;, RXR-_).  We have been using zebrafish as a developmental model to study the role of retinoic acid in vertebrate development and have isolated cDNA's which closely correspond in sequence to their mouse and human receptors.  Zebrafish RAR's (zfRAR-a, -&szlig;, -_) and RXR's (zfRXR-a, -&szlig;, -_) are also functionally conserved and form RAR/RXR heterodimers, and exhibit similar target gene specific activation compared to their mouse counterparts. Developmental patterns of zebrafish RAR and RXR expression are also conserved.  For example, by whole mount <i>in situ</i> hybridization, zebrafish RAR-_ expression is detectable in head and tail bud mesenchyme, and at later stages, is in the mesenchyme of the developing fin bud.  Thus, the zebrafish retinoid signal transduction system is highly conserved, strongly supporting the relevance of the zebrafish model to study the developmental role of retinoic acid.<p>

<H3><A NAME="19">MEETING ON ZEBRAFISH DEVELOPMENT AND GENETICS AT COLD SPRING HARBOR LABORATORY</A></H3> <p>
April 27 - May 1, 1994 
<p>
Organized by:<i>
Wolfgang Driever</i>, Massachusetts General Hospital;<i> Judith Eisen</i>, University of Oregon;<i>
 David Grunwald</i>, University of Utah;<i>
 Charles Kimmel</i>, University of Oregon
 <p>
      This is the first open-invitation meeting dedicated to research on the zebrafish. The
meeting will cover a broad range of topics.  Anticipated sessions and chairpersons include: 
 <p>
 <i>	Gastrulation, body patterning, and morphogenesis:</i> C. Nusslein-Volhard and R. Ho 	<i><p>
Determination of cell fate:</i> N. Holder and M.-A. Akimenko  	<i><p>
Development of the nervous system: </i>J. Campos-Ortega and J. Kuwada  	<i><p>
Growth control: </i>M. Schartl  	<i><p>
Organogenesis: </i>M. Fishman 	<i><p>
Genetics and Genomics: </i>M. Westerfield  	<i><p>
Gene transfer, gene expression, new methodologies:</i> N. Hopkins <p>

      Submitted abstracts will be considered for oral or poster presentations.  Abstract deadline is Feb. 9, 1994.  For further information and registration materials, contact Meetings Coordinator, Cold Spring Harbor Laboratory, 1 Bungtown Road, Cold Spring Harbor, New York 11724-2213. TEL:516-367-8346  FAX:516-367-8845 <p>
 	<i>
Everyone who receives an individual copy of the Zebrafish Science Monitor should be receiving registration materials directly from Cold Spring Harbor Laboratory.</i><b><p>
<H3><A NAME="20">ZF ADDRESS UPDATE</A></H3></b><b><br>
Since 8/09/93</b>
<b><p>
ADDITIONS</b><p>
<p>
Rein Aasland<br>
Gene Expression Programme<br>
EMBL<br>
Meyerhofstrasse 1<br>
D-6900 Heidelberg, GERMANY<br>
49-6221-387-538/ FAX 49-6221-387-518<br>
aasland@embl-heidelberg.de
<p>
Bruce Appel<br>
Institute of Neuroscience<br>
1254 University of Oregon<br>
Eugene, OR 97403-1254<br>
(503) 346-4539/ FAX (503) 346-4548<br>
appel@uoneuro.uoregon.edu
<p>
Christine Beattie<br>
Institute of Neuroscience<br>
1254 University of Oregon<br>
Eugene, OR 97403-1254<br>
(503) 346-4539/ FAX (503) 346-4548<br>
beattie@uoneuro.uoregon.edu
<p>
Giacomo Bernardi<br>
Hopkins Marine Station<br>
Stanford University<br>
Pacific Grove, CA 93950<br>
(408) 655-6210/ FAX (408) 375-0793<br>
giacomo.bernardi@nitelog.com
<p>
Russell J. Buono<br>
Dept/Anatomy &amp; Dev Biology<br>
Thomas Jefferson Medical College<br>
1020 Locust St.<br>
Philadelphia, PA 19107-6799<br>
(215) 955-7820/ FAX (215) 923-3808
<p>
Barbara X. Chapman<br>
Division of Biology 139-74<br>
Caltech<br>
Pasadena, CA 91125<br>
(818) 395-2863/ FAX (818) 449-5163<br>
bchapman@egg.gg.caltech.edu
<p>
Philip S. Crosier<br>
Department of Molecular Medicine<br>
University of Auckland Medical School<br>
Private Bag 92019<br>
Auckland, NEW ZEALAND<br>
64 9 373 7599 ext. 6387<br>
FAX 64 9 373 7492<br>
ps.crosier@auckland.ac.nz
<p>
Cunming Duan<br>
School of Fisheries HF-15<br>
University of Washington<br>
Seattle, WA 98195
<p>
Denis Duboule<br>
Depts/Zoology &amp; Animal Biology<br>
Sciences III<br>
30 quai Ernest-Ansermet<br>
1204 Geneva, SWITZERLAND<br>
41-22-702-6770/FAX 41-22-781-1747<br>
duboule@sc2a.unige.ch
<p>
Stephen C. Ekker<br>
The Johns Hopkins University<br>
School of Medicine<br>
Dept/Mol Biology &amp; Genetics<br>
714 PCTB, 725 N. Wolfe St.<br>
Baltimore, MD  21205<br>
(410) 955-1863/ FAX (410) 955-9124<br>
ekker@jhuigf.med.jhu.edu<br>
<p>
Scott E. Fraser<br>
Division of Biology 139-74<br>
Caltech<br>
Pasadena, CA 91125<br>
(818) 395-2790/ FAX (818) 449-5163<br>
sefraser@egg.gg.caltech.edu<br>
<p>
Karl J. Fryxell<br>
Department of Biology<br>
University of California<br>
900 University Avenue<br>
Riverside, CA  92521<br>
(909) 787-5908/ FAX (909) 787-4286<br>
fryxell@ucrac1.ucr.edu
<p>
Peter Good<br>
Laboratory of Molecular Genetics<br>
NICHD<br>
Bldg. 6B, Room 4B/420<br>
Bethesda, MD  20892<br>
good@helix.nih.gov<br>
<p>
Daniel Gros<br>
Lab/Biologie/Diff. Cellulaire<br>
Parc Scientifique de Luminy<br>
13288 Marseilles, Cedex 09<br>
FRANCE
<p>
Miyuki Hashimoto<br>
Biomolecular Medicine, Evans 603<br>
University Hospital<br>
88 E. Newton St.<br>
Boston, MA 02118-2393<br>
(617) 638-6027/ FAX (617) 638-6009<br>
hashimo@mbcrr.harvard.edu
<p>
Philippe Herbomel<br>
D&eacute;partement de Biologie Mol&eacute;culaire<br>
Institut Pasteur<br>
25 rue du Dr Roux<br>
75724 Paris Cedex 15<br>
FRANCE<br>
33-1-45-688490/FAX 33-1-45-688521
<p>
Corinne Houart<br>
Institute of Neuroscience<br>
1254 University of Oregon<br>
Eugene, OR 97403-1254<br>
(503) 346-4596/ FAX (503) 346-4548<br>
houart@uoneuro.uoregon.edu<br>
<p>
Richard Hsieh<br>
211 Waugh Street, Apt. 4<br>
Columbia, MO 65201<br>
FAX &amp; Phone (314) 442-7695
<p>
Abigail Jensen<br>
IRC Lab of Mol &amp; Cell Biology<br>
University College London<br>
Gower Street<br>
London WC1E 6BT, UNITED KINGDOM
<p>
Barry Knox<br>
Department of Biochemistry &amp;<br>
  Molecular Biology<br>
SUNY HSC<br>
750 E. Adams St.<br>
Syracuse, NY 13210
<p>
Christopher C. Kohler<br>
Cooperative Fisheries Research Lab<br>
Southern Illinois University<br>
Carbondale, IL 62901-6511<br>
(618) 536-4126/ FAX (618) 536-7761
<p>
Baoquan Liu<br>
Fisheries Research Lab<brp>
Southern Illinois University<br>
Carbondale, IL 62901<br>
(618) 453-4126/ FAX (618) 536-7761<br>
gr9088@siucvmb.siu.edu
<p>
De-Yu Lu<br>
374 East 13th Avenue  Apt E<br>
Columbus, OH 43201
<p>
Ir&egrave;ne Marics<br>
Lab/Biologie/Diff. Cellulaire<br>
Parc Scientifique de Luminy<br>
13288 Marseilles, Cedex 09<br>
FRANCE<br>
(33)91-26-9240/FAX (33)91-26-9386
<p>
John McCrory<br>
Department of Biology<br>
University of Victoria<br>
Victoria, B.C.<br>
CANADA  V8W 2Y2<br>
(604) 721-7143/ FAX (604) 721-7120<br>
jemcr@uvvm.uvic.ca
<p>
William C. Michel<br>
University of Utah School of Medicine<br>
Dept/Physiology<br>
410 Chipeta Way<br>
Salt Lake City, UT  84108<br>
(801) 585-5420/ FAX (801) 581-3476<br>
wcmichel@cc.utah.edu
<p>
William Muhlach<br>
Department of Zoology<br>
Southern Illinois University<br>
Carbondale, IL 62901-6511
<p>
L.W. Nawrocki<br>
Department of Biology<br>
University College London<br>
Gower Street<br>
London WC1E 6BT<br>
UNITED KINGDOM<br>
44-81-387-7050/ FAX 44-81-380-7026<br>
m.blaxter@ic.ac.uk
<p>
Marcia L. O'Connell<br>
NICHD, LMG<br>
Bldg. 6B, Room 4B/422<br>
Bethesda, MD  20892<br>
o_connell%lmgvax.dnet@dxi.nih.gov
<p>
Michael Pack<br>
CVRC - 149 13th Street<br>
MGH-East, 4th Floor<br>
Charlestown, MA 02129<br>
pack@helix.mgh.harvard.edu
<p>
Debra A. Peattie<br>
1716 Cambridge St.<br>
Cambridge, MA 02138<brp>
(617) 547-1829
<p>
Nigel Pringle<br>
IRC Lab of Mol &amp; Cell Biology<br>
University College London<br>
Gower Street<br>
London WC1E 6BT, UNITED KINGDOM
<p>
Michael Rebagliati<br>
Bldg 6B, Rm 4B-422<br>
LMG-NICHD/NIH<br>
9000 Rockville Pike<br>
Bethesda, MD 20892<br>
(301) 496-9689
<p>
Urs Rutishauser<br>
Dept/Genetics  School of Medicine<br>
Case Western Reserve University<br>
10900 Euclid Avenue<br>
Cleveland, OH  44106<br>
(216) 368-2428/ FAX (216) 368-3182<br>
uxr@po.cwru.edu
<p>
Charles Sagerstr&ouml;m<br>
Whitehead Institute 401<br>
Nine Cambridge Center<br>
Cambridge, MA 02142<br>
(617) 258-5200/ FAX (617) 258-6505<br>
sagerstrom@wi.mit.edu
<p>
John Schmidt<br>
Dept/Biological Sciences<br>
SUNY @ Albany<br>
1400 Washington Avenue<br>
Albany, NY 12222<br>
(518) 442-4309/ FAX (518) 442-4767<br>
js213@albny.1bx.bitnet
<p>
P.J. Seeley<br>
King Alfred's College<br>
Sparkford Road<br>
Winchester, Hampshire <br>
UNITED KINGDOM  S022 4NR<br>
44 962 841515/ FAX 44 962 842280
<p>
Nancy Sherwood<br>
Department of Biology<br>
University of Victoria<br>
Victoria, B.C.<br>
CANADA  V8W 2Y2<br>
(604) 721-7143/ FAX (604) 721-7120<br>
jemcr@uvvm.uvic.ca
<p>
Yoshihiro Takemoto<br>
Nippon Syntex K.K./Inst Immunology<br>
157, Nagai<br>
Niiharimura, Niiharigun, Ibaraki<br>
JAPAN 300-41<br>
81-298-30-6135/ FAX 81-298-30-6150
<p>
Giselle Thibadeau<br>
Department of Molecular Biology<br>
Princeton University<br>
Princeton, NJ 08544<br>
(609) 258-3983/ FAX (609) 258-5323
<p>
Katie Thompson<br>
12991 Caminito Bodega<br>
Del Mar, CA  92014-3818<br>
(619) 792-5060/ FAX (619) 793-7630
<p>
Alan Trimble<br>
Department of Zoology, NJ-15<br>
University of Washington<br>
Seattle, WA 98195<br>
(206) 543-8932<br>
trimblea@zoology.washington.edu
<p>
Sara E. Zalik	Dept/Zoology  Fac of Science<br>
University of Alberta<br>
CW-312 Biological Sciences Centre<br>
Edmonton, Alberta, CANADA T6G 2E9<br>
(403) 492-3308/ FAX (403) 492-7033
<p>
<b><p>
CORRECTIONS</b>
<p>
Danielle Cahaan<br>
14053 113th Avenue N.E.<br>
Kirkland, WA 98034-1039<br>
<p>
William L. Clapp<br>
VAMC (113E) Laboratory Service<br>
University of Florida<br>
1601 S.W. Archer Rd.<br>
Gainesville, FL  32608-1197<br>
(904) 376-1611 ext 6821<br>
clapp.pathlogy@mail.health.ufl.edu
<p>
S.W. de Laat<br>
Netherlands Inst/Devel Biol<br>
Uppsalalaan 8<br>
3584 CT Utrecht<br>
THE NETHERLANDS<br>
31-30-510211/FAX 31-30-516464
<p>
Suzanne Giordano<br>
c/o Prof. Dr. Claudio St&uuml;rmer<br>
Fac/Biology  University of Konstanz<br>
D-7750 Konstanz<br>
GERMANY
<p>
Pascal Haffter<br>
D-72076 T&uuml;bingen  GERMANY
<p>
Andrew Heyward<br>
Dept/Obstetrics &amp; Gynacology<br>
University of Adelaide<br>
G.P.O. Box 498<br>
Adelaide, SA 5001, AUSTRALIA<br>
618-303-5100/FAX 618-303-4099<br>
andrew_heyward@medicine.ccmail.adelaide.edu.au
<p>
Robert Ho<br>
Department of Molecular Biology<br>
Princeton University<br>
Princeton, NJ 08544<br>
(609) 258-2887/ FAX (609) 258-5323<br>
rho@molbiol.princeton.edu
<p>
Don Kane<br>
D-72076 Tubingen 1, GERMANY
<p>
Sigrun Korsching<br>
D-72076 T&uuml;bingen  GERMANY
<p>
Dennis Liu<br>
(206) 543-3117/ FAX (206) 543-0754<br>
dliu@genetics.washington.edu
<p>
James B. Lorens<br>
jim.lorens@pki.uib.no
<p>
Brian Metscher<br>
metscher@toad.bio.uci.edu
<p>
Paul Z. Myers<br>
Department of Biology<br>
Temple University<br>
Philadelphia, PA 19122
<p>
Ruthann Nichols<br>
(313) 764-4461/ FAX (313) 764-0884
<p>
Christiane N&uuml;sslein-Volhard<br>
D-72076 T&uuml;bingen 1  GERMANY
<p>
Laszlo Orb&aacute;n<br>
36-28-330-600 / FAX 36-28-320-096 <b><br>
or</b> 36-28-330-416
<p>
James B. Rand<br>
(405) 271-7681/ FAX (405) 271-3153
<p>
Nisson Schechter<br>
nschechter@epo.hsc.sunysb.edu
<p>
Reiko Toyama<br>
NICHD, LMG<br>
Bldg. 6B  Room 420<br>
Bethesda, MD  20892<br>
toyama%lmgvax.dnet@dxi.nih.gov
<p>
Rachel Warga<br>
D-72076 Tubingen 1, GERMANY
<p>
Pam Yelick<br>
Forsyth Dental Center<br>
140 The Fenway<br>
Boston, MA 02115<br>
linda@neu.edu
<b><p>
<H3><A NAME="21">ZEBRAFISH REFERENCE UPDATE</A></H3><br>
Added since August 9, 1993</b>
<p>
Burns, J.C., T. Friedmann, W. Driever, M. Burrascano, and J.-K. Yee (1993) Vesicular stomatitis virus G glycoprotein pseudotyped retroviral vectors: concentration to very high titer and efficient gene transfer into mammalian and nonmammalian cells.  <i>Proc. Natl. Acad. Sci. USA</i>, <b>90</b>:8033-8037.<p>
Eisen, J.S. and J.A. Weston (1993) Development of the neural crest in the zebrafish.  <i>Devel. Biol.</i> <b>159</b>:50-59.<p>
Fjose, A., S. Nornes, U. Weber, and M. Mlodzik (1993) Functional conservation of vertebrate seven-up related genes in neurogenesis and eye development.  <i>EMBO J.</i> <b>12</b>:1403-1414.<p>
Groth, G., K. Schreeb, V. Herdt, and K.J. Freundt (1993) Toxicity studies in fertilized zebrafish eggs treated with N-methylamine, N,N-dimethylamine, 2-aminoethanol, isopropylamine, aniline, N-methylaniline, N,N-dimethylaniline, quinone, chloroacetaldehyde, or cyclohexanol.  <i>Bull. Environ. Contam. Toxicol.</i> <b>50</b>:878-882.<p>
Kelly, G.M., C.-J. Lai, and R.T. Moon (1993) Expression of <i>Wnt10a</i> in the central nervous system of developing zebrafish. <i>Devel. Biol.</i> <b>158</b>:113-121.<p>
Meyer, A., C.H. Biermann, and G. Ort&iacute; (1993) The phylogenetic position of the zebrafish (<i>Danio rerio</i>), a model system in developmental biology: an invitation to the comparative method.  <i>Proc. R. Soc. Lond.</i> <b>252</b>:231-236.<p>
Miranda, C.L., P. Collodi, X. Zhao, D.W. Barnes, and D.R. Buhler (1993) Regulation of cytochrome P450 expression in a novel liver cell line from zebrafish (<i>Brachydanio rerio</i>).  <i>Arch. Biochem. Biophys.</i> <b>305</b>:320-327.<p>
Mullins, M.C. and C. N&uuml;sslein-Volhard (1993) Mutational approaches to studying embryonic pattern formation in the zebrafish.  <i>Curr. Op. Genet. Devel.</i> <b>3</b>:648-654.<p>
Ono, H., C. O'Huigin, V. Vincek, R.J. Stet, F. Figueroa, and J. Klein (1993) New beta chain-encoding Mhc class II genes in the carp.  <i>Immunogenetics</i> <b>38</b>:146-149.<p>
Platt, C. (1993) Zebrafish inner ear sensory surfaces are similar to those in goldfish.  <i>Hear. Res.</i> <b>65</b>:133-140.<p>
Robinson, J., E.A. Schmitt, F.I. H&aacute;rosi, R.J. Reece, and J.E. Dowling (1993) Zebrafish ultraviolet visual pigment: absorption spectrum, sequence, and localization.  <i>Proc. Natl. Acad. Sci. USA</i> <b>90</b>:6009-6012.<p>
Str&auml;hle, U., P. Blader, D. Henrique, and P.W. Ingham (1993) <i>Axial</i>, a zebrafish gene expressed along the developing body axis, shows altered expression in <i>cyclops</i> mutant embryos.  <i>Genes Dev.</i> <b>7</b>:1436-1446.
<p><hr>
<A HREF="/index.html">ZFIN  <img src="fish_net.gif"></A>
</body>
