<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html>
	<head>
		<title>
			ZFIN News 2007
		</title>
	</head>

<script language="JavaScript" src="/header.js"></script>

<table width=100% cellspacing=0>
  <tr>
   <td width=80% align=left>
     <h1>2007 Zebrafish News </h>
   </td>
   <td width=10% align=right>
    View news for
   </td>
   <td width=10% align=left> 	
     <form name="news">
       <DIV>
          <select name="year" onChange=
            "document.location.href=
            document.news.year.options[document.news.year.selectedIndex].value">
            <option value=""SELECTED>Select year
	    <option value="siteNews.html">2007
            <option value="siteNews_2006.html">2006
            <option value="siteNews_2005.html">2005 
          </select></DIV>
      </form>
    </td>
  </tr>
</table>

<p>This page contains news relevant to the zebrafish research community.

If you have any questions, comments or would like to submit an entry, please 
contact <a href="mailto:zfinadmn@zfin.org">Jonathan Knight</a>.
<br><br>

<h3><a name="newsletter">"ZFIN NEWS" is the ZFIN Newsletter, published bi-annually at the University of Oregon.  </h3>

<a href="ZFIN_news_Spring_07.pdf"> Vol. 4, No. 1 - Spring 2007</a>
		 <blockquote>
		 Expanded support for Genotypes and Phenotypes <br>
		 ZFIN gets a Makeover <br>
		 New ZIRC Home and Shopping Pages</blockquote>
		 
		 		 <p>Click here for our <a href="/zf_info/news/Newsletters.html">ZFIN Newsletter Archive.</a></p><br>


		 
      <h2><a name="vega0718">New Vega release</a></h2>
      
      Wednesday, July 18<br>
      <p>The Sanger Institute has released a Vega update on the 12th of July 2007. The Vega database contains finished clone sequence with manual annotation, placed on chromosomes according to data from the physical mapping project. The current release features 9,254 clones, 2,411 of which have been annotated. 10,374 genes have been manually annotated on over 370 Mb of annotated sequence. </p>
<p>
You can browse/search Vega at <a href="http://vega.sanger.ac.uk/Danio_rerio">http://vega.sanger.ac.uk/Danio_rerio</a>. If you have comments on the annotation or arrangements of clones, or want a clone to be annotated, please contact <a href="mailto:zfish-help@sanger.ac.uk">zfish-help@sanger.ac.uk</a>. 
</p><br>

		 
      <h2><a name="zv7">Zv7 Assembly and Pre Ensembl</a></h2>
      
      Wednesday, July 18<br>
      
      <p>The Sanger Institute has released the seventh zebrafish genome assembly, Zv7, on the 13th of July 2007. This assembly comprises 1.44 Gbp in 5,036 scaffolds. It was generated starting with 1.02 Gbp from 7,823 BAC clones and supplemented with a 5.5x whole genome shotgun assembly. The sequences were integrated using alignments as well as BAC end and marker placement. Over 200 Mbp of duplicated sequence has been removed compared to Zv6. The assembly can be downloaded at <a href="ftp://ftp.ensembl.org/pub/assembly/zebrafish/Zv7release">ftp://ftp.ensembl.org/pub/assembly/zebrafish/Zv7release</a> and more information can be found at <a href="http://www.sanger.ac.uk/Projects/D_rerio/Zv7_assembly_information.shtml">http://www.sanger.ac.uk/Projects/D_rerio/Zv7_assembly_information.shtml</a>.
      
      <p>The Zv7 sequence has been used to generate an Ensembl database available at <a href="http://pre.ensembl.org/Danio_rerio/index.html">http://pre.ensembl.org/Danio_rerio/index.html</a>. This is an extended Pre Ensembl site, providing a full Ensembl gene set for Zv7. A full Ensembl site, providing comparative genomic and data mining tools, will be available for the next Ensembl release in August. 

<br><br><br>
      
      <h2><a name="papers">Call for Papers - Disease-based review to the American Journal of Medical Genetics</a></h2>
      
      Friday, March 16<br>
      
      <p>Dear ZFIN community,

 <p>We would like to request a call for papers concerning a disease-based review to the American Journal of Medical Genetics.  The focus of this manuscript is understanding human malformation phenotypes using the zebrafish model.   We believe such a review, written for an audience of primarily non-zebrafish scientists, would be helpful in communicating the value of this model organism for scientists and policy-makers considering alternate vertebrate model systems. We had started the process of drafting this review late last year, and at the Investigator's Workshop, several members of the community expressed interest in such a document for the whole community.

 <p>We would like to make this review as comprehensive as possible and don't want to miss any of the great papers that have come out in this area. If you have a publication that is out or about to be out (i.e. in press), we would love to receive the citation and/or preprint so we can include your science in this review. All materials received prior to publication release will be held in strict confidence until the appropriate paper has come out in the respective journal.

 <p>Sincerely,
Shelly Myers, Genetics, Cell Biology and Genetics - University of Minnesota
Lisa Schimmenti, MD, Pediatrics- University of Minnesota
Stephen C. Ekker, Genetics, Cell Biology and Genetics - University of Minnesota

 <p>Please send your information to:<br>
<br>Stephen C. Ekker  <a href=mailto:ekker001@umn.edu>ekker001@umn.edu</a>
<br>Lisa Schimmenti  <a href=mailto:las@umn.edu>las@umn.edu</a>

 <p>Thanks! 
<br><br><br>

<h2><a name="antibodies">Suggest Targets for Anti-Zebrafish Antibodies</a></h2>
Friday, March 9<br>
<p>The need for more anti-zebrafish antibodies was highlighted at the
recent conference of zebrafish investigators at Asilomar. To try and
address this, I have made some preliminary enquiries to an antibody
supplier to see if this niche market would be commercially viable. The
company that has shown some interest has expertise in making
polyclonal affinity-purified anti-peptide antibodies in goats. To make the most
of this opportunity, I need to know which antibodies would initially be
the most useful to the zebrafish community as a whole.</p>

<p>To obtain these data, please email me, <a href="mailto:gw2@sanger.ac.uk">gw2@sanger.ac.uk</a>, by Friday 6th April with a list of preferred targets identified by a GenBank or
ENSembl accession number. Preference would be given according to the
following criteria:</p>
<ol>
<li>
Popularity (number of requests for a particular antibody);
</li>
<li>
Willingness of the requestor to test a small aliquot of the
antibody and provide these data *quickly* to the company so that they can show
it works; please indicate in your reply if you'd be willing to do this.<br>
</li>
</ol>
Thanks,<br>

Gavin<br><br>

Gavin J. Wright<br>
Cell Surface Signalling Laboratory<br>
<br><br><br>


<h2><a name="amster">Deadlines for the 5th European Zebrafish Genetics and Development Meeting</a></h2>
Wednesday, March 7
<p>Please note that the abstract submission deadline for the 5th European Zebrafish Genetics and Development Meeting is March 9th, 2007. Also, March 9th is the deadline for early registrations, which are somewhat cheaper than later registrations.
 
 <p>For more information, see the official Meeting website: <a href="http://www.zebrafish2007.org/">http://www.zebrafish2007.org/</a>
		 
<br><br><br>

<h2><a name="genetools">Quality control testing of Gene Tools morpholinos</a></h2>
Friday, January 26<br>
<p><a href="http://zfin.org/zf_info/catch/moqc.pdf">Efficacy Test of Chordin Morpholinos</a> 

<br><br><br>


<h2><a name="hgnc">Zebrafish genes now included in HGNC Comparison of Orthology Predictions (HCOP)</a></h2>
Thursday, January 25

<p>Zebrafish genes have been added to the HGNC HCOP search tool (<a href="http://www.gene.ucl.ac.uk/hcop">http://www.gene.ucl.ac.uk/hcop</a>) which compares orthology assertions from different databases. 

<br><br>


</body>


<script language="JavaScript" src="/footer.js"></script>
