<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html>
 <head>
  <title>ZFIN News</title>
  <link rel=alternate type="application/rss+xml" href="siteNews.rss" title="RSS feed for Zfin News">
  <script language="JavaScript" src="/javascript/header.js" type="text/javascript"></script>

<table width=100% cellspacing=0>
  <tr>
   <td width=80% align=left>
     <h1>Zebrafish News</h1>
   </td>
   <td width=10% align=right>
    View news for
   </td>
   <td width=10% align=left>
     <form name="news">
       <DIV>
          <select name="year" onChange=
            "document.location.href=
            document.news.year.options[document.news.year.selectedIndex].value">
            <option value=""SELECTED>Select year
            <option value="siteNews.html#2009">2009
            <option value="siteNews.html#2008">2008
            <option value="siteNews.html#2007">2007
            <option value="siteNews.html#2006">2006
            <option value="siteNews.html#2005">2005
          </select></DIV>
      </form>
    </td>
  </tr>
</table>

<p>This page contains news relevant to the zebrafish research community.

If you have any questions, comments or would like to submit an entry, please
contact <a href="mailto:zfinadmn@zfin.org"> Jonathan Knight </a>.
<br><br>

<!-- BEGIN TEMPLATE

<h2><a name="NAME">Headline</a></h2>
DD MMM 2010<br>
<div><p>
</div>

---- END TEMPLATE -->

<h2><a name="rnaseq">RNAseq gene models in Ensembl 58</a></h2>
23 Jun 2010<br>
<div><p>
Ensembl release 58 contains a new set of gene models using Illumina RNAseq data from 9 zebrafish tissues and developmental stages.   Fragmented polyA+ RNA was paired end sequenced to read lengths of 37, 54 or 76 bases at the Wellcome Trust Sanger Institute as part of the Zebrafish Transcriptome Sequencing Project, European Read Archive ERP000016.   Gene models have been automatically produced by an initial local alignment to Zv8 to identify exons, then read pair information was used to call adjacent exons and approximate transcripts were assembled.     Reads were realigned to these transcripts with a splice model.   Reads which splice relative to the genomic reference sequence defined canonical and non-canonical intron.      All variants were called using the approximate transcripts and spliced read information and Ensembl displays the gene models with the most supporting reads.    For further details see

<p><a href="http://www.ensembl.org/info/docs/genebuild/zebrafish_rnaseq_annotation.html">http://www.ensembl.org/info/docs/genebuild/zebrafish_rnaseq_annotation.html</a>

<p><a href="http://ensembl.blogspot.com/2010/06/zebrafish-rna-seq-gene-models.html">http://ensembl.blogspot.com/2010/06/zebrafish-rna-seq-gene-models.html</a>

 
</div>


<h2><a name="gbrowse">GBrowse genome browser now available at ZFIN</a></h2>
22 Apr 2010<br>
<div>
<p>GBrowse at ZFIN

<p>The zebrafish genomic sequence and annotated data can now be explored
at ZFIN using the Generic Genome Browser, or GBrowse. GBrowse at ZFIN
hosts the Sanger Institute's genome assembly, based on fully sequenced
BAC and PAC clones from the Tuebingen strain with Vega gene and
transcript annotation. The data in the tracks show genome annotation,
including the position of genes on the linkage groups, alternatively
spliced transcripts, BACs used as assembly components, and genes that
have expression and/or phenotype data stored in ZFIN.

<p>ZFIN will continue to add tracks to GBrowse incrementally, based on
availability of data and community interest.

<p>GBrowse at ZFIN can be launched by clicking on the GBrowse link on the
ZFIN home page (<a href="http://zfin.org">http://zfin.org</a>) or directly at <a href="http://zfin.org/genome-browser">http://zfin.org/genome-browser</a>
. GBrowse can also be accessed by clicking on the image on ZFIN
transcript pages, and on assembly BAC pages. A link to a GBrowse
tutorial from Open Helix is available on the ZFIN GBrowse page.
</div>



<h2><a name="mol">New article about curation published in Molecular Genetics and
Genomics  - March 24, 2010
</a></h2>
09 Apr 2010<br>
<div><p>
The process of literature curation is described in <a href="http://dx.doi.org/10.1007/s00438-010-0525-8
">Hirschman, et al., A MOD(ern) perspective on literature curation,
Mol Genet Genomics (2010),</a> written by four curators from  <a href="http://www.yeastgenome.org/
">SGD</a> (The Saccharomyces Genome Database), <a href="http://www.arabidopsis.org/
">TAIR</a> (The Arabidopsis Information Resource), <a href="http://zfin.org
">ZFIN</a> (The Zebrafish Information Network), and <a href="http://www.informatics.jax.org/
">MGI</a> (Mouse Genome Informatics).  This perspective discusses how
papers are identified and prioritized for curation, some of the
challenges MOD curators face, and ways in which researchers and
publishers can support and facilitate the work of curators.

</div>

<h2><a name="outage">Expected Downtime, Saturday, March 20.</a></h2>
16 Mar 2010<br>
<div>
<p>ZFIN will experience periods of downtime next Saturday, March 20.  Interruptions in service can be expected between the hours of 2 pm and 10 pm, PST. We apologize for the inconvenience.
</div>

<h2><a name="vega17">New Vega Release</a></h2>
17 Dec 2009<br>
<div>
<p>
The Sanger Institute has released a Vega update on the 17th of December 2009. The Vega database contains finished clone sequence with manual annotation, placed on chromosomes according to data from the physical mapping project. The current release features 10,755 clones with 15,761 manually annotated genes.

<p>You can browse/search Vega at <a href="http://vega.sanger.ac.uk/Danio_rerio">http://vega.sanger.ac.uk/Danio_rerio</a>. If you have comments on the annotation or arrangements of clones, or want a clone to be annotated, please let us know at <a href="mailto:zfish-help@sanger.ac.uk">zfish-help@sanger.ac.uk</a>

</div>


<h2><a name="chem">Chemical Biology Session</a></h2>
01 DEC 2009<br>
<div>
<p>We were very excited by the large number of people who attended the Chemical Biology session in Rome, and want to follow up on the idea of an on-line discussion group. We hope this forum may provide an opportunity to discuss chemical biology techniques, to help each other develop the field, share ideas, and identify unmet needs in the community.

<p>We would like to invite you to join the next Chemical Biology discussion using free broadcasting (<a href="http://www.ustream.tv/">http://www.ustream.tv/</a>)  on January 20 at 11:00 am EST.

<p>You will need to login, and then join Len Zon's "Chemical Biology Zebrafish" sessions. We recommend signing up a few days before, so that you are able to be sure to be able to access the live broadcast. For this broadcast Len Zon and Randy Peterson will host the video, and we can join in for the discussion via the "chat room". This provides a rapid way to ask questions, to which Len and Randy can respond.

<p>The Ustream technology will allow us to ask questions live. However, please also feel free to send Liz Patton (<a href="mailto:epatton@staffmail.ed.ac.uk">epatton@staffmail.ed.ac.uk</a>) some questions before the broadcast, and she can send them on to Len and Randy in advance.

<p>The broadcast will be at 11am from Boston (EST). This is 8am for the West Coast (California), 4pm for the UK, and 5pm for Holland, France and Germany.
</div>

<h2><a name="react">Reactome Pathway Database User Survey</a></h2>
01 Oct 2009<br>
<div>
<p>Reactome is committed to providing access to high-quality pathway information and helpful data analysis tools.  With this in mind, we are actively soliciting comments from the research community in order to assess community needs.  We are interested to hear about your experience with Reactome, and would like to know a bit about your background and research interests so that we can continue to improve the Reactome site and tools.

<p>You can access the survey at: <a href="http://tinyurl.com/l48zzq">http://tinyurl.com/l48zzq</a>
</div>



<a name="2009"></a>

<h2><a name="NIHguide">NIH Guide Notice Published </a></h2>
21 Aug 2009<br>
<div>
<p>Revised Resource Sharing Plan Instructions for PAR-08-138 and
PAR-08-139, Genetic Screens to Enhance Zebrafish Research and
Enhancing Zebrafish Research with Research Tools and Techniques (R01)
(NOT-HD-09-019)
Eunice Kennedy Shriver National Institute of Child Health and Human
Development
National Institute of Diabetes and Digestive and Kidney Diseases
<a href="http://grants.nih.gov/grants/guide/notice-files/NOT-HD-09-019.html">http://grants.nih.gov/grants/guide/notice-files/NOT-HD-09-019.html</a>


<p>Notice Number: NOT-HD-09-019

<p>Key Dates
<p>Release Date: August 7, 2009

<p>Issued by
Eunice Kennedy Shriver National Institute of Child Health and Human
Development (NICHD) (<a href="http://www.nichd.nih.gov">http://www.nichd.nih.gov</a>)
National Institute of Diabetes and Digestive and Kidney Diseases
(NIDDK), (<a href="http://www2.niddk.nih.gov">http://www2.niddk.nih.gov</a>)

<p>Purpose

<p>Applicants responding to PAR-08-138 (<a href="http://grants.nih.gov/grants/guide/pa-files/PAR-08-138.html">http://grants.nih.gov/grants/
guide/pa-files/PAR-08-138.html</a>), Genetic Screens to Enhance Zebrafish
Research, and/or PAR-08-139 (<a href="http://grants.nih.gov/grants/guide/pa-files/PAR-08-139.html">http://grants.nih.gov/grants/guide/pa-
files/PAR-08-139.html</a>), Enhancing Zebrafish Research with Research
Tools and Techniques, when preparing their Resource Sharing Plan are
strongly encouraged to contact the Zebrafish International Resource
Center (ZIRC, <a href="http://zfin.org/zirc/home/stckctr.php">http://zfin.org/zirc/home/stckctr.php</a>) to discuss their
plans for sharing resources created under their proposed application
and to receive a cost estimate for deposition of materials at ZIRC.
As per NOT-OD-04-042 (<a href="http://grants.nih.gov/grants/guide/notice-files/NOT-OD-04-042.html">http://grants.nih.gov/grants/guide/notice-files/
NOT-OD-04-042.html</a>), investigators may request funds in their
application to defray reasonable costs associated with sharing
materials or data or transfer of model organisms and associated data
to appropriate repositories.  These costs should be considered in
developing the budget for the project.  For applications with modular
budgets, this cost estimate should be included as part of the resource
sharing plan.  For non-modular applications with detailed budgets, the
cost estimate should be justified as part of the requested budget.

<p>All other aspects of these FOAs remain the same.

<p>Inquiries

<p>Questions about this Notice should be directed to:

<blockquote>
<br>Lorette C. Javois, Ph.D.
<br>Eunice Kennedy Shriver National Institute of Child Health and Human
Development
<br>National Institutes of Health
<br>Building 6100, Room 4B01, MSC 7500
<br>Bethesda, Maryland 20892-7500
<br>Phone: 301-496-5541
<br>Email: <a href="mailto:lj89j@nih.gov">lj89j@nih.gov</a> (e-mail is the strongly preferred method for
inquiries)
</blockquote>


</div>

<h2><a name="wikipath">New at WikiPathways</a></h2>
01 Aug 2009<br>
<div>
<p>
WikiPathways has recently quadrupled its zebrafish pathway collection, which now totals over 100 pathways representing over 2,400 unique genes. At WikiPathways, you can browse, edit, download and publish pathways that are annotated for computational analysis and illustrated for human consumption. Check out the new pathways <a href="http://www.wikipathways.org/index.php?title=Special%3ABrowsePathwaysPage&amp;browse=Danio+rerio&amp;http://www.wikipathways.org/index.php?title=Special%3ABrowsePathwaysPage&amp;browse=Danio+rerio&amp;browseCat=All+Categories">here</a>.
<p>
WikiPathways has also recently published an article describing the new web service API, which allows programmatic access to search, download, and data mapping functions. Check out the PLoS One article <a href="http://www.plosone.org/article/info%3Adoi%2F10.1371%2Fjournal.pone.0006447">here</a>.
</div>

<h2><a name="vega2">New Vega Release</a></h2>
08 Apr 2009<br>
<div>
<p>The Sanger Institute has released a Vega update on the 30th of March 2009. The Vega database contains finished clone sequence with manual annotation, placed on chromosomes according to data from the physical mapping project. The current release features 8,954 clones with 14,707 manually annotated genes.

<p>You can browse/search Vega at <a href="http://vega.sanger.ac.uk/Danio_rerio">http://vega.sanger.ac.uk/Danio_rerio</a>. If you have comments on the annotation or arrangements of clones, or want a clone to be annotated, please let us know at <a href="mailto:zfish-help@sanger.ac.uk">zfish-help@sanger.ac.uk</a>
</div>

<h2><a name="zfpoll">Antibodies Poll</a></h2>
06 Mar 2009<br>
<div>

<p>Dear Zebrafish researchers,

<p>The need for more anti-zebrafish antibodies was again highlighted at the recent conference of zebrafish investigators at Asilomar. To facilitate the involvement of the commercial sector in addressing this problem, I am conducting another community poll to identify the most popular antibody targets. AnaSpec, a commercial antibody manufacturer has committed to producing 100 rabbit anti-peptide polyclonal antibodies with some level of validation.

<p>Email me: <a href="mailto:gw2@sanger.ac.uk">gw2@sanger.ac.uk</a> by Monday 6th April, with a list of preferred targets unambiguously identified by a GenBank accession number. Please forward this message to your local mailing lists.

<p>Thanks,

<p>Gavin

<blockquote>
<br>Gavin J. Wright
<br>Cell Surface Signalling Laboratory
 <br>Wellcome Trust Sanger Institute
<br>Cambridge CB10 1HH
 <br>T: 01223 496852
</blockquote>



</div>

<h2><a name="lnmap">LN54 panel interactive mapping</a></h2>
28 Jan 2009<br>
<div>

<p>The LN54 panel interactive mapping service can be accessed through ZFIN by following the "Interactive Mapping" link at <a href="http://zfin.org/cgi-bin/webdriver?MIval=aa-crossview.apg&amp;OID=ZDB-REFCROSS-990426-6">http://zfin.org/cgi-bin/webdriver?MIval=aa-crossview.apg&OID=ZDB-REFCROSS-990426-6</a>
Every attempt will be made to return a mapping location to you within 2 business days.  Graphics will not be available.
</div>

<h2><a name="tilling">TILLING online submission site available</a></h2>
22 Jan 2009<br>
<div>

<p>The Zebrafish TILLING Consortium is pleased to announce the launch of a submission site for TILLING requests. TILLING is a method for finding unique ENU-induced mutations in specific genes in a library of mutagenized fish. This project is a joint effort of the Moens lab at the Fred Hutchinson Cancer Research Center, the Solnica-Krezel lab at Vanderbilt University, and the Postlethwait lab at the University of Oregon. The goal of the project is to generate and distribute loss-of-function mutations in genes of interest to the zebrafish community. Our labs have established large cryopreserved libraries of ENU-mutagenized zebrafish and have identified loss-of-function mutations in over 70 zebrafish genes to date. Our consortium-based approach will be to screen genes of interest sequentially at the three locations. This work is supported by a multi-PI NIH Grant HG002995-04, which provides salary support for TILLING personnel. The cost of TILLING consumables is passed on to the requesters.

<p>To learn more about the project and to submit your requests online, go to:

<p><a href="http://www.fhcrc.org/tilling">http://www.fhcrc.org/tilling</a>

</div>

<h2><a name="2008">2008 News Archive</a><br><br>
<a name="Zv8">Zv8 Assembly, Pre Ensembl and Community Annotation Tool Released</a></h2>
12 Dec 2008<br>
<div>

<h3><p>Zv8 Assembly and Pre Ensembl</h3>

<p>The Sanger Institute has released the eighth zebrafish genome assembly, Zv8, on the 11th of December 2008. This assembly comprises 1.48 Gbp in 11,623 scaffolds. It was generated starting with 1.1 Gbp from 9,816 BAC clones and supplemented with a 7x whole genome shotgun assembly. Particular attention has been paid to sort the clone and contig order according to the genetic and radiation hybrid maps as explained in detail at <a href="http://www.sanger.ac.uk/Projects/D_rerio/assembly_notes/Zv8_assembly_notes.shtml">http://www.sanger.ac.uk/Projects/D_rerio/assembly_notes/Zv8_assembly_notes.shtml</a>.

<p>More information is available from <a href="http://www.sanger.ac.uk/Projects/D_rerio/wgs.shtml">http://www.sanger.ac.uk/Projects/D_rerio/wgs.shtml</a>.

<p>The Zv8 sequence has been used to generate an Ensembl database available at <a href="http://pre.ensembl.org/Danio_rerio">http://pre.ensembl.org/Danio_rerio</a> featuring sequence and automated feature annotation plus search and download facilities. A full Ensembl site, providing a gene build and comparative genomic and data mining tools, will be available in spring 2009.

<h3><p>Community Annotation Tool</h3>

<p>If you want to comment on certain regions in the Zv8 assembly, you can do so now using the webform at <a href="http://www.sanger.ac.uk/cgi-bin/Projects/D_rerio/Annotation/submitAnnotation.pl">http://www.sanger.ac.uk/cgi-bin/Projects/D_rerio/Annotation/submitAnnotation.pl</a>. Your report will be automatically saved to a database and the comments can be viewed as a feature track on the Zv8 pre-ensembl web site at http://www.pre-ensembl.org/Danio_rerio and later in the full Ensembl release. Please follow the instructions on the community annotation web site. Your annotation is visible to the public to inform them about issues with the assembly. We will be using your information to improve the clone path and the next assembly Zv9.

<p>Please send your questions/comments to <a href="mailto:zfish-help@sanger.ac.uk">zfish-help@sanger.ac.uk</a>.
</div>

<h2><a name="zfn">Zinc Finger Protein Search Algorithms On-line</a></h2>
03 Oct 2008<br>
<div>

<p>We would like to let the zebrafish community know about the availability of reagents and resources relating to our recent paper describing the use of zinc finger nucleases to generate targeted mutations in zebrafish.  These tools will enable labs to perform bacterial 1-hybrid selections to make zinc finger proteins against a particular target site of interest.  These resources include the plasmids and the bacterial selection strain that are now available through www.addgene.org, along with an on-line tool that will help identify good sites and generate libraries for screening.

<p>To download a PDF that includes a list of plasmids and strains, go to:
<a href="http://lawsonlab.umassmed.edu/PDFs/B1Hrec.pdf">http://lawsonlab.umassmed.edu/PDFs/B1Hrec.pdf</a>

<p>For tips getting started with the bacterial 1-hybrid:
<a href="http://lawsonlab.umassmed.edu/PDFs/B1HStart.pdf">http://lawsonlab.umassmed.edu/PDFs/B1HStart.pdf</a>

<p>For the on-line search tool to identify target sites and design library oligos, go to:
<a href="http://pgfe.umassmed.edu/ZFPsearch.html">http://pgfe.umassmed.edu/ZFPsearch.html</a>

<p>Related references:<br>
Noyes et al. (2008) Nucleic Acids Res.  36(8):2547-60<br>
Meng et al. (2008) Nat Biotechnol.  26(6):695-701

<p>We will try to make additional documentation on these available as soon as possible.  Check back at the Lawson (<a href="http://lawsonlab.umassmed.edu">http://lawsonlab.umassmed.edu</a>) and Wolfe Lab (<a href="http://labs.umassmed.edu/WolfeLab">http://labs.umassmed.edu/WolfeLab</a>) websites for updates.</div>


<h2><a name="ab">Antibodies join ZFIN</a></h2>
03 Oct 2008<br>
<div>

<p>Finding an antibody that works on zebrafish is about to get easier.  The new <a href="http://zfin.org/action/antibody/search">antibody search page</a> lets you search for antibodies by name or labeling pattern.  Antibody marker pages show a summary of labeling in wild-type fish along with other useful data.

We are starting with the full set of antibodies available at <a href="http://zebrafish.org/zirc/home/guide.php">ZIRC</a>, and will be adding many more antibodies described in the literature.  You can help by <a href="mailto:zfinadmn@zfin.org">telling us</a> which antibodies are most useful to you.</div>

<h2><a name="clones">1819 ZGC clones recloned in pENTR223.1</a></h2>
28 May 2008<br>
<div>

<p>ZGC has added 1819 full length ZGC clones recloned into a Gateway Entry vector (pENTR223.1; <a href="http://mgc.nci.nih.gov/Vectors/prot_pENTR223.1">http://mgc.nci.nih.gov/Vectors/prot_pENTR223.1</a>) for expression analysis and other applications.

 <p>These clones can be found by searching Entrez nucleotide for "NIH_ZGC_36".</div>

<h2><a name="snp">Retrieve zebrafish SNPS in ZFIN</a></h2>
08 May 2008<br>
<div><p>

<p>Zebrafish SNPs can be retrieved in ZFIN. This is the first phase of implementation.  We provide the ability to access SNPs mapped to a genomic clone.  Options are provided to retrieve a text file or conduct a Batch Query at NCBI. For example, see:
<a href="http://zfin.org/cgi-bin/webdriver?MIval=aa-markerview.apg&amp;OID=ZDB-BAC-041007-170">BAC DKEY-5N18</a>

   <p>and

<p>the <a href="http://zfin.org/cgi-bin/webdriver?MIval=aa-snpview.apg&amp;OID=ZDB-BAC-041007-170&amp;snppubs=1">SNP retrieval page</a> for DKEY-5N18.

 <p>The full implementation including the ability to search for and retrieve SNP information directly as well as association between SNPs and genes will happen after the implementation of a Genome Browser in ZFIN. This will allow the SNP information to be placed in a genomic context.
</p>
</div>



<h2><a name="NIH">Two funding opportunities at NIH</a></h2>
14 Apr 2008<br>
<div>
 <p>NIH and the Trans-NIH Coordinating Committee have published two new

Program Announcements with special Receipt dates (PARs) soliciting R01
applications for research using zebrafish.  Please note that not all
NIH Institutes and Centers are participating in both of these
solicitations; therefore, only applications relevant to the missions
of participating Institutes and Centers will be considered responsive
to the announcements.  These two new PAR's can be found in the NIH
Guide at the following urls:

	<p><a href="http://grants.nih.gov/grants/guide/pa-files/PAR-08-138.html">http://grants.nih.gov/grants/guide/pa-files/PAR-08-138.html</a>

	<p>Genetic Screens to Enhance Zebrafish Research (R01)

	<p><a href="http://grants.nih.gov/grants/guide/pa-files/PAR-08-139.html">http://grants.nih.gov/grants/guide/pa-files/PAR-08-139.html</a>

	<p>Enhancing Zebrafish Research with Research Tools and Techniques (R01)
</div>


<h2><a name="newsletter">"ZFIN NEWS" is the ZFIN Newsletter, published bi-annually at the University of Oregon.</a></h2>
25 Mar 2008<br>
<div>
<a href="ZNEWS_Spring_08.pdf"> Vol. 5, No. 1 - Spring 2008</a>
		 <blockquote>
		 Maximizing Data Impact  <br>
		 Anatomy Pages Expanded <br>
		 Full Text publications <br>
		 Morpholino Database (MODB) and ZFIN <br>
		 Microarray Data at ZFIN<br>
		 Nomenclature Guidelines Updated<br>
		 </blockquote>

<p>Click here for our <a href="/zf_info/news/Newsletters.html">ZFIN Newsletter Archive.</a></p><br>
</div>

<h2><a name="madmtg">Meeting Website for the 2008 International Zebrafish meeting has now gone live</a></h2>
19  Feb 2008<br>
<div>
<p>We are pleased to announce that the meeting website for the 2008 International Zebrafish meeting has now gone live:

<p><a href="http://www.union.wisc.edu/zebrafish/index.html">http://www.union.wisc.edu/zebrafish/index.html</a>

<p>A description of the meeting follows below, and more information is available on the website. Important deadlines: Abstract submission, April 14, 2008; Lodging reservations, May 19, 2008.

<p>8th International Meeting on Zebrafish Development and Genetics<br>
June 25-29, 2008 - Madison, WI, USA<br>
The organizers of the 2008 International Zebrafish Development and Genetics Conference welcome all interested scientists to join us for discussion of their latest results and ideas. The schedule will feature a variety of oral sessions, posters, and vendors, as well as opportunities to socialize on the shores of Lake Mendota.

<p>This year's schedule will feature extra time for free discussion, as well as a workshop format featuring topics suggested by the community. (Thanks to all of those who wrote in to suggest workshop topics.) We have also provided extra information about bus connections, as an alternate way to reach Madison from Chicago or Milwaukee airports.
</div>

<h2><a name="zfmtg">Call for Suggestions for workshops at 2008 Madison meeting</a></h2>
8 Jan 2008<br>
<div>
<p>Planners are working hard on the 2008 international zebrafish meeting (June 25-29,
Madison), and would like to solicit your help. This year's meeting will introduce workshops on specialized topics proposed by the community. These will be organized in two sessions, each with up to 5 workshops running concurrently. Workshops will last two hours, and might consist of 3 or 4 short ( e.g. 15 min) talks, with extensive time for discussion.

<p>We are now asking for the community to propose the topics for these workshops. Topics will be posted on a blog: <a href="http://zf2008workshops.blogspot.com/">http://zf2008workshops.blogspot.com/</a> and we encourage everyone in the community to visit this blog and comment on proposals.

To propose a topic, email <a href="mailto:zfmeeting@gmail.com">zfmeeting@gmail.com</a>. Please include your name and institution. We will invite you to be a blog author so that you can describe your proposal in a new post. If you want to comment on a topic, or propose a change, simply comment on the relevant post (please include your name and institution in your comment).
</div>

<h2><a name="2007">2007 News Archive</a><br><br>
<a name="vega">New Vega release</a></h2>
27 Sep 2007<br>
<div>
<p>The Sanger Institute has released a Vega update on the 27th of September 2007. The Vega database contains finished clone sequence with manual annotation, placed on chromosomes according to data from the physical mapping project. The current release features 9,535 clones, 2,800 of which have been annotated. 11,115 genes have been manually annotated on over 415 Mb of annotated sequence.

<p>You can browse/search Vega at <a href="http://vega.sanger.ac.uk/Danio_rerio">http://vega.sanger.ac.uk/Danio_rerio</a>. If you have comments on the annotation or arrangement of clones, or would like a particular clone to be annotated, please contact <a href=mailto:zfish-help@sanger.ac.uk>zfish-help@sanger.ac.uk</a>.
</div>

<h2><a name="ZFIQ">Zebrafish Image Quantitator available</a></h2>
23 Aug 2007<br>
<div><p>The <a href="http://www.cbi-platform.net/">ZFIQ</a> (Zebrafish Image Quantitator) software
toolkit provides a set of image analysis tools for quantitative, reproducible and
accurate interpretation of zebrafish imaging data.
</div>

<h2><a name="fishnet">FishNet database now available</a></h2>
21  Aug 2007<br>
<div><p>We are pleased to announce that FishNet, an interactive database of zebrafish anatomy, is
now available at <a href="http://www.FishNet.org.au">www.FishNet.org.au</a>.

FishNet contains more than 36,000 images of zebrafish development, with more than 1.500 of these being annotated. The 3D models can be manipulated on screen or virtually sectioned.

A detailed description of the database is available at
<a href="http://www.biomedcentral.com/1741-7007/5/34/">http://www.biomedcentral.com/1741-7007/5/34/</a>

Comment and suggestions can be submitted via the website or to <a href=mailto:fishnet@victorchang.edu.au>fishnet@victorchang.edu.au</a>
</div>

<h2><a name="zv7-8-07">Zv7 Released in Ensembl</a></h2>
20 Aug 2007<br><div>
     <p>The Sanger Institute has released the seventh zebrafish genome assembly, Zv7, in Ensembl on the 16th of August 2007. The database features repeatmasking, marker placement, alignments with cDNAs, ESTs and peptides, mapped BAC ends, expression patterns, as well as comparative genomics data and data mining tools. The gene build resulted in 21,322 genes, 98 pseudogenes and 4,053 non-coding genes.

<p>Please note that previous Ensembl data built on assembly Zv6 are still available in the Ensembl archives.

<p>The assembly can be downloaded at <a href="ftp://ftp.ensembl.org/pub/assembly/zebrafish/Zv7release">ftp://ftp.ensembl.org/pub/assembly/zebrafish/Zv7release</a> and more information can be found at <a href="http://www.sanger.ac.uk/Projects/D_rerio/Zv7_assembly_information.shtml">http://www.sanger.ac.uk/Projects/D_rerio/Zv7_assembly_information.shtml</a>.</p><br>
</div>

<h2><a name="vega0718">New Vega release</a></h2>
18  Jul 2007<br>
<div> <p>The Sanger Institute has released a Vega update on the 12th of July 2007. The Vega
database
contains finished clone sequence with manual annotation, placed on chromosomes according to data from the physical mapping project. The current release features 9,254 clones, 2,411 of which have been annotated. 10,374 genes have been manually annotated on over 370 Mb of annotated sequence. </p>
<p>
You can browse/search Vega at <a href="http://vega.sanger.ac.uk/Danio_rerio">http://vega.sanger.ac.uk/Danio_rerio</a>. If you have comments on the annotation or arrangements of clones, or want a clone to be annotated, please contact <a href="mailto:zfish-help@sanger.ac.uk">zfish-help@sanger.ac.uk</a>.
</p><br>
</div>


<h2><a name="zv7">Zv7 Assembly and Pre Ensembl</a></h2>
18 Jul 2007<br>
<div>
      <p>The Sanger Institute has released the seventh zebrafish genome assembly, Zv7, on the 13th of July 2007. This assembly comprises 1.44 Gbp in 5,036 scaffolds. It was generated starting with 1.02 Gbp from 7,823 BAC clones and supplemented with a 5.5x whole genome shotgun assembly. The sequences were integrated using alignments as well as BAC end and marker placement. Over 200 Mbp of duplicated sequence has been removed compared to Zv6. The assembly can be downloaded at <a href="ftp://ftp.ensembl.org/pub/assembly/zebrafish/Zv7release">ftp://ftp.ensembl.org/pub/assembly/zebrafish/Zv7release</a> and more information can be found at <a href="http://www.sanger.ac.uk/Projects/D_rerio/Zv7_assembly_information.shtml">http://www.sanger.ac.uk/Projects/D_rerio/Zv7_assembly_information.shtml</a>.

      <p>The Zv7 sequence has been used to generate an Ensembl database available at <a href="http://pre.ensembl.org/Danio_rerio/index.html">http://pre.ensembl.org/Danio_rerio/index.html</a>. This is an extended Pre Ensembl site, providing a full Ensembl gene set for Zv7. A full Ensembl site, providing comparative genomic and data mining tools, will be available for the next Ensembl release in August.

<br><br><br>
</div>

<!--
 over a year old
-->

 <h2><a name="papers">Call for Papers - Disease-based review to the American Journal of Medical Genetics</a></h2>
16 Mar 2007<br>
 <div>
 p>Dear ZFIN community,

 <p>We would like to request a call for papers concerning a disease-based review to the American Journal of Medical Genetics.  The focus of this manuscript is understanding human malformation phenotypes using the zebrafish model.   We believe such a review, written for an audience of primarily non-zebrafish scientists, would be helpful in communicating the value of this model organism for scientists and policy-makers considering alternate vertebrate model systems. We had started the process of drafting this review late last year, and at the Investigator's Workshop, several members of the community expressed interest in such a document for the whole community.

 <p>We would like to make this review as comprehensive as possible and don't want to miss any of the great papers that have come out in this area. If you have a publication that is out or about to be out (i.e. in press), we would love to receive the citation and/or preprint so we can include your science in this review. All materials received prior to publication release will be held in strict confidence until the appropriate paper has come out in the respective journal.

 <p>Sincerely,
Shelly Myers, Genetics, Cell Biology and Genetics - University of Minnesota
Lisa Schimmenti, MD, Pediatrics- University of Minnesota
Stephen C. Ekker, Genetics, Cell Biology and Genetics - University of Minnesota

 <p>Please send your information to:<br>
<br>Stephen C. Ekker  <a href=mailto:ekker001@umn.edu>ekker001@umn.edu</a>
<br>Lisa Schimmenti  <a href=mailto:las@umn.edu>las@umn.edu</a>

 <p>Thanks!
<br><br><br>
</div>

<h2><a name="antibodies">Suggest Targets for Anti-Zebrafish Antibodies</a></h2>
9 Mar 2007<br>
<div>
<p>The need for more anti-zebrafish antibodies was highlighted at the
recent conference of zebrafish investigators at Asilomar. To try and
address this, I have made some preliminary enquiries to an antibody
supplier to see if this niche market would be commercially viable. The
company that has shown some interest has expertise in making
polyclonal affinity-purified anti-peptide antibodies in goats. To make the most
of this opportunity, I need to know which antibodies would initially be
the most useful to the zebrafish community as a whole.</p>

<p>To obtain these data, please email me, <a href="mailto:gw2@sanger.ac.uk">gw2@sanger.ac.uk</a>, by Friday 6th April with a list of preferred targets identified by a GenBank or
ENSembl accession number. Preference would be given according to the
following criteria:</p>
<ol>
<li>
Popularity (number of requests for a particular antibody);
</li>
<li>
Willingness of the requestor to test a small aliquot of the
antibody and provide these data *quickly* to the company so that they can show
it works; please indicate in your reply if you'd be willing to do this.<br>
</li>
</ol>
Thanks,<br>

Gavin<br><br>

Gavin J. Wright<br>
Cell Surface Signalling Laboratory<br>
<br><br><br>
</div>

<h2><a name="amster">Deadlines for the 5th European Zebrafish Genetics and Development Meeting</a></h2>
7 Mar 2007<br>
<div>
 <p>Please note that the abstract submission deadline for the 5th European Zebrafish Genetics and Development Meeting is March 9th, 2007. Also, March 9th is the deadline for early registrations, which are somewhat cheaper than later registrations.

 <p>For more information, see the official Meeting website: <a href="http://www.zebrafish2007.org/">http://www.zebrafish2007.org/</a>

 <br><br><br>
</div>

<h2><a name="genetools">Quality control testing of Gene Tools morpholinos</a></h2>
25 Jan 2007<br>
<div>
 <p><a href="http://zfin.org/zf_info/catch/moqc.pdf">Efficacy Test of Chordin Morpholinos</a>
 <br><br><br>
</div>

<h2><a name="hgnc">Zebrafish genes now included in HGNC Comparison of Orthology Predictions (HCOP)</a></h2>
25 Jan 2007<br>
<div>
 <p>Zebrafish genes have been added to the HGNC HCOP search tool (<a href="http://www.gene.ucl.ac.uk/hcop">http://www.gene.ucl.ac.uk/hcop</a>) which compares orthology assertions from different databases.
 <br><br>
</div>


<h2><a name="2006">2006 News Archive</a><br><br>
New Vega Release, December 21, 2006</h2>

<p>The Sanger Institute has released a Vega update on the 21st of December 2006. The Vega database contains finished clone sequence with manual annotation, placed on chromosomes according to data from the physical mapping project. The current release features 6,569 clones, 1,854 of them being annotated so far. Single loci already reported by ZFIN have been identified and are currently undergoing annotation. This resulted in 6,452 coding, 406 non-coding genes and 100 pseudogenes. 2,875 of the coding genes already had records in ZFIN, the rest are newly identified genes. All genes have been submitted to ZFIN where new records were generated, allowing you to move from ZFIN record to Vega gene and vice versa. A DAS source has been set up to display Ensembl genes that could be mapped to the Vega clones.

 <p>You can browse/search Vega at <a href="http://vega.sanger.ac.uk/Danio_rerio">http://vega.sanger.ac.uk/Danio_rerio</a>. If you have comments on the annotation or arrangements of clones, or want a clone to be annotated, please let us know at <a href=mailto:zfish-help@sanger.ac.uk>zfish-help@sanger.ac.uk</a>. </p>

 <br>




<h2>Reissue of NIH PAR "Tools for Zebrafish Research", December 20, 2006</h2>

<p> Please note the publication in the NIH Guide of PAR-07-145 "Tools for Zebrafish Research" with one receipt date of September 19, 2007 (letters of intent due August 19, 2007).

<p>This reissue of PAR-05-080 was necessitated by the switch to electronic only submission of R01 applications to NIH in February 2007.  The Funding Opportunity Announcement will be available for download and subsequent submission as of January 5, 2007 from the Grants.gov website.

<p>Here is the URL to PAR-07-145 -- <a href="http://grants.nih.gov/grants/guide/pa-files/PAR-07-145.html">http://grants.nih.gov/grants/guide/pa-files/PAR-07-145.html</a></p><br>

<h2>The 5th European Zebrafish Genetics and Development Meeting
will take place from 12-15 July 2007 in Amsterdam, the Netherlands, October 30, 2006</h2>

<p>If you would like to receive further information please visit our Meeting site: <a href="http://www.zebrafish2007.org/">http://www.zebrafish2007.org/</a>

<p><h4>Key dates</h4>

<p>15 December 2006
Web site open for registration and abstract submission

<p>9 March 2007
Deadline for abstract submission + early registration

<p>30 April 2007
Notification of acceptance of abstracts

<p>12-15 July 2007
5th European Zebrafish Meeting in Amsterdam

<p><h4>Location</h4>

<p>The meeting will be held at the Amsterdam RAI Convention and Exhibition Centre, ideally located from both the city centre of the Dutch Capital and Amsterdam International Airport Schiphol.

<p><h4>Keynote speakers</h4>

<p>* Ronald Plasterk
Hubrecht Laboratory, Utrecht, NL

<p>* Alexander Schier
Harvard University, Cambridge, US

<p><h4>Organizing committee</h4>

<p>* Stefan Schulte-Merker
Hubrecht Laboratory, Utrecht, NL

<p>* Jeroen den Hertog
Hubrecht Laboratory, Utrecht, NL

<p>* Dana Zivkovic
Hubrecht Laboratory, Utrecht, NL

<p>* Suresh Jesuthasan
Temasek Life Sciences Laboratoy, Singapore</p><br>

<h2>Nimblegen Expression Arrays, October 26, 2006</h2>
<p>
I have been in contact with some people at Nimblegen (<a href="http://www.nimblegen.com">http://www.nimblegen.com</a>) regarding making a more extensive expression array (using the latest data available). The advantages of Nimblegen chips (which are now available to everyone) include their high density (380,000 oligos per chip). One can buy chips from them for processing in your own lab or core facility, or have them do the hyb and data analysis.

<p>Question is whether there is sufficient interest in the community to justify their initial investment.
The price of the chip would to some extent depend on demand.

<p>Let me know if you might be interested.

<p>Didier Stainier<br>
<a href=mailto:dstainier@biochem.ucsf.edu>dstainier@biochem.ucsf.edu</a><br><br>

<h2>RFI: Soliciting Input and Ideas for Roadmap Trans-NIH Strategic Initiatives, October 25, 2006</h2>

<p>NIH Guide Notice NOT-OD-07-011 was recently published.  It is a Request for Information (RFI) soliciting input and ideas for Roadmap Trans-NIH Strategic Initiatives.  Please take this opportunity to submit ideas of interest to the Zebrafish Community.</p>


<p><a href="http://grants.nih.gov/grants/guide/notice-files/NOT-OD-07-011.html">http://grants.nih.gov/grants/guide/notice-files/NOT-OD-07-011.html</a></p><br>


<h2>New VEGA and ENSEMBL Gene Build Release at the Sanger Institute, October 25, 2006</h2>

<p><h4>VEGA</h4>

<p>The Sanger Institute has released a Vega update on the 16th of
October 2006. The Vega database contains finished clone sequence with
manual annotation, placed on chromosomes according to data from the
physical mapping project. The current release features 6,357 clones,
1,874 of them being annotated so far. This resulted in 5,213 coding,
420 non-coding genes and 98 pseudogenes. 1,450 of the coding genes
already had records in ZFIN, the rest are newly identified genes. All
genes have been submitted to ZFIN where new records were generated,
allowing you to move from ZFIN record to Vega gene and vice versa. A
DAS source has been set up to display Ensembl genes that could be
mapped to the Vega clones.
You can browse/search Vega at http://vega.sanger.ac.uk/Danio_rerio.
If you have comments on the annotation or arrangements of clones, or
want a clone to be annotated, please let us know at <a href=mailto:zfish-help@sanger.ac.uk>zfish-help@sanger.ac.uk</a>.


 <p><h4>ENSEMBL</h4>

 <p>On the 12th of October, the Sanger Institute released a new Ensembl
gene build on the Zv6 assembly to reflect the increase in available
data. This resulted in 24,948 coding, 3,446 non-coding and 112
pseudogenes. You can view/search/download these genes alongside other
data mapped to genome, e.g.

<blockquote>
<br>- markers
<br>- ESTs, cDNAs and peptide sequences from EMBL/Genbank
<br>- zebrafish BAC ends
<br>- expression patterns linked to ZFIN
<br>- morpholinos linked to ZFIN
<br>- ZF-Models data (enhancer detection, TILLING)
<br>- microarrays - SNPs
</blockquote>

 <p>Some of these data are available as  DAS tracks which can be switched
on in the 'DAS sources' menu on ContigView. Also, a DAS source has
been set up to display Vega genes that could be mapped to the Zv6
assembly.

 <p>Comparative genomics data can be viewed in ContigView (switch on
under 'Comparative') and also as MutiContigView, displaying the
alignment between two selected species (go to 'view alongside...' on
left side of page). You can find an example at <a href="http://www.ensembl.org/
Danio_rerio/multicontigview?s1=Homo_sapiens;c=21:13857328.5;w=153630">http://www.ensembl.org/
Danio_rerio/multicontigview?s1=Homo_sapiens;c=21:13857328.5;w=153630</a>

 <p>You can browse/search Ensembl at <a href="http://www.ensembl.org/Danio_rerio">http://www.ensembl.org/Danio_rerio</a>

 <p>Advanced data searches can be performed using BioMart at <a href="http://
www.ensembl.org/Danio_rerio/martview">http://
www.ensembl.org/Danio_rerio/martview</a>. </p><br>




<h2>Trans-NIH Zebrafish Initiative Update, August 15, 2006 </h2>
<p>The Trans-NIH Zebrafish Coordinating Committee would like the community to know that the zebrafish site on the NIH Model Organisms for Biomedical Research web page has been updated. <a href="http://www.nih.gov/science/models/zebrafish/">http://www.nih.gov/science/models/zebrafish/</a>.</p>

<p>We would like to remind everyone of the upcoming August 19 deadline for letters of intent for Tools for Zebrafish Research grants as described on the NIH site:

<a href="http://grants.nih.gov/grants/guide/pa-files/PAR-05-080.html">http://grants.nih.gov/grants/guide/pa-files/PAR-05-080.html</a>
</p><br>

<h2>Zebrafish Genome Resources Tutorial, June 12, 2006 </h2>
<p>
ZFIN, the Sanger Institute, and NCBI provide a broad range of zebrafish web resources. These resources include genome browsers, map viewers, gene expression, mutant, morpholino, sequence and orthology data, BLAST, data mining tools and much more. These <a href="/zf_info/2006_tutorial/tutorial.html">tutorials</a> offer an opportunity to receive assistance with your questions.</p><br>

<h2>"The Zebrafish: Cellular and Developmental Biology" now available online,  June 9, 2006</h2>
<p>The second zebrafish methods book, "The Zebrafish: Cellular and Developmental Biology", is now available <a href="http://www.sciencedirect.com/science/bookseries/009167">online</a>.</p><br>

<h2>Zebrafish Zv6 assembly available in UCSC Genome Browser, June 9, 2006 </h2>
<p>The latest zebrafish assembly -- Zv6 (UCSC version danRer4, March 2006) -- is now available in the UCSC Genome Browser. The Zv6 assembly was produced by The Wellcome Trust Sanger Institute in collaboration with the Max Planck Institute for Developmental Biology in Tuebingen, Germany, and the Netherlands Institute for Developmental Biology (Hubrecht Laboratory), Utrecht, The Netherlands.</p>

</p>This assembly consists of 1,626,077,335 bp in 6.653 scaffolds (N50 = 1,247,221 bp) with a sequence coverage of approximately 6.5-7x. For more information about this assembly, see the Sanger Institute web page for the <a href="http://www.sanger.ac.uk/Projects/D_rerio/Zv6_assembly_information.shtml/">Danio rerio Sequencing Project</a>. The danRer4 sequence and annotation data can be downloaded from the UCSC Genome Browser <a Href="ftp://hgdownload.cse.ucsc.edu/goldenPath/danRer4/"> FTP server</a> or <a href="http://hgdownload.cse.ucsc.edu/downloads.html#zebrafish">downloads page</a>. Please review the <a Href="http://www.sanger.ac.uk/Projects/use-policy.shtml"> Sanger guidelines</a> for using these data.</p>

<p>We'd like to thank the Wellcome Trust Sanger Institute, the Max Planck Institute for Developmental Biology, Hubrecht Laboratory and the other institutions who contributed to the sequencing, mapping, and annotation of this release. Special thanks to the Zebrafish Genome Initiative at Children's Hospital in Boston for their collaboration. The UCSC zebrafish Genome Browser was produced by Rachel Harte, Archana Thakkapallayil, Robert Kuhn, Ann Zweig, and Donna Karolchik. See the <a href="http://genome.ucsc.edu/goldenPath/credits.html#zebrafish_credits"> UCSC credits page</a> for a detailed list of the organizations and individuals who contributed to the release of this browser.</p><br>




<h2>The Sanger Institute has released Zv6, April 5, 2006</h2>

<p>The Sanger Institute has released the sixth zebrafish genome assembly,
Zv6, on the 31st of March 2006. This assembly comprises 1.6 Gb in 6,653
scaffolds. It was generated starting with 1.02 Gb from 7,615 BAC clones
and supplemented with a 7x whole genome shotgun. The sequences were
integrated using alignments as well as BAC end and marker placement,
with particular focus on the HS and MGH panel. The assembly can be
downloaded at <a href="ftp://ftp.ensembl.org/pub/assembly/zebrafish/Zv6release">ftp://ftp.ensembl.org/pub/assembly/zebrafish/Zv6release</a>
and more information can be found at
<a href="http://www.sanger.ac.uk/Projects/D_rerio/Zv6_assembly_information.shtml">http://www.sanger.ac.uk/Projects/D_rerio/ Zv6_assembly_information.shtml.</a>

<p>The Zv6 sequence has been used to generate an Ensembl database and a
pre-release is available at
http://pre.ensembl.org/Danio_rerio/index.html. This pre-release
features analyses including repeat masking, homology searches and ab
initio gene predictions. The full Ensembl gene set is currently in
preparation and will be released in June. </p><br>

<h2>ZF-MODELS knock-out service, March 31, 2006</h2>
<p>The ZF-MODELS consortium would like to draw attention to its public
and free knock-out service for zebrafish, based on the resequencing  of
mutagenized fish (TILLING). Up to 60 knock-outs will be performed  for
researchers outside the consortium, whose requests are considered  on
the basis of their scientific merit. The workload is distributed
between the laboratories of Sanger Institute (D. Stemple), Hubrecht
Laboratory (E. Cuppen), and TU Dresden (M. Brand). To request a
zebrafish KO mutant, please send a brief (approx. one page) summary  to <A href="mailto:ecuppen@niob.knaw.nl">Dr. Edwin Cuppen</a>. This summary  should
contain the following information:</p>

<p>Information on the gene to be knocked out: Name of the gene and
Ensembl ID. Please note that the ZF-MODELS consortium will only  accept
requests for KOs based on Ensembl IDs. The requester should  also check
the Ensembl annotation. In case the annotation is  incorrect, the
requester has to make sure that it is changed. The ZF- MODELS
consortium will not screen for a gene as long as the  annotation is not
correct.
Scientific relevance: Explain why is it important to knock this gene
out.
Zebrafish background of the lab requesting the KO mutant: Explain how
any work on the KO mutant provided would be embedded in an  environment
in your lab where people can actually work with zebrafish  mutants.</p>

<p>The summaries received by Dr. Edwin Cuppen form the basis for
prioritized lists drafted up by the ZF-MODELS Executive Committee on  a
regular basis.</p>

<p>Once a request has been accepted, you will be asked to design an
amplicon(s) for your region(s) of interest through the LIMSTILL  online
tool <a href="http://limstill.niob.knaw.nl/">http://limstill.niob.knaw.nl/</a>. Oligos will be ordered  and
tested by the ZF-MODELS consortium. The progress of requests can  be
tracked through the internal knock-out website of Hubrecht  Laboratory
(login required).</p>

<p>Information on the knock-out mutants generated by the ZF-MODELS
consortium is made publicly available through the knock-out website
<a href="http://www.niob.knaw.nl/researchpages/cuppen/zfmodels/">http://www.niob.knaw.nl/researchpages/cuppen/zfmodels/</a>of Hubrecht
Laboratory, immediately after a knock-out is generated. Knock-out  fish
are immediately delivered to the requesters.</p>

<p>"ZF-MODELS - Zebrafish Models for Human Development and Disease" is  an
Integrated Project funded by the Sixth Framework Programme of the
European Commission. For further information please see the project
website <a href="http://www.zf-models.org">http://www.zf-models.org</a>.</p><br>


<h2>Site Search Available, March 24, 2006 </h2>

<p>ZFIN Site Search provides a quick and easy glance at data available on the ZFIN web site. Simply enter your topic of interest and press the enter key.

<p>Results from Site Search and advanced ZFIN search forms may vary. Advanced ZFIN search forms query a particular class of data while Site Search searches snapshots of all ZFIN pages for an occurrence of your query string.


<p>Site Search does not take advantage of the data relationships and ontologies that are defined in the ZFIN database. For a comprehensive search of expression or anatomical structures we recommend you use the Expression or Anatomy search forms.

<p>If you do not find what you are looking for using this search, please be sure to try one of ZFIN's advanced search forms.

</p>
<p><br>
</p>

<h2><a name="2005">2005 News Archive</a><br><br>

<h2>Online Zebrafish Genome Resources Tutorial from Dresden workshop available August 1, 2005</h2>
<p>NCBI, ZFIN and the Sanger Institute presented a joint workshop on genome resources at the 4th European Zebrafish Meeting, held in Dresden, Germany July 14 and 15, 2005. A tutorial is available at:  <a href="http://www.sanger.ac.uk/Projects/D_rerio/workshop_Dresden/">http://www.sanger.ac.uk/Projects/D_rerio/workshop_Dresden/</a>.</p><br>


<h2>Figures and Morpholinos, June 30, 2005</h2>
<p>We've redesigned ZFIN's gene expression pages to display figures from  journal articles.  We've added support for morpholinos, too, so you  can find expression of your favorite genes in morphant and mutant  backgrounds.
<br>
It's still 'early days' - it takes time and effort to annotate  figures and load them into the database, and some journals still  impose copyright restrictions, but we're keeping up with new  publications and plan to gradually add figures from older papers.  <a href="/cgi-bin/webdriver?MIval=aa-xpatselect.apg">Try out the new gene expression search page.</a>


<script type="text/javascript" language="JavaScript" src="/javascript/footer.js"></script>
