<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html>
	<head>
		<title>
			ZFIN Site News & Status
		</title>
	</head>

<script language="JavaScript" src="/header.js"></script>

<center>
<h1>Zebrafish News</h1>
</center>

<p>This page contains news relevant to the zebrafish research community.

If you have any questions, comments or would like to submit an entry, please 
contact <a href="mailto:zfinadmn@zfin.org">Jonathan Knight</a>.
<br><br>


<h3>"ZFIN NEWS" is the ZFIN Newsletter, published bi-annually at the University of Oregon.  </h3>

<a href="Newsletter_Summer2006.pdf"> Vol. 3, No. 2 - Summer 2006</a>
		 <blockquote>
		 ZFIN BLAST Search Is Here! <br>
		 Site Search: A Global View Of ZFIN Data </blockquote>
      
         <a href="Newsletter_Winter06.pdf"> Vol. 3, No. 1 - Winter 2006</a>
	  <blockquote>
		 Find Great Probes Fast! <br>
		 From Ensembl To ZFIN Expression Patterns </blockquote>
		 
		 <p>Click here for our <a href="http://zfin.org/zf_info/news/Newsletters.html">ZFIN Newsletter Archive.</a></p>
		 
		 
		 
		 <hr><br>
		 
		 <h2>RFI: Soliciting Input and Ideas for Roadmap Trans-NIH Strategic Initiatives, 2006/10/26</h2>
		 
		 <p>NIH Guide Notice NOT-OD-07-011 was recently published.  It is a Request for Information (RFI) soliciting input and ideas for Roadmap Trans-NIH Strategic Initiatives.  Please take this opportunity to submit ideas of interest to the Zebrafish Community.</p>  


<p><a href="http://grants.nih.gov/grants/guide/notice-files/NOT-OD-07-011.html">http://grants.nih.gov/grants/guide/notice-files/NOT-OD-07-011.html</a></p><br>
		 
		 
		 <h2>New VEGA and ENSEMBL Gene Build Release at the Sanger Institute, 2006/10/25</h2>
		 
		 <p><h4>VEGA</h4>

 <p>The Sanger Institute has released a Vega update on the 16th of
October 2006. The Vega database contains finished clone sequence with
manual annotation, placed on chromosomes according to data from the
physical mapping project. The current release features 6,357 clones,
1,874 of them being annotated so far. This resulted in 5,213 coding,
420 non-coding genes and 98 pseudogenes. 1,450 of the coding genes
already had records in ZFIN, the rest are newly identified genes. All
genes have been submitted to ZFIN where new records were generated,
allowing you to move from ZFIN record to Vega gene and vice versa. A
DAS source has been set up to display Ensembl genes that could be
mapped to the Vega clones.
You can browse/search Vega at http://vega.sanger.ac.uk/Danio_rerio.
If you have comments on the annotation or arrangements of clones, or
want a clone to be annotated, please let us know at <a href=mailto:zfish-help@sanger.ac.uk>zfish-help@sanger.ac.uk</a>.


 <p><h4>ENSEMBL</h4>

 <p>On the 12th of October, the Sanger Institute released a new Ensembl
gene build on the Zv6 assembly to reflect the increase in available
data. This resulted in 24,948 coding, 3,446 non-coding and 112
pseudogenes. You can view/search/download these genes alongside other
data mapped to genome, e.g.

<blockquote>
<br>- markers
<br>- ESTs, cDNAs and peptide sequences from EMBL/Genbank
<br>- zebrafish BAC ends
<br>- expression patterns linked to ZFIN
<br>- morpholinos linked to ZFIN
<br>- ZF-Models data (enhancer detection, TILLING)
<br>- microarrays - SNPs
</blockquote>

 <p>Some of these data are available as  DAS tracks which can be switched
on in the 'DAS sources' menu on ContigView. Also, a DAS source has
been set up to display Vega genes that could be mapped to the Zv6
assembly.

 <p>Comparative genomics data can be viewed in ContigView (switch on
under 'Comparative') and also as MutiContigView, displaying the
alignment between two selected species (go to 'view alongside...' on
left side of page). You can find an example at <a href="http://www.ensembl.org/
Danio_rerio/multicontigview?s1=Homo_sapiens;c=21:13857328.5;w=153630">http://www.ensembl.org/
Danio_rerio/multicontigview?s1=Homo_sapiens;c=21:13857328.5;w=153630</a>

 <p>You can browse/search Ensembl at <a href="http://www.ensembl.org/Danio_rerio">http://www.ensembl.org/Danio_rerio</a>

 <p>Advanced data searches can be performed using BioMart at <a href="http://
www.ensembl.org/Danio_rerio/martview">http://
www.ensembl.org/Danio_rerio/martview</a>. </p><br>


	

<h2>Trans-NIH Zebrafish Initiative Update, 2006/08/15 </h2>
<p>The Trans-NIH Zebrafish Coordinating Committee would like the community to know that the zebrafish site on the NIH Model Organisms for Biomedical Research web page has been updated. <a href="http://www.nih.gov/science/models/zebrafish/">http://www.nih.gov/science/models/zebrafish/</a>.</p>

<p>We would like to remind everyone of the upcoming August 19 deadline for letters of intent for Tools for Zebrafish Research grants as described on the NIH site:

<a href="http://grants.nih.gov/grants/guide/pa-files/PAR-05-080.html">http://grants.nih.gov/grants/guide/pa-files/PAR-05-080.html</a>
</p><br>
		 
<h2>Zebrafish Genome Resources Tutorial, 2006/06/12 </h2>
<p>
ZFIN, the Sanger Institute, and NCBI provide a broad range of zebrafish web resources. These resources include genome browsers, map viewers, gene expression, mutant, morpholino, sequence and orthology data, BLAST, data mining tools and much more. These <a href="http://zfin.org/zf_info/2006_tutorial/tutorial.html">tutorials</a> offer an opportunity to receive assistance with your questions.</p><br>

<h2>"The Zebrafish: Cellular and Developmental Biology" now available online,  2006/06/09</h2>
<p>The second zebrafish methods book, "The Zebrafish: Cellular and Developmental Biology", is now available <a href="http://www.sciencedirect.com/science/bookseries/009167">online</a>.</p><br>

<h2>Zebrafish Zv6 assembly available in UCSC Genome Browser, 2006/06/09 </h2>
<p>The latest zebrafish assembly -- Zv6 (UCSC version danRer4, March 2006) -- is now available in the UCSC Genome Browser. The Zv6 assembly was produced by The Wellcome Trust Sanger Institute in collaboration with the Max Planck Institute for Developmental Biology in Tuebingen, Germany, and the Netherlands Institute for Developmental Biology (Hubrecht Laboratory), Utrecht, The Netherlands.</p>

</p>This assembly consists of 1,626,077,335 bp in 6.653 scaffolds (N50 = 1,247,221 bp) with a sequence coverage of approximately 6.5-7x. For more information about this assembly, see the Sanger Institute web page for the <a href="http://www.sanger.ac.uk/Projects/D_rerio/Zv6_assembly_information.shtml/">Danio rerio Sequencing Project</a>. The danRer4 sequence and annotation data can be downloaded from the UCSC Genome Browser <a Href="ftp://hgdownload.cse.ucsc.edu/goldenPath/danRer4/"> FTP server</a> or <a href="http://hgdownload.cse.ucsc.edu/downloads.html#zebrafish">downloads page</a>. Please review the <a Href="http://www.sanger.ac.uk/Projects/use-policy.shtml"> Sanger guidelines</a> for using these data.</p>

<p>We'd like to thank the Wellcome Trust Sanger Institute, the Max Planck Institute for Developmental Biology, Hubrecht Laboratory and the other institutions who contributed to the sequencing, mapping, and annotation of this release. Special thanks to the Zebrafish Genome Initiative at Children's Hospital in Boston for their collaboration. The UCSC zebrafish Genome Browser was produced by Rachel Harte, Archana Thakkapallayil, Robert Kuhn, Ann Zweig, and Donna Karolchik. See the <a href="http://genome.ucsc.edu/goldenPath/credits.html#zebrafish_credits"> UCSC credits page</a> for a detailed list of the organizations and individuals who contributed to the release of this browser.</p><br>


      



<h2>The Sanger Institute has released Zv6, 2006/04/05</h2>

<p>The Sanger Institute has released the sixth zebrafish genome assembly,
Zv6, on the 31st of March 2006. This assembly comprises 1.6 Gb in 6,653
scaffolds. It was generated starting with 1.02 Gb from 7,615 BAC clones
and supplemented with a 7x whole genome shotgun. The sequences were
integrated using alignments as well as BAC end and marker placement,
with particular focus on the HS and MGH panel. The assembly can be
downloaded at <a href="ftp://ftp.ensembl.org/pub/assembly/zebrafish/Zv6release">ftp://ftp.ensembl.org/pub/assembly/zebrafish/Zv6release</a>
and more information can be found at
<a href="http://www.sanger.ac.uk/Projects/D_rerio/Zv6_assembly_information.shtml">http://www.sanger.ac.uk/Projects/D_rerio/ Zv6_assembly_information.shtml.</a>

<p>The Zv6 sequence has been used to generate an Ensembl database and a
pre-release is available at
http://pre.ensembl.org/Danio_rerio/index.html. This pre-release
features analyses including repeat masking, homology searches and ab
initio gene predictions. The full Ensembl gene set is currently in
preparation and will be released in June. </p><br> 

<h2>ZF-MODELS knock-out service, 2006/03/31</h2>
<p>The ZF-MODELS consortium would like to draw attention to its public
and free knock-out service for zebrafish, based on the resequencing  of
mutagenized fish (TILLING). Up to 60 knock-outs will be performed  for
researchers outside the consortium, whose requests are considered  on
the basis of their scientific merit. The workload is distributed
between the laboratories of Sanger Institute (D. Stemple), Hubrecht
Laboratory (E. Cuppen), and TU Dresden (M. Brand). To request a
zebrafish KO mutant, please send a brief (approx. one page) summary  to <A href="mailto:ecuppen@niob.knaw.nl">Dr. Edwin Cuppen</a>. This summary  should
contain the following information:</p>

<p>Information on the gene to be knocked out: Name of the gene and
Ensembl ID. Please note that the ZF-MODELS consortium will only  accept
requests for KOs based on Ensembl IDs. The requester should  also check
the Ensembl annotation. In case the annotation is  incorrect, the
requester has to make sure that it is changed. The ZF- MODELS
consortium will not screen for a gene as long as the  annotation is not
correct.
Scientific relevance: Explain why is it important to knock this gene
out.
Zebrafish background of the lab requesting the KO mutant: Explain how
any work on the KO mutant provided would be embedded in an  environment
in your lab where people can actually work with zebrafish  mutants.</p>

<p>The summaries received by Dr. Edwin Cuppen form the basis for
prioritized lists drafted up by the ZF-MODELS Executive Committee on  a
regular basis.</p>

<p>Once a request has been accepted, you will be asked to design an
amplicon(s) for your region(s) of interest through the LIMSTILL  online
tool <a href="http://limstill.niob.knaw.nl/">http://limstill.niob.knaw.nl/</a>. Oligos will be ordered  and
tested by the ZF-MODELS consortium. The progress of requests can  be
tracked through the internal knock-out website of Hubrecht  Laboratory
(login required).</p>

<p>Information on the knock-out mutants generated by the ZF-MODELS
consortium is made publicly available through the knock-out website
<a href="http://www.niob.knaw.nl/researchpages/cuppen/zfmodels/">http://www.niob.knaw.nl/researchpages/cuppen/zfmodels/</a>of Hubrecht
Laboratory, immediately after a knock-out is generated. Knock-out  fish
are immediately delivered to the requesters.</p>

<p>"ZF-MODELS - Zebrafish Models for Human Development and Disease" is  an
Integrated Project funded by the Sixth Framework Programme of the
European Commission. For further information please see the project
website <a href="http://www.zf-models.org">http://www.zf-models.org</a>.</p><br>


<h2>Site Search Available, 2006/03/24 </h2>

<p>ZFIN Site Search provides a quick and easy glance at data available on the ZFIN web site. Simply enter your topic of interest and press the enter key.

<p>Results from Site Search and advanced ZFIN search forms may vary. Advanced ZFIN search forms query a particular class of data while Site Search searches snapshots of all ZFIN pages for an occurrence of your query string.


<p>Site Search does not take advantage of the data relationships and ontologies that are defined in the ZFIN database. For a comprehensive search of expression or anatomical structures we recommend you use the Expression or Anatomy search forms.

<p>If you do not find what you are looking for using this search, please be sure to try one of ZFIN's advanced search forms.

<a href="http://zfin.org/SearchApp/syntax_help.jsp">Additional Site Search Tips</a></p>
<p><br>  
</p>
<h2>Online Zebrafish Genome Resources Tutorial from Dresden workshop available 2005/08/01</h2>
<p>NCBI, ZFIN and the Sanger Institute presented a joint workshop on genome resources at the 4th European Zebrafish Meeting, held in Dresden, Germany July 14 and 15, 2005. A tutorial is available at:  <a href="http://www.sanger.ac.uk/Projects/D_rerio/workshop_Dresden/">http://www.sanger.ac.uk/Projects/D_rerio/workshop_Dresden/</a>.</p><br>  


<h2>Figures and Morpholinos, 2005/06/30</h2>
<p>We've redesigned ZFIN's gene expression pages to display figures from  journal articles.  We've added support for morpholinos, too, so you  can find expression of your favorite genes in morphant and mutant  backgrounds.
<br>
It's still 'early days' - it takes time and effort to annotate  figures and load them into the database, and some journals still  impose copyright restrictions, but we're keeping up with new  publications and plan to gradually add figures from older papers.  <a href="/cgi-bin/webdriver?MIval=aa-xpatselect.apg">Try out the new gene expression search page.</a> 


</body>


<script language="JavaScript" src="/footer.js"></script>
