<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html>
	<head>
		<title>
			ZFIN Zebrafish Science Monitor, Vol. 6 (1)
		</title>
	</head>

<script language="JavaScript" src="/javascript/header.js"></script>

<div align="center"><center>

<table border="3" cellpadding="0" cellspacing="0" width="80%" bordercolorlight="#C0C0C0" bordercolordark="#000080">
  <tr>
    <td width="100%" bgcolor="#80FFFF"><p align="center"><big><big><font face="Times New Roman"><strong><big><em>The Zebrafish Science Monitor</em>, </big>Vol. 6
    (1)</strong></font></big></big></td>
  </tr>
</table>
</center></div>

<h3 ALIGN="CENTER"><em>&quot;<font face="Comic Sans MS">All the Zebrafish News That's Fit
to Print</em>&quot;</font></h3>

<hr>

<h4><font face="Comic Sans MS"><big><a NAME="December 1, 1999">December 8, 1999</a></big></font></h4>

<hr>

<ul>
  <li><font face="Comic Sans MS"><big><a HREF="vol6.1.html#BMP ACTIVITY ESTABLISHES A GRADIENT OF POSITIONAL">BMP Activity
    Establishes a Gradient of Positional Information Throughout the Entire Neural Plate</a> </big></font></li>
  <li><font face="Comic Sans MS"><big><a HREF="vol6.1.html#ZEBRAFISH PATHOLOGY AND HEALTH SERVICES">Zebrafish Pathology and Health
    Services</a></big></font></li>
  <li><font face="Comic Sans MS"><big><a HREF="vol6.1.html#THE WASHINGTON UNIVERSITY ZEBRAFISH GENOME RESOURCES">The Washington
    University Zebrafish Genome Resources Project</a></big></font></li>
  <li><font face="Comic Sans MS"><big><a HREF="vol6.1.html#KEEPING AND BREEDING ZEBRAFISH (DANIO RERIO) ON SMALL SCALE">Keeping
    and Breeding Zebrafish (<em>Danio rerio</em>) on Small Scale</a></big></font></li>
  <li><font face="Comic Sans MS"><a HREF="vol6.1.html#ANTERIOR MOVEMENT OF VENTRAL DIENCEPHALIC PRECURSORS"><big>Anterior
    Movement of Ventral Diencephalic Precursors Separates the Primordial Eye Field in the
    Neural Plate and Requires <em>Cyclops</em></big></a></font></li>
  <li><font size="4" color="#0000FF" face="Comic Sans MS"><a href="vol6.1.html#VENTRAL DIENCEPHALON PRECURSORS DRIVE MORPHOGENESIS OF THE ZEBRAFISH FOREBRAIN AND RETINA">Ventral
    Diencephalon Precursors Drive Morphogenesis of the Zebrafish Forebrain and Retina</a></font></li>
  <li><font face="Comic Sans MS"><big><a HREF="vol6.1.html#DECHORIONATION MAY SERIOUSLY DISTURB ZEBRAFISH">Dechorionation May
    Seriously Disturb Zebrafish Development</a></big></font></li>
  <li><font face="Comic Sans MS"><big><a HREF="vol6.1.html#THE TIGR ZEBRAFISH GENE INDEX (ZGI): ESTIMATING THE">The TIGR Zebrafish
    Gene Index (ZGI): Estimating the Transcript Content of the <em>Danio rerio</em> Genome</a></big></font></li>
  <li><font face="Comic Sans MS"><big><a HREF="vol6.1.html#ZEBRAFISH NOMENCLATURE COMMITTEE">Zebrafish
    Nomenclature Committee</a></big></font></li>
  <li><font face="Comic Sans MS"><big><a HREF="vol6.1.html#A SIMPLE METHOD FOR OBTAINING VERY YOUNG EMBRYOS DURING AN EXTENDED TIME PERIOD">A
    Simple Method for Obtaining Very Young Embryos During an Extended Time Period</a></big></font></li>
  <li><font color="#0000FF" face="Comic Sans MS"><big><a href="vol6.1.html#Zebrafish Development &amp; Genetics">4th Biennial Meeting on Zebrafish
    Development &amp; Genetics</a></big></font></li>
  <li><font size="4" face="Comic Sans MS"><a href="vol6.1.html#Check out these zebrafish websites!">Check Out These Zebrafish
    Websites!</a></font></li>
</ul>

<hr>

<p><font face="Comic Sans MS"><a href="vol6.1.html#December 1, 1999">Return to Monitor
Contents</a></font></p>



<hr>

<p><font face="Times New Roman" size="6"><a NAME="BMP ACTIVITY ESTABLISHES A GRADIENT OF POSITIONAL"><strong>BMP ACTIVITY ESTABLISHES
A GRADIENT OF POSITIONAL INFORMATION THROUGHOUT THE ENTIRE NEURAL PLATE</strong></a></font></p>

<p><font size="2"><font face="Comic Sans MS">By <a href="mailto:a.barth@ucl.ac.uk">K.A.
Barth</a><sup>1</sup>, Y. Kishimoto<sup>2</sup>, <a href="mailto:k.rohr@ucl.ac.uk">K.B.
Rohr</a><sup>1</sup>, C. Seydler<sup>2</sup>, <a href="mailto:s.schulte@artemis-pharmaceuticals.de">S. Schulte-Merker</a><sup>2</sup>, and <a href="mailto:ucgawil@ucl.ac.uk">S.W. Wilson</a><sup>1</sup>, <sup>1</sup>Department of
Anatomy &amp; Developmental Biology, University College London, Gower Street, London WC1E
6BT United Kingdom; <sup>2</sup>Max-Planck Institut f&uuml;r Entwicklungsbiologie, Abteilung
Genetik, Spemannstrasse 35, 72076 T&uuml;bingen, Germany</font></font></p>

<table border="0" cellpadding="0" cellspacing="0" width="100%">
  <tr>
    <td width="78%"><font face="Comic Sans MS">Bone morphogenetic proteins (Bmps) are key
    regulators of dorsoventral (DV) patterning. Within the ectoderm, Bmp activity has been
    shown to inhibit neural development, promote epidermal differentiation and influence the
    specification of dorsal neurons and neural crest. In this study, we examine the patterning
    of neural tissue in mutant zebrafish embryos with compromised Bmp signaling activity. We
    find that although Bmp activity does not influence anteroposterior (AP) patterning, it
    does affect DV patterning in all AP levels of the neural plate. Thus, we show that Bmp
    activity is required for specification of cell fates around the margin of the entire
    neural plate, including forebrain regions that do not form neural crest.</font><p><font face="Comic Sans MS">Surprisingly, we find that Bmp activity is also required for
    patterning neurons at all DV levels of the CNS. In <em>swirl/bmp2b<sup>-</sup></em> (<em>swr<sup>-</sup>)
    </em>embryos, laterally positioned sensory neurons are absent whereas more medial
    interneuron populations are hugely expanded. However, in <em>somitabun</em> (<em>sbn<sup>-</sup></em>)
    embryos, which probably retain higher residual Bmp activity, it is the sensory neurons and
    not the interneurons that are expanded. Conversely, in severely Bmp depleted embryos, both
    interneurons and sensory neurons are absent and it is the most medial neurons that are
    expanded. These results are consistent with there being a gradient of Bmp-dependent
    positional information extending throughout the entire neural and non-neural ectoderm.</font></td>
    <td width="22%"><font face="Comic Sans MS"><img src="ak_fig_for_Monte.jpg" width="196" height="379" alt="ak_fig_for_Monte (979184 bytes)"></font></td>
  </tr>
</table>

<hr>
<a HREF="vol6.1.html#December 1, 1999">

<p><font face="Comic Sans MS">Return to Table of Contents</font></p>
</a>

<hr>

<p><font face="Times New Roman" size="6"><a NAME="ZEBRAFISH PATHOLOGY AND HEALTH SERVICES"><strong>ZEBRAFISH
PATHOLOGY AND HEALTH SERVICES</strong></a></font></p>

<p><font face="Comic Sans MS"><font size="2">By <a href="mailto:monte@uoneuro.uoregon.edu">M.
Westerfield</a>, Zebrafish International Resource Center, 5274 University of Oregon,
Eugene, OR 97405-5274 USA</font><big><big> </big></big></font></p>

<p><font face="Comic Sans MS"><strong><em>Got some sick fish?</em></strong> The Zebrafish
International Resource Center in Eugene may be able to help. </font></p>

<p><font face="Comic Sans MS">The Resource Center now provides diagnostic services and
health status testing for laboratory zebrafish. The histopathology services can be
utilized to analyze specific or suspected disease problems. Additional diagnostic services
will be available in the near future and include necropsy, bacteriology and virology.
Based on clinical history and test results specific recommendations for control or
treatment of the problem will be provided. To analyze a disease problem within a colony, a
sampling of 5-7 moribund fish is preferred. We are also interested in examining fish with
suspected neoplasias or other problems that may involve only an individual fish. </font></p>

<p><font face="Comic Sans MS"><strong><em>Is your local animal use committee breathing
down your neck?</em></strong> The Resource Center can also provide routine sentinel or
quality control testing of zebrafish from healthy laboratory colonies. This information
may be useful to fulfill institutional animal care and use committee (IACUC) health status
monitoring requirements.</font></p>

<p><font face="Comic Sans MS">More information on the Zebrafish International Resource
Center and Pathology Service is available at <a HREF="/">ZFIN</a> (<a HREF="http://zebrafish.org/zirc/home/guide.php" TARGET="new">http://zebrafish.org/zirc/home/guide.php</a>).</font></p>

<p><font face="Comic Sans MS">Contact Jennifer Matthews, DVM, Ph.D, (Email: <font COLOR="#0000ff">jmatthews@uoneuro.uoregon.edu</font>). A submission form and detailed
information on sample preparation, submission and shipping will be provided.</font></p>

<p><font face="Comic Sans MS">Nominal fees are charged for these services.</font></p>

<hr>
<a HREF="vol6.1.html#December 1, 1999">

<p><font face="Comic Sans MS">Return to Table of Contents</font></p>
</a>

<hr>

<p><font face="Times New Roman" size="6"><strong><a NAME="THE WASHINGTON UNIVERSITY ZEBRAFISH GENOME RESOURCES">THE WASHINGTON UNIVERSITY
ZEBRAFISH GENOME RESOURCES PROJECT</a></strong></font> </p>

<p><font size="2" face="Comic Sans MS">By <a href="mailto:sjohnson@genetics.wustl.edu">S.L.
Johnson</a>, Department of Genetics, Box 8232, Washington University School of Medicine,
4566 Scott Avenue, St. Louis, MO 63110 USA; <a HREF="http://dbbs.wustl.edu/RIB/Johnson_Stephen.html">http://dbbs.wustl.edu/RIB/Johnson_Stephen.html</a></font></p>

<p><font face="Comic Sans MS">The goal of the Zebrafish Genome Initiative is to provide
tools that will better facilitate the identification of genes associated with mutant
phenotypes. Projects funded by the initiative include development of chromosomal
deletions, improvement of meiotic maps, gene discovery by ESTs, development of high
density transcript maps and the improvement of inbred strains. Below I have outlined the
contribution of Washington University Zebrafish Genomic Resources' (WUZGR) effort in EST
and transcript mapping projects, including a brief progress report, how to use the
resources, and how to acknowledge WUZGR data. </font></p>

<table border="0" cellpadding="6" cellspacing="1" width="100%">
  <tr>
    <td width="41%" bgcolor="#A8FFFF" style="border: 2px solid rgb(0,0,0)"><font face="Comic Sans MS"><img src="gscheader1.jpg" width="383" height="70" alt="gscheader1.jpg (11783 bytes)"></font><b><p><font face="Comic Sans MS">WashU Zebrafish
    EST Project</font></b><font size="2"><ul>
      <li></font><font size="2"><font face="Comic Sans MS">Clone availablity and Library
        Information - The I.M.A.G.E. Consortium Resources page</font></li>
      <li><font face="Comic Sans MS">NCBI Blast Server - Blast your sequence against the NCBI EST
        database</font></li>
      <li><font face="Comic Sans MS">Computer Programs - Local Trace viewing, EST clustering etc. </font></li>
      <li><font face="Comic Sans MS">Weekly zebrafish library reports - View or download weekly
        zebrafish library summary reports</font></li>
      <li><font face="Comic Sans MS">Zebrafish EST production spreadsheets - Weekly copies of our
        production status spreadsheets in txt format</font></li>
      <li><font face="Comic Sans MS">Java trace viewer - See and scroll along a trace on your
        screen</font></li>
      <li><font face="Comic Sans MS">FTP service - Need to download many traces ?</font></li>
      <li><font face="Comic Sans MS">Zebrafish ESTace - EST ACeDB</font></font></li>
    </ul>
    </td>
    <td width="59%"><font face="Comic Sans MS"><big><strong>The Washington University
    zebrafish EST project</strong>. </big></font><p><font face="Comic Sans MS">ESTs (Expressed
    Sequence Tags) are high-throughput, single-pass sequence reads from cDNA clones. These
    reads can be from 5' or 3' ends of the cDNA, or both. As of Nov. 7, 1999, Washington
    University has submitted more than 38,000 zebrafish ESTs to GenBank, from ~ 25,000
    independent cDNA clones. Our analysis suggests that we have identified between 10,000 and
    15,000 genes to date. By the end of the project, we hope to have generated 100,000 ESTs,
    identifying 25,000 to 30,000 genes. </font></p>
    <p><font face="Comic Sans MS">The majority of clones in the Washington University
    Zebrafish EST project are provided by Matt Clark at the Max-Planck-Institute for Molecular
    Genetics (MPIMG) in Berlin. Matt uses oligonucleotide hybridization fingerprinting of
    arrayed cDNAs to detect redundancy in arrayed cDNA libraries, then rearrays minimally
    redundant clones for the Washington University-MPIMG EST collaboration. Matt is surveying
    nearly 300,000 cDNAs from a variety of cDNA libraries from different tissues and stages </font></td>
  </tr>
</table>

<p><font face="Comic Sans MS">of development. In addition, we are using several
non-normalized libraries for the EST project, including brain and nasal epithelium
libraries (provided by <a href="mailto:jngai@socrates.berkeley.edu">John Ngai</a>), kidney
libraries (provided by <a href="mailto:zon@rascal.med.harvard.edu">Len Zon</a>) and fin
regeneration libraries (provided by <a href="mailto:raymond@genetics.wustl.edu">Raymond
Lee</a>) that are serving smaller roles in the EST project. <strong>Additional cDNA
libraries needed.</strong> Are you worried that your tissue or organ of interest is not
well represented in the zebrafish EST project? We anticipate needing an additional 50,000
to 100,000 clones from normalized libraries as we obtain additional funding to extend the
EST project. If you are interested in making normalized zebrafish libraries, contact Steve
at <a HREF="mailto:sjohnson@genetics.wustl.edu">sjohnson@genetics.wustl.edu</a>. </font></p>

<p><font face="Comic Sans MS"><strong>Searching for ESTs</strong>. As EST data are
generated, they are submitted to the dbEST division of NCBI. They can then be searched for
sequence similarity using the NCBI BLAST server. It is important to use the correct
database ("dbEST" or "other ESTs", rather than "NR").
Additional ESTs or clones from the same gene can be identified by BLASTing the EST
database with the EST of interest, and then selecting for hits with nearly perfect
matches. Alternatively, NCBI has begun a Unigene catalogue for zebrafish (<a HREF="http://www.ncbi.nlm.nih.gov/UniGene/Dr.Home.html">http://www.ncbi.nlm.nih.gov/UniGene/Dr.Home.html</a>).
You can search this database with Zebrafish EST Accession numbers. The WUZGR site also
presents zebrafish ESTs arranged by cluster (ESTs likely to belong to the same gene; see <a HREF="http://zfish.wustl.edu">http://zfish.wustl.edu</a>). </font></p>

<p><font face="Comic Sans MS"><strong>Clone retrieval.</strong> Most WUZGR EST clones are
available from Research Genetics, Genome Systems, or the RZPD in Berlin (see dbEST
submissions for links). Because errors are common in EST clone retrieval (see below), we
recommend resequencing to confirm clone identity. From time-to-time, clones will be
unavailable because the distributor has not yet received replicas of the library for clone
retrieval. If you are in a hurry, you might try identifying additional EST clones from the
same gene or cluster by searching the databases, or alternatively, use the EST sequence to
design primers for 3' or 5' RACE to isolate your own clone. Clones are not available from
WUZGR. </font></p>

<p><font face="Comic Sans MS"><strong>Errors in EST data.</strong> Because of the nature
of high throughput sequencing projects, ESTs are subject to several types of error that
need to be considered when using the data or retrieving EST clones. First, because the
submissions result from single pass sequence, they are not subject to the same quality
control as genomic sequence submissions. Typically, EST sequence is &gt;95% accurate, but
it is wrong to assume perfect sequence. Second, EST submission names may not correctly
correspond to clone wells. For instance, errors in lane tracking of the sequencing gel may
result in 3' or 5' reads that don't correspond to the same clone. Comparison of
registration of 3' and 5' read BLAST hits on the WUZRG EST pages may help identify
out-of-register ESTs. Finally, sometimes a set of sequencing reactions is associated with
the wrong plate of clones. In collaboration with Research Genetics, we are rearraying and
resequencing a fraction of the zebrafish EST clones to identify such errors in
submission-to-plate correspondence. Resequencing data is used to help the distributors
improve the accuracy of clone retrieval, or identify plates from which clones cannot be
reliable retrieved. </font></p>

<p><font face="Comic Sans MS"><strong>RH - Transcript map of the zebrafish genome.</strong>
A second goal of WUZGR project is to develop a high density transcript map of the
zebrafish genome. This will provide candidate genes for positional cloning projects,
nearby markers to initiate walks and information about the syntenic correspondence of the
zebrafish and human genomes. WUZGR is funded to place 2500 ESTs or genes on the LN54 RH
map. Our collaborators in the Ekker and Zon labs are placing many more thousand ESTs on RH
maps. To date, WUZGR has mapped greater than 750 ESTs and genes on the LN54 panel (see <a href="http://zfish.wustl.edu">http://zfish.wustl.edu</a>). </font></p>

<p><font face="Comic Sans MS"><strong>Citations using unpublished WUZGR EST and RH data
from the WWW</strong>. We recommend following Genome Research's guidelines on use of
unpublished high throughput genomic data (<a href="http://www.genome.org/misc/ifora.shtml">http://www.genome.org/misc/ifora.shtml</a>).
Specifically, we suggest investigations that make use of WUZGR EST sequences or clones
should acknowledge <a href="mailto:clark@mpimg-berlin-dahlem.mpg.de">M. Clark</a> and <a href="mailto:sjohnson@genetics.wustl.edu">S. Johnson</a>, Washington University Zebrafish
Genome Resource EST project, grant DK55379. When using WUZGR web-based data, care should
be taken to verify conclusions such as clone identity or mapping information
independently. We anticipate that many investigators will use WUZGR RH mapping data as a
starting point in cloning projects. By their nature, these efforts will tend to verify the
conclusions drawn from our prepublication. Such investigations should acknowledge B.
Barbazuk, J. McPherson and <a href="mailto:sjohnson@genetics.wustl.edu">S. Johnson</a>,
Washington University Zebrafish Genome Resources RH mapping project, grant DK55379.
Investigators wishing to use prepublished mapping data (for instance, maps or portions of
maps) without independent verification should contact <a href="mailto:sjohnson@genetics.wustl.edu">S. Johnson</a> to arrange collaborations. </font></p>

<hr>
<a HREF="vol6.1.html#December 1, 1999">

<p><font face="Comic Sans MS">Return to Table of Contents</font></p>
</a>

<hr>
<font face="Times New Roman" size="6"><strong><a NAME="KEEPING AND BREEDING ZEBRAFISH (DANIO RERIO) ON SMALL SCALE">

<p>KEEPING AND BREEDING ZEBRAFISH (<em>DANIO RERIO</em>) ON SMALL SCALE</a></strong></font>
</p>

<p><font size="2" face="Comic Sans MS">By <a href="mailto:a.oberemm@bgvv.de">A. Oberemm</a>,
Institute of Freshwater Ecology and Inland Fisheries (present address: BgVV, P.O. Box
330013, 14191 Berlin, GERMANY)</font></p>

<p><font face="Comic Sans MS"><big><strong>Introduction</strong></big></font></p>

<p><font face="Comic Sans MS">In the years 1994-1998, I successfully maintained a small
zebrafish breeding colony for the use of embryos in a bioassay. During this period, I have
been asked several times to aid other working groups to improve their breeding set-up or
to provide embryos for their experimental work. So I got the idea that it could be useful
to summarize my personal experiences on zebrafish breeding in a brief article that might
be used as a guide for successful breeding of zebrafish in small working groups.</font></p>

<p><font face="Comic Sans MS">Our system was run by 1-2 people. It consisted of 4-6
medium-sized tanks (90-100 L) arranged in two levels that required only a few square
meters. The use of several separated tanks prevented the risk of spreading diseases and
minimized effects of technical accidents. We required from several hundred to 2,000 eggs
per day. Usually, fish were co-operative and we got what we needed! We never had problems
due to the outbreak of diseases.</font></p>

<p><font face="Comic Sans MS">Since the method described by Creaser (1934), many
techniques have been described to maintain and breed zebrafish. The technique described
here depends on some refinement of common techniques and has been proven to work well for
several years in a small lab. In a former issue of <a href="/zf_info/monitor/mon.html"><em>The Zebrafish Science Monitor</em></a>,
Brand et al. described a method designed for large institutions (<a HREF="/zf_info/monitor/vol3.5/4.html">Volume 3, Issue 5, 1997</a>).
Additional technical information on zebrafish may be obtained from <a HREF="/zf_info/zfbook/zfbk.html"><em>The Zebrafish Book</em> </a>and
<a HREF="/">ZFIN</a> (<a HREF="/">http://zfin.org</a>).
Zebrafish are a tough, highly adaptable lab species. However, some of my minor refinements
which are easy to perform, address the natural demands of this species and may contribute
to the well-being of the fish. These refinements may also be applicable to larger breeding
colonies.</font></p>

<p><font face="Comic Sans MS"><big><strong>Fish and fish selection</strong></big></font></p>

<p><font face="Comic Sans MS">One of our breeding groups consisted of about 12 wild-type
females and 15 males. Fish may be obtained from a pet shop, a wholesale trader or,
preferably, as offspring from large labs working with zebrafish. We obtained our fish,
labeled as a Singapore import, from a wholesale trader. Some experience and patience is
required when selecting and sorting young, lean fish for a breeding group. If a spacious
tank is available, a large group of fish can be kept under favorable conditions for a few
days; then selecting and sexing is much more convenient and successful.</font></p>

<p><font face="Comic Sans MS"><big><strong>Tank system and maintenance</strong></big></font></p>

<p><font face="Comic Sans MS">Breeding fish were kept in glass aquaria (70-80 x 40 x 35 cm
L x W x H, 90-100 L). To provide some shelter, bottom, back, and side panels of the tanks
were shielded with dark paperboard. The fish showed much more pronounced coloration and
schooling behavior when they were transferred from a bright, open tank to a dark, shielded
tank. Berlin tap water (conductivity app. 700 &micro;S, pH 8) was used to fill the tanks. Prior
to use, the water was aerated overnight. If appropriate tap water is unavailable
(chlorine-free and carbonate rich, conductivity range about 500 - 800 &micro;S) or the
conditions are not well defined, the use of artificial tank water is recommended: 200 mg
CaCl<sub>2</sub> x 2 H<sub>2</sub>O, 103 mg NaHCO<sub>3</sub>, 100 mg seasalt (Instant
Ocean<sup>&reg;</sup>) added to 1 L water. The water should be reverse osmosis or better
quality, do not use water from crude ion exchange systems. The resulting water is
approximately 800 &micro;S at pH 8 and is well buffered. This recipe was prepared according to
a recommendation from B. Trevarrow, Zebrafish Facility of the University of Oregon and was
successfully applied for several years.</font></p>

<p><font face="Comic Sans MS">The water temperature was maintained at 26&deg; C. In my
experience, this temperature is sufficient to enable a high embryo production rate and
contributes to the robustness of the animals. If there is no air-conditioned room
available, use high quality immersion heaters (50 W at an ambient temperature of 18-22&deg;
C). A 14 hour light/10 hour dark cycle kept the fish in a reproducing state. The water
quality was maintained using an internal bioactive filter system (foam parts; operated by
air or motor; i.e. &quot;Juwel-Standard&quot;, Germany; monthly cleaning by removing and
rinsing the top foam layer) and changing about &frac14; of the water volume weekly. Filters were
adjusted to enhance surface turbulence to optimise gas exchange and a current for the
rheophile fish. The use of an appropriate filter system is essential for maintenance of
the system and seems often to be a major cause of failure. Many common filter systems have
too small a filter surface with too high a circulation rate, which does not provide
sufficient microbial activity for denitrification. Combined with intensive feeding, these
conditions will often cause high levels of toxic nitrogenous waste products.</font></p>

<p><font face="Comic Sans MS">Tanks were outfitted with baskets of Java fern (<em>Microsorium
pteropus</em>) which were kept at the bottom of the tank by attaching them to a glass rod.
These plants are useful as a spawning substrate and provide shelter as well as maintaining
water quality. Because they do not demand ground material, it is easy to maintain clean
conditions by occasionally rinsing the baskets (refer to discussion at <a HREF="http://zebra.sc.edu/methods/miscmeth.html">http://zebra.sc.edu/methods/miscmeth.html</a>).
Snails (<em>Melanoides sp</em>.) were used to reduce algae on the sides and remaining food
particles. However, when using snails, ensure they originate from a safe source, because
they often bear parasites and diseases.</font></p>

<p><font face="Comic Sans MS">Fish were fed 3 times a day, alternating among frozen white
worm (<em>Chaoborus spp</em>.), <em>Daphnia</em> <em>spp.</em> and adult brine shrimp (<em>Artemia
spp</em>.). It has been shown that these nutritional conditions are in accordance with
feeding habits in the natural environment (Dutta 1993). Be aware that the quality of
frozen food varies considerably depending on the source. In some cases, diseases may be
introduced by the food. Flake food (i.e. Tetra Min<sup>&reg;</sup>) may be used occasionally
as a supplemental food, for example with automatic feeders on the weekend. Feeding must be
carefully adjusted to the demand, because overfeeding will reduce the water quality and
cause overweight fish. In general, all the food per feeding should be eaten within a few
minutes (about 3-4 g frozen material per 25-30 adult fish). Reduced appetite and activity
during feeding is always a sign that the fish do not feel well. If conditions are well
maintained, each breeding stock will produce from several hundred to more than 1000 eggs
per day. When egg production decreases after about a year, younger fish should replace the
stock.</font></p>

<p><font face="Comic Sans MS">This maintenance regime is comparably less time-consuming
and care-intensive than other methods and also offers the advantage that the fish are less
disturbed by regular maintenance activities like tank cleaning and water changes. However,
continuous attention and action is needed for its successful operation.</font></p>

<p><font face="Comic Sans MS"><big><strong>Embryo production</strong></big></font></p>

<p><font face="Comic Sans MS">To collect eggs, a mesh (steel, width 3-4 mm) covered
rectangular, translucent plastic dish (spawning box, 20 x 20 x 6 cm) is placed into the
tank and a fern bushel arranged on the mesh in the evening. Remove surplus plants, because
this will greatly reduce the amount of eggs in the dish. The fish start their spawning
activities at the onset of light the following morning, using the fern as a substrate for
mating. Eggs drop through the mesh into the dish, which should be removed after 1-2 hours.
The content of the dish is poured through a strainer, the eggs and debris are gently
rinsed using aerated tap water or artificial water, and the cleaned eggs are transferred
to medium-sized Petri dishes. Fertilized eggs are selected using a dissecting microscope
and an Eppendorf pipette with a fine-polished tip (opening of the tip 1.5-2 mm, rounded by
flame). If no Eppendorf pipette is available, you can use a polished Pasteur pipette.</font></p>

<p><font face="Comic Sans MS">Because well-conditioned females will lay eggs approximately
every three days and fish show some group-determined behavior, the number of eggs may vary
considerably from day to day. When eggs are needed on a particular day, egg production may
be controlled to some extent by adding plants and boxes only on the evening prior to the
egg collection date. However, the presence of the substrate/ shelter contributes to the
well being of the fish, so I always leave the plants in the tanks. </font></p>

<p><font face="Comic Sans MS"><big><strong>Raising fish for stock maintenance</strong></big></font></p>

<p><font face="Comic Sans MS">To maintain a healthy self-sustaining stock, raising your
own fish is recommended. However, compared to the maintenance of the adult breeding
system, controlled raising of zebrafish demands more effort and some experience, which is
largely due to the nutritional demands of larvae and juveniles. Unless continuous, intense
care is provided, attempts may be ineffective resulting in low survival and growth rates
and frequent malformations. In small colonies, inbreeding may also occur after few
generations. For more information on this subject refer to <a HREF="/zf_info/zfbook/zfbk.html"><em>The Zebrafish Book</em></a>.</font></p>

<hr>
<a HREF="vol6.1.html#December 1, 1999">

<p><font face="Comic Sans MS">Return to Table of Contents</font></p>
</a>

<hr>
<a NAME="ANTERIOR MOVEMENT OF VENTRAL DIENCEPHALIC PRECURSORS"><font face="Times New Roman" size="6"><strong>

<p>ANTERIOR MOVEMENT OF VENTRAL DIENCEPHALIC PRECURSORS SEPARATES THE PRIMORDIAL EYE FIELD
IN THE NEURAL PLATE AND REQUIRES <em>CYCLOPS</em>.</strong></font><font size="3">(<em>Development</em>,
in press.)</font></a> </p>

<p><font size="2" face="Comic Sans MS">By <a href="mailto:varga@uni-freiburg.de">Z.M.
Varga</a>, <a href="mailto:wegner@uoneuro.uoregon.edu">J. Wegner</a>, and <a href="mailto:monte@uoneuro.uoregon.edu">M. Westerfield</a>, Institute of Neuroscience,
University of Oregon, Eugene, OR 97403 USA</font></p>

<p><font face="Comic Sans MS">A currently favored hypothesis postulates that a single
field of cells in the neural plate forms bilateral retinas. To learn how retinal
precursors segregate, we followed individual labeled neural plate cells in zebrafish. In
the late gastrula, a single field of <em>odd-paired like</em> expressing cells contributed
to both retinas, bordered posteriorly by diencephalic precursors expressing <em>mariposa</em>.
Median <em>mariposa </em>expressing cells moved anteriorly, separating the eyes, and
formed ventral anterior diencephalon, the presumptive hypothalamus. In <em>cyclops</em>
mutants, corresponding cells failed to move anteriorly, a ventral diencephalon never
formed, and the eyes remained fused. Ablation of the region containing these cells induced
cyclopia in wild types. Our results indicate that movement of a median subpopulation of
diencephalic precursors separates retinal precursors into left and right eyes. Wild-type <em>cyclops</em>
gene function is required for these morphogenetic movements.</font></p>

<hr>
<a HREF="vol6.1.html#December 1, 1999">

<p><font face="Comic Sans MS">Return to Table of Contents</font></p>
</a>

<hr>

<p><a name="VENTRAL DIENCEPHALON PRECURSORS DRIVE MORPHOGENESIS OF THE ZEBRAFISH FOREBRAIN AND RETINA"><font face="Times New Roman" size="6"><strong>VENTRAL DIENCEPHALON PRECURSORS DRIVE
MORPHOGENESIS OF THE ZEBRAFISH FOREBRAIN AND RETINA</strong></font></a></p>

<p><font SIZE="1" face="Comic Sans MS">By <a href="mailto:kwoo@itsa.ucsf.edu">K. Woo</a><sup>2</sup>,
<a href="mailto:jshih@druggist.gg.caltech.edu">J. Shih</a><sup>3</sup>, and <a href="mailto:sefraser@gg.caltech.edu">S.E. Fraser</a><sup>1</sup>; <sup>1</sup>Biological
Imaging Center, Beckman Institute, MC139-74, Division of Biology, California Institute of
Technology, Pasadena, CA 91125 USA; <sup>2</sup>Department of Anatomy, University of
California, San Francisco, 513 Parnassus Ave, San Francisco, CA 94143 USA; <sup>3</sup>Ontogeny,
Inc., 45 Moulton Street, Cambridge, MA 02138 USA</font></p>

<p><font face="Comic Sans MS">Despite dramatic progress in understanding the mechanisms
that pattern the hindbrain and spinal cord, relatively little is known about the initial
patterning and morphogenesis of the vertebrate forebrain. Recent studies suggest that
cells in the ventral die</font><img src="fraser.gif" width="359" height="284" alt="fraser.gif (92677 bytes)" align="right"><font face="Comic Sans MS">ncephalon provide
signals that influence forebrain pattern formation, analogous to the role of the axial
mesoderm and floorplate in patterning the trunk neuraxis. In this study, we have examined
forebrain morphogenesis in the zebrafish, extending our earlier fatemap studies of the
developing neurectoderm. The changes in the fatemap during normal gastrulation suggest
that ventral diencephalic precursors might physically bisect the unitary retina field to
generate the paired retinal primordia. Here, we present direct evidence supporting this
hypothesis. Time-lapse analysis of labeled ventral diencephalon precursors demonstrates
the proposed &quot;bisecting&quot; movement <i>in vivo</i>. This extension of ventral
diencephalon precursors appears to be intrinsic, as the cells undergo the analogous
motions when grafted to the ventral germring. In addition, these grafted ventral
diencephalon precursors are able to direct the formation of a secondary axis, similar to
those generated by transplanting the embryonic shield, except they lack notochordal
tissue. A number of zebrafish mutations cause fusion of the retinas and are thought to be
caused by abnormalities in the inductive and patterning activities of the prechordal plate
mesoderm. Our results suggest that the motions of the ventral diencephalon progenitors may
play as significant of a role as the signals they emit in normal development and mutant
phenotypes. </font></p>

<hr>
<a HREF="vol6.1.html#December 1, 1999">

<p><font face="Comic Sans MS">Return to Table of Contents</font></p>
</a>

<hr>

<p><font face="Times New Roman" size="6"><strong><a NAME="DECHORIONATION MAY SERIOUSLY DISTURB ZEBRAFISH">DECHORIONATION MAY SERIOUSLY DISTURB
ZEBRAFISH DEVELOPMENT</a></strong></font> </p>

<p><font face="Comic Sans MS" size="2">By L.F. Jaffe and R. Cr&eacute;ton, Marine Biological
Laboratory, Woods Hole and Brown University Medical School, Providence, RI USA; <a HREF="mailto:LJaffe@MBL.edu">ljaffe@MBL.edu</a>; <a HREF="http://waves.mbl.edu">http://waves.mbl.edu</a></font></p>

<p><font face="Comic Sans MS">We intend to pursue a serious discrepancy between Cr&eacute;ton <em>et
al</em>. (1998) Patterns of free calcium in zebrafish embryos. <em>J. Cell Sci.</em> <strong>111</strong>:1613-1622
and Gilland <em>et al</em>. (1999) Imaging of multicellular large-scale rhythmic calcium
waves during zebrafish gastrulation. <em>PNAS</em> <strong>98</strong>:157-161. </font></p>

<p><font face="Comic Sans MS">In brief, Cr&eacute;ton <em>et al.</em> found no such waves and
suppose that this was because we used intact embryos while Gilland <em>et al</em>. used
dechorionated ones. However, the difference might also lie in the more advanced imaging
system employed by Gilland <em>et al</em>. We intend to resolve this by redoing these
experiments in Andrew Miller's lab in Hong Kong with their equipment. Moreover, as
reviewed by Woo <em>et al. </em>(1995) differences in the variability of early cell
lineages in zebrafish have been reported by Strehlow <em>et al.</em> (1993, 1994), Helde <em>et
al.</em> (1994), and Abdelilah <em>et al.</em> (1994). David Strehlow still attributes
these differences to the fact that he studied intact embryos while the other two groups
studied dechorionated ones. This controversy likewise remains unresolved and we would
appreciate informal comments as we plan our work.</font></p>

<p><font SIZE="1" face="Comic Sans MS">Woo, K., Shih, J., and Fraser, S.E. (1995) Fate
maps of the zebrafish embryo. <em>Curr. Op. Genet. Dev.</em> <strong>5</strong>:439-443. <br>
Strehlow, D., Heinrich, G., and Gilbert, W. (1994) The fates of the blastomeres of the
16-cell zebrafish embryo. <em>Development</em> <strong>120</strong>:1791-1798. <br>
Strehlow, D. and Gilbert, W. (1993) A fate map for the first cleavages of the zebrafish. <em>Nature</em>
<strong>361</strong>:451-453. <br>
Helde, K.A., Wilson, E.T., Cretekos, C.J., and Grunwald, D.J. (1994) Contribution of early
cells to the fate map of the zebrafish gastrula. <em>Science</em> <strong>265</strong>:517-520.
<br>
Abdelilah, S., Solnica-Krezel, L., Stainier, D.Y.R., and Driever, W. (1994) Implications
for dorsoventral axis determination from the zebrafish mutation <em>janus</em>. <em>Nature</em>
<strong>370</strong>:468-471. </font></p>

<hr>
<a HREF="vol6.1.html#December 1, 1999">

<p><font face="Comic Sans MS">Return to Table of Contents</font></p>
</a>

<hr>

<p><strong><a NAME="THE TIGR ZEBRAFISH GENE INDEX (ZGI): ESTIMATING THE"><font face="Times New Roman" size="6">THE TIGR ZEBRAFISH GENE INDEX (ZGI): ESTIMATING THE
TRANSCRIPT CONTENT OF THE <em>DANIO RERIO</em> GENOME</font></a></strong> </p>

<p><font face="Comic Sans MS" size="2">By <a href="mailto:jupton@tigr.org">J.F. Upton</a>,
T.S. Hansen, <a href="mailto:ieholt@tigr.org">I.E. Holt</a>, S. Karamycheva, <a href="mailto:fliang@tigr.org">F. Liang</a>, G. Pertea, and <a href="mailto:johnq@tigr.org">J.
Quackenbush</a>; The Institute for Genomic Research 9712 Medical Center Dr., Rockville,
MD, 20850, USA; http://www.tigr.org/~jupton</font></p>

<p><font face="Comic Sans MS">The Zebrafish Gene Index (ZGI) is maintained at TIGR and
updated on a quarterly basis. It represents a best guess of transcript content in the
zebrafish genome through the meticulous cataloguing of Expressed Sequence Tags (ESTs) and
coding sequences, which are downloaded from GenBank on a continual basis.</font></p>

<p><font face="Comic Sans MS">ESTs are subjected to vector trimming and removal of other
extraneous DNA fragments, such as adaptors and linkers. A clear range of useable sequence
is determined for each EST (sequences which are composed of non-useable sequences are not
considered further). ESTs and coding sequences are then placed into clusters through
pair-wise comparison; sequences with &gt;95% identity over 40 bp and overhangs no longer
than 20 bases are assigned to the same cluster. These clusters are then subjected to
high-stringency assembly using CAP3 (Huang, et al. 1999), producing Tentative Consensus
(TC) sequences. The first-pass Tcs are subjected to a second round of clustering and
assembly in order to reduce redundancy. The resulting TCs are compared against a
nonredundant amino acid database and top hits are recorded and used for provisional
functional assignments.</font></p>

<p><font face="Comic Sans MS">ZGI is available on the internet at &lt;</font><font size="3" face="Comic Sans MS" color="#0000FF"><u> <a href="http://www.tigr.org/tdb/zgi/">http://www.tigr.org/tdb/zgi/</a></u></font><font face="Comic Sans MS">&gt; and features EST sequence information, TC information, and
tentative functional assignments. The index is searchable by EST, TC, library identifier,
tissue, gene product name, and keyword. Additionally, a BLAST search utility is provided.</font></p>

<p><font face="Comic Sans MS">One valuable feature of ZGI (and all other TIGR Gene
Indices) is the heritability of TC identifiers. With each rebuild, TCs can change by
splitting or merging. However, previous consensus incarnations are tracked and annotated;
ESTs and TCs may therefore be followed by users across subsequent index releases. </font></p>

<p><font face="Comic Sans MS">As of this writing, ZGI is undergoing an update, with a
release expected by late November. The current ZGI was released September 2, 1999 and
features 3211 TCs and 13,148 single ESTs, for a total of 16,359 unique sequences in the
index.</font></p>

<p><font face="Comic Sans MS">TIGR also maintains nine other indices, featuring human,
rat, mouse, <em>Drosophila</em>, <em>Arabidopsis</em>, Maize, Rice, Soybean, and Tomato.
An eleventh index for potato is expected in 2000.</font></p>

<p><font face="Comic Sans MS"><strong>Reference</strong></font></p>

<p><font face="Comic Sans MS" size="2">Huang X, Madan A (1999) CAP3: A DNA sequence
assembly program. <em>Genome Res.</em> <strong>9</strong>(9), 868-877. </font></p>

<hr>
<a HREF="vol6.1.html#December 1, 1999">

<p><font face="Comic Sans MS">Return to Table of Contents</font></p>
</a>

<hr>
<font face="Times New Roman" size="6"><a NAME="ZEBRAFISH NOMENCLATURE COMMITTEE"><strong>

<p>ZEBRAFISH NOMENCLATURE COMMITTEE</strong></a></font> </p>

<p><font face="Comic Sans MS" size="2">By <a href="mailto:kfs@zfin.org">K. Fraser</a><sup>1</sup>,
<a href="mailto:mullins@mail.med.upenn.edu">M. Mullins</a><sup>2</sup>, <a href="mailto:jpostle@oregon.uoregon.edu">J. Postlethwait</a><sup>1</sup>, and <a href="mailto:monte@uoneuro.uoregon.edu">M. Westerfield</a><sup>1</sup>; <sup>1</sup>
Institute of Neuroscience, University of Oregon, Eugene, OR 97403-1254 USA; <sup>2</sup>
Department of Cell and Developmental Biology, University of Pennsylvania, 1211 BRB II/III,
421 Curie Blvd, Philadelphia, PA 19104-6058 USA</font></p>

<p><font face="Comic Sans MS">The zebrafish nomenclature committee, appointed at the Cold
Spring Harbor meeting a couple of years ago, is alive and well. Our job is to provide
guidelines and advice for naming zebrafish genes and mutants. The guidelines are based on
input from the zebrafish research community and on the rules used for naming human and
mouse genes. </font></p>

<p><font face="Comic Sans MS">During the past three years, we have participated in the
International Nomenclature Workshops and have published updated guidelines in <em>Trends
in Genetics</em> (Genetic Nomenclature Guide, 1998). The current guidelines are available
on <a HREF="/">ZFIN</a> (<a HREF="/zf_info/nomen.html">http://zfin.org/zf_info/nomen.html</a>).
</font></p>

<p><font face="Comic Sans MS">The committee continuously reviews zebrafish gene names to
consolidate multiple names for the same gene, to assign uniform names to all members of
gene families, and to clarify the relationships between genes in zebrafish and other
vertebrates. These reviews often require BLAST, cluster and mapping analysis. We also work
closely with the nomenclature committees for mouse and human throughout this process. The
approved name and all other previous names of all published zebrafish genes are listed on <a HREF="/">ZFIN</a>.</font></p>

<p><font face="Comic Sans MS">We encourage all of you to consult the committee before
publishing new gene names. <a HREF="/">ZFIN</a> currently supports a
mechanism for researchers to reserve names for newly discovered mutants. These names are
reviewed by the committee to ensure they conform to accepted nomenclature conventions and
then given a tentative approval that reserves the name until the mutant is published. This
mechanism can easily be extended to provide an automatic means for checking names of genes
identified by molecular means. Are you interested in such a service?</font></p>

<p><font face="Comic Sans MS">The mouse and human research communities now require
approval from their respective nomenclature committees before new names can be published.
Most of the major journals support this process by requiring a statement of the
committee's approval for all new names before the manuscript is published. This helps
avoid changing names later because of nomenclature problems. Is the zebrafish research
community ready for this?</font></p>

<p><font face="Comic Sans MS">Please give us your feedback at: <a HREF="mailto:nomenclature@zfin.org">nomenclature@zfin.org</a></font></p>
<a HREF="vol6.1.html#December 1, 1999">

<hr>

<p><font face="Comic Sans MS">Return to Table of Contents</font></a></p>

<hr>

<p><a name="A SIMPLE METHOD FOR OBTAINING VERY YOUNG EMBRYOS DURING AN EXTENDED TIME PERIOD"><font face="Times New Roman" size="6"><strong>A SIMPLE METHOD FOR OBTAINING VERY YOUNG EMBRYOS
DURING AN EXTENDED TIME PERIOD</strong></font></a></p>

<p><font face="Comic Sans MS" size="2">By <a href="mailto:razerez@sun2.ruf.uni-freiburg.de">E. Raz</a>; Department of Developmental
Biology, Freiburg Hauptstr. 1, 79104 Freiburg, Germany </font></p>

<p><font face="Comic Sans MS">Injection of large number of embryos with RNA or DNA during
early development is limited because most fish lay eggs during the first 30-60 minutes
after the lights turn on. To solve this problem, some people grow fish in closets with
different time regimes, a solution that requires extra tank space and complicates the
feeding schedule. Another alternative is to separate the males from the females early in
the morning and to perform several rounds of <em>in vitro</em> fertilization. Yet, this
procedure is labor intensive and is stressful for the fish. </font></p>

<table border="0" cellpadding="0" cellspacing="0" width="104%">
  <tr>
    <td width="82%"><font face="Comic Sans MS">We are using small breeding cages similar to
    those used in T&uuml;bingen which are composed of clear plastic boxes (19cm x 9cm x 8cm LxWxH)
    and white soft plastic inserts with wired-mesh on the bottom. Two inserts can be put one
    on top of the other, and if they are not pressed tightly together, a space is left between
    the meshes. As shown in the figure, we put the two fish together in two separate
    compartments so that they cannot mate until the desired time. We then put the fish
    together in a cage with fresh water and typically get eggs 5-30 minutes later. Setting up
    10-15 crosses in this way allows one to obtain a constant supply of very early embryos for
    two to three hours each morning. Other advantages of this method are that the fish spend
    very little time together, eliminating cases of death due to evening/night fish fights,
    and the problem of fish laying eggs during the evening.</font></td>
    <td width="24%"><img src="raz.gif" width="236" height="253" alt="raz.gif (17878 bytes)"></td>
  </tr>
</table>

<hr>
<a HREF="vol6.1.html#December 1, 1999">

<p><font face="Comic Sans MS">Return to Table of Contents</font></p>

<table border="0" cellpadding="0" cellspacing="0" width="100%">
  <tr>
    <td width="100%"></td>
  </tr>
</table>

<hr>
<div align="center"><center>

<table border="3" cellpadding="0" cellspacing="0" width="60%" bordercolorlight="#00FFFF" bordercolordark="#0000FF" bgcolor="#F4F791" background="lgrey033.jpg">
  <tr>
    <td width="100%" bgcolor="#F4F791">&nbsp;<h5 ALIGN="CENTER"></a><big><font face="Comic Sans MS"><strong>4th
    Biennial Meeting on</strong></font></big></h5>
    <h1 ALIGN="CENTER"><font face="Comic Sans MS" color="#FF00FF"><a name="Zebrafish Development &amp; Genetics"><strong>Zebrafish Development &amp; Genetics</strong></a></font></h1>
    <h4 ALIGN="CENTER"><font face="Comic Sans MS"><font color="#0000FF">Cold Spring Harbor
    Laboratory<br>
    Cold Spring Harbor, NY</font><br WP="BR1">
    </font></h4>
    <h2 ALIGN="CENTER"><font face="Comic Sans MS"><strong>April 26-30, 2000</strong></font></h2>
    <h5 ALIGN="CENTER"><font face="Comic Sans MS" color="#0000FF">Abstracts due: February 2,
    2000</font></h5>
    <p align="center"><a HREF="vol6.1.html#December 1, 1999">&nbsp;</td>
  </tr>
</table>
</center></div></a>

<hr>
<a HREF="vol6.1.html#December 1, 1999">

<p><font face="Comic Sans MS">Return to Table of Contents</font></p>
</a>

<hr>
<align="center">
<a name="Check out these zebrafish websites!"><font face="Comic Sans MS" size="6">

<p align="center">Check out these zebrafish websites!</font></a> </p>

<hr size="6">
<div align="center"><center>

<table border="2" cellpadding="5" cellspacing="0" width="69%" bordercolorlight="#00FFFF" bordercolordark="#0000FF" bgcolor="#F4F791">
  <tr>
    <td width="34%"><font face="Comic Sans MS" size="4"><strong>ZFIN</strong> </font></td>
    <td width="66%"><font face="Comic Sans MS"><a HREF="/">http://zfin.org</a></font></td>
  </tr>
  <tr>
    <td width="34%"><strong><font face="Comic Sans MS" size="4">Zebrafish K-12 </font><font COLOR="#f80080" face="Comic Sans MS" size="4">(new!)</font></strong></td>
    <td width="66%"><font face="Comic Sans MS"><a HREF="http://www.neuro.uoregon.edu/k12/zfk12.html">http://www.neuro.uoregon.edu/k12/zfk12.html</a></font></td>
  </tr>
  <tr>
    <td width="34%"><strong><font face="Comic Sans MS" size="4">Zebrafish Info Server </font></strong></td>
    <td width="66%"><font face="Comic Sans MS"><a HREF="http://zebra.biol.sc.edu/">http://zebra.biol.sc.edu/</a></font></td>
  </tr>
  <tr>
    <td width="34%"><strong><font face="Comic Sans MS" size="4">Zebrafish News Group</font></strong></td>
    <td width="66%"><font COLOR="#0000ff" face="Comic Sans MS">bionet.organisms.zebrafish</font></td>
  </tr>
</table>
</center></div>

<hr>
<a HREF="vol6.1.html#December 1, 1999">

<p><font face="Comic Sans MS">Return to Table of Contents</font></p>

<hr>
</a>
</body>

<script language="JavaScript" src="/javascript/footer.js"></script>
