<html>

<head>
<meta NAME="GENERATOR" CONTENT="Microsoft FrontPage 3.0">
<title>The Zebrafish Science Monitor, Vol. 7 (1)</title>
</head>

<body BACKGROUND="../vol6.1/lgrey033.jpg" TEXT="#000000" LINK="#0000FF" VLINK="#0000FF" ALINK="#FF0000">
<div align="center"><center>

<table border="3" cellpadding="0" cellspacing="0" width="80%" bordercolorlight="#C0C0C0" bordercolordark="#000080">
  <tr>
    <td width="100%" bgcolor="#80FFFF"><p align="center"><big><big><font face="Times New Roman"><strong><big><em>The Zebrafish Science Monitor</em>, </big>Vol. 7
    (1)</strong></font></big></big></td>
  </tr>
</table>
</center></div>

<h3 ALIGN="CENTER"><em>&quot;<font face="Comic Sans MS">All the Zebrafish News That's Fit
to Print</em>&quot;</font></h3>

<hr>

<h4><a NAME="August 14, 2000"><font face="Comic Sans MS"><big>August 14, 2000</big></font></a></h4>

<hr>

<ul>
  <li><font SIZE="4" face="Comic Sans MS"><a href="#REDUCED LEVELS OF MUTANT mRNAS IN ZEBRAFISH MUTANTS MAY NOT BE DUE TO LOSS OF POSITIVE AUTOREGULATION">Reduced
    Levels of Mutant mRNAs in Zebrafish Mutants May Not Be Due to Loss of Positive
    Autoregulation</a></font></li>
  <li><font SIZE="4" face="Comic Sans MS"><a href="#HEALTHY, HAPPY FISH: DISEASE PREVENTION AND USEFUL POINTERS">Healthy, Happy Fish:
    Disease Prevention and Useful Pointers</a></font></li>
  <li><font face="Comic Sans MS"><a href="#TO ORTHOLOGUE OR NOT TO ORTHOLOGUE, THAT IS THE QUESTION"><big>To Orthologue or Not
    to Orthologue, That Is The Question</big></a></font></li>
  <li><font face="Comic Sans MS"><a href="#WHATS IN A NAME"><big>What&#146;s in a Name?</big></a></font></li>
  <li><a href="#ANATOMICAL NOMENCLATURE FOR ZEBRAFISH"><font SIZE="4" face="Comic Sans MS">Anatomical
    Nomenclature for Zebrafish</font></a></li>
  <li><a href="#EB 2001 MEETING"><font face="Comic Sans MS"><big>EB 2001 Meeting</big></font></a></li>
</ul>

<hr>

<table border="0" cellpadding="0" cellspacing="0" width="100%">
  <tr>
    <td width="13%"><font face="Comic Sans MS"><a HREF="/">ZFIN</a></font> </td>
    <td width="87%"><a href="/"><img src="../vol6.1/Fish_net.gif" alt="Fish_net.gif (2779 bytes)" WIDTH="100" HEIGHT="37"></a></td>
  </tr>
</table>

<hr>
<b><font SIZE="4">

<p align="left"></font><a name="REDUCED LEVELS OF MUTANT mRNAS IN ZEBRAFISH MUTANTS MAY NOT BE DUE TO LOSS OF POSITIVE AUTOREGULATION"><font face="Times New Roman" size="6">REDUCED LEVELS OF MUTANT mRNAS IN ZEBRAFISH MUTANTS MAY
NOT BE DUE TO LOSS OF POSITIVE AUTOREGULATION</font></a></b></p>
<i>

<p></i><small><font face="Comic Sans MS">By C.B. Moens, Division of Basic Science, Fred
Hutchinson Cancer Research Center, Seattle WA 98109-1024. <a href="mailto:cmoens@fhcrc.org">cmoens@fhcrc.org</a></font></small></p>

<p><font face="Comic Sans MS">With the revelations of the recent zebrafish meeting, to
date about 50 chemically-induced mutant loci have been cloned, either positionally or by
candidate gene approaches. In many cases, analysis of the expression of these cloned genes
has revealed strongly reduced levels of mutant mRNA in homozygous mutant embryos. This
observation is often interpreted as suggesting that in wild-type embryos the gene is under
positive autoregulatory control, whether direct or indirect. However this has not been
directly demonstrated to be the case, and alternative explanations for the loss of mutant
transcripts exist. </font></p>
<div align="center"><center>

<table border="0" cellpadding="0" cellspacing="0" width="45%">
  <tr>
    <td width="100%"><img src="moens_figure.jpg" width="447" height="447" alt="moens figure (273379 bytes)" align="right"></td>
  </tr>
  <tr>
    <td width="100%"><font face="Comic Sans MS" SIZE="2"><b>Figure 1</b>. A,B: <i>lzr/pbx4 </i>is
    expressed broadly and at high levels in wild-type embryos at 24 hours (A), but expression
    is strongly reduced in <i>lzr- </i>embryos at the same stage (B). C,D: <i>lzr<sup>-</sup> </i>embryos
    have reduced expression of <i>krox20 </i>in r3, and fused streams of <i>dlx2 </i>expression
    in the migrating cranial crest (C). These phenotypes are rescued by injection of
    full-length <i>lzr<sup>+</sup></i> mRNA at the 1-cell stage (D). Insets show PCR-dHPLC
    genotyping: one peak indicates a <i>lzr<sup>-/-</sup></i> genotype, two peaks a <i>lzr<sup>+/-</sup></i>
    genotype. E,F: RNA <i>in situs</i> using three digoxigenin-labelled probes, <i>krox20</i>,
    <i>dlx2</i> and <i>lzr/pbx4</i> demonstrate that rescue of the <i>lzr<sup>-</sup></i>
    phenotype does not prevent elimination of <i>lzr<sup>-</sup></i> mRNA. Thus a
    genotypically heterozygous embryo (E) exhibits broad expression of <i>lzr/pbx4</i>, while
    a genotypically mutant embryo (F) that has been rescued as in (D) nevertheless eliminates
    its <i>lzr<sup>-</sup> </i>mRNA.</font></td>
  </tr>
</table>
</center></div>

<p><font face="Comic Sans MS">We recently identified the zebrafish <i>pbx</i> family
member <i>pbx4</i> as being mutated in <i>lazarus </i>mutant embryos (Pöpperl et al., in
press). When we observed that in <i>lzr<font SIZE="4"><sup>-</sup></font> </i>embryos the
levels of <i>pbx4</i> mRNA was substantially reduced (Fig. 1 A,B), we decided to test
directly whether the lack of up-regulation of the mutant transcript was due to the loss of
autoregulatory input. In <i>lzr<font SIZE="4"><sup>-</sup></font> </i>embryos, <i>krox20 </i>expression
in r3 is reduced and <i>dlx2</i> expression in the neural crest fails to be subdivided
into three streams corresponding with the first three pharyngeal arches (Fig. 1C).
Injection of full-length <i>pbx4</i> mRNA rescues both of these aspects of the <i>lzr<font SIZE="4"><sup>-</sup></font> </i>phenotype (Fig. 1D), while injection of mutant <i>pbx4</i>
mRNA does not (Fig. 1C).</font></p>

<p><font face="Comic Sans MS">In order to test the autoregulation hypothesis, we rescued <i>lzr<font SIZE="4"><sup>-</sup></font> </i>embryos by injection of an mRNA encoding full-length <i>pbx4</i>
in which the <i>pbx4</i> 3&#146;UTR was replaced with the SV40 polyadenylation signal. We
then performed RNA in situ hybridizations on these injected embryos at 18 hpf using <i>krox20</i>
and <i>dlx2</i> to assay rescue of the <i>lzr<font SIZE="4"><sup>-</sup></font> </i>phenotype,
and a probe generated from the <i>pbx4</i> 3&#146;UTR to detect the levels of endogenous <i>pbx4</i>
mRNA. While 100% (n=48) of injected embryos were phenotypically wild-type with respect to <i>krox20
</i>and <i>dlx2</i>, we could readily identify two classes of embryos with respect to <i>pbx4</i>
expression. The first class, comprising 73% of injected embryos, had normal levels of pbx4
mRNA (Fig. 1E). The second class (27% of injected embryos) had undetectable levels of
endogenous <i>pbx4</i> mRNA (Fig. 1F). Subsequent genotyping of these embryos by
denaturing HPLC demonstrated that the latter&nbsp; class were the <i>lzr<font SIZE="4"><sup>-</sup></font>
</i>embryos (Fig. 1E,F insets). Thus endogenous <i>pbx4</i> mRNA fails to be up-regulated
in <i>lzr<font SIZE="4"><sup>-</sup></font> </i>embryos regardless of the presence of
rescuing wild-type Pbx4 protein. </font></p>

<p><font face="Comic Sans MS">This experiment demonstrates that a mechanism other than the
loss of autoregulatory input is primarily responsible for the loss of mutant <i>pbx4</i>
mRNA in <i>lzr<font SIZE="4"><sup>-</sup></font> </i>embryos. A process called
nonsense-mediated mRNA decay, whereby the cell recognizes and eliminates mRNAs encoding
truncated proteins, has been described in yeast, C. elegans and in humans (Hentze and
Kulozik, 1999; Culbertson, 1999). While the molecular genetics of this process are just
beginning to be elucidated, its adaptive nature is easy to imagine, as it protects the
cell from expressing potentially deleterious dominant-negative forms of proteins. The
C-to-T mutation in <i>pbx4</i> responsible for the <i>lzr<font SIZE="4"><sup>-</sup></font></i>
phenotype generates a premature stop codon that truncates the Pbx4 protein after 45 amino
acids, making this mutant message a prime candidate for being a target of
nonsense-mediated RNA decay in the zebrafish. This analysis raises the possibility that
mutant mRNAs encoding truncated proteins in other zebrafish mutants may similarly be
targets for nonsense mediated decay. </font></p>
<b>

<p><font face="Comic Sans MS">References:</font></b></p>

<p><font face="Comic Sans MS">Pöpperl H., Rikhof, H., Chang, H., Haffter, P., Kimmel,
C.B., and Moens, C.B. (2000). <i>lazarus</i> is a novel <i>pbx</i> gene that globally
mediates <i>hox </i>gene function in zebrafish. Mol. <i>Cell</i>, <i>in press</i>.</font></p>

<p><font face="Comic Sans MS">Hentze, M.W., Kulozik, A.E. (1999) A perfect message: RNA
surveillance and nonsense-mediated decay. <i>Cell</i> <b>96</b>:307-10.</font></p>

<p><font face="Comic Sans MS">Culbertson, M.R. (1999) RNA surveillance. Unforeseen
consequences for gene expression, inherited genetic disorders and cancer. <i>Trends Genet.</i>
<b>15</b>: 74-80.</font></p>

<hr>
<a HREF="../vol6.1/vol6.1.html#December 1, 1999">

<p></a><a href="#August 14, 2000"><font face="Comic Sans MS">Return to Table of Contents</font></a><a HREF="../vol6.1/vol6.1.html#December 1, 1999"></p>
</a>

<hr>
<font FACE="Times New Roman" SIZE="4">

<p></font><a name="HEALTHY, HAPPY FISH: DISEASE PREVENTION AND USEFUL POINTERS"><strong><font face="Times New Roman" size="6">HEALTHY, HAPPY FISH: DISEASE PREVENTION AND USEFUL
POINTERS</font></strong></a></p>
<font FACE="Times New Roman" size="3">

<p></font><font face="Comic Sans MS" SIZE="2">By J. Matthews and B. Trevarrow; Zebrafish
International Resource Center, University of Oregon, Eugene, OR 97403 USA</font><font SIZE="3"></p>

<p><font face="Comic Sans MS">For any large colony of fish, precautions should be taken to
prevent the introduction and spread of disease. The easiest strategy for combating disease
is to ensure good water quality and strict sanitation and quarantine procedures.</font></p>
<b>

<p><font face="Comic Sans MS">General Procedures:</font></b></p>

<blockquote>
  <ul>
    <li><font face="Comic Sans MS">Minimize transfer of fish and water between tanks as much as
      possible.</font></li>
    <li><font face="Comic Sans MS">Periodically sterilize all equipment that comes in contact
      with fish or tanks.</font></li>
    <li><font face="Comic Sans MS">Use fishnets, siphons, and cleaning sponges in only one tank
      at a time and autoclave or sanitize them before using them in a different tank.</font></li>
    <li><font face="Comic Sans MS">Water sterilization (e.g. ultra-violet, ozone, sub-micron
      filtration) is recommended in recirculating systems.</font></li>
    <li><font face="Comic Sans MS">Monitor water quality parameters routinely.</font></li>
    <li><font face="Comic Sans MS">Remove sick and dead fish from tanks to minimize disease
      carriers and avoid cannibalism.</font></li>
    <li><font face="Comic Sans MS">Optimize general husbandry practices (nutrition, stress,
      stocking density, etc.).</font></li>
    <li><font face="Comic Sans MS">Quarantine all fish from outside sources and bring only
      properly sanitized embryos into the colony. A procedure for sanitizing the embryos with
      bleach solution is described below.</font></li>
  </ul>
</blockquote>
<b>

<p><font face="Comic Sans MS">Quarantine Procedures </font></b></p>

<p><font face="Comic Sans MS">Fish brought in from the outside world (i.e. not born in
your facility) must be isolated in a quarantine area to reduce the risk of contamination
of existing stocks with infectious diseases. The ideal quarantine area is a room
completely separated from the main fish facility and on a separate flow-through water
system. Depending on the size and arrangement of your facility this may or may not be
practical. Small facilities can use individual aquaria or static tanks in a designated
area of the facility. In either case, procedures should be designed and implemented to
prevent the spread of diseases from the quarantine area.</font></p>

<blockquote>
  <ul>
    <li><font face="Comic Sans MS">Fish that were not born in the facility, that have had an
      out-of-facility experience, or that have been in contact with fish from outside of the
      facility should never enter or return to the main fish facility. Quarantine these fish.</font></li>
    <li><font face="Comic Sans MS">Examine all incoming fish to the quarantine area for signs of
      disease. If the fish have any illness or are coming from a facility with known disease
      problems, treat them (if possible). </font></li>
    <li><font face="Comic Sans MS">Observe fish in quarantine for a minimum of two weeks prior
      to breeding. If fish show signs of disease within this period, treat or euthanize them,
      depending on the value of the particular fish. Dispose of dead fish in a manner that will
      not contaminate the main facility.</font></li>
    <li><font face="Comic Sans MS">Never move live fish out of the quarantine area unless they
      leave the building entirely. Only surface-sanitized embryos obtained from the fish in the
      quarantine area can be transferred into the main facility (see Embryo Bleaching Procedure
      below) </font></li>
    <li><font face="Comic Sans MS">Bleach or autoclave all equipment leaving the quarantine
      area. Transport, clean, and store quarantine room equipment in containers that are kept
      separate from the central facility. Label these containers clearly.</font></li>
    <li><font face="Comic Sans MS">Consider everything in the quarantine area contaminated,
      including personnel. Wash your hands and arms thoroughly after working in the area. </font></li>
  </ul>
</blockquote>
<b>

<p><font face="Comic Sans MS">Embryo Bleaching Procedure</font></b></p>

<blockquote>
  <p></font><font face="Comic Sans MS" SIZE="3">1. Prepare the bleach solution: 100 Fl (0.1
  ml) of bleach in 170 ml of clean system water. Mixing it up ahead of time guarantees that
  the bleach is evenly dispersed. The bleach we use is 5% sodium hypochlorite (NaOCl) from
  Baker (Order No. 9416-01; 5.3% available chlorine; PO</font><font face="Comic Sans MS"><font SIZE="5"><sub>4</sub></font> &lt;0.0005%; Ca &lt;0.001%). The use of chemically pure
  sodium hypochlorite is recommended. Some inexpensive brands of bleach contain contaminants
  or additives that are potentially toxic to embryos.</font></p>
  <p><font face="Comic Sans MS">2. Place the embryos in a clean dish, remove the water they
  are in and add the bleach solution, or use a pipette to add the embryos to the bleach
  solution in a small volume of water.</font></p>
  <p><font face="Comic Sans MS">3. Gently swirl the embryos in the bleach solution and allow
  them to stand for 5 minutes.</font></p>
  <p><font face="Comic Sans MS">4. Rinse the embryos two to three times in sterile system
  water and transfer the eggs to a clean disposable petri dish.</font></p>
  <p><font face="Comic Sans MS">5. The bleaching procedure can be repeated if desired.</font></p>
  <p><font face="Comic Sans MS">6. There are many opinions regarding the timing of the
  bleaching procedure for optimal survival. Some people bleach the embryos within several
  hours of fertilization (3-8 h), others wait until epiboly is complete (10 h) and still
  others wait until they are 24-36 h. In general, we recommend bleaching the embryos within
  the first day. Bleaching toughens the chorion to the point that fish may not be able to
  hatch. Because of this, the chorion should be removed either mechanically or with pronase
  treatment prior to normal hatching.</font></p>
  <p><font face="Comic Sans MS">7. Embryos that have been properly bleached can be removed
  safely from the quarantine room. Be careful not to recontaminate the embryos by contact
  with nonsterile water or glassware.</font></p>
</blockquote>
<i>

<p><font face="Comic Sans MS">Suggestions for future topics relating to zebrafish health
and husbandry can be sent to </i><a href="mailto:fish_health@zfin.org">fish_health@zfin.org</a><i>.</i></font></p>

<hr>
<a HREF="../vol6.1/vol6.1.html#December 1, 1999">

<p></a><a href="#August 14, 2000"><font face="Comic Sans MS">Return to Table of Contents</font></a><a HREF="../vol6.1/vol6.1.html#December 1, 1999"></p>
</a>

<hr>
<font FACE="Times New Roman"><b>

<h1><a name="TO ORTHOLOGUE OR NOT TO ORTHOLOGUE, THAT IS THE QUESTION">TO ORTHOLOGUE OR
NOT TO ORTHOLOGUE, THAT IS THE QUESTION</a></b></font><font FACE="Times New Roman" SIZE="1"></h1>
</font><font SIZE="1">

<p></font><font size="2" face="Comic Sans MS">By the Zebrafish Nomenclature Committee (J.
Postlethwait*, M. Ekker**, K. Frazer*, M. Mullins***, and M. Westerfield*) *Institute of
Neuroscience, University of Oregon, Eugene, OR 97403 USA; **Loeb Institute of Medical
Research, Ottawa Civic Hospital, Ottawa, Ontario, CANADA; ***Department of Cell and
Developmental Biology, University of Pennsylvania, Philadelphia, PA 19104 USA</font></p>

<p><font face="Comic Sans MS">When are genes orthologues? To answer this question we first
need to agree on the definition of orthologues. From an evolutionary perspective,
orthologues are a pair of genes, one in each species, that are descended from a single
gene in the last common ancestor of those two species. Thus, orthologue assignments are
purely historical and do not involve function. Functions of genes may change during
evolution as lineal descendants acquire new functions or lose old functions.</font></p>

<p><font face="Comic Sans MS">This is like the science of taxonomy before cladistics. Do
you classify the lesser panda in a taxonomic group with raccoons because it serves a
raccoon-like function (acts like a raccoon) and has a raccoon-like expression pattern (it
is found in the same ecosystem)? Or, do you classify it as a bear because it shares a more
recent common ancestor with grizzlies than with raccoons?</font></p>

<p><font face="Comic Sans MS">The proper assignment of orthologous relationships is
especially important with fish because of the extra genome duplication event that occurred
in the ray finned fish lineage. Take for example, two of the <i>hedgehog</i> genes, <i>sonic
hedgehog </i>(<i>shh</i>) and <i>tiggy wiggle hedgehog </i>(<i>twhh</i>). Mapping and
phylogenetic analyses show that zebrafish <i>twhh</i> and human <i>SHH</i> are a pair of
genes descended from a single gene in the last common ancestor of human and zebrafish.
Thus, they are orthologues even though they do not have the same expression pattern.
Unfortunately, our nomenclature affects the way the research community thinks about the
relationships between genes. In some cases a novel name may obfuscate the evolutionary
relationship between the human and zebrafish genes, and hence make it difficult to relate
work in zebrafish to work in tetrapods.</font></p>

<p><font face="Comic Sans MS">Genetic drift after duplication, as proposed by the DDC
model of Force et al. (Genetics 151:1531-1545, 1999), will usually cause zebrafish gene
duplicates to have expression patterns that differ from their human orthologue. This is
especially true in cases in which the gene subsequently duplicated tandemly in the human
lineage and drift occurred independently in both lineages.</font></p>

<p><font face="Comic Sans MS">So, we propose that expression patterns should be used with
caution as criteria for assigning orthologies. Moreover, orthologues should be identified
on the basis of lines of descent whenever possible.</font></p>

<p><font face="Comic Sans MS">But how does one correctly infer the lines of descent?
First, all related genes in both species must be known, second, phylogenetic trees must be
constructed by unambiguous alignments and third, syntenic chromosomal relationships should
be known. </font></p>

<p><font face="Comic Sans MS">For example, if a gene is discovered in zebrafish whose true
orthologue has not yet been discovered in human, then the closest human match by sequence
analysis alone, may not be the orthologue. Similarly, in cases in which there has been
independent gene duplication in both lineages, a pair of genes in zebrafish will be
orthologous to a pair of genes in human and there will be no unique orthologue.</font></p>

<p><font face="Comic Sans MS">To construct meaningful phylogenetic trees, appropriate
outgroups must be available to root the trees. The best outgroup would be a species that
diverged from the vertebrate tree shortly before the divergence of the human and zebrafish
lineages. Sharks are probably a good outgroup, although there is some controversy about
that. Lamprey would be the next best, then amphioxus or a Urochordate. Often sequences are
not available from these species, so the phylogenetic tree cannot be properly rooted and
the analysis is, thus, provisional.</font></p>

<p><font face="Comic Sans MS">One problem with using trees to define orthologies is that
after duplication, the two members of the gene pair may have evolved at different rates.
This is especially possible when gene functions change after duplication and evolutionary
constraints on sequence may be less for one of the gene copies. The faster evolution of
one gene copy can result in long branch attraction causing that copy to appear to have
diverged earlier in the tree than is appropriate for its true history.</font></p>

<p><font face="Comic Sans MS">Conserved synteny is an important criterion for identifying
gene orthology independently of sequence relationships. If a zebrafish gene with sequence
similarity to a human gene maps in a chromosome segment that contains many other genes in
which syntenies are conserved, then the hypothesis of orthology is supported.</font></p>

<p><font face="Comic Sans MS">Even when using all these criteria, mistakes will be made.
In general, there are two types of mistakes. The first type of mistake occurs when we give
a zebrafish gene a name that is novel when in fact the human orthologue is known and thus
we hide the true gene relationships. The second type occurs when we give a zebrafish gene
a name that indicates orthology when none exists. An error of the first type makes the
zebrafish literature difficult to read and compare to other species. An error of the
second type implies relationships that do not really exist. Neither case is a total
disaster, of course, and each can be changed when more data become available. </font></p>

<p><font face="Comic Sans MS">Thus, the problem of assigning orthologies is difficult.
With current limitations of our techniques, we just have to do our best by constructing
the best possible phylogenetic tree, investigating syntenic relationships, and drawing
tentative conclusions that best fit our understanding of gene phylogeny as we know it
today.</font><font FACE="Times New Roman"></p>
</font>

<hr>
<a HREF="../vol6.1/vol6.1.html#December 1, 1999">

<p></a><a href="#August 14, 2000"><font face="Comic Sans MS">Return to Table of Contents</font></a><a HREF="../vol6.1/vol6.1.html#December 1, 1999"></p>
</a>

<hr>
<font FACE="Times New Roman"><b>

<h1><a name="WHATS IN A NAME">WHAT&#146;S IN A NAME?</a></b></font><font FACE="Times New Roman" SIZE="1"></h1>

<p></font><font size="2" face="Comic Sans MS">By the Zebrafish Nomenclature Committee (K.
Frazer<sup>1</sup>, M. Ekker<sup>2</sup>, M. Mullins<sup>3</sup>, J. Postlethwait<sup>1</sup>,
M. Westerfield<sup>1</sup>), <sup>1</sup>Zebrafish International Resource Center, 5274
University of Oregon, Eugene, OR 97403-5274 USA, <sup>2</sup>Loeb Health Research
Institute at The Ottawa Hospital, 725 Parkdale Ave., Ottawa, On, Canada, K1Y 4E9, <sup>3</sup>Department
of Cell and Developmental Biology, University of Pennsylvania, 1211 BRB II/III, 421 Curie
Blvd, Philadelphia, PA 19104-6058, USA</font></p>

<p><font face="Comic Sans MS">Finding an appropriate gene name and symbol to use for your
newly discovered gene can be a frustrating endeavor! There are so many factors to consider
like whether homologues or orthologues are known in other species, whether it is a member
of a gene family, and whether it has a duplicate. We hope the following tips will guide
you and make the experience of gene naming more pleasant.</font></p>
<b>

<p><font face="Comic Sans MS">How to propose a good gene name and symbol:</font></b></p>

<p><font face="Comic Sans MS">1) Become familiar with the zebrafish nomenclature
guidelines: <a href="/zf_info/nomen.html">http://zfin.org/zf_info/nomen.html</a></font>.</p>

<p><font face="Comic Sans MS">2) Search ZFIN (<a href="/">http://zfin.org/</a>)
to ensure that your gene hasn't already been named. A good way to do this is to search
ZFIN by GenBank or other sequence database accession number. It is important to identify
accession numbers of other sequences in GenBank that may correspond to your gene and then
use those numbers to search for genes in ZFIN.</font></p>

<p><font face="Comic Sans MS">You can identify other related sequences with a sequence
comparison algorithm like BLAST (<a href="http://www.ncbi.nlm.nih.gov/BLAST/">http://www.ncbi.nlm.nih.gov/BLAST/</a>).
Using the BLASTN algorithm to compare nucleotide sequences is preferred as a near exact
match of the amino acid sequence may identify the duplicate gene created by the genome
duplication or a subsequent tandemly duplicated gene.</font></p>

<p><font face="Comic Sans MS">Duplicate genes often have differences in the third
nucleotide or &quot;wobble&quot; position in a codon. Closely related sequences with
mismatches separated by stretches of 2, 5, or 8 identical nucleotides may indicate that
the genes are duplicates.</font></p>

<p><font face="Comic Sans MS">3) If you believe your gene is the genome duplicate of a
named zebrafish gene, propose the gene symbol comprised of the existing, named gene symbol
with a &quot;b&quot; attached to the end. If shown to be the duplicate, the existing gene
will be renamed by adding an &quot;a&quot; to the original symbol.</font></p>

<p><font face="Comic Sans MS">If you are reporting both duplicates and have mapped them to
duplicate chromosome segments, make an effort to keep the &quot;a&quot; and &quot;b&quot;
terminology consistent with other duplicates already mapped on the chromosome.</font></p>

<blockquote>
  <blockquote>
    <blockquote>
      <b><p><font face="Comic Sans MS">Example: </b>You identify the two zebrafish duplicates of
      the human gene <i>ZZZ</i>. One duplicate you map to LG3 and the other duplicate you map to
      LG12. Propose <i>zzza</i> for the gene on LG3 and <i>zzzb</i> for the gene on LG12 to
      conform to the hoxb&quot;a&quot; and hoxb&quot;b&quot; genes already mapped to these
      chromosomes.</font></p>
    </blockquote>
  </blockquote>
</blockquote>

<p><font face="Comic Sans MS">4) If the gene is new to zebrafish but has a human or mouse
orthologue, check the Human Nomenclature Database or the Mouse Genome Informatics
resources for the name and symbol used in those species:</font></p>
<div align="center"><center>

<table border="3" cellpadding="3" cellspacing="3" width="50%">
  <tr>
    <td width="100%"><font face="Comic Sans MS"><b>Human:</b> <a href="http://www.gene.ucl.ac.uk/nomenclature/">http://www.gene.ucl.ac.uk/nomenclature/</a></font>
    (Select the &quot;Nomenclature Database&quot; link to get to the search form)<b><p><font face="Comic Sans MS">Mouse:</b> <a href="http://www.informatics.jax.org/">http://www.informatics.jax.org/</a></font>
    (Select the &quot;Genetic and Phenotypic Data&quot; link, then the &quot;Genes, Markers
    and Phenotypes&quot; link to get to the search form) </td>
  </tr>
</table>
</center></div>

<p><font face="Comic Sans MS">Human and mouse gene names in these two resources are
approved by nomenclature committees.</font></p>

<p><font face="Comic Sans MS">5) If there is an established name and symbol in human, use
the same symbol for zebrafish (of course following zebrafish nomenclature guidelines like
all lower case italics, no &quot;z&quot; prefix, etc.). If there is no human gene, but
there is a mouse gene, use the same name and symbol used in mouse.</font></p>

<blockquote>
  <blockquote>
    <blockquote>
      <b><p><font face="Comic Sans MS">Example:</b> Suppose you isolate the zebrafish orthologue
      of the human &quot;<i>myelin associated glycoprotein</i>&quot;. Searching the Human
      Nomenclature Database returns the gene symbol <i>MAG</i>. Searching the Mouse Genome
      Informatics resources returns the gene symbol <i>Mag</i>. Therefore, an appropriate symbol
      adhering to the zebrafish nomenclature guidelines would be <i>mag</i>. </font></p>
    </blockquote>
  </blockquote>
</blockquote>

<p><font face="Comic Sans MS">This is only a starting point for proposing a good symbol.
You aren't done yet!</font></p>

<p><font face="Comic Sans MS">6) If the gene you are naming is a member of a gene family,
check to see what is used for the &quot;root&quot; symbol of the gene family in zebrafish.
If there is no established gene family in zebrafish but there is in human or mouse, use
the same &quot;root&quot; symbol established in human and mouse.</font></p>

<blockquote>
  <blockquote>
    <blockquote>
      <b><p><font face="Comic Sans MS">Example:</b> You isolate a new <i>distal-less homeobox</i>
      gene. Searching ZFIN using the gene query form, you find that all the known <i>distal-less
      homeobox</i> genes begin with &quot;<i>dlx</i>&quot;. The last number in the <i>dlx</i>
      series is <i>dlx8</i>. Therefore, a good starting point would be to propose <i>dlx9</i>. </font></p>
    </blockquote>
  </blockquote>
</blockquote>

<p><font face="Comic Sans MS">7) We've already seen how to propose a zebrafish gene symbol
in cases where it is a duplicate, an orthologue of a known human or mouse gene, or when it
is a member of a gene family. What if your gene is none of these? </font></p>

<blockquote>
  <blockquote>
    <blockquote>
      <b><p><font face="Comic Sans MS">Example:</b> You isolate a novel gene expressed in the
      retina that allows zebrafish to see in the dark. You want to call it &quot;<i>retina
      expressed, night vision</i>&quot;. <i>ren</i> would be a starting point.</font></p>
    </blockquote>
  </blockquote>
</blockquote>

<p><font face="Comic Sans MS">8) Check ZFIN to ensure that the symbol you want to use is
not already used for a different gene or mutant. Also, check the Human Nomenclature
Database and the Mouse Genome Informatics resources to ensure the combination is not
already in use in human or mouse for a different gene.</font></p>

<blockquote>
  <blockquote>
    <blockquote>
      <b><p><font face="Comic Sans MS">Example:</b> <i>myelin associated glycoprotein</i>, <i>mag</i>.
      You search ZFIN for a gene already named <i>mag</i> as well as for a mutant whose
      abbreviation is <i>mag</i>. A quick search shows that <i>mag</i> is not already used in
      zebrafish. We know <i>MAG</i> is <i>myelin associated glycoprotein</i> in human and mouse.
      The symbol <i>mag</i> is still OK.</font></p>
      <b><p><font face="Comic Sans MS">Example:</b> <i>distal-less homeobox 9</i>, <i>dlx9</i>.
      You know <i>dlx9</i> isn't already used for a zebrafish gene or mutant. However, when you
      search the Human Nomenclature Database, you find a <i>DLX9</i> gene in humans. If your
      zebrafish <i>dlx</i> gene is not the orthologue of the human <i>DLX9</i> gene, you need to
      propose a different symbol.</font></p>
      <b><p><font face="Comic Sans MS">Example:</b> <i>retina expressed, night vision</i>, <i>ren</i>.
      You find no zebrafish gene or mutant called <i>ren</i>. However, when you search the Human
      Nomenclature Database and the Mouse Genome Informatics resources, you find that <i>REN</i>
      is already used for <i>renin</i> genes. You need to come up with a different symbol for
      your night vision gene. </font></p>
    </blockquote>
  </blockquote>
</blockquote>

<p><font face="Comic Sans MS">9) Make sure the symbol you select does not interrupt an
existing gene family. The easiest way to check this is to enter a truncated form of your
symbol in a gene query form and search the database using the &quot;begins with&quot;
operator.</font></p>

<blockquote>
  <blockquote>
    <blockquote>
      <b><p><font face="Comic Sans MS">Example:</b> You isolate a heart expressed cytokeratin
      and you want to propose <i>hexc</i>. <i>hexc</i> is not used for zebrafish, human, or
      mouse genes. However, you search the Human Nomenclature Database and the Mouse Genome
      Informatics resources for &quot;<i>HEX</i>&quot; using the &quot;begins with&quot;
      operator. You find that in both human and mouse, there are <i>HEXA</i> and <i>HEXB</i>
      genes for <i>hexosaminidase A</i> and <i>hexosaminidase B</i>. <i>hexc</i> for your
      zebrafish gene is not a good symbol because if another human or mouse <i>hexosaminidase</i>
      gene is found, it will be assigned <i>HEXC</i>. </font></p>
    </blockquote>
  </blockquote>
</blockquote>

<p><font face="Comic Sans MS">10) Check other resources to ensure that the symbol you
select is not a very commonly used, unapproved symbol. Useful resources to check are:</font></p>

<blockquote>
  <blockquote>
    <blockquote>
      <b><p><font face="Comic Sans MS">LocusLink:</b> <a href="http://www.ncbi.nlm.nih.gov/LocusLink/index.html">http://www.ncbi.nlm.nih.gov/LocusLink/index.html</font>
      </a></p>
      <b><p><font face="Comic Sans MS">The Genome Database:</b> <a href="http://gdbwww.gdb.org/gdb/">http://gdbwww.gdb.org/gdb/</a></font></p>
      <b><p><font face="Comic Sans MS">OMIM:</b><a href="http://www.ncbi.nlm.nih.gov/Omim/">
      http://www.ncbi.nlm.nih.gov/Omim/</a></font></p>
      <b><p><font face="Comic Sans MS">PubMed:</b><a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?db=PubMed">
      http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?db=PubMed</a> </font></p>
    </blockquote>
  </blockquote>
</blockquote>

<p><font face="Comic Sans MS">11) When you have found a suitable symbol, send it to the
zebrafish nomenclature committee for confirmation and final approval (<font COLOR="#0000ff">nomenclature@zfin.org</font>). We will then reserve this symbol for you. </font></p>

<p><font face="Comic Sans MS">Although we typically have a quick turn around time, please
don't wait until the last minute to request approval of a gene name and symbol. Often a
seemingly simple request is more complicated than anticipated and may require
correspondence with other nomenclature committees or independent analyses on our part.</font></p>

<p><font face="Comic Sans MS">To help us provide quick turn around time for your requests,
the following information is greatly appreciated and will be held in strict confidence:</font>

<ul>
  <li><font face="Comic Sans MS">If possible, please provide a phylogenetic tree or
    description of the relationships between your zebrafish gene and orthologous or related
    genes in other species.</font></li>
  <li><font face="Comic Sans MS">If you have mapped the gene, provide the map position and on
    which mapping panel the gene is mapped. Preferably indicate its location with respect to
    the two nearest flanking &quot;z&quot; markers (MGH SSLPs
    [http://zebrafish.mgh.harvard.edu/]).</font></li>
  <li><font face="Comic Sans MS">If you believe the zebrafish gene is the orthologue of a gene
    in another species, provide the evidence (amino acid sequence comparison, nucleotide
    sequence comparison, conserved synteny, probe cross hybridization, etc.).</font></li>
  <li><font face="Comic Sans MS">Provide a sequence accession number if the sequence is public
    in an on-line sequence database. It is very helpful if you include the amino acid
    sequence, nucleotide sequence, or both with your request if they are not publicly
    available.</font></li>
  <li><font face="Comic Sans MS">Include any other information or commentary you view as
    pertinent in assigning proper gene nomenclature.</font></li>
</ul>

<p><font face="Comic Sans MS">If you need help with any of this analysis, contact us (<font COLOR="#0000ff">nomenclature@zfin.org</font>). </font></p>

<p><font face="Comic Sans MS">Happy Gene Naming!</font></p>

<hr>
<a HREF="../vol6.1/vol6.1.html#December 1, 1999">

<p></a><a href="#August 14, 2000"><font face="Comic Sans MS">Return to Table of Contents</font></a><a HREF="../vol6.1/vol6.1.html#December 1, 1999"></p>
</a>

<hr>
<b><font SIZE="1">

<h1></font><font size="6"><a name="ANATOMICAL NOMENCLATURE FOR ZEBRAFISH"><font face="Times New Roman">ANATOMICAL NOMENCLATURE FOR ZEBRAFISH</font></a></b></h1>

<p></font><font size="2" face="Comic Sans MS">By S.J. Moorman, Department of Anatomy,
School of Medicine, Case Western Reserve University, 10900 Euclid Ave., Cleveland, OH
44106-4930 (<a href="mailto:sjm8@po.cwru.edu">sjm8@po.cwru.edu</a>)</font><font SIZE="4"></p>

<p></font><font face="Comic Sans MS" size="3">The current popular use of the term
nomenclature is in reference to gene names. Historically, the term has been in use with
reference to anatomical terminology for so long that this reference is included in the
definition of the word nomenclature in most dictionaries. A good deal of effort is going
into making the nomenclature used for naming genes in the zebrafish consistent with that
used in other animals. Until recently, no similar effort has been applied to the zebrafish
anatomical nomenclature </font></p>

<p><font face="Comic Sans MS" SIZE="4">(see: </font><font face="Comic Sans MS"><a href="/zf_info/anatomy/dict/meeting_sum.html"><font size="3">/zf_info/anatomy/dict/meeting_sum.html</font></a>).
</font></p>

<p><font face="Comic Sans MS">Anatomical terminology is the foundation of medical
communication and it is important that medical and scientific communities throughout the
world use the same name for each structure. However, I don't advocate making all the
anatomical terminology for zebrafish consistent with that of domestic animals (see: Nomina
Anatomica Veterinaria (NAV)) or humans (see: Terminologia Anatomica (TA)). The terminology
should be consistent with that used for other fishes (with annotations indicating the
equivalent mammalian structures). The consistent use of correct anatomical terminology by
the community of zebrafish researchers can only foster communication. </font></p>

<p><font face="Comic Sans MS">Zebrafish development, from fertilization to maturity, can
be divided into phases with each phase subdivided into periods and each period further
subdivided into named stages. The periods and stages have been described in detail
(Kimmel, at al., 1995). However, the phases of development lack definition especially with
regard to anatomical landmarks that can be used to determine when to change phase names.
In this article, I would like to define the phases that occur during zebrafish development
and then define standard 'terms of direction' for the zebrafish.</font></p>

<p><font face="Comic Sans MS">The terms for the developmental phases should be
pre-embryonic, embryonic, larval, juvenile, and adult. The pre-embryonic phase begins at
fertilization and ends when there is a recognizable neural keel. This phase encompasses
the zygotic, blastula, and gastrula periods and ends at approximately 10 hours
post-fertilization. The pre-embryonic phase is followed by the embryonic phase of
development. The embryonic phase, encompassing the segmentation, pharyngula, and hatching
periods, ends at 72 hours with the appearance of the cleithrum, the first visible bone
(see Kimmel, et al.). These definitions of pre-embryonic and embryonic phases are
consistent with the definitions of the same phases in the domestic animals and human.
Defining a pre-embryonic phase introduces a problem. It is not unusual for zebrafish
researchers to say that they work on &#145;gastrula stage embryos&#146;. Embryologically,
it would be more accurate to say they work on the zebrafish gastrula. It is always better
to use the stage name alone. After the embryonic phase, the terminology diverges from that
of higher vertebrates because there are no equivalents of fetal and neonatal periods in
the zebrafish.</font></p>

<p><font face="Comic Sans MS">The embryonic phase is followed by the larval phase of
development. At the beginning of the larval phase, the pelvic fins have not begun to
develop and the gonads are not differentiated. All of the other organ systems have at
least begun to develop and are recognizable either using Nomarski optics or on
histological section. The gonads are recognizable and identifiable on histological section
as either ovaries or testes by 21 days post-fertilization. Development of the pelvic fins
presents a convenient landmark because, as external structures, their development is more
easily documented. The bones of the pelvic fin begin to ossify at approximately 30 days
post-fertilization (Grandel and Schulte-Merker 1998). This marks the end of the larval
phase and the beginning of the juvenile phase. The juvenile phase ends with sexual
maturity at 3-4 months post-fertilization. A sexually mature zebrafish should be called an
adult and this term should be used until the fish dies. </font></p>

<p><font face="Comic Sans MS">Some of the proposed 'terms of direction' for the zebrafish
might be more controversial than the definitions of the developmental phases outlined
above. In the adult zebrafish, the most useful terms of direction are the following:
rostral, caudal, dorsal, ventral, medial, lateral, transverse, superficial, and deep. If
we adopt the same definitions of these terms used in the domestic animals (NAV) and human
(TA), then I do not need to define them here. I am suggesting the term rostral be used
instead of the term cranial. Cranial means &quot;toward the head&quot;. Within the head,
this meaning is ambiguous at best. Rostral, meaning &quot;toward the nose&quot;, is more
accurate. If the term rostral were used instead of cranial, then the terminology would not
change based on region of the body. Notable in their absence, are the terms anterior and
posterior. These terms should not be used in the adult zebrafish because of the confusion
that arises from their meaning in the human. For similar reasons, they are not used in the
domestic animals (NAV). Anterior and posterior are still used when describing or naming
structures found in the head of humans. This is acceptable because the orientation of the
human head with respect to the trunk makes the terms dorsal and ventral ambiguous. This is
not true in the quadrupeds and certainly not true in the fishes. </font></p>

<p><font face="Comic Sans MS">In embryology, historical use of the terms anterior and
posterior has been controversial. In the human, these terms referred to different
directions during development and in the adult. Some time ago, the terms cranial and
caudal were adopted to replace anterior and posterior during human (Nomina Anatomica,
1989) and domestic animal (Nomina Embryologica Veterinaria) development. This made the
terminology consistent. The terminology in the zebrafish should also be consistent. Thus,
rostral and caudal should be used in both the adult and during development in the
zebrafish. Interestingly, textbooks are normally the last to incorporate new information
or changes in convention. All of the current editions of embryology textbooks have adopted
the terms cranial and caudal yet, the terms anterior and posterior are almost universally
used in the primary literature. </font></p>

<p><font face="Comic Sans MS">As scientists, we strive for accuracy. This should be
reflected in all aspects of our work including our manuscripts and scientific
presentations. We don&#146;t hesitate to change the name of a gene when it is cloned and
discovered to have already been identified and given a different name. We should be
willing to make the same concessions for the sake of accuracy in our anatomical
nomenclature. The consistent use of correct anatomical terminology by the community of
zebrafish researchers can only foster communication.</font></p>

<blockquote>
  <b><p><font face="Comic Sans MS">References:</font></p>
  </b><p><font face="Comic Sans MS">Grandel H, Schulte-Merker S (1998) The development of
  the paired fins in the zebrafish (<i>Danio rerio</i>) <i>Mech. Dev.</i> <b>79</b>:99-120</font></p>
  <p><font face="Comic Sans MS">Kimmel CB, Ballard WW, Kimmel SR, Ullmann B, Schilling TF
  (1995) Stages of embryonic development of the zebrafish. <i>Dev. Dyn.</i> <b>203</b>:253-310</font></p>
  <p><font face="Comic Sans MS">Nomina Anatomica Veterinaria, 4<sup>th</sup> Edition (1994)
  World Association of Veterinary Anatomists, Zurich and Ithaca, NY</font></p>
  <p><font face="Comic Sans MS">Nomina Embryologica Veterinaria, 1<sup>st</sup> Edition
  (1994) World Association of Veterinary Anatomists, Zurich and Ithaca, NY</font></p>
  <p><font face="Comic Sans MS">Terminologia Anatomica (1998) Federative Committee on
  Anatomical Terminology: Thieme, Stuttgart, Germany</font></p>
</blockquote>

<hr>
<a HREF="../vol6.1/vol6.1.html#December 1, 1999">

<p></a><a href="#August 14, 2000"><font face="Comic Sans MS">Return to Table of Contents</font></a><a HREF="../vol6.1/vol6.1.html#December 1, 1999"></p>
</a>

<hr>

<p><big><strong><font face="Times New Roman"><big><big><a name="EB 2001 MEETING">EB 2001
MEETING</a><big><br>
</big></big>March 31 through April 4, 2001</big></font></strong></big><font size="6"></p>
</font>

<p><font SIZE="2" face="Comic Sans MS">B</font><font face="Comic Sans MS">y </font><font face="Arial" size="2">Stephen J. Moorman, Ph.D., Case Western Reserve University, 1900
Euclid Ave., Cleveland, OH 44106-4930 (</font><small><a href="mailto:sjm8@cwru.edu">sjm8@cwru.edu</a>)</small></p>
<font SIZE="2">

<p>&nbsp;</p>

<p></font><font size="3"><font face="Comic Sans MS">EB 2001 (formerly FASEB) will be held
in Orlando, Florida from March 31 through April 4, 2001. The American Association of
Anatomists (AAA) is sponsoring a &quot;Zebrafish Meeting&quot; as part of EB 2001. Their
sponsorship includes 1 symposium, 4 mini-symposia (slide sessions), and a full day of
posters separate from the oral sessions. The people who have agreed to participate as
'chairs' along with the session titles are as follows:</font></p>

<p><font face="Comic Sans MS"><strong>Symposium Title:</strong></font></font><font face="Comic Sans MS" size="4"> </font><font face="Comic Sans MS" size="5"><strong>Patterning
During Development: Insights from Zebrafish</strong></font><font size="3"> 

<ul>
  <li><font face="Comic Sans MS"><strong>Christine Beattie</strong>, The Ohio State University<br>
    Mini Title: Early Development </font></li>
</ul>

<ul>
  <li><font face="Comic Sans MS"><strong>Mary Mullins</strong>, Ph.D., Department of Cell and
    Developmental Biology, University of Pennsylvania<br>
    Mini Title: Developmental Neurobiology</font></li>
</ul>

<ul>
  <li><font face="Comic Sans MS"><strong>Chi-Bin Chien</strong>, Ph.D., Department of
    Neurobiology and Anatomy, University of Utah Medical Center<br>
    Mini Title: Sensory Systems</font></li>
</ul>

<ul>
  <li><font face="Comic Sans MS"><strong>Stephen J. Moorman</strong>, Ph.D., Department of
    Anatomy, School of Medicine, Case Western Reserve University<br>
    Mini Title: Cell Signalling</font></li>
</ul>

<ul>
  <li><font face="Comic Sans MS"><strong>Bruce Riley</strong>, Ph.D., Department of Biology,
    Texas A&amp;M University</font></li>
</ul>

<p><font face="Comic Sans MS"><a href="mailto:sjm8@cwru.edu">Stephen Moorman</a> is
serving as the overall meeting organizer and contact person. The symposium consists of
invited speakers each receiving an honoraria for their participation. The costs of the
symposium and the meeting in general are being covered by the AAA. The mini-symposia
speakers will be selected by the chairs from submitted abstracts. We are also
investigating the possibility of getting the 'proceedings' of the meeting published in
Developmental Dynamics (an AAA journal).</font></font><font SIZE="2"></p>
</font>

<hr>

<p><a href="#August 14, 2000"><font face="Comic Sans MS">Return to Table of Contents</font></a></p>

<hr>
</body>
</html>
