<HTML>
<HEAD>
<META HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=iso-8859-1">
<META NAME="Generator" CONTENT="Microsoft Word 98">
<TITLE>A Quick Summary:  May 11,1999 meeting at NIH</TITLE>
</HEAD>
<body bgcolor="#FFFFFF" background="/zf_info/fish_bgd.gif">

<B><FONT FACE="Helvetica" SIZE=4>
<P ALIGN="CENTER">Summary of May 11,1999 organizational meeting at NIH</P>
<P ALIGN="CENTER">to discuss establishment of an anatomical dictionary for zebrafish</P>
</font></B><FONT FACE="Times,Times New Roman"><P ALIGN="CENTER">(prepared by Eck Doerry)</P>
</FONT><B><I><FONT FACE="Helvetica"><P>Overview</P>
</font></i></B><FONT FACE="Times,Times New Roman"><P>Development of a complete and uniform anatomical dictionary for the zebrafish is vital to the success of zebrafish science.  Specifically, the dictionary is necessary for:</P>

<UL>
<LI><U>Effective data dissemination and informatics.</u> Reference to anatomical structures is the central to all phenotypic description.  For example, in situ expression patterns are described in terms of the structures in which expression is observed; mutations are described in terms of the anatomical structure that are affected. In this way, anatomical structure is the common ground on which expressed genotype meets mutational phenotype, yielding powerful insight into gene function.  A concise and consistent anatomical atlas is necessary to support this correlation, particularly in an automated or computer-aided fashion.</LI>
<LI><U>A reference framework.</u>  By defining what anatomical structures exist, the dictionary establishes an ontological framework for a variety of reference resources, including a concise staging series (i.e., defined by development of structures, not time), an atlas of reference images indexed by anatomical structure and developmental age, and annotated 3-D reconstructions.</LI>
<LI><U>Interoperability</u>.  A stable, concise anatomical atlas is the key to leveraging work done in other species, particularly through semi-automated mapping between zebrafish data (stored in ZFIN) and data stored in other species&#146; databases (e.g. MGI,Flybase).  Given dictionaries for two species, mapping relationships between analogous structures (e.g. fins in fish=legs in flies) can be developed; searches based on mutations to zebrafish fins could yield genes expressed in drosophila legs. </LI></UL>

<P>&nbsp;</P>
<P>Although the need for an anatomical dictionary has been recognized for some time, development been slow.  Bill Trevarrow made a tremendous start in 1996, but progress stalled due to anatomical quandaries and lack of time to resolve them.  In addition, several pragmatic efforts have yielded rudimentary dictionaries to characterize specific sets of in situ or mutant data.   It is vital that these efforts be coordinated before they conflict and/or diverge too far;  untangling the chaos post hoc would be a nightmare.</P>

<P>To make a start at coordinating development of the dictionary, an informal breakfast meeting of interested parties was arranged on May 11,1999,  at the annual NIH-sponsored workshop on zebrafish genomic and genetic tools.  Although not all interested parties were able to attend, this provided a convenient, low-cost, low-commitment opportunity for introductions and quick overviews of projects, perspectives on the dictionary, and planning for future development.  </P>
</FONT><B><I><FONT FACE="Helvetica"><P>People and Projects</P>

</font></i></B><UL><FONT FACE="Helvetica"><I><B>
</b></i></font><FONT FACE="Times,Times New Roman"><LI>Matt Clark and Pia Aanstad (Berlin).  Project:  Characterize a large number of in situs of early (0-24) embryos, focusing around Shield stage. </LI>
<LI>Eck Doerry and Monte Westerfield (Eugene). Project:  Develop dictionary as infrastructure for zebrafish informatics; develop data model to support archiving and searching of in situ data within ZFIN.</LI>
<LI>Wolfgang Driever (Freiburg). Project:  Collaborating with Jonathan Bard and Fons Verbeek to develop a dictionary and, ultimately, a detailed anatomical atlas for zebrafish. </LI>
<LI>Jonathan Epstein (NIH).  Project:  Develop informatics support and software tools to support a number of projects producing in situ data.</LI>
<LI>Bernard and Christine Thisse (Strasbourg).  Project:  Characterize large in situ dataset ranging from 0-48 hour embryos.</LI>
  <LI>Reiko Toyama (NIH). Project: Characterize in situs generated from a normalized 
    cDNA library derived from tailbud to 10 somite stage embryos. Focus is on 
    expression in 0-24hr embryos.</LI>
<LI>Brant Weinstein (NIH).  Project: Interest is in vessel development. Has developed sophisticated 3-D reconstructions from confocal image that will need to be annotated to denote distinct structures.</LI></font></UL><FONT FACE="Times,Times New Roman">

<P>&nbsp;</P>
<P>Others who have expressed interest but could not attend:</P>


<UL>
<LI>Jonathan Bard (Edinburgh).  Jonathan was the driving force behind the development of the mouse anatomical dictionary.  He has recently begun doing the same for the zebrafish.  A draft of a zebrafish atlas Jonathan has developed was circulated at the meeting.</LI>
<LI>Fons Verbeek (Utrecht).  Fons is developing annotated 3-D reconstructions (from serial sections) of zebrafish embryos.  When fully annotated, it is expected that this effort will eventually yield a definitive 3-D atlas for the zebrafish.</LI></UL>

</FONT><B><I><FONT FACE="Helvetica"><P>Outcomes and Plans</P>
</font></i></B><FONT FACE="Times,Times New Roman">
<P>Predictably, everyone in attendance agreed that developing the anatomical dictionary 
  is a high priority. The need is so urgent, in fact, that we find ourselves in 
  the uncomfortable position of having to develop the dictionary, while at the 
  same time using it to characterize data being produced today. The following 
  points sketch out specific conclusions:</P>
<UL>
  <LI>There was overall agreement and commitment to work collaboratively on the 
    dictionary, making every effort to synchronize development. </LI>
  <LI>No one person or source of expertise adequate for developing the dictionary 
    exists. Thus, the dictionary should be developed in a distributed fashion, 
    drawing on expertise wherever it exists. For example, experts on CNS structure 
    might establish a subgroup to discuss and develop this aspect of the dictionary; 
    similar subgroups would form around other specialties. </LI>
  <LI>We will establish a central site as the clearinghouse for information related 
    to the dictionary; the latest &quot;version&quot; of the dictionary will always 
    be available here. Although rudimentary at the outset, this site might eventually 
    provide tools and resources to support the distributed, collaborative development 
    of the dictionary. In particular, the site will provide a mailing list for 
    the working group, and will archive the critical discussion on the list as 
    a way for preserving development rationale. ZFIN will host this site, at least 
    initially.</LI>
  <LI>We will pursue funding for regular meetings (at least annually) of the working 
    group and for 1-2 dedicated staff to organize the effort, solicit input from 
    experts where needed, and curate contributions. development of the dictionary. 
  </LI>
  <LI>The first meeting is scheduled for December 10, 1999. <a href="mailto:toyama@box-t.nih.gov">Reiko 
    Toyama</a> and <a href="mailto:monte@uoneuro.uoregon.edu">Monte Westerfield</a> 
    are co-organizers.</LI>
</UL>
</FONT>
<hr>
For questions or comments, contact: 
<address> Pat Edwards <br>
Institute of Neuroscience <br>
University of Oregon <br>
Eugene, OR 97403 USA <br>
fax: 541-346-4548 <br>
<a href="mailto:edwards@uoneuro.uoregon.edu">edwards@uoneuro.uoregon.edu</a> <br>
</address>
</BODY>
</HTML>
