<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html>
	<head>
		<title>
			ZFIN Zebrafish anatomical dictionary December 10,1999 meeting
		</title>
	</head>

<script language="JavaScript" src="/javascript/header.js"></script>


<h1 align="center"><FONT FACE="Helvetica" SIZE=4> Summary of the Zebrafish Anatomical
  Dictionary Workshop <br>
  December 10, 1999 </FONT></h1>
<h3>Overview</h3>
<p>The use of zebrafish as a model organism for genetic, genomic, developmental
  and other studies continues to grow. As a result, there is great need for the
  zebrafish research community to establish a standardized anatomical vocabulary.
  A dictionary defining the vocabulary and anatomical atlases illustrating the
  vocabulary will allow consistent use of terminology and will foster cross-species
  comparisons of gene expression patterns, mutant phenotypes and other studies
  of gene function. </p>
<p>An initial informal gathering to formulate plans for generating an anatomical
  dictionary for zebrafish was held during the workshop on Genomic and Genetic
  Tools for the Zebrafish at the NIH on May 10-11,1999. People attending that
  meeting agreed on the high priority of developing an anatomical dictionary.
  The<a href="meeting_sum.html"> report of this meeting</a> is posted on ZFIN.</p>
<p> To move forward toward the goal of establishing the anatomical dictionary,
  a second meeting was organized by M. Westerfield and R. Toyama at the Natcher
  Conference Center at NIH on December 10, 1999. The purpose of this meeting was
  to establish a more detailed plan with specific goals for the dictionary and,
  ultimately, the anatomical atlases. </p>
<h3>Presentations: </h3>
<p><b>Overview:</b> Westerfield gave an overview of the zebrafish anatomical database
  supported by ZFIN. Currently, some of the tools required for the anatomical
  dictionary and atlas are already provided by ZFIN. For example, detailed <a href="../../zfbook/stages/index.html">descriptions
  of developmental stages</a> and a list of anatomical terms at different stages
  is available on line and can be modified by workgroup members. The original
  data and new data are archived; thus, each revision can be reviewed as necessary.</p>
<p> <b>cDNA Expression Patterns: </b>Three groups (Aanstad, Thisse, and Toyama)
  who are working on large scale in situ cDNA screening projects presented their
  data organization systems and emphasized the necessity of standardized anatomical
  terms to annotate expression domains. To search cDNA expression databases on
  the basis of gene expression patterns, key word descriptions of expression domains
  are required because it is currently not possible to search by photographic
  images. Therefore, a standardized anatomical vocabulary is essential. This vocabulary
  should be organized in a hierarchical manner to indicate relationships among
  tissue structures and anatomical and functional systems. Several points were
  made:</p>
<ul>
  <li>Some mRNAs may be expressed in a subset of cells that have no corresponding
    anatomical term. In these cases, identifying and anotating detailed morphological
    structures by histology will add precision to the definition of the gene expression
    patterns. </li>
  <br>
  <br>
  <li>Gene expression patterns may not respect morphological tissue boundaries.
    Particularly in early embryogenesis, some gene expression patterns can be
    expected to define areas before histological or anatomical structures are
    visible. These areas may be evident by gene expression patterns alone and,
    thus, will require precise topological vocabularies. Gene expression domains
    can be used as labels/landmarks to name subsets of cells in the embryo prior
    to the formation of visible structures. </li>
  <br>
  <br>
  <li>Based on his experience from screening and analyzing axonal patterning mutants,
    Wilson argued strongly for the creation and organization of a high quality,
    well-labeled neurohistological and anatomical atlas. Specifically, he suggested
    that a listing of well-characterized gene markers of brain structures is a
    required resource for the research community. </li>
</ul>
<p><b>Three-dimensional (3D) atlas of the zebrafish</b>: </p>
<p>1) Verbeek presented a 3D-digital zebrafish atlas reconstructed from confocal
  microscopic and serial section images. He described a computer automated technique
  to support this project, although some steps still require manual operation.
  Incorporating gene expression data into the 3D atlas is technically still challenging
  because current whole-mount in situ techniques are unsuitable for acquiring
  images by confocal microscopy. The current system uses 20x magnification; it
  is unclear whether this will provide sufficiently high-resolution for an atlas.
</p>
<p>2) Weinstein presented a WEB accessible database of a 3D atlas of blood vessel
  formation during zebrafish embryogenesis. Individual vessels are named based
  on comparisons with other model systems (human and other fish) Vasculature is
  quite variable between individuals possibly because of polymorphisms among different
  strains. Thus, annotations need to be assigned with these issues in mind. This
  work has been published in Dev Biol. 2001; 230:278-301 and is currently <a href=http://mgchd1.nichd.nih.gov:8000/zfatlas/Intro%20Page/intro1.html>
  available on line</a>. </p>
<p>3) Cheng described his agarose-embedded tissue array system to handle large
  numbers of zebrafish larvae simultaneously for histological and genetic analysis.
  His group has accumulated cytological data of adult and larval zebrafish. Based
  on his experience in anatomic pathology, he proposed, and Bard reiterated, that
  the details evident from histological paraffin and plastic sections should allow
  identification of anatomical parts during embryonic and larval development that
  correlate with gene expression patterns. Moorman pointed out that these sections
  could be used to generate 3D images.</p>
<p>4) Bard discussed his experience in establishing the mouse anatomical dictionary
  and current efforts to generate a zebrafish dictionary. The group agreed that
  the zebrafish community should model their database after the best features
  of other model organism databases, including Drosophila, C. elegans, and mouse,
  and should incorporate links, where appropriate, among anatomical data from
  different model systems. </p>
<h3><b>Conclusions: </b></h3>
<p>1. The group agreed unanimously that a zebrafish anatomical dictionary and
  atlas need to be established with high priority. Some of the specific goals
  are already under way. </p>
<p>2. Our ultimate goal is to generate 3D atlases with histological and gene expression
  data incorporated in a searchable fashion (virtual zebrafish). This will require
  development of a standardized anatomical dictionary and atlases that define
  structures and systems. However, we are aware of the current, urgent need for
  a dictionary in the research community. Therefore, it ismost important to elucidate
  the community's needs and to provide information in a timely manner. </p>
<p>3. The dictionary of anatomical terms needs to be flexible and to support future
  modifications. </p>
<p>4. To collect sufficient data to start organizing the dictionary and atlases,
  we agreed to ask the zebrafish community to contribute to this project. It was
  suggested that we contact people who have expertise in particular tissues and
  organs and ask them to provide their best histological and gene expression (mRNA
  in situ or antibody labeling) data (slides, prints, electronic files, etc.)
  with precise annotations. We will provide a standardized form to be completed
  by these individuals to facilitate entry of the data into ZFIN. We hope to gather
  the first round of data within a reasonably short period and will consider obtaining
  further data if necessary (e.g. additional developmental stages). </p>
<p>5. Finally, data should be reviewed by outside experts to maintain quality
  and accuracy. </p>

<h3>Tentative roles for meeting participants were assigned as follows:<br>
</h3>
<ul>
  <li>3D embryonic atlas: Verbeek<br>
  </li>
  <li>Embryo, Larval, and adult histology from hematoxylin and eosin-stained paraffin
    sections: Cheng </li>
  <br>
  <li>3D Atlas of larva and adult: Moorman from Cheng sections </li>
  <br>
  <li>Tumor Atlas: Cheng </li>
  <br>
  <li>Organization of dictionary: Westerfield and Bard </li>
</ul>
<h3><a href="mem.html">Members of research community who have agreed to contribute
  to the dictionary</a></h3>
<h3> Meeting participants:</h3>
<ul>
  <li> <a href="/<!--|CGI_BIN_DIR_NAME|-->/ZFIN_jump?record=ZDB-PERS-960805-12">Pia Aanstad</a>,
    Abt. Lehrach, Max Planck Institut f&uuml;r Molekulare Genetik, Ihnestr. 73, 14195
    Berlin (Dahlem), GERMANY </li>
  <li><a href="/<!--|CGI_BIN_DIR_NAME|-->/ZFIN_jump?record=ZDB-PERS-970411-7">Jonathan Bard</a>,
    Anatomy Department, Edinburgh University, Edinburgh EH8 9AG, UNITED KINGDOM </li>
  <li>Jack Chen, Building 10 Room 9D52, NICHD/NIH, Bethesda, MD 20892
  </li>
  <li><a href="/<!--|CGI_BIN_DIR_NAME|-->/ZFIN_jump?record=ZDB-PERS-960805-673">Keith Cheng</a>,
    The Jake Gittlen Cancer Research Institute H059, Penn State College of Medicine,
    500 University Drive, HO59 Hershey, PA 17033</li>
  <li> <a href="/<!--|CGI_BIN_DIR_NAME|-->/ZFIN_jump?record=ZDB-PERS-960805-134">Igor Dawid</a>, Laboratory of
    Molecular Genetics, National Institutes of Health, NICHHD LMG, Bldg 6B, Rm
    413, 9000 Rockville Pike, Bethesda, MD 20892 </li>
  <li><a href="/<!--|CGI_BIN_DIR_NAME|-->/ZFIN_jump?record=ZDB-PERS-980615-1">Jonathan Epstein</a>,
    Unit on Biological Computation, Office of Scientific Director, Nat. Institute
    of Child Health and Human Development, National Institutes of Health, Bethesda,
    MD 20892</li>
  <li><a href="/<!--|CGI_BIN_DIR_NAME|-->/ZFIN_jump?record=ZDB-PERS-960805-640">Chuck Kimmel</a>,
    Institute of Neuroscience, 1254 University of Oregon, Eugene, OR 97403-1254 </li>
  <li><a href="/<!--|CGI_BIN_DIR_NAME|-->/ZFIN_jump?record=ZDB-PERS-961106-1">Stephen Moorman</a>,
    Department of Anatomy, School of Medicin,e Case Western Reserve University,
    10900 Euclid Ave., Cleveland, OH 44106-4930</li>
  <li><a href="/<!--|CGI_BIN_DIR_NAME|-->/ZFIN_jump?record=ZDB-PERS-960805-556">Bernard Thisse</a>,
    Institut de Genetique et de Biologie Moleculaire et Cellulaire, CNRS INSERM
    ULP, BP163 67404 Illkirch Cedex, C.U. de Strasbourg, FRANCE
  </li>
  <li><a href="/<!--|CGI_BIN_DIR_NAME|-->/ZFIN_jump?record=ZDB-PERS-960805-563">Reiko Toyama</a>,
    NICHD, LMG Bldg. 6B, Room 420, Bethesda, MD 20892</li>
  <li><a href="/<!--|CGI_BIN_DIR_NAME|-->/ZFIN_jump?record=ZDB-PERS-970422-1">Fons Verbeek</a>,
    NIOB Hubrecht Laboratory, Uppsalalaan 8, 3584 CT Utrecht, THE NETHERLANDS</li>
  <li><a href="/<!--|CGI_BIN_DIR_NAME|-->/ZFIN_jump?record=ZDB-PERS-960805-604">Brant Weinstein</a>,
    Unit on Vertebrate Organogenesis, Laboratory of Molecular Genetics, NICHD,
    NIH Building 6B, Room 309 Bethesda, MD 20892</li>
  <li><a href="/<!--|CGI_BIN_DIR_NAME|-->/ZFIN_jump?record=ZDB-PERS-960805-676">Monte Westerfield</a>,
    Institute of Neuroscience, 1254 University of Oregon, Eugene, OR 97403-1254</li>
  <li><a href="/<!--|CGI_BIN_DIR_NAME|-->/ZFIN_jump?record=ZDB-PERS-960805-612">Steve Wilson</a>,
    Department of Anatomy and Developmental Biology, University College London,
    Gower Street, London WC1E 6BT, UNITED KINGDOM</li>
</ul>

</BODY>

<script language="JavaScript" src="/javascript/footer.js"></script>
