<HTML>
<!--This file created 11/2/98 9:49 PM by Claris Home Page version 3.0 60 Day Trial-->
<HEAD>
   <TITLE>Zebrafish Informatics</TITLE>
   <META NAME=GENERATOR CONTENT="Claris Home Page 3.0 60 Day Trial">
   <X-CLARIS-WINDOW TOP=68 BOTTOM=768 LEFT=19 RIGHT=549>
   <X-CLARIS-TAGVIEW MODE=minimal>
</HEAD>
<BODY BGCOLOR="#FFFFFF" VLINK="#551A8B">
<P>Copy of original proposal to the NIH<BR wp=br1>
</P>

<H2><CENTER><STRONG>Funded Proposal for Meeting on
</STRONG></CENTER></H2>

<H1><CENTER><STRONG>Zebrafish Informatics</STRONG></CENTER></H1>

<CENTER>(September 12, 1998 in Eugene, OR)</CENTER>

<P><A HREF="mtgsum.html"><B>View meeting summary</B></A></P>

<H3>Background:</H3>

<P>The last several years have witnessed an explosion in our
understanding of vertebrate development, largely based on work from a
few model genetic organisms. The zebrafish is the newest of these
model organisms. Because the basic principles of body patterning
appear similar during embryogenesis of all vertebrates, insights
gained from work on embryonic zebrafish will have implications for
human health and disease. Moreover, research on this organism meets
the intent of the Animal Welfare Act because use of many higher
vertebrates can now be replaced by use of this lower vertebrate.<BR wp=br1>
<BR wp=br2>
Systematic genetic research on zebrafish has led to the recent
identification of over 7,000 genetic mutations. For a thorough
understanding of the developmental processes disrupted by zebrafish
mutations, we must study not only the mutant phenotype, but also
learn the structure of the mutated gene's protein product, its normal
function during development, and how the gene's expression is
regulated. To achieve this, we must ascertain the gene's sequence and
make molecular probes for the gene and its transcripts; i.e., we must
clone genes identified by mutations. Geneticists have developed two
main strategies for cloning mutated genes, the candidate gene
approach and the positional cloning approach. <U>Both approaches
require an extensive genetic linkage map.</U><BR wp=br1>
<BR wp=br2>
A genetic map has been established for zebrafish; each chromosome is
represented by a single linkage group (Postlethwait et al., 1994;
Johnson et al., 1996; Knapik et al., 1996). Among vertebrates, only
human (Murray et al., 1994), mouse (Dietrich et al., 1994), rat
(Yamada et al., 1994; Jacob et al., 1995) and zebrafish have closed
linkage maps. The significance of this achievement is that any newly
discovered mutation can be placed at a unique location on the map
near previously localized DNA polymorphisms. These polymorphisms can
be used to generate DNA probes for cloning the gene responsible for
the mutant phenotype.<BR wp=br1>
<BR wp=br2>
Two methods have been developed to date to map zebrafish mutations.
One of the methods utilizes DNA pooling protocols to identify DNA
polymorphisms closely linked to mutations in haploid embryos
(Postlethwait et al., 1994). The other method exploits the high
amount of chromosomal interference in zebrafish meiosis to map
embryonic lethal or adult visible mutations by tetrad analysis
(Johnson et al., 1995, 1996). The current map contains approximately
1800 markers. Because the linkage map is about 3000 cM long, and the
genome is about 1.7 x 10<FONT SIZE="-1"><SUP>9</SUP> </FONT>base
pairs, the average inter-marker interval is about 1.7 cM or 0.9 kb.
The NIH has recently released an RFA (DK-98-006) for proposals to
increase the resolution of the map to 0.3 cM or better. <U>To reach
the goal of using the genetic map for efficient cloning of mutated
genes, the density of markers needs to be significantly
increased.</U><BR wp=br1>
<BR wp=br2>
<BR wp=br1>
<FONT SIZE="+1"><STRONG>The Problem:</STRONG></FONT></P>

<P>To achieve the proposed increase in map density, a large number of
markers need to be mapped. This will be accomplished by complementary
efforts carried out in several laboratories. This generates a problem
for the research community because map information is spread among
these laboratories.<BR wp=br1>
<BR wp=br2>
This problem was recognized and discussed at a meeting at Cold Spring
Harbor earlier this month where, for the first time, the PIs from
essentially all the main laboratories that are contributing to the
genetic map met and discussed current and upcoming needs for
informatics support. Everyone agreed that <U>a major goal is to
provide information on markers and the map in an integrated form with
a uniform interface</U>, rather than having various types of map data
spread around the world in different laboratories in different
databases on different servers. The research community needs an
integrated genetic map, linked to the physical map, that combines
data from all mapping experiments. The mapping laboratories decided
that the ZFIN database being developed here in Oregon is the logical
choice for the "centralized" site.<BR wp=br1>
<BR wp=br2>
The RFA to generate new markers states that integration (of new data
generated in response to the RFA) with the current bioinformatics
zebrafish database at the University of Oregon should be
facilitated." We currently hold a grant to establish a "National
Resource for Zebrafish" (P40 RR12546) which includes salary support
for 1 informatics person. The very significantly larger amount of
data and the faster time-line for acquisition of data generated in
response to the RFA were not anticipated at the time we submitted the
Resource application. We currently lack the staff and resources to
process the anticipated data, making coordinated planning and
development of efficient automated procedures for integrating new
data even more vital. Because several of the laboratories have
already started to generate new markers, the data are accumulating
and will get out of hand if we fail to act in the very near
future.<BR wp=br1>
<BR wp=br2>
Although we've made good progress with ZFIN to date, a number of
conceptual and technical problems still remain to be solved. For
example, it is unclear how mutations and cloned genes should be
integrated into the microsatellite map and how markers mapped on
different mapping panels can be integrated. Similarly, representing
information from the physical map in the data model and visually
relating it to the genetic map represents a significant challenge. On
a more pragmatic level, we need to agree on uniform data formats,
data submission procedures, and mechanisms for updating information
once it has been submitted to ensure smooth transfer of information
into the integrated database.<BR wp=br1>
<BR wp=br2>
In summary, <U>to provide adequate informatics support for the
genetic map, we need to develop a plan in the very near future to
integrate data from the mapping laboratories into ZFIN</U>.<BR wp=br1>
</P>

<P><BR wp=br2>
<FONT SIZE="+1"><STRONG>Proposed solution:</STRONG></FONT></P>

<P>These problems are primarily organizational; if we can agree on
what data are important and how to get them into ZFIN, the creation
of an integrated map should be fairly straight forward. The most
effective way to create consensus and to develop a plan of attack is
to bring together key researchers from each of the laboratories that
are contributing to the mapping project to discuss the problems
associated with coordinating informatics support. Both biologists
producing the mapping data and informatics personnel should be
represented. It is important that representatives of the human and
mouse genome projects also attend to provide input on methods used
for those maps and to work towards the long term goal of linking map
information between species for studies of functional genomics,
syntenic conservation, and evolution.<BR wp=br1>
We propose that a committee, representing the mapping labs, meet to
develop solutions to the following problems:</P>

<P>1. Is the current ZFIN data model adequate for representing all
relevant data from existing and proposed mapping experiments?</P>

<P>2. How can data from different mapping panels be integrated
seamlessly?</P>

<P>3. How should ZFIN be modified to integrate data related to the
physical map (e.g. radiation hybrids)?</P>

<P>4. What common tools and data collection formats should be adopted
by the mapping labs to streamline submission of data to ZFIN?</P>

<P>5. How quickly should newly generated data by transferred to ZFIN?
When should it become publicly visible?</P>

<P>6. How can ZFIN be extended to support cross-species genomic
analysis, i.e., inter-operability with other species' databases?<BR wp=br1>
<BR wp=br2>
After the meeting we will draft a report describing the decisions and
recommendations of the committee. This report will be submitted to
you. We will also distribute the report to the mapping laboratories
and we will post the report on the zebrafish WWW server.<BR wp=br1>
<BR wp=br2>
<FONT SIZE="+1"><STRONG>References:</STRONG></FONT></P>

<OL>
   <LI>Dietrich, W.F., J.C. Miller, R.G. Steen, M. Merchant, D.
   Damron et al. (1994) A genetic map of the mouse with 4,006 simple
   sequence length polymorphisms. <EM>Nat. Genet</EM>.
   <STRONG>7</STRONG>:220-245.</LI>
   
   <LI>Jacob, H., D.M. Brown, R.K. Bunker, M.J. Daly, V.J. Dzau et
   al. (1995) A genetic linkage map of the laboratory rat, <EM>Rattus
   novegicus</EM>. <EM>Nat. Genet</EM>.
   <STRONG>9</STRONG>:63-69.</LI>
   
   <LI>Johnson, S.L., D. Africa, S. Horne, and J.H. Postlethwait
   (1995) Half-tetrad analysis in zebrafish: mapping the ros mutation
   and the centromere of linkage group I. <EM>Genetics</EM>
   <STRONG>139</STRONG>:1727-1737.</LI>
   
   <LI>Johnson, S.L., M.A. Gates, M. Johnson, W.S. Talbot, S. Horne,
   K. Baik, S. Rude, J.R. Wong, and J.H. Postlethwait (1996)
   Centromere-linkage analysis and consolidation of the zebrafish
   genetic map. <EM>Genetics</EM>
   <STRONG>142</STRONG>:1277-1288.</LI>
   
   <LI>Knapik, E.W., A. Goodman, O.S. Atkinson, C.T. Roberts, M.
   Shiozawa, C.U. Sim, S. Weksler-Zangen, M.R. Trolliet, C. Futrell,
   B.A. Innes, G. Koike, M. McLaughlin, L. Pierre, J.S. Simon, E.
   Vilallonga, M. Roy, P.-W. Chiang, M.C. Fishman, W. Driever, and
   H.J. Jacob (1996) A reference cross DNA panel for zebrafish
   (<EM>Danio rerio</EM>) anchored with simple sequence length
   polymorphisms<EM>. Development</EM>
   <STRONG>123</STRONG>:451-460.</LI>
   
   <LI>Murray, J.C., K.H. Buetow, J.L. Weber, S. Ludwigsen, T.
   Scherpbier-Heddema et al., (1994) A comprehensive human linkage
   map with centimorgan density. <EM>Science</EM>
   <STRONG>265</STRONG>:2049-2054.</LI>
   
   <LI>Postlethwait, J., S. Johnson, C.N. Midson, W.S. Talbot, M.
   Gates, E.W. Ballenger, D. Africa, R. Andrews, T. Carl, J.S. Eisen,
   S. Horne, C.B. Kimmel, M. Hutchinson, M. Johnson, and A. Rodriguez
   (1994) A genetic map for the zebrafish. <EM>Science
   </EM><STRONG>264</STRONG>:699-703.</LI>
   
   <LI>Yamada, J., T. Kuramoto and T. Serikawa (1994) A rat genetic
   linkage map and comparative maps for mouse or human homogous rat
   genes. <EM>Mamm. Genome</EM> <STRONG>5</STRONG>:63-83.</LI>
</OL>

<P><A HREF="mtgsum.html"><B>View meeting summary</B></A></P>

<P>

<HR>

For questions or comments, contact:</P>

<ADDRESS>&nbsp;</ADDRESS>

<ADDRESS>Pat Edwards<BR>
Institute of Neuroscience<BR>
University of Oregon<BR>
Eugene, OR 97403 USA<BR>
fax: 541-346-4548<BR>
<A HREF="mailto:edwards@uoneuro.uoregon.edu">
edwards@uoneuro.uoregon.edu</A></ADDRESS>

<P>

<HR>

<A HREF="/index.html">ZFIN <IMG SRC="fish_net.gif" X-CLARIS-USEIMAGEWIDTH X-CLARIS-USEIMAGEHEIGHT ALIGN=bottom></A></P>
</BODY>
</HTML>
