<HTML>
<!--This file created 11/2/98 9:49 PM by Claris Home Page version 3.0 60 Day Trial-->
<HEAD>
   <TITLE>Zebrafish Mapping Bioinformatics</TITLE>
   <META NAME=GENERATOR CONTENT="Claris Home Page 3.0 60 Day Trial">
   <X-CLARIS-WINDOW TOP=42 BOTTOM=757 LEFT=4 RIGHT=534>
   <X-CLARIS-TAGVIEW MODE=minimal>
</HEAD>
<BODY BGCOLOR="#FFFFFF" VLINK="#551A8B">
<H1><CENTER><STRONG>Summary of Meeting on Zebrafish Mapping
Bioinformatics</STRONG></CENTER></H1>

<H3><CENTER><STRONG>September 12, 1998</STRONG></CENTER></H3>

<H3><CENTER><STRONG>University of Oregon, Eugene,
OR</STRONG></CENTER></H3>

<P>Representatives of zebrafish laboratories, human and mouse
bioinformatics resources, and the NIH met to discuss the current
status of genetic and physical mapping of the zebrafish genome and
bioinformatics support for the resulting maps. The purposes of the
meeting were 1) to assess the current state of mapping resources, 2)
to summarize future plans for mapping more markers, genes and
mutations, 3) to assess the adequacy of ZFIN as a centralized
database for mapping information, and 4) to develop a plan for
integrating future data from mapping laboratories into ZFIN.</P>

<P><A HREF="mtgprop.html"><STRONG>Background and purpose of the
meeting</STRONG></A><BR wp=br1>
</P>

<H3>The following resources are currently being developed:</H3>

<H2>Maps:</H2>

<UL>
   <LI>Reference crosses (4)</LI>
   
   <LI>Radiation hybrids (2)</LI>
   
   <LI>Mutant mapping crosses (many)</LI>
</UL>

<H2>Mapped markers:</H2>

<UL>
   <LI>Anonymous (SSLP, RAPD, AFLP; ~10,000)</LI>
   
   <LI>ESTs (~12,000)</LI>
   
   <LI>Cloned Genes (~2,000)</LI>
   
   <LI>Mutations (~1,000)</LI>
</UL>

<H2>Databases:</H2>

<UL>
   <LI>ZFIN integrated developmental, genetic, and map data
   resource</LI>
   
   <LI>Individual mapping laboratory data servers (~6)</LI>
   
</UL>

<H3>The following issues were discussed:</H3>

<P><STRONG>1. How should ZFIN be modified to integrate data related
to the physical map?</STRONG></P>

<P>ZFIN should provide links from gene, EST, and other marker records
to the PAC clones that contain them in the RZPD (Berlin database)
arrayed library. Similar links should be added to YAC, BAC, and
cosmid clones as they become available.<BR wp=br1>
<BR wp=br2>
</P>

<P><STRONG>2. How can data from different mapping panels be
integrated seamlessly? Should this be done automatically or do we
need a chromosome committee?</STRONG></P>

<P>For each map marker, complete matrices of the original raw data
from all mapping experiments should be integrated into ZFIN and made
available in a format compatible with most common mapping programs.
This will allow investigators to verify marker locations given in
ZFIN by generating their own maps from the original data.</P>

<P>ZFIN should provide a mechanism for easy and direct comparisons of
different maps.</P>

<P>ZFIN should calculate 'best guess' maps for each reference cross
and RH panel dynamically using the currently available data or
subsets of the data selected by users. The most recently computed map
locations should be added to each marker record.</P>

<P>It is very important that ZFIN provides a mechanism to link the RH
maps with the meiotic maps in a manner that allows easy
comparisons.</P>

<P>Fully-automated computer generation of a single, definitive,
integrated map is not practical due to ambiguities and conflicts
inherent in the data from different mapping experiments. Producing
such a map would require extensive curation by a chromosome
committee.<BR wp=br1>
<BR wp=br2>
</P>

<P><STRONG>3. What common tools and data collection formats should be
adopted by the mapping labs to streamline submission of data to
ZFIN?</STRONG></P>

<P>ZFIN is not a suitable resource for a laboratory database system.
Thus, mapping laboratories need to develop their own in-house
databases for organizing and curating locally-generated map data
prior to submission to ZFIN.</P>

<P>The various laboratories that are developing RH maps should
coordinate their local software resources to the greatest extent
possible. It would be especially fruitful if a common program (e.g.
RHMapper) could be adopted by all groups. This will facilitate data
sharing among the mapping laboratories. It will also ease the work
involved in developing scripts for exporting data to ZFIN. Similarly,
laboratories contributing data to the meiotic maps should all adopt
the same software. MapManager seems to be the program of choice.
MapManager should be modified to provide a text output that can be
read by ZFIN.</P>

<P>RH mapping labs should provide regular updates to the centralized
maps in ZFIN.</P>

<P>People outside RH mapping laboratories who map new markers onto an
RH panel should submit the newly mapped markers to ZFIN including the
scoring data. ZFIN will calculate and return a mapping report with
the positions of the new markers. ZFIN will also send a report to the
appropriate RH mapping lab(s). The new markers will be held in ZFIN
as provisional until the RH mapping lab(s) review and approve
them.<BR wp=br1>
<BR wp=br2>
</P>

<P><STRONG>4. How quickly should newly generated data by transferred
to ZFIN? When should it become publicly visible?</STRONG></P>

<P>In general, map data should be submitted to ZFIN and made public
as quickly as it is generated.</P>

<P>In some cases, (i.e. mapped genes and mutations that have not yet
been published), a 6 month waiting period after submission to ZFIN,
but before journal publication should be provided. During this
waiting period, information on the marker submitted to ZFIN can be
kept private.<FONT COLOR="#FF0000"> </FONT> This will allow
researchers to register and reserve mutant and gene names before
publication.</P>

<P>Investigators should be allowed to make the map locations of
mapped, but unpublished genes and mutations available publicly under
a pseudoname during the 6 month waiting period. This will allow
researchers to make contacts and establish collaborations with each
other while protecting their data before publication.</P>

<P>Journals should be encouraged to require ZFIN accession numbers
for genes, mutations and map markers by the time of publication. This
will provide incentives for researchers to submit their data to the
common database.<BR wp=br1>
<BR wp=br2>
</P>

<P><STRONG>5. How can ZFIN be extended to support cross-species
genomic analysis, i.e., inter-operability with other species'
databases?</STRONG></P>

<P>Wherever possible, ZFIN should provide links to records of
orthologous genes in other species databases.</P>

<P>ZFIN should avoid storing sequence information internally and,
instead, should provide links to the external databases (i.e.
GenBank, SWISPROT, etc.) that are responsible for curation of
sequences.</P>

<P>ZFIN should provide a mechanism for displaying and comparing
syntenic relationships between zebrafish and other vertebrates. Links
to MGD (Jackson Laboratory mouse database) which provides mouse/human
syntenies should be provided.<BR wp=br1>
<BR wp=br2>
</P>

<P><STRONG>6. Is the current ZFIN data model adequate for
representing all relevant data from existing and proposed mapping
experiments? </STRONG></P>

<P>Refinement of the ZFIN data model should be an on-going process as
new types of map information are developed and as new laboratories
begin to contribute map data. Detailed discussion of the various
types of maps and map markers that ZFIN should represent led to
recommendations for a number of changes:</P>

<P>Class <STRONG>CLONED_GENE</STRONG>. Add additional information
about the clone, in particular, the name of the plasmid (and its
source) used for each mapping experiment.</P>

<P>Class <STRONG>EST</STRONG>. The name of the EST should generally
be the GenBank accession number of the EST sequence. In some cases,
there will be multiple ESTs obtained for the same sequence some of
which may or may not be mapped. It makes sense to store only the
mapped ESTs in ZFIN. However, a mechanism should be developed to
identify all duplicates and provide access to a complete listing and
links to each regardless of whether they are listed in ZFIN or dbEST.
dbEST provides such match searches; links to this resource should be
investigated. Perhaps a 'BLAST' button can be added so investigators
can perform searches on the fly.</P>

<P>Class <STRONG>SSLP</STRONG>. It is probably useful to store the
primer sequences in ZFIN and provide them to users even though this
may duplicate data stored elsewhere (i.e. GenBank). Information about
allele size should be included for each strain in which the marker
has been mapped. Information about the 'quality' of the marker should
be added. For example, the percent of successful amplifications could
be used to indicate reliability. This information should also be
included for all other markers that are scored by PCR
amplification.</P>

<P>Class <STRONG>MUTANTS.</STRONG> Add a list of linked markers to
each mutant record; raw data for the mapping experiments that
provided these links should also be included. These could be
displayed graphically as a summary 'stick figure' representing the
local map.</P>

<P>Class <STRONG>DEFICIENCIES</STRONG>. This class which describes
deficiency chromosomes needs to be added. Its attributes should
include: segregation, Linkage Group, mapping information (markers
that are missing and flanking markers that are present), mutant
phenotype, sources, etc.<BR wp=br1>
<BR wp=br2>
<FONT SIZE="+1"><STRONG>Attendees:</STRONG></FONT></P>

<DL>
   <DT><STRONG>David Badman</STRONG></DT>
   
   <DT>Hematology Program Director</DT>
   
   <DT>DKUHD, NIDDK, NIH</DT>
   
   <DT>6AS-13C, Building 45 MSC 6600</DT>
   
   <DT>Bethesda, MD 20892-6600 USA</DT>
   
   <DT>&nbsp;</DT>
   
   <DT><STRONG>Jill Carrington</STRONG></DT>
   
   <DT>Director, Biological Models and Materials</DT>
   
   <DT>Research Program, NCRR, NIH</DT>
   
   <DT>One Rockledge Center, Rm 6164</DT>
   
   <DT>6705 Rockledge Drive, MSC 7965</DT>
   
   <DT>Bethesda, MD 20892 USA</DT>
   
   <DT>&nbsp;</DT>
   
   <DT><STRONG>Matthew Clark</STRONG></DT>
   
   <DT>Max-Planck-Institut f&uuml;r Molekulare Genetik</DT>
   
   <DT>Ihnestrasse 73</DT>
   
   <DT>D-14195 Berlin (Dahlem), GERMANY</DT>
   
   <DT>&nbsp;</DT>
   
   <DT><STRONG>Eck Doerry</STRONG></DT>
   
   <DT>Department of Computer &amp; Information Science</DT>
   
   <DT>1202 University of Oregon</DT>
   
   <DT>Eugene, OR 97403-1202 USA</DT>
   
   <DT>&nbsp;</DT>
   
   <DT><STRONG>Sarah Douglas</STRONG></DT>
   
   <DT>Department of Computer &amp; Information Science</DT>
   
   <DT>1202 University of Oregon</DT>
   
   <DT>Eugene, OR 97403-1202 USA</DT>
   
   <DT>&nbsp;</DT>
   
   <DT><STRONG>Marc Ekker</STRONG></DT>
   
   <DT>Loeb Institute of Medical Research</DT>
   
   <DT>Ottawa Civic Hospital</DT>
   
   <DT>725 Parkdale</DT>
   
   <DT>Ottawa, Ont, CANADA K1Y 4E9</DT>
   
   <DT>&nbsp;</DT>
   
   <DT><STRONG>Janan Eppig</STRONG></DT>
   
   <DT>Jackson Lab</DT>
   
   <DT>600 Main Street</DT>
   
   <DT>Bar Harbor, ME 04609 USA</DT>
   
   <DT>&nbsp;</DT>
   
   <DT><STRONG>Jonathan Epstein</STRONG></DT>
   
   <DT>Unit on Biologic Computation</DT>
   
   <DT>Office of the Scientific Director</DT>
   
   <DT>NIH, NICHD</DT>
   
   <DT>Bldg. 10, Rm 10D14</DT>
   
   <DT>10 Center Dr</DT>
   
   <DT>Bethesda, MD 20892 USA</DT>
   
   <DT>&nbsp;</DT>
   
   <DT><STRONG>Robert Geisler</STRONG></DT>
   
   <DT>Max-Planck-Institut f&uuml;r Entwicklungsbiologie</DT>
   
   <DT>Spemannstrasse 35/III</DT>
   
   <DT>D-72076 T&uuml;bingen, GERMANY</DT>
   
   <DT>&nbsp;</DT>
   
   <DT><STRONG>Pascal Haffter</STRONG></DT>
   
   <DT>Max-Planck-Institut f&uuml;r Entwicklungsbiologie</DT>
   
   <DT>Spemannstrasse 35/III</DT>
   
   <DT>D-72076 T&uuml;bingen, GERMANY</DT>
   
   <DT>&nbsp;</DT>
   
   <DT><STRONG>Marnie Halpern</STRONG></DT>
   
   <DT>Department of Embryology</DT>
   
   <DT>Carnegie Institution of Washington</DT>
   
   <DT>115 W. University Parkway</DT>
   
   <DT>Baltimore, MD 21210 USA</DT>
   
   <DT>&nbsp;</DT>
   
   <DT><STRONG>Debbie Henken</STRONG></DT>
   
   <DT>Developmental Biology, Genetics &amp; Teratology Branch</DT>
   
   <DT>Center for Research for Mothers and Children</DT>
   
   <DT>NICHHD, NIH</DT>
   
   <DT>6100 Executive Boulevard, Room 4B01</DT>
   
   <DT>Bethesda, MD 20892-7510 USA</DT>
   
   <DT>&nbsp;</DT>
   
   <DT><STRONG>Don Jackson</STRONG></DT>
   
   <DT>Cardiovascular Research Center</DT>
   
   <DT>MGH, Mailcode 1494100</DT>
   
   <DT>149-13th St. 4th Floor</DT>
   
   <DT>Charlestown, MA 02129 USA
   
   <P><STRONG>Steve Johnson</STRONG></P></DT>
   
   <DT>Department of Genetics</DT>
   
   <DT>Box 8232</DT>
   
   <DT>Washington University School of Medicine</DT>
   
   <DT>4566 Scott Ave</DT>
   
   <DT>St. Louis, MO 63110 USA</DT>
   
   <DT>&nbsp;&nbsp;</DT>
   
   <DT><STRONG>Frank Li</STRONG></DT>
   
   <DT>Department of Genetics</DT>
   
   <DT>Box 8232</DT>
   
   <DT>Washington University School of Medicine</DT>
   
   <DT>4566 Scott Ave</DT>
   
   <DT>St. Louis, MO 63110 USA</DT>
   
   <DT>&nbsp;&nbsp;</DT>
   
   <DT><STRONG>John Postlethwait</STRONG></DT>
   
   <DT>Institute of Neuroscience</DT>
   
   <DT>1254 University of Oregon</DT>
   
   <DT>Eugene, OR 97403-1254</DT>
   
   <DT>&nbsp;&nbsp;</DT>
   
   <DT><STRONG>David Ransom</STRONG></DT>
   
   <DT>Howard Hughes Medical Institute</DT>
   
   <DT>Children's Hospital, Enders 650</DT>
   
   <DT>300 Longwood Ave</DT>
   
   <DT>Boston, MA 02115-5737 USA</DT>
   
   <DT>&nbsp;&nbsp;</DT>
   
   <DT><STRONG>Greg Schuler</STRONG></DT>
   
   <DT>National Center for Biotechnology Information</DT>
   
   <DT>NIH Bldg 38A</DT>
   
   <DT>8600 Rockville Pike</DT>
   
   <DT>Bethesda, MD 20894 USA</DT>
   
   <DT>&nbsp;</DT>
   
   <DT><STRONG>Will Talbot</STRONG></DT>
   
   <DT>Developmental Genetics Program</DT>
   
   <DT>Skirball Institute</DT>
   
   <DT>NYU Medical Center</DT>
   
   <DT>540 First Avenue</DT>
   
   <DT>New York, NY 10016 USA</DT>
   
   <DT>&nbsp;</DT>
   
   <DT><STRONG>Reiko Toyama</STRONG></DT>
   
   <DT>Bldg. 6B, Room 420</DT>
   
   <DT>Laboratory of Molecular Genetics</DT>
   
   <DT>NICHD/NIH</DT>
   
   <DT>Bethesda, MD 20892 USA</DT>
   
   <DT>&nbsp;&nbsp;</DT>
   
   <DT><STRONG>Monte Westerfield</STRONG></DT>
   
   <DT>Institute of Neuroscience</DT>
   
   <DT>1254 University of Oregon</DT>
   
   <DT>Eugene, OR 97403-1254</DT>
</DL>

<P>

<HR>

For questions or comments, contact:</P>

<ADDRESS>&nbsp;</ADDRESS>

<ADDRESS>Pat Edwards<BR>
Institute of Neuroscience<BR>
University of Oregon<BR>
Eugene, OR 97403 USA<BR>
fax: 541-346-4548<BR>
<A HREF="mailto:edwards@uoneuro.uoregon.edu">
edwards@uoneuro.uoregon.edu</A></ADDRESS>

<P>

<HR>

<A HREF="/index.html">ZFIN <IMG SRC="fish_net.gif" X-CLARIS-USEIMAGEWIDTH X-CLARIS-USEIMAGEHEIGHT ALIGN=bottom></A></P>
</BODY>
</HTML>
