<html>
<TITLE>Zebrafish Database</TITLE>
<H3>The Zebrafish Database Project</H3>
<HR>
<B>THE PROBLEM STATEMENT</B> <p> To obtain a comprehensive understanding of
embryonic development, we need to characterize the anatomical, physiological,
developmental, and molecular genetic properties of a wide range of different
types of cells and test hypotheses that relate these properties functionally: How
do genes regulate the specific properties of cells?  For example, in the case of
neurons, what determines the shape of a neuron and how does shape affect
function? Do specific genes regulate where the first neuronal axons grow?
Presently there is no straight forward method for obtaining all the data relevant
to answering these questions. 
<p> 
Information about these diverse properties of
embryos is being obtained from a variety of organisms. However, systematic
organization of this information into databases and methods for rapid and easy
access to such databases are still lacking; this hampers relating one property to
another. Moreover, this is a world-wide problem [Fox94]: to integrate information
from diverse labs in a timely fashion and to make it easily accessible to
everyone in a meaningful manner that will truly promote scientific discovery.
Such databases have been established for the gene sequence data generated by
molecular biologists, but none adequately integrate the image data types required
to characterize the morphology, antigenicity, and gene expression patterns of
developing embryos. Progress will be greatly promoted by solutions to this
problem.
<p>
<HR><PRE>
<b><u><a name="1">Numbers of images</a></u> <u>available within the next
year:</u> </b><p>
<i>Developmental</i>		whole-mounts 		    40 images  	   8 Mb
<i>staging atlas:</i>  	sections (10 &#181;m)	   400 images  	  80 Mb
<br>
<i>Anatomical atlas</i> 	50 sagital, 50 horizontal
<i>(10&#181;m sections)</i>:	50 transverse		   150 images	  30 Mb<br>
<i>Adult brain atlas</i>:	25 &#181;m 3 views 		   600 images	 120 Mb<br>
<i>Morphology of</i>
<i>identified neurons</i>:	45 types x 10 views	   450 images	  90 Mb<br>
<i>Mutant (genetic loci)   </i>500 loci x 40 views  	 20000 images   4000 Mb
<i>anatomy atlases</i>:       (whole-mounts + sections)
<br>
<i>Gene expression</i>        150 genes x 40 views	  6000 images   1200 Mb
<i>pattern atlases</i>:       (whole-mounts + sections)
<p>
<b><u>Numbers of text files</u>:</b><br>
<i>Publications:  	 </i>3120; over 100 in the past year 		 280 Mb
<i>Genes</i>: 		  150						  80 Mb
<i>Mapped markers</i>:   560; 3600 anticipated within 2 years		 100 Mb
<i>People</i>:	  	  709 presently					  30 Mb
<i>Mutant alleles:</i>	  282; &gt;2,000* expected by next year		  25 Mb
</PRE>
<b>Table 1. </b>Estimates of disk space for zebrafish data within the next
year. Total may double within 2-3 years. *Number of mutant alleles estimated at
350 from Eugene, 740 from Boston, 1450 from T&uuml;bingen. Storage of images
estimated for 1300x1300 32-bit resolution, compressed with JPEG "good"
compression level.</HR>
<p>
The zebrafish is a vertebrate in which embryonic development can be studied with
very high precision; yet there is still a complete lack of a comprehensive
database for  zebrafish. Although zebrafish use in genetic research is relatively
new, the number of labs and the amount of data generated by these labs are
increasing at a phenomenal rate; the information already far exceeds the ability
of individual scientists to track and organize it (Table 1). Moreover, the
absence of a consensus for representing different types of data is leading to
inconsistencies in the way that data is stored by different labs. An organized
and concerted effort is desperately needed to establish a database for the
zebrafish community.<p>
<HR>
<B>The Zebrafish Database</B><p>
Continue on to <a href="aims.html">The Solution Statement</a>
<HR>
Return to <a href="cont.html">Table of Contents</a>
<HR>
<A HREF="/index.html">ZFIN  <img src="fish_net.gif"></A>
</html>
