<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html>
	<head>
		<title>
			ZFIN Author Guidelines
		</title>
	</head>

<script language="JavaScript" src="/javascript/header.js"></script>


<h1>ZFIN Author Guidelines</h1>

<p>For guidance on nomenclature issues, please refer to the <a href='http://zfin.org/zf_info/nomen.html'
 >Nomenclature Conventions</a> page, or contact the zebrafish 
nomenclature coordinator at <a href="mailto:nomenclature@zfin.org">nomenclature@zfin.org</a>. 

<p>Unambiguous identification of the biological entities described in published literature is critical for accurate curation of your published data.  When a publication is ambiguous or unclear, we must curate to a less detailed record, or not curate the data at all, which in turn degrades the usefulness of published data in ZFIN and for our downstream data collaborators.  We suggest including a section of the methods or a supplemental table dedicated to identification of the genes, alleles, genotypes, probes and antibodies used in your publications.  The utility of your published data is greatly enhanced when these are explicitly described:

<p><b>Genes:</b>
<p>Gene names change.  To unambiguously define what gene is being discussed in your papers, locate the proper gene record in ZFIN through a gene search or a BLAST analysis.  Then include in your paper the ZDB-GENE ID found at the top of the corresponding gene page.  If no gene record is present in ZFIN, include a nucleotide sequence accession number with your gene to ensure unambiguous identification.  We also recommend contacting the zebrafish nomenclature coordinator (<a href="mailto:nomenclature@zfin.org">nomenclature@zfin.org</a>) to get an approved gene name for a new gene before publication.
  
<p>For example:
<br>ift57 (ZDB-GENE-040614-1)
<br>mynewgene1 (BC056817)

<p><b>Alleles:</b>
<p>Different alleles of a single gene can have distinct phenotypes.  Consequently, it is essential that papers include the alleles used rather than just reporting the locus/gene that is mutated.  Without allele details, we can only curate data to an 'unspecified' allele of a gene/locus, which is a catch all for any data involving a mutant gene/locus where we do not have allele details.  Use of the 'unspecified' allele greatly diminishes the utility of your data as the "unknown" alleles can relate to data from any allele of that gene/locus.  The best way to unambiguously identify the allele is to include it's designation in your paper and associate it with the publication from which it was first derived.  Contact the ZFIN (<a href="mailto:zfinadmn@zfin.org">zfinadmn@zfin.org</a>) if you generate new mutants and you do not yet have an institutional line designation for your mutants/transgenics.

<p>For example:
<br>tx218 (Haffter et al., 2006)

<p><b>Genotypes:</b>
<p>Variations in allele zygosity, breeding strategy and genetic background can all give rise to unique phenotypes.  Consequently it is important to have the genotypes of fish clearly defined for each figure in your publications.  The full description of a genotype should include the alleles/transgenes present, their zygosity, the breeding strategy used to generate the fish (such as homozygous mutants generated from heterozygous parents), and the genetic background(s) involved.  Indicating in the methods that AB and WIK fish were used is not sufficient as we will not know which was used for which experiment, or if it was a random mix of AB and WIK.  If this information is not provided clearly, we can not include it in our curation, and any impact of the background used (which may not be known until a later date) is lost.  Descriptions of genotypes are perhaps best accomplished in the methods section of a paper or in figure legends if several genotypes are used in a single figure.

<p>For example:

<p>Methods: 
<br>"Homozygous and heterozygous wnt2 hu2848 fish were generated by crossing heterozygous parents.  All fish were on an AB background."

<p>Figure Legend:
<br>"Heterozygous (A) and homozygous (B) hu2848 littermates were examined by in situ hybridization for expression of wnt2."

<p><b>Probes:</b>
<p>We strongly encourage providing a sequence accession number for the sequence from which probes are generated.  This ensures correct association of the resulting data with the proper gene and avoids confusion if a gene record (and associated data) must be split in the future.

<p><b>Antibodies:</b>

<p>Because ZFIN curates antibodies, it is extremely helpful to have very specific information for unambiguously identifying the antibodies used in your papers.  As a minimum, please include the gene(s) known to be targetted by the antibody, the source of the antibody, and the catalog number when reporting the antibodies used in your papers.  Providing this data in the supplementary table or methods is the best way to convey this information.


<p>Example of a supplementary table:

<table border="1">
<tr>
<td><b>Gene / Probe</b></td>
<td><b>ZFIN ID / Accession #</b></td>
</tr>
<tr>
<td>wnt2</td>
<td>ZDB-GENE980526-416</td>
</tr>
<tr>
<td>newgene1</td>
<td>BC056817</td>
</table>

<p>
<table border="1">
<tr>
<td><b>Probe Target</b></td>
<td><b>Probe Accession #</b></td>
</tr>
<tr>
<td>wnt2</td>
<td>BC056277</td>
</tr>
</table>


<p>
<table border="1">
<tr>
<td><b>Ab Name</b></td>
<td><b>Source</b></td>
<td><b>Catalog Number</b></td>
</tr>
<tr>
<td>Anti-pax2a</td>
<td>Chemicon</td>
<td>xx-yyyy</td>
</tr>
</table>
	



<script language="JavaScript" src="/javascript/footer.js"></script>
