
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">

<html>
<head>
<title>ZFIN data reports</title>

<style type="text/css">
  <!--
  td.downloads_small {
    font-size: smaller
  }
  .downloads {
    font-size: larger;
    margin: 5px;
    background-color: <!--|BACKGROUND_COLOR|-->;
  }
  -->
</style>


</head>

<script language="JavaScript" src="/header.js" type="text/javascript"></script>

<!--
  :KLUDGE: The last thing header.js does is open a table, row, and column
  with WIDTH=100%.  This is often a good thing, but on this page, because
  of the wide GO table below, this causes the text at the top of the page
  to wrap off to the right, unless you make the page really wide.
  Therefore, close the table, run the whole page outside of a table and
  then reopen table at the end.
-->

</td>
</tr>
</table>

<div class="downloads">

<table width="100%">
  <tr>
    <td><br><h2>ZFIN Data Reports</h2></td>
    <td align="right"><a HREF="/warranty.html">Warranty Disclaimer and Copyright Notice</a>&nbsp;&nbsp;</td>
  </tr>
</table>

<p>
ZFIN data reports are updated every day of the week at 10:10pm PST.
The format for each file is described below.
Each record contains a ZFIN ID which may be used to link data between files.
All files use tab as the field delimiter.
The 'download' links will display the file on your browser.  Use your browser
'save as' feature to make a local copy of the file.  If you choose to link to
a ZFIN data page from your web site, append the ZFIN ID to the URL
http://zfin.org/cgi-bin/ZFIN_jump?record=<p>
 Please direct questions or requests for additional data to <a href="mailto:zfinadmn@zfin.org">ZFIN.</a><p>


<table BORDER CELLPADDING=2 BGCOLOR="<!--|HIGHLIGHT_COLOR|-->" WIDTH="100%"><tr><th>Table of Contents</th></tr></table>
<br>
<table>
<tr><td><a HREF="#seq_notes">Notes with Sequences</a></td></tr>
<tr><td><a HREF="#seq_notes_no_accession"> Notes with Sequences No Other Accession</a></td></tr>
<tr><td><a HREF="#marker"> Genetic Markers</a></td></tr>
<tr><td><a HREF="#alias"> Previous Names</a></td></tr>
<tr><td><a HREF="#map" >Mapping Data</a></td></tr>
<tr><td><a HREF="#seq"> Sequence Data</a></td></tr>
<tr><td><a HREF="#orthology"> Orthology Data</a></td></tr>
<tr><td><a HREF="#relationships"> Genetic Marker Relationships</a></td></tr>
<tr><td><a HREF="#go"> Gene Ontology Data</a></td></tr>
<tr><td><a HREF="#expression"> Gene Expression</a></td></tr>
<tr><td><a HREF="#phenotype"> Phenotype Data</a></td></tr>
<tr><td><a HREF="#genotype"> Genotype Data</a></td></tr>
<tr><td><a HREF="#wildtypes"> Wildtype Lines</a></td></tr>
<tr><td><a HREF="#ad"> Anatomical Ontologies</a></td></tr>
<tr><td><a HREF="#mo"> Morpholino Data</a></td></tr>
<tr><td><a HREF="#ot"> Ontology Translations</a></td></tr>
<tr><td><a HREF="#pmt"> PubMed Translations</a></td></tr>
</table>

<!-- -->

<a NAME="marker"></a><br>
<table BORDER CELLPADDING=2 BGCOLOR="<!--|HIGHLIGHT_COLOR|-->" WIDTH="100%"><tr><th>Genetic Markers</th></tr></table>

<p><b> Notes with sequences</b>
<table BORDER CELLPADDING=2><tr><td>ZFIN ID</td><td>Note</td></tr></table>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a HREF="/data_transfer/Downloads/seq_in_notes.txt">Download</a>

<p><b> Notes with sequences and no other accessions</b>
<table BORDER CELLPADDING=2><tr><td>ZFIN ID</td><td>Note</td></tr></table>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a HREF="/data_transfer/Downloads/seq_in_notes_no_accessions.txt">Download</a>


<p><b> Genetic Markers</b>
<table BORDER CELLPADDING=2><tr><td>ZFIN ID</td><td>Symbol</td><td>Name</td><td>Marker type</td></tr></table>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a HREF="/data_transfer/Downloads/genetic_markers.txt">Download</a>

<p><b> Previous ZFIN IDS</b> includes former and current ZFIN IDs for merged data<table BORDER CELLPADDING=2><tr><td>Former ZFIN ID</td><td>Current ZFIN ID</td></tr></table>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a HREF="/data_transfer/Downloads/zdb_history.txt">Download</a>

<!-- -->
<a NAME="alias"></a><br>
<table BORDER CELLPADDING=2 BGCOLOR="<!--|HIGHLIGHT_COLOR|-->" WIDTH="100%"><tr><th>Previous Names</th></tr></table>

<p><b> Previous names</b>
<table BORDER CELLPADDING=2><tr><td>Current ZFIN ID</td><td>Current Name</td><td>Current Symbol</td><td>Previous Name</td></tr></table>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a HREF="/data_transfer/Downloads/aliases.txt">Download</a>

<!-- -->

<a NAME="map"></a><br><br>
<table BORDER CELLPADDING=2 BGCOLOR="<!--|HIGHLIGHT_COLOR|-->" WIDTH="100%"><tr><th>Mapping Data</th></tr></table>
<p><b> Mapping Data from the 6 Zebrafish Mapping Panels</b>
<table BORDER CELLPADDING=2><tr><td>ZFIN ID</td><td>Symbol</td><td>Panel Symbol</td><td>LG</td><td>Location</td><td>Metric</td></tr></table>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a HREF="/data_transfer/Downloads/mappings.txt">Download</a>

<!-- -->

<a NAME="seq"></a><br><br>
<table BORDER CELLPADDING=2 BGCOLOR="<!--|HIGHLIGHT_COLOR|-->" WIDTH="100%"><tr><th>Sequence Data</th></tr></table>

<p><b> ZFIN Markers associations to GenBank sequence information</b>
<table BORDER CELLPADDING=2><tr><td>ZFIN ID</td><td>Name</td><td>GenBank ID</td></tr></table>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a HREF="/data_transfer/Downloads/genbank.txt">Download</a>

<p><b> ZFIN Markers associations to RefSeq sequence information</b>
<table BORDER CELLPADDING=2><tr><td>ZFIN ID</td><td>Symbol</td><td>RefSeq ID</td></tr></table>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a HREF="/data_transfer/Downloads/refseq.txt">Download</a>

<p><b> ZFIN Markers associations to Entrez Gene sequence information</b>
<table BORDER CELLPADDING=2><tr><td>ZFIN ID</td><td>Symbol</td><td>Entrez Gene ID</td></tr></table>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a HREF="/data_transfer/Downloads/entrezgene.txt">Download</a>

<p><b> ZFIN Markers associations to UniGene sequence information</b>
<table BORDER CELLPADDING=2><tr><td>ZFIN ID</td><td>Symbol</td><td>UniGene ID</td></tr></table>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a HREF="/data_transfer/Downloads/unigene.txt">Download</a>


<p><b> ZFIN Marker associations to UniProt protein IDs</b>
<table BORDER CELLPADDING=2><tr><td>ZFIN ID</td><td>Symbol</td><td>UniProt ID</td></tr></table>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a HREF="/data_transfer/Downloads/uniprot.txt">Download</a>

<p><b> ZFIN Marker associations to InterPro protein IDs</b>
<table BORDER CELLPADDING=2><tr><td>ZFIN ID</td><td>Symbol</td><td>InterPro ID</td></tr></table>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a HREF="/data_transfer/Downloads/interpro.txt">Download</a>

<p><b> ZFIN Marker associations to GenPept protein IDs</b>
<table BORDER CELLPADDING=2><tr><td>ZFIN ID</td><td>Symbol</td><td>GenPept ID</td></tr></table>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a HREF="/data_transfer/Downloads/genpept.txt">Download</a>

<p><b> ZFIN Marker associations to Sanger Vega IDs</b>
<table BORDER CELLPADDING=2><tr><td>ZFIN ID</td><td>Symbol</td><td>Vega ID</td></tr></table>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a HREF="/data_transfer/Downloads/vega.txt">Download</a>

<p><b> ZFIN Markers associations to Ensembl IDs in ZFIN</b>
<br>(Only associations with 1:1 relationships between ZFIN marker IDs and Ensembl IDs are stored in ZFIN.
<br>All associations can be obtained using BioMart http://www.ensembl.org/biomart/martview)
<table BORDER CELLPADDING=2><tr><td>ZFIN ID</td><td>Symbol</td><td>Ensembl ID</td></tr></table>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a HREF="/data_transfer/Downloads/ensembl_1_to_1.txt">Download</a>

<p><b> ZFIN Gene IDs indirectly associated with Sequence Accessions via cDNA & EST</b>
<table BORDER CELLPADDING=2><tr><td>ZFIN Gene ID</td><td>ZFIN Gene Symbol</td><td>Sequence Accession</td></tr></table>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a HREF="/data_transfer/Downloads/gene_seq.txt">Download</a>



<!-- -->

<a NAME="orthology"></a><br><br>
<table BORDER CELLPADDING=2 BGCOLOR="<!--|HIGHLIGHT_COLOR|-->" WIDTH="100%"><tr><th>Orthology</th></tr></table>

<p><b> Human and Zebrafish Orthology</b>
<table BORDER CELLPADDING=2><tr><td>ZFIN ID</td><td>ZFIN Symbol</td><td>ZFIN Name</td><td>Human Symbol</td><td>Human Name</td><td>OMIM ID</td><td>Entrez Gene ID</td></tr></table>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a HREF="/data_transfer/Downloads/human_orthos.txt">Download</a>

<p><b> Mouse and Zebrafish Orthology</b>
<table BORDER CELLPADDING=2><tr><td>ZFIN ID</td><td>ZFIN Symbol</td><td>ZFIN Name</td><td>Mouse Symbol</td><td>Mouse Name</td><td>MGI ID</td><td>Entrez Gene ID</td></tr></table>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a HREF="/data_transfer/Downloads/mouse_orthos.txt">Download</a>

<p><b> Drosophila and Zebrafish Orthology</b>
<table BORDER CELLPADDING=2><tr><td>ZFIN ID</td><td>ZFIN Symbol</td><td>ZFIN Name</td><td>Fly Symbol</td><td>Fly Name</td><td>Flybase ID</td></tr></table>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a HREF="/data_transfer/Downloads/fly_orthos.txt">Download</a>

<p><b> Yeast and Zebrafish Orthology</b>
<table BORDER CELLPADDING=2><tr><td>ZFIN ID</td><td>ZFIN Symbol</td><td>ZFIN Name</td><td>Yeast Symbol</td><td>Yeast Name</td><td>SGD ID</td></tr></table>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a HREF="/data_transfer/Downloads/yeast_orthos.txt">Download</a>


<!-- -->

<a NAME="relationships"></a><br><br>
<table BORDER CELLPADDING=2 BGCOLOR="<!--|HIGHLIGHT_COLOR|-->" WIDTH="100%"><tr><th>Genetic Marker Relationships</th></tr></table>
<p><b>Genetic Marker Relationships</b>
<table BORDER CELLPADDING=2><tr><td>Gene ID</td><td>Gene Symbol</td><td>Marker ID</td><td>Marker Symbol</td><td>Relationship</td></tr></table>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a HREF="/data_transfer/Downloads/gene_marker_relationship.txt">Download</a>

<!-- -->


<a NAME="go"></a><br><br>
<table BORDER CELLPADDING=2 BGCOLOR="<!--|HIGHLIGHT_COLOR|-->" WIDTH="100%"><tr><th>Gene Ontology Data</th></tr></table>

<p><b> Gene Ontology (GO) Annotations of Zebrafish Markers
<br>(<i>see also</i> <a HREF="http://www.geneontology.org/doc/GO.annotation.html#file">GO Annotation File format</a> at the Gene Ontology Consortium)</b>
<table BORDER CELLPADDING=2>
  <tr>
    <td class="downloads_small">Database<br>Desig-<br>nation</td>
    <td class="downloads_small">Marker ID</td>
    <td class="downloads_small">Gene Symbol</td>
    <td class="downloads_small">Quali-<br>fiers</td>
    <td class="downloads_small">GO<br>Term ID</td>
    <td class="downloads_small">Refer-<br>ence ID</td>
    <td class="downloads_small">GO<br>Evidence Code</td>
    <td class="downloads_small">Inferred From</td>
    <td class="downloads_small">Ontology:
      <span style="font-size: smaller">
        <br>P=Biological Process
        <br>F=Molecular Function
        <br>C=Cellular Component
      </span>
    </td>
    <td class="downloads_small">Marker Name</td>
    <td class="downloads_small">Marker Synonyms<br>(if any)</td>
    <td class="downloads_small">Marker Type:
      <span style="font-size: smaller">
        <br>gene<br>transcript<br>protein
      </span>
    </td>
    <td class="downloads_small">Taxon</td>
    <td class="downloads_small">Modifi-<br>cation Date</td>
    <td class="downloads_small">Assigned By</td>
  </tr>
</table>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a HREF="ftp://ftp.geneontology.org/pub/go/gene-associations/gene_association.zfin.gz">Download from www.geneontology.org</a>


<!-- -->

<a NAME="expression"></a><br><br>
<table BORDER CELLPADDING=2 BGCOLOR="<!--|HIGHLIGHT_COLOR|-->" WIDTH="100%"><tr><th>Gene Expression</th></tr></table>

<p><b> ZFIN Genes with Expression Assay Records</b>
<table BORDER CELLPADDING=2><tr><td>Gene ID</td><td>Gene Symbol</td><td>EST ID(optional)</td><td>EST Symbol(optional)</td><td>Expression Type</td><td>Expression ID</td><td>Publication ID</td><td>Genotype ID</td><td>Environment ID</td></tr></table>

<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a HREF="/data_transfer/Downloads/xpat.txt">Download</a>

<p><b> Expression Environment Description </b>
<table BORDER CELLPADDING=2><tr><td>Environment ID</td><td>Condition Group</td><td>Condition Content</td><td>Value</td><td>Unit</td><td>Comment</td></tr></table>

<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a HREF="/data_transfer/Downloads/xpat_environment.txt">Download</a>


<!-- -->
<a NAME="genotype"></a><br><br>
<table BORDER CELLPADDING=2 BGCOLOR="<!--|HIGHLIGHT_COLOR|-->" WIDTH="100%"><tr><th>Genotype Data</th></tr></table>

<p><b> Genotype Features</b>
<table BORDER CELLPADDING=2><tr><td>ZFIN ID</td><td>Genotype Name</td><td>Genotype Unique Name</td><td>Allele ID</td><td>Allele Name</td><td>Allele Abbreviation</td><td>Allele Type</td><td>Gene or Construct Symbol</td><td>Corresponding ZFIN Gene ID/Construct ID</td></tr></table>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a HREF="/data_transfer/Downloads/genotype_features.txt">Download</a>

<p><b> Genotype Backgrounds </b>
<table BORDER CELLPADDING=2><tr><td>ZFIN ID</td><td>Genotype Name</td><td>Background</td></tr></table>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a HREF="/data_transfer/Downloads/genotype_backgrounds.txt">Download</a>

<!-- -->
<a NAME="wildtypes"></a><br><br>
<table BORDER CELLPADDING=2 BGCOLOR="<!--|HIGHLIGHT_COLOR|-->" WIDTH="100%"><tr><th>Wildtype Lines</th></tr></table>

<p><b> Wildtype Lines </b>
<table BORDER CELLPADDING=2><tr><td>ZFIN ID</td><td>Genotype Name</td><td>Genotype Abbreviation</td></tr></table>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a HREF="/data_transfer/Downloads/wildtypes.txt">Download</a>

<a NAME="phenotype"></a><br><br>
<table BORDER CELLPADDING=2 BGCOLOR="<!--|HIGHLIGHT_COLOR|-->" WIDTH="100%"><tr><th>Phenotype Data</th></tr></table>

<p><b> Genotypes with Phenotypes</b>
<table BORDER CELLPADDING=2><tr>
<td>Genotype ID</td>
<td>Genotype Name</td>
<td>Start Stage ID</td>
<td>Start Stage Name</td>
<td>End Stage ID</td>
<td>End Stage Name</td>
<td>Affected Structure or Process 1 ID</td>
<td>Affected Structure or Process 1 Name</td>
<td>Affected Structure or Process 2 ID</td>
<td>Affected Structure or Process 2 Name</td>
<td>Phenotype Keyword ID</td>
<td>Phenotype Keyword Name</td>
<td>Phenotype Modifier</td>
<td>Publication ID</td>
<td>Environment ID</td>
</tr>
</table>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a HREF="/data_transfer/Downloads/phenotype.txt">Download</a>

<p><b> Genotypes with Phenotypes (obo format)</b>
<table BORDER CELLPADDING=2><tr>
<td>Genotype ZFIN ID</td>
<td>Genotype Name</td>
<td>Start Stage Obo ID</td>
<td>End Stage Obo ID</td>
<td>Affected Structure or Process 1 obo ID</td>
<td>Affected Structure or Process 2 obo ID</td>
<td>Phenotype Keyword obo ID</td>
<td>Phenotype Modifier</td>
<td>Publication ZFIN ID</td>
<td>Environment ZFIN ID</td>
</tr>
</table>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a HREF="/data_transfer/Downloads/pheno_obo.txt">Download</a>

<p><b> Phenotype Environment Description </b>
<table BORDER CELLPADDING=2><tr><td>Environment ID</td><td>Condition Group</td><td>Condition Content</td><td>Value</td><td>Unit</td><td>Comment</td></tr></table>

<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a HREF="/data_transfer/Downloads/pheno_environment.txt">Download</a>


<!-- -->


<a NAME="ad"></a><br><br>
<table BORDER CELLPADDING=2 BGCOLOR="<!--|HIGHLIGHT_COLOR|-->" WIDTH="100%"><tr><th>Anatomical Ontologies</th></tr></table>

<p><a href="http://obo.cvs.sourceforge.net/*checkout*/obo/obo/ontology/anatomy/gross_anatomy/animal_gross_anatomy/fish/zebrafish_anatomy.obo"> <b> Zebrafish Anatomical Ontology in OBO Format </b> </a>

<p><b> Zebrafish Anatomical Terms</b>
<table BORDER CELLPADDING=2><tr> <td>Entire Staged Anatomy (tab indented)</td></tr></table>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a HREF="/data_transfer/Downloads/staged_anatomy.txt">Download</a>
        (*note: format changed April 26, 2005)

<p><b> Zebrafish Anatomy Term Relationships</b>
<table BORDER CELLPADDING=2><tr> <td>Parent Item ID</td><td>Child Item ID</td><td>Relationship Type ID</td></tr></table>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a HREF="/data_transfer/Downloads/anatomy_relationship.txt">Download</a>
        (*note: format changed April 26, 2005)

<p><b> Zebrafish Stage Series</b>
<table BORDER CELLPADDING=2><tr> <td>Stage ID</td><td>Stage Name</td><td>Begin Hours</td><td>End Hours</td></tr></table>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a HREF="/data_transfer/Downloads/stage_ontology.txt">Download</a>


<p><b> Zebrafish Anatomy Term Stages</b>
<table BORDER CELLPADDING=2><tr> <td>Anatomy Item ID</td><td>Anatomy Item Name</td><td>Begin Stage ID</td><td>End Stage ID</td></tr></table>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a HREF="/data_transfer/Downloads/anatomy_item.txt">Download</a>

<p><b> Zebrafish Anatomy Term Synonyms</b>
<table BORDER CELLPADDING=2><tr> <td>Anatomy Item ID</td><td>Anatomy Item Name</td><td>Anatomy Item Synonym</td></tr></table>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a HREF="/data_transfer/Downloads/anatomy_synonyms.txt">Download</a>

<p><b> Zebrafish Gene Expression by Stage and Anatomy Term </b>
<table BORDER CELLPADDING=2><tr> <td>Expression ID</td><td>Begin Stage ID</td><td>End Stage ID</td><td>Anatomy Item ID</td><td>Expression Found</td></tr></table>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a HREF="/data_transfer/Downloads/xpat_stage_anatomy.txt">Download</a>

<a NAME="mo"></a><br><br>
<table BORDER CELLPADDING=2 BGCOLOR="<!--|HIGHLIGHT_COLOR|-->" WIDTH="100%"><tr><th>Morpholino Data</th></tr></table>

<p><b> Morpholino </b>
<table BORDER CELLPADDING=2><tr> <td>Gene ID</td><td>Gene Symbol</td><td>MO ID</td><td>MO Symbol</td><td>MO Sequence</td><td>Note</td></tr></table>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a HREF="/data_transfer/Downloads/Morpholinos.txt">Download</a>

<a NAME="ot"></a><br><br>
<table BORDER CELLPADDING=2 BGCOLOR="<!--|HIGHLIGHT_COLOR|-->" WIDTH="100%"><tr><th>Ontology Translations</th></tr></table>

<p><b> Gene Ontology IDs: ZFIN to GO ids </b>
<br>(<i>see also</i> <a HREF="http://www.geneontology.org/ontology/gene_ontology_edit.obo">GO in OBO format</a> at the Gene Ontology Consortium)</b>
<table BORDER CELLPADDING=2><tr> <td>GO term ZFIN ID</td><td>GO term GO ID</td></tr></table>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a HREF="/data_transfer/Downloads/gene_ontology_translation.txt">Download</a>

<p><b> Anatomy Ontology IDs: ZFIN to AO ids </b>
<br>(<i>see also</i> <a HREF="http://obo.cvs.sourceforge.net/*checkout*/obo/obo/ontology/anatomy/gross_anatomy/animal_gross_anatomy/fish/zebrafish_anatomy.obo">AO File in OBO format </a> at the OBO foundry )</b><br>
<table BORDER CELLPADDING=2><tr> <td>AO term ZFIN ID</td><td>AO term AO ID</td></tr></table>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a HREF="/data_transfer/Downloads/anatomy_ontology_translation.txt">Download</a>

<p><b> Quality Ontology IDs: ZFIN to Quality ids </b>
<br>(<i>see also</i> <a HREF="http://obo.cvs.sourceforge.net/*checkout*/obo/obo/ontology/phenotype/quality.obo">Quality file in OBO format</a> at the OBO foundry</b><br>
<table BORDER CELLPADDING=2><tr> <td>Quality term ZFIN ID</td><td>Quality term Quality ID</td><td>Quality term name</td></tr></table>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a HREF="/data_transfer/Downloads/phenotype_quality_ontology_translation.txt">Download</a>

<a NAME="pmt"></a><br><br>
<table BORDER CELLPADDING=2 BGCOLOR="<!--|HIGHLIGHT_COLOR|-->" WIDTH="100%"><tr><th>PubMed Translations</th></tr></table>

<p><b> ZFIN Publication IDs: ZFIN IDs to PubMed ids </b>
<br>
<table BORDER CELLPADDING=2><tr> <td>Publication ZFIN ID</td><td>PubMed ID or None when not available</td></tr></table>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a HREF="/data_transfer/Downloads/pub_to_pubmed_id_translation.txt">Download</a>



</div>

<table>
<tr>
<td>

<script language="JavaScript" src="/footer.js"></script> 
</HTML>
