<HTML>
 <HEAD>
  <TITLE>ZFIN Glossary</TITLE>
 </HEAD>
<script language="JavaScript" src="http://<!--|DOMAIN_NAME|-->/header.js"></script>
<HR>
<h2 align="center">ZFIN Glossary</h2>

<P>This glossary provides definitions for terms that are useful in understanding 
  zebrafish development, anatomy, genetics and bioinformatics. Terms are defined 
  in a general sense as they apply to eukaryotic genetics, particularly zebrafish 
  genetics; some terms are also defined as they are specifically used in ZFIN 
  (e.g., <A HREF="glossary.html#gene">gene</A>). This list will grow as time 
  goes on. If there is a term you think should be included in this list, please 
  contact <A HREF="mailto:zfinadmin@zfin.org">ZFIN User Support</A>. 
<P><em><font size="-1">Anatomical terms were adapted from <a href="zfbook/stages/stages.html">Kimmel 
  et al., 1995</a>. Bioinformatic terms were adapted from the <A HREF="http://www.informatics.jax.org/mgihome">MGI 
  mouse glossary</A> and we thank <A HREF="http://www.informatics.jax.org/mgihome">MGI</A> 
  for its generous help.</font></em> 
<P>There are a number of other useful online glossaries: 
<UL>
  <LI>The <A HREF="anatomy/dict/sum.html">Zebrafish Anatomical Dictionary</A> 
    displays anatomical terms of the zebrafish in a hierarchical manner for a 
    series of developmental stages, using standard anatomical nomenclature.</LI>
  <LI>The <A HREF="zfbook/stages/index.html">Zebrafish Developmental Staging Series</A> 
    describes the standard criteria used for identifying the developmental age 
    of zebrafish.</LI>
  <LI>The <A HREF="http://www.geneontology.org/doc/GO.defs">GO Term Definitions</A> 
    provides current definitions of all terms used in the Gene Ontology Project.</LI>
<LI>The <A HREF="http://www.nhgri.nih.gov/DIR/VIP/Glossary/pub_glossary.cgi">Glossary of Genetic Terms</A> at the National Human Genome Research Institute defines basic genetic terms most relevant to human genetics. This glossary includes audio explanations and illustrations.</LI>
<LI>The <A HREF="http://www.ornl.gov/TechResources/Human_Genome/publicat/glossary.html">Genome Glossary</A> for the Human Genome Project defines many basic genetic terms.</LI>
<LI>The <A HREF="http://www.ncbi.nlm.nih.gov:80/entrez/meshbrowser.cgi">MeSH 
Browser</A> at <A HREF="glossary.html#NCBI">NCBI</A> contains useful definitions of 
many terms, organized as a hierarchical controlled vocabulary.</LI>
<LI>The <A HREF="http://www.graylab.ac.uk/omd/index.html">On-line Medical Dictionary</A> provides definitions for hundreds of medical terms used in describing human phenotypes.</LI>
<LI>The <A HREF="http://www.ncbi.nlm.nih.gov/Education/BLASTinfo/glossary2.html">BLAST 
Glossary</A> at NCBI defines terms used in searching sequence databases.</LI>
  <LI>The <A HREF="http://www.weihenstephan.de/%7Eschlind/genglos.html">Hypermedia 
    Glossary of Genetic Terms</A> by Birgid Schlindwein contains over 600 genetic 
    terms.</LI>
</UL>

<p>If you find an online glossary that you think should be included here, please 
  contact <A HREF="mailto:zfinadmin@zfin.org">ZFIN User Support</A>. </p>
<HR>
<DL>
  <CENTER>
    <H3> <A HREF="glossary.html#a">A</A>&nbsp; <A HREF="glossary.html#b">B</A>&nbsp; 
      <A HREF="glossary.html#c">C</A>&nbsp; <A HREF="glossary.html#d">D</A>&nbsp; 
      <A HREF="glossary.html#e">E</A>&nbsp; <A HREF="glossary.html#f">F</A>&nbsp; 
      <A HREF="glossary.html#g">G</A>&nbsp; <A HREF="glossary.html#h">H</A>&nbsp; 
      <A HREF="glossary.html#i">I</A>&nbsp; <A HREF="glossary.html#j">J</A>&nbsp; 
      <A HREF="glossary.html#k">K</A>&nbsp; <A HREF="glossary.html#l">L</A>&nbsp; 
      <A HREF="glossary.html#m">M</A>&nbsp; <A HREF="glossary.html#n">N</A>&nbsp; 
      <A HREF="glossary.html#o">O</A>&nbsp; <A HREF="glossary.html#p">P</A>&nbsp; 
      <A HREF="glossary.html#q">Q</A>&nbsp; <A HREF="glossary.html#r">R</A>&nbsp; 
      <A HREF="glossary.html#s">S</A>&nbsp; <A HREF="glossary.html#t">T</A>&nbsp; 
      <A HREF="glossary.html#u">U</A>&nbsp; <A HREF="glossary.html#v">V</A>&nbsp; 
      <A HREF="glossary.html#w">W</A>&nbsp; <A HREF="glossary.html#x">X</A>&nbsp; 
      <A HREF="glossary.html#y">Y</A>&nbsp; <A HREF="glossary.html#z">Z</A>&nbsp; </H3>
    <BR>
    Click on a letter above to jump to that section of the ZFIN Glossary. 
  </CENTER>
  <HR>
  <P> 
  <DT><B><A NAME="3'">3' (3-prime)</A></B> 
  <DD>A term that identifies one end of a single-stranded <A HREF="glossary.html#nucleic_acid">nucleic 
    acid</A> molecule. The 3' end is the end of the molecule that terminates in 
    a 3' hydroxyl group. The 3' direction is the direction toward the 3' end. 
    Nucleic acid sequences are written with the <A HREF="glossary.html#5'">5' 
    end</A> to the left and the 3' end to the right, in reference to the direction 
    of <A HREF="glossary.html#DNA">DNA</A> synthesis during <A HREF="glossary.html#replication">replication</A> 
    (from 5' to 3'), <A HREF="glossary.html#RNA">RNA</A> synthesis during <A HREF="glossary.html#transcription">transcription</A> 
    (from 5' to 3'), and the reading of <A HREF="glossary.html#mRNA">mRNA</A> 
    sequence (from 5' to 3') during <A HREF="glossary.html#translation">translation</A>. 
    See the <A HREF="http://www.nhgri.nih.gov/DIR/VIP/Glossary/Illustration/base_pair.html">Figure</A> 
    at <A HREF="glossary.html#NHGRI">NHGRI</A>. See also <A HREF="glossary.html#5'">5' 
    (5-prime)</A>, <A HREF="glossary.html#central_dogma">central dogma</A>. 
    <P> 
  <DT><B><A NAME="3'_UTR">3' UTR</A></B> 
  <DD>3' Untranslated Region. That portion of an <A HREF="glossary.html#mRNA">mRNA</A> 
    from the <A HREF="glossary.html#3'">3' end</A> to the position of the last 
    <A HREF="glossary.html#codon">codon</A> used in <A HREF="glossary.html#translation">translation</A>. 
    See also <A HREF="glossary.html#5'_UTR">5'UTR</A>. 
    <P> 
  <DT><B><A NAME="5'">5' (5-prime)</A></B> 
  <DD>A term that identifies one end of a single-stranded <A HREF="glossary.html#nucleic_acid">nucleic 
    acid</A> molecule. The 5' end is the end of the molecule that terminates in 
    a 5' phosphate group. The 5' direction is the direction toward the 5' end. 
    Nucleic acid sequences are written with the 5' end to the left and the <A HREF="glossary.html#3'">3' 
    end</A> to the right, in reference to the direction of <A HREF="glossary.html#DNA">DNA</A> 
    synthesis during <A HREF="glossary.html#replication">replication</A> (from 
    5' to 3'), <A HREF="glossary.html#RNA">RNA</A> synthesis during <A HREF="glossary.html#transcription">transcription</A> 
    (from 5' to 3'), and the reading of <A HREF="glossary.html#mRNA">mRNA</A> 
    sequence (from 5' to 3') during <A HREF="glossary.html#translation">translation</A>. 
    See the <A 

HREF="http://www.nhgri.nih.gov/DIR/VIP/Glossary/Illustration/base_pair.html">Figure</A> 
    at <A HREF="glossary.html#NHGRI">NHGRI</A>. See also <A HREF="glossary.html#3'">3' 
    (3-prime)</A>, <A HREF="glossary.html#central_dogma">central dogma</A>. 
    <P> 
  <DT><B><A NAME="5'_UTR">5' UTR</A></B> 
  <DD>5' Untranslated Region. That portion of an <A HREF="glossary.html#mRNA">mRNA</A> 
    from the <A HREF="glossary.html#5'">5' end</A> to the position of the first 
    <A HREF="glossary.html#codon">codon</A> used in <A HREF="glossary.html#translation">translation</A>. 
    See also <A HREF="glossary.html#3'_UTR">3'UTR</A>. 
    <!-- AAAAAAAAAAAAAAAAAAAAAAAAAAA -->
    <A NAME="a"></A> 
    <P> 
  <DT><B><A NAME="accession_ID">accession ID</A></B> 
  <DD>A unique alphanumeric character string that is used to identify unambiguously 
    a particular record in a database. Examples include <A HREF="glossary.html#ZFIN">ZFIN</A> 
    accession IDs, <A HREF="glossary.html#genbank">GenBank</A> accession IDs, 
    and MedLine accession IDs. 
    <P> 
  <DT><A NAME="allele"><B>allele</B></A> 
  <DD>Any of the alternative forms of a <A HREF="glossary.html#gene">gene</A> 
    occupying a given <A HREF="glossary.html#locus">locus</A>; any one of several 
    <A HREF="glossary.html#mutation">mutational</A> forms of a gene. In <A HREF="glossary.html#ZFIN">ZFIN</A>, 
    allelic variants are associated with<a href="#mutant"> mutants</a> and mutant 
    <A HREF="glossary.html#phenotype">phenotypes</A>. 
    <!--new (1/19/01) allele name from ps-->
    <P> 
  <DT><B><A NAME="allele_designation">allele designation</A></B> 
  <DD>In <A HREF="glossary.html#ZFIN">ZFIN</A>, an "allele disignation" is a 
    set of letters and numbers that uniquely identifies an <A HREF="glossary.html#allele">allele</A> 
    of a given <A HREF="glossary.html#gene">gene</A>. The allele designation 
    is added as a superscript to the <A HREF="glossary.html#gene_symbol">gene 
    symbol</A>. For details see the <a href="nomen.html">Zebrafish Nomenclature 
    Guidelines</a> for <a href="nomen.html#allele">alleles</a>. See also <A HREF="glossary.html#gene_name">gene 
    name</A>. 
  <dt>&nbsp;</dt>
  <DD> 
    <P> 
  <DT><B><A NAME="alternative_splicing">alternative splicing</A></B> 
  <DD>The production of two or more distinct <A HREF="glossary.html#mRNA">mRNAs</A> 
    by differences in <a href="glossary.html#splicing">splicing</a> (using different 
    <a href="glossary.html#exon">exons</a>) of <A HREF="glossary.html#RNA">RNA</A> 
    <A HREF="glossary.html#transcript">transcripts</A> having the same sequence. 
    <P> 
  <DT><B><A NAME="amino_acid">amino acid</A></B> 
  <DD>A molecule of the general formula NH<SUB>2</SUB>-CHR-COOH, where "R" is 
    one of a number of different side chains. Amino acids are the building blocks 
    of <A HREF="glossary.html#protein">proteins</A>. The sixty-four <A HREF="glossary.html#codon">codons</A> 
    of the <A HREF="glossary.html#genetic_code">genetic code</A> allow the use 
    of twenty different amino acids (the primary amino acids) in the synthesis 
    of proteins. Other nonprimary amino acids occur in proteins by <A HREF="glossary.html#enzyme">enzymatic</A> 
    modification of amino acids in mature proteins, and as metabolic intermediates. 
    See the <A 

HREF="http://www.nhgri.nih.gov/DIR/VIP/Glossary/Illustration/amino_acid.html">Figure</A> 
    at <A HREF="glossary.html#NHGRI">NHGRI</A>. For Figures showing the strucutre 
    of each of the twenty primary animo acids, see <A 

HREF="http://www.accessexcellence.org/AB/GG/aminoAcids1.html">Figure1</A> and 
    <A 

HREF="http://www.accessexcellence.org/AB/GG/aminoAcids2.html">Figure2</A> from 
    "Molecular Biology of the Cell" by Alberts et al. 
    <P> 
  <DT><B><A NAME="amino_terminus">amino terminus</A></B> 
  <DD>A term that identifies one end of a <A HREF="glossary.html#protein">protein</A> 
    molecule. The amino terminus is that end of the molecule that terminates in 
    a free amino group. See the <A 

HREF="http://www.nhgri.nih.gov/DIR/VIP/Glossary/Illustration/amino_acid.html">Figure</A> 
    at <A HREF="glossary.html#NHGRI">NHGRI</A>. See also <A HREF="glossary.html#amino_acid">amino 
    acid</A>, <A HREF="glossary.html#central_dogma">central dogma</A>. 
    <P> 
  <DT><B><A NAME="amorphic">amorphic mutation</A></B> 
  <DD>A type of <A HREF="glossary.html#mutation">mutation</A> in which the altered 
    <A HREF="glossary.html#gene_product">gene product</A> lacks the <A HREF="glossary.html#molecular_function">molecular 
    function</A> of the <A HREF="glossary.html#wild_type">wild-type</A> gene. 
    Synonyms: <A HREF="glossary.html#loss-of-function">loss-of-function mutation</A>, 
    <A HREF="glossary.html#null_mutation">null mutation</A>. <BR>
    See also: 
    <UL>
      <LI><A HREF="glossary.html#antimorphic">antimorphic mutation</A></LI>
      <LI><A HREF="glossary.html#gain-of-function">gain-of-function mutation</A></LI>
      <LI><A HREF="glossary.html#hypermorphic">hypermorphic mutation</A></LI>
      <LI><A HREF="glossary.html#hypomorphic">hypomorphic mutation</A></LI>
      <LI><A HREF="glossary.html#neomorphic">neomorphic mutation</A></LI>
    </UL>
    <P> 
  <DT><B><A NAME="ancestor">ancestor</A></B> 
  <DD>In bioinformatics, this term refers to terms in a hierarchical <A HREF="glossary.html#controlled_vocabulary">controlled 
    vocabulary</A> like ones containing <A HREF="glossary.html#gene_ontology">Gene 
    Ontology (GO)</A> terms. An "ancestor" of a term is a term any number of levels 
    above it in the hierarchy from which it is descended. For example, the GO 
    term <A HREF="glossary.html#enzyme">enzyme</A> [GO:0003824] is an <B>ancestor</B> 
    to the GO term <B>alcohol dehydrogenase</B> [GO:0004022]. See also: <A HREF="glossary.html#children">children</A>, 
    <A HREF="glossary.html#parent">parent</A>. <A HREF="glossary.html#sibling">sibling</A>. 
    <P> 
  <DT><B><A NAME="aneuploid">aneuploid</A></B> 
  <DD>Not <A HREF="glossary.html#euploid">euploid</A>. 
    <P> 
  <DT><B><A NAME="annotation">annotation</A></B> 
  <DD>Note added to a document or image to provide additional needed information. 
    See also <A HREF="glossary.html#sequence_annotation">sequence annotation</A>. 
    <P> 
  <DT><B><A NAME="anonymous_DNA">anonymous DNA segment</A></B> 
  <DD>A segment of <A HREF="glossary.html#DNA">DNA</A> not known to correspond 
    to a named <A HREF="glossary.html#gene">gene</A> that can be used as a <A HREF="glossary.html#marker">marker</A> 
    in the construction of <A HREF="glossary.html#genetic_map">genetic maps</A>. 
    See also <A HREF="glossary.html#STS">STS</A>. 
    <P> 
  <DT><B><A NAME="antibody">antibody</A></B> 
  <DD>A <A HREF="glossary.html#protein">protein</A> produced by cells of the 
    immune system that binds to an <A HREF="glossary.html#antigen">antigen</A>. 
    See the <A 

HREF="http://www.nhgri.nih.gov/DIR/VIP/Glossary/Illustration/antibody.html">Figure</A> 
    at <A HREF="glossary.html#NHGRI">NHGRI</A>. See also <A HREF="glossary.html#monoclonal">monoclonal 
    antibody</A>. 
    <P> 
  <DT><B><A NAME="antigen">antigen</A></B> 
  <DD>A <A HREF="glossary.html#protein">protein</A> or other molecule that can 
    elicit an immune response; the <A HREF="glossary.html#antibody">antibody</A> 
    protein that is produced binds to the antigen. 
    <P> 
  <DT><B><A NAME="antimorphic">antimorphic mutation</A></B> 
  <DD>A type of <A HREF="glossary.html#mutation">mutation</A> in which the altered 
    <A HREF="glossary.html#gene_product">gene product</A> possesses an altered 
    <A HREF="glossary.html#molecular_function">molecular function</A> that acts 
    antagonistically to the <A HREF="glossary.html#wild_type">wild-type</A> <A HREF="glossary.html#allele">allele</A>. 
    Antimorphic mutations are always <A HREF="glossary.html#dominant">dominant</A> 
    or <A HREF="glossary.html#semidominant">semidominant</A>. <BR>
    See also: 
    <UL>
      <LI><A HREF="glossary.html#amorphic">amorphic mutation</A></LI>
      <LI><A HREF="glossary.html#gain-of-function">gain-of-function mutation</A></LI>
      <LI><A HREF="glossary.html#hypermorphic">hypermorphic mutation</A></LI>
      <LI><A HREF="glossary.html#hypomorphic">hypomorphic mutation</A></LI>
      <LI><A HREF="glossary.html#loss-of-function">loss-of-function mutation</A></LI>
      <LI><A HREF="glossary.html#neomorphic">neomorphic mutation</A></LI>
      <LI><A HREF="glossary.html#null_mutation">null mutation</A></LI>
    </UL>
    <P> 
  <DT><B><A NAME="antisense">antisense</A></B> 
  <DD> 1. In molecular biology, that strand of a <A HREF="glossary.html#DNA">DNA</A> 
    molecule whose sequence is <A HREF="glossary.html#complementary">complementary</A> 
    to the strand represented in <A HREF="glossary.html#mRNA">mRNA</A>.<BR>
    2. In molecular biology, an <A HREF="glossary.html#RNA">RNA</A> molecule 
    <A HREF="glossary.html#complementary">complementary</A> to the strand normally 
    processed into mRNA and <A HREF="glossary.html#translation">translated</A>. 
    <P> 
  <DT><B><A NAME="apoptosis">apoptosis</A></B> 
  <DD>Programmed cell death, that is, the death of cells by a specific sequence 
    of events triggered in the course of normal development (e.g. cells between 
    digits in the limb bud) or as a means of preserving normal function (e.g. 
    in response to <A HREF="glossary.html#virus">viral</A> infection). 
    <P> 
  <DT><B><A NAME="approved">approved</A></B> 
  <DD>With respect to a <A HREF="glossary.html#gene_symbol">gene symbol</A>, 
    <A HREF="glossary.html#gene_name">gene name</A>, or <A HREF="glossary.html#allele_designation">allele 
    designation </A> within ZFIN, an "approved" symbol or name is one that has 
    been assigned by the <a href="nomen_comm.html">Zebrafish Nomenclature Committee</a>. 
    <P> 
  <DT><B><A NAME="ATCC">ATCC</A></B>  
  <DD>American Type Culture Collection. A large collection of microbial stocks, 
    including microbes containing DNA segments. See the <A HREF="http://www.atcc.org/">ATCC 
    home page</A> for further information. 
    <P> 
  <DT><B><A NAME="autoradiography">autoradiography</A></B> 
  <DD>The detection of a unstable isotope that emits <A HREF="glossary.html#radiation">radiation</A> 
    by a photographic emulsion. In the case of <A HREF="glossary.html#in_situ"><I>in 
    situ</I> hybridization</A>, this involves dipping microscope slides in liquid 
    emulsion. In the case of <A HREF="glossary.html#southern_blot">Southern blots</A>, 
    <A HREF="glossary.html#northern_blot">Northern blots</A>, or <A HREF="glossary.html#western_blot">Western 
    blots</A>, the membrane is placed next to a sheet of X-ray film. 
    <P> 
  <DT><B><A NAME="autosome">autosome</A></B> 
  <DD>Any <A HREF="glossary.html#chromosome">chromosome</A> that is not a <A HREF="glossary.html#sex_chromosome">sex 
    chromosome</A>. 
    <!-- BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB -->
    <A NAME="b"></A> 
    <P> 
  <DT><B><A NAME="BAC">BAC</A></B> 
  <DD>Bacterial Artificial Chromosome. A type of <A HREF="glossary.html#cloning_vector">cloning 
    vector</A> derived from the naturally-occurring F factor episome. A BAC can 
    carry 100 - 200 kb of foreign <A HREF="glossary.html#DNA">DNA</A>. 
    <P> 
  <DT><B><A NAME="bac_end">BAC/YAC end</A></B> 
  <DD> BAC/YAC end refers to sequences at the end of foreign <A HREF="glossary.html#DNA">DNA</A> 
    inserts in a <A HREF="glossary.html#BAC">BAC</A> or <A HREF="glossary.html#YAC">YAC</A>. 
    These sequences are a source of <A HREF="glossary.html#STS">STSs</A> to determine 
    the extent of overlap between BACs or YACs and to aid in the alignment of 
    sequence <A HREF="glossary.html#contig">contigs</A>. 
    <P> 
  <DT><B><A NAME="backcross">backcross</A></B> 
  <DD>A type of genetic <A HREF="glossary.html#cross">cross</A> in which a <A HREF="glossary.html#hybrid">hybrid</A> 
    strain is crossed to one of the two parental strains. <BR>
    See also: 
    <UL>
      <LI><A HREF="glossary.html#incross">incross</A></LI>
      <LI><A HREF="glossary.html#intercross">intercross</A></LI>
      <LI><A HREF="glossary.html#outcross">outcross</A></LI>
      <LI><A HREF="glossary.html#testcross">testcross</A></LI>
    </UL>
    <P> 
  <DT><B><A NAME="bacteriophage">bacteriophage</A></B> 
  <DD>A <A HREF="glossary.html#virus">virus</A> that infects bacteria. 
    <P> 
  <DT><B><A NAME="base">base</A></B> 
  <DD>One of a set of nitrogenous compounds attached to the sugar-phosphate backbone 
    in a <A HREF="glossary.html#nucleic_acid">nucleic acid</A>. In <A HREF="glossary.html#DNA">DNA</A>, 
    the <A HREF="glossary.html#purine">purine</A> bases are adenine (A) and guanine 
    (G), whereas the <A HREF="glossary.html#pyrimidine">pyrimidine</A> bases 
    are cytosine (C) and thymine (T). In <A HREF="glossary.html#RNA">RNA</A>, 
    the purine bases are adenine (A) and guanine (G), whereas the pyrimidine bases 
    are cytosine (C) and uracil (U). See the <A 

HREF="http://www.nhgri.nih.gov/DIR/VIP/Glossary/Illustration/rna.html">Figure</A> 
    at <A HREF="glossary.html#NHGRI">NHGRI</A>. 
    <P> 
  <DT><B><A NAME="base_pair">base pair (bp)</A></B> 
  <DD>In double-stranded <A HREF="glossary.html#nucleic_acid">nucleic acids</A>, 
    a "base pair" is the structure formed between two complementary <A HREF="glossary.html#nucleotide">nucleotides</A> 
    by hydrogen bonding. In <A HREF="glossary.html#DNA">DNA</A>, adenine (A) 
    pairs with thymine (T) and cytosine (C) pairs with guanine (G). In <A HREF="glossary.html#RNA">RNA</A>, 
    adenine (A) pairs with uracil (U) and cytosine (C) pairs with guanine (G). 
    See the <A 

HREF="http://www.nhgri.nih.gov/DIR/VIP/Glossary/Illustration/base_pair.html">Figure</A> 
    at <A HREF="glossary.html#NHGRI">NHGRI</A>. 
    <P> 
  <DT><B><A NAME="bioinformatics">bioinformatics</A></B> 
  <DD>The acquisition, storage, arrangement, analysis, display and communication 
    of information related to the biology of living things, generally assisted 
    by the use of computers. 
    <P> 
  <DT><B><A NAME="biological_process">biological process</A></B> 
  <DD>Refers to a broad category of biological tasks accomplished via one or more 
    ordered assemblies of <A HREF="glossary.html#molecular_function">molecular 
    functions</A>. Usually there is some temporal aspect to it, although a process 
    event may be essentially instantaneous. It often involves transformation, 
    in the sense that something goes into a process and something different comes 
    out of it. Examples of biological processes included in this category are 
    cell growth and maintenance, signal transduction, <A HREF="glossary.html#pyrimidine">pyrimidine</A> 
    metabolism, and <A HREF="glossary.html#cAMP">cAMP</A> <A HREF="glossary.html#biosynthesis">biosynthesis</A>. 
    In the GO Project vocabularies, Biological Process is a primary class of terms. 
    See the <A HREF="http://www.geneontology.org/">GO Consortium</A> site for 
    further information. 
    <P> 
  <DT><B><A NAME="biosynthesis">biosynthesis</A></B> 
  <DD>Synthesis of chemical compounds by <A HREF="glossary.html#enzyme">enzymatic</A> 
    processes in living organisms. 
    <P> 
  <DT><B><A NAME="biotin">biotin</A></B> 
  <DD>One of the water-soluble B vitamins. It is useful in molecular biology as 
    a chemical tag on <A HREF="glossary.html#nucleic_acid">nucleic acid</A> <A HREF="glossary.html#probe">probes</A> 
    or <A HREF="glossary.html#antibody">antibodies</A>, because the biotin-scavenging 
    <A HREF="glossary.html#protein">proteins</A> avidin and streptavidin bind 
    biotin with high affinity. These biotin-binding proteins can be coupled to 
    fluorescent dyes, <A HREF="glossary.html#enzyme">enzymes</A> that can be 
    detected using <A HREF="glossary.html#chromogenic">chromogenic</A> reactions, 
    or <A HREF="glossary.html#colloidal_gold">colloidal gold</A>, allowing detection 
    of biotin-labeled probes or antibodies on <A HREF="glossary.html#southern_blot">Southern 
    blots</A>, <A HREF="glossary.html#northern_blot">Northern blots</A>, <A HREF="glossary.html#western_blot">Western 
    blots</A>, or cytological preparations. 
    <P> 
  <DT><B><A NAME="BLAST">BLAST</A></B> 
  <DD>Basic Local Alignment and Searching Tool. A sequence comparison <A 

HREF="http://www.ncbi.nlm.nih.gov/Education/BLASTinfo/BLAST_algorithm.html">algorithm</A> 
    optimized for speed, which is used to search sequence databases for optimal 
    local alignments to a query sequence. More information is available at <A HREF="http://www.ncbi.nlm.nih.gov/BLAST/blast_overview.html">NCBI</A>. 
    <!-- CCCCCCCCCCCCCCCCCCCCCCCCCCCCCC -->
    <A NAME="c"></A> 
    <P> 
  <DT><B><A NAME="cAMP">cAMP</A></B> 
  <DD>Cyclic AMP. A form of the <A HREF="glossary.html#nucleotide">nucleotide</A> 
    adenosine monophosphate that serves as a signaling molecule within and between 
    cells. 
    <P> 
  <DT><B><A NAME="carboxyl_terminus">carboxyl terminus</A></B> 
  <DD>A term that identifies one end of a <A HREF="glossary.html#protein">protein</A> 
    molecule. The carboxyl terminus is the end of the molecule that terminates 
    in a free carboxyl group. See the <A 

HREF="http://www.nhgri.nih.gov/DIR/VIP/Glossary/Illustration/amino_acid.html">Figure</A> 
    at <A HREF="glossary.html#NHGRI">NHGRI</A>. See also <A HREF="glossary.html#amino_acid">amino 
    acid</A>, <A HREF="glossary.html#central_dogma">central dogma</A>. 
    <P> 
  <DT><B><A NAME ="catabolism">catabolism</A></B> 
  <DD>Degradation of chemical compounds into compounds having a lower molecular 
    weight by <A HREF="glossary.html#enzyme">enzymatic</A> processes in living 
    organisms. 
    <P> 
  <DT><B><A NAME="cDNA">cDNA</A></B> 
  <DD>Complementary DNA. A <A HREF="glossary.html#DNA">DNA</A> copy of an <A HREF="glossary.html#mRNA">mRNA</A> 
    or complex sample of mRNAs, made using <A HREF="glossary.html#reverse_transcriptase">reverse 
    transcriptase</A>. 
    <P> 
  <DT><B><A NAME="cellcomponent">cellular component</A></B> 
  <DD>Refers to subcellular structures, locations, and macromolecular complexes. 
    Some examples are <A HREF="glossary.html#nucleus">nucleus</A>, <A HREF="glossary.html#telomere">telomere</A> 
    and <B>origin recognition complex</B>. In the GO Project vocabularies, Cellular 
    Component is a primary class of terms. See the <A HREF="http://www.geneontology.org/">GO 
    Consortium</A> site for further information. 
    <P> 
  <DT><B><A NAME="centimorgan">centimorgan</A></B> 
  <DD>A unit of length in a <A HREF="glossary.html#genetic_map">genetic map</A>. 
    Two <A HREF="glossary.html#locus">loci</A> are 1 cM apart if <A HREF="glossary.html#recombination">recombination</A> 
    is detected between them in 1% of <A HREF="glossary.html#meiosis">meioses</A>. 
    <P> 
  <DT><B><A NAME="central_dogma">central dogma</A></B> 
  <DD>The principal statement of the molecular basis of inheritance. In its simplest 
    form:<BR>
    <BR>
    " <A HREF="glossary.html#DNA">DNA</A> makes <A HREF="glossary.html#RNA">RNA</A> 
    makes <A HREF="glossary.html#protein">protein</A>."<BR>
    <BR>
    This means that (generally) genetic information is stored in and transmitted 
    as DNA. <A HREF="glossary.html#gene">Genes</A> are <A HREF="glossary.html#gene_expression">expressed</A> 
    by being copied as RNA ( <A HREF="glossary.html#transcription">transcription</A>), 
    which is processed into <A HREF="glossary.html#mRNA">mRNA</A> via <A HREF="glossary.html#splicing">splicing</A> 
    and <A HREF="glossary.html#polyadenylation">polyadenylation</A>. The information 
    in mRNA is <A HREF="glossary.html#translation">translated</A> into a protein 
    sequence using a <A HREF="glossary.html#genetic_code">genetic code</A> to 
    interpret three-base <A HREF="glossary.html#codon">codons</A> as instructions 
    to add one of twenty <A HREF="glossary.html#amino_acid">amino acids</A> or 
    to stop translation. See the <A 

HREF="http://www.accessexcellence.org/AB/GG/central.html">Figure</A> at Access 
    Excellence or the <A 

HREF="http://www.nhgri.nih.gov/DIR/VIP/Glossary/Illustration/mrna.html">Figure</A> 
    at <A HREF="glossary.html#NHGRI">NHGRI</A>. 
    <P> 
  <DT><B><A NAME="centromere">centromere</A></B> 
  <DD>In zebrafish genetics, the primary constriction of a <A HREF="glossary.html#chromosome">chromosome</A> 
    separating it into the short arm (p) and the long arm (q). The centromere 
    is the chromosomal region over which the <A HREF="glossary.html#kinetochore">kinetochore</A> 
    is organized. See the <A 

HREF="http://www.nhgri.nih.gov/DIR/VIP/Glossary/Illustration/chromosome.html">Figure</A> 
    at <A HREF="glossary.html#NHGRI">NHGRI</A>. 
    <P> 
  <DT><B><A NAME="chiasma">chiasma</A></B> 
  <DD>The cytologically visible consequence of a <A HREF="glossary.html#recombination">reciprocal 
    recombination</A> event in <A HREF="glossary.html#meiosis">meiosis</A>, observable 
    in the later stage of meiotic prophase. Chiasmata hold <A HREF="glossary.html#homolog">homologous</A> 
    <A HREF="glossary.html#chromosome">chromosomes</A> together prior to anaphase 
    of the first meiotic division. 
    <P> 
  <DT><B><A NAME="children">children</A></B> 
  <DD>In bioinformatics, this term refers to terms in a hierarchical <A HREF="glossary.html#controlled_vocabulary">controlled 
    vocabulary</A> like ones containing <A HREF="glossary.html#gene_ontology">Gene 
    Ontology (GO)</A> terms. A "child" of a term is a term any number of levels 
    below it in the hierarchy that is a descendant of the term. For example, the 
    GO term <B>alcohol dehydrogenase</B> [GO:0004022] is a child of the GO term 
    <A HREF="glossary.html#enzyme">enzyme</A> [GO:0003824]. See also: <A HREF="glossary.html#ancestor">ancestor</A>, 
    <A HREF="glossary.html#sibling">sibling</A>. 
    <P> 
  <DT><B><A NAME="chimera">chimera</A></B> 
  <DD>1. An animal formed from two different animals, that is from two different 
    embryonic sources. See also <a href="glossary.html#mosaic">mosaic</a>.<BR>
    2. A clone containing <A HREF="glossary.html#genome">genomic</A> <A HREF="glossary.html#DNA">DNA</A> 
    from nonadjacent genomic segments or <A HREF="glossary.html#cDNA">cDNA</A> 
    from two different <A HREF="glossary.html#mRNA">mRNAs</A> (see <A HREF="glossary.html#cloning_artifact">cloning 
    artifact</A>). 
    <P> 
  <DT><B><A NAME="chlorambucil">chlorambucil</A></B> 
  <DD>A chemical <A HREF="glossary.html#mutagen">mutagen</A>, also called nitrogen 
    mustard. 
    <P> 
  <DT><B><A NAME="chromatin">chromatin</A></B> 
  <DD>The <A HREF="glossary.html#nucleus">nuclear</A> material that makes up 
    <A HREF="glossary.html#chromosome">chromosomes</A>, consisting of <A HREF="glossary.html#DNA">DNA</A> 
    and <A HREF="glossary.html#protein">protein</A>. See also <A HREF="glossary.html#euchromatin">euchromatin</A>, 
    <A HREF="glossary.html#heterochromatin">heterochromatin</A>. 
    <P> 
  <DT><B><A NAME="chromogenic">chromogenic</A></B> 
  <DD>Color-generating. A chromogenic <A HREF="glossary.html#substrate">substrate</A> 
    is colorless until acted upon by an <A HREF="glossary.html#enzyme">enzyme</A>; 
    it then becomes an insoluble pigment. 
    <P> 
  <DT><B><A NAME="chromosome">chromosome</A></B> 
  <DD>A structural unit within a eukaryotic <A HREF="glossary.html#nucleus">nucleus</A> 
    that carries <A HREF="glossary.html#gene">genes</A>. A chromosome consists 
    of a long, continuous strand of <A HREF="glossary.html#DNA">DNA</A> and associated 
    <A HREF="glossary.html#protein">proteins</A>. See the <A 

HREF="http://www.nhgri.nih.gov/DIR/VIP/Glossary/Illustration/chromosome.html">Figure</A> 
    at <A HREF="glossary.html#NHGRI">NHGRI</A>. 
    <P> 
  <DT><B><A NAME="chromosome_aberration">chromosome aberration</A></B> 
  <DD>1. A <A HREF="glossary.html#chromosome_rearrangement">chromosome rearrangement</A>.<BR>
    2. Less commonly, a <A HREF="glossary.html#numerical_aberration">numerical 
    aberration</A> in the number of chromosomes. 
    <P> 
  <DT><B><A NAME="chromosome_rearrangement">chromosome rearrangement</A></B> 
  <DD>A kind of <A HREF="glossary.html#mutation">mutation</A> in which there 
    is a change in the arrangement of the <A HREF="glossary.html#genome">genome</A> 
    into <A HREF="glossary.html#chromosome">chromosomes</A>; this term usually 
    applies to those changes that are visible <A HREF="glossary.html#cytogenetic">cytogenetically</A>. 
    <BR>
    Classes of chromosome rearrangements include: 
    <UL>
      <LI><A HREF="glossary.html#deficiency">deficiency</A></LI>
      <LI><A HREF="glossary.html#duplication">duplication</A></LI>
      <LI><A HREF="glossary.html#insertion">insertion</A></LI>
      <LI><A HREF="glossary.html#inversion">inversion</A></LI>
      <LI><A HREF="glossary.html#translocation">translocation</A></LI>
      <LI><A HREF="glossary.html#transposition">transposition</a></LI>
    </UL>
    <P> 
  <DT><B><A NAME="clone">clone</A></B> 
  <DD>1. A segment of <A HREF="glossary.html#DNA">DNA</A> contained within a 
    <A HREF="glossary.html#cloning_vector">cloning vector</A>.<BR>
    2. An organism derived from a founding individual by asexual means that is 
    genetically identical to the founding individual. 
    <P> 
  <DT><B><A NAME="cloning_artifact">cloning artifact</A></B> 
  <DD>A <A HREF="glossary.html#DNA">DNA</A> <A HREF="glossary.html#clone">clone</A> 
    whose structure does not accurately represent <A HREF="glossary.html#genome">genomic</A> 
    or <A HREF="glossary.html#mRNA">mRNA</A> sequence, due to errors in the cloning 
    process. For example, two noncontiguous genomic fragments may be joined by 
    <A HREF="glossary.html#ligate">ligation</A> prior to being incorporated into 
    the <A HREF="glossary.html#cloning_vector">cloning vector</A>. 
    <P> 
  <DT><B><A NAME="cloning_vector">cloning vector</A></B> 
  <DD>A <A HREF="glossary.html#DNA_construct">DNA construct</A> capable of <A HREF="glossary.html#replication">replication</A> 
    within a bacterial or yeast host that can harbor foreign <A HREF="glossary.html#DNA">DNA</A>, 
    facilitating experimental manipulation of that DNA segment. 
    <P> 
  <DT><B><A NAME="cM">cM</A></B> 
  <DD>See <A HREF="glossary.html#centimorgan">centimorgan</A>. 
    <P> 
  <DT><B><A NAME="coding_region">coding region</A></B> 
  <DD>That part of a <A HREF="glossary.html#gene">gene</A> whose sequence is 
    read as <A HREF="glossary.html#codon">codons</A> during <A HREF="glossary.html#translation">translation</A>. 
    See also <A HREF="glossary.html#central_dogma">central dogma</A>, <A HREF="glossary.html#transcription">transcription</A>, 
    <A HREF="glossary.html#translation">translation</A>, <A HREF="glossary.html#intron">intron</A>, 
    e <A HREF="glossary.html#exon">Exon</A>. 
    <!--addition of x-refs- 1/19 - codominant-->
    <P> 
  <DT><B><A NAME="codominant">codominant</A></B> 
  <DD>One of a series of terms applied to the <A HREF="glossary.html#phenotype">phenotypic</A> 
    effect of a particular <A HREF="glossary.html#allele">allele</A> in reference 
    to another allele (usually the standard <A HREF="glossary.html#wild_type">wild-type</A> 
    allele) with respect to a given <A HREF="glossary.html#trait">trait</A>. 
    An allele "a" is said to be codominant with respect to the wild-type allele 
    "A" if the A/a <A HREF="glossary.html#heterozygote">heterozygote</A> fully 
    expresses both of the <A HREF="glossary.html#phenotype">phenotypes</A> associated 
    with the a/a and A/A <A HREF="glossary.html#homozygote">homozygotes</A>. 
    An example of codominance is the <A HREF="http://www3.ncbi.nlm.nih.gov/htbin-post/Omim/dispmim?110300">ABO</A> 
    blood type <A HREF="glossary.html#antigen">antigens</A> in humans, where 
    AA individuals are type A, BB individuals are type B, and AB individuals are 
    type AB. See also <A HREF="glossary.html#dominant">dominant</A>, <A HREF="glossary.html#recessive">recessive</A>, 
    s <A HREF="glossary.html#semidominant">Semidominant</A>. 
    <P> 
  <DT><B><A NAME="codon">codon</A></B> 
  <DD>Three <A HREF="glossary.html#base">bases</A> in a <A HREF="glossary.html#DNA">DNA</A> 
    or <A HREF="glossary.html#RNA">RNA</A> sequence that specify an <A HREF="glossary.html#amino_acid">amino 
    acid</A> or a termination signal (stop codon). See the <A 

HREF="http://www.nhgri.nih.gov/DIR/VIP/Glossary/Illustration/codon.html">Figure</A> 
    at <A HREF="glossary.html#NHGRI">NHGRI</A>. See also <A HREF="glossary.html#central_dogma">central dogma</A>. 
    <P> 
  <DT><B><A NAME="coisogenic">coisogenic</A></B> 
  <DD>A strain that differs from a particular <A HREF="glossary.html#inbred_strain">inbred 
    strain</A> at only one <A HREF="glossary.html#locus">locus</A>. A coisogenic 
    strain arises when a <A HREF="glossary.html#mutation">mutation</A> occurs 
    in an inbred strain. The coisogenic strain can be propagated by <A HREF="glossary.html#intercross">intercrossing</A> 
    <A HREF="glossary.html#heterozygote">heterozygotes</A> to produce <A HREF="glossary.html#homozygote">homozygotes</A>; 
    if these are nonviable the strain can be maintained by <A HREF="glossary.html#backcross">backcrossing</A> 
    heterozygotes to the original <A HREF="glossary.html#inbred_strain">inbred 
    strain</A>. 
    <P> 
  <DT><B><A NAME="colloidal_gold">colloidal gold</A></B> 
  <DD>Fine particles of gold (on the order of 5-20 nm diameter) that can be coupled 
    to <A HREF="glossary.html#antibody">antibodies</A> or other <A HREF="glossary.html#protein">proteins</A>, 
    allowing the detection of the binding of the labeled proteins by electron 
    microscopy. 
    <P> 
  <DT><B><A NAME="complementary">complementary sequence</A></B> 
  <DD>A single-stranded <A HREF="glossary.html#nucleic_acid">nucleic acid</A> 
    that would bind to a given single-stranded nucleic acid by <A HREF="glossary.html#base_pair">base 
    pairing</A>. 
    <P> 
  <DT><B><A NAME="complementation">complementation</A></B> 
  <DD>The appearance of only <A HREF="glossary.html#wild_type">wild-type</A> 
    <A HREF="glossary.html#phenotype">phenotypes</A> in <A HREF="glossary.html#hybrid">hybrid</A> 
    offspring from two <A HREF="glossary.html#mutant">mutant</A> individuals 
    <A HREF="glossary.html#homozygous">homozygous</A> or <a href="glossary.html#heterozygous">heterozygous</a> 
    for <A HREF="glossary.html#recessive">recessive</A> <A HREF="glossary.html#mutation">mutations</A>. 
    Complementation shows that the two parental mutant individuals have mutant 
    <A HREF="glossary.html#allele">alleles</A> of different <A HREF="glossary.html#gene">genes</A>, 
    even if they are phenotypically similar. 
    <P> 
  <DT><B><A NAME="complex">complex/cluster/region</A></B> 
  <DD>In <A HREF="glossary.html#ZFIN">ZFIN</A>, the <A HREF="glossary.html#marker">marker</A> 
    type "Complex/Cluster/Region" is used to refer to any of the following: 
    <OL>
      <LI><A HREF="glossary.html#gene_complex">Gene complex</A>; a group of <A HREF="glossary.html#gene">genes</A> 
        linked closely together that are related <A HREF="glossary.html#evolution">evolutionarily</A> 
        or <A HREF="glossary.html#molecular_function">functionally</A>. Interspersed 
        unrelated genes located within the group are included. 
      <LI>A segment of the zebrafish <A HREF="glossary.html#genome">genome</A> 
        defined by comparison to an <A HREF="glossary.html#ortholog">orthologous</A> 
        segment in the genome of another species, or by some specific characteristic, 
        such as loss of <A HREF="glossary.html#heterozygous">heterozygosity</A>. 
      <LI>A <A HREF="glossary.html#marker">marker</A> repository for information 
        pertaining to a specific <A HREF="glossary.html#gene_family">gene family</A>, 
        where such information lacks precise family member resolution. 
    </OL>
    <P> 
  <DT><B><A NAME="conserved_synteny">conserved synteny</A></B> 
  <DD>The occurrence of <A HREF="glossary.html#synteny">synteny</A> of <A HREF="glossary.html#ortholog">orthologous</A> 
    <A HREF="glossary.html#gene">genes</A> in two different organisms. Conserved 
    synteny between zebrafish and human or mouse does not extend over entire <A HREF="glossary.html#chromosome">chromosomes</A>. 
    <P> 
  <DT><B><A NAME="contig">contig</A></B> 
  <DD>1. A physical map of contiguous genomic <A HREF="glossary.html#DNA">DNA</A> 
    assembled using overlapping cloned segments (see <A HREF="glossary.html#STS">STS</A>).<BR>
    2. A contiguous DNA sequence assembled using overlapping DNA sequences. 
    <P> 
  <DT><B><A NAME="controlled_vocabulary">controlled vocabulary</A></B> 
  <DD>A restricted set of defined terms allowing the representation of complex 
    information in a database. See, for example, <A HREF="glossary.html#gene_ontology">Gene 
    Ontology</A>. 
    <P> 
  <DT><B><A NAME="cosmid">cosmid</A></B> 
  <DD>A type of <A HREF="glossary.html#cloning_vector">cloning vector</A> derived 
    from <A HREF="glossary.html#bacteriophage">bacteriophage</A> lambda. A cosmid 
    can carry about 40 kb of foreign <A HREF="glossary.html#DNA">DNA</A>. 
    <P> 
  <DT><B><A NAME="Cre">cre recombinase</A></B> 
  <DD>A <A HREF="glossary.html#site-specific_recombination">site-specific recombination</A> 
    <A HREF="glossary.html#enzyme">enzyme</A> that recognizes the 34 base pair 
    <A HREF="glossary.html#loxP">loxP</A> sequence. 
    <P> 
  <DT><B><A NAME="cross">cross</A></B> 
  <DD>An experimental mating of two genetically distinct sexually reproducing 
    organisms. <BR>
    See also: 
    <UL>
      <LI><A HREF="glossary.html#backcross">backcross</A></LI>
      <LI><A HREF="glossary.html#incross">incross</A></LI>
      <LI><A HREF="glossary.html#intercross">intercross</A></LI>
      <LI><A HREF="glossary.html#outcross">outcross</A></LI>
      <LI><A HREF="glossary.html#testcross">testcross</A></LI>
    </UL>
    <P> 
  <DT><B><A NAME="cross-hybridization">cross-hybridization</A></B> 
  <DD>With respect to <A HREF="glossary.html#nucleic_acid">nucleic acids</A>, 
    "cross-hybridization" refers to the formation of double-stranded <A HREF="glossary.html#DNA">DNA</A>, 
    <A HREF="glossary.html#RNA">RNA</A>, or DNA/RNA hybrids by complementary 
    <A HREF="glossary.html#base_pair">base pairing</A> between two molecules 
    that are not identical in sequence. Cross-hybridization may be observed between 
    nucleic acids derived from <A HREF="glossary.html#ortholog">orthologous</A> 
    or <A HREF="glossary.html#paralog">paralogous</A> genes. 
    <P> 
  <DT><B><A NAME="crossover">crossover</A></B> 
  <DD>A reciprocal <A HREF="glossary.html#recombination">recombination</A> event. 
    <P> 
  <DT><B><A NAME="Cy5">Cy5</A></B> 
  <DD>A fluorescent dye used to label <A HREF="glossary.html#DNA">DNA</A> <A HREF="glossary.html#probe">probes</A> 
    for <A HREF="glossary.html#FISH">FISH</A> or <A HREF="glossary.html#antibody">antibodies</A> 
    for <A HREF="glossary.html#immunofluorescence">immunofluorescence</A> or 
    <A HREF="glossary.html#western_blot">Western blots</A>. Also used to label 
    nucleic acid probes for <a href="glossary.html#microarray">microarray</a> 
    analysis. 
    <P> 
  <DT><B><A NAME="cytogenetic">cytogenetic</A></B> 
  <DD>Refers to the correlation of genetic and cytological information through 
    the microscopic analysis of stained preparations of <A HREF="glossary.html#chromosome">chromosomes</A>, 
    including those from individuals carrying <A HREF="glossary.html#mutation">mutations</A>. 
    <P> 
  <DT><B><A NAME="cytogenetic_band">cytogenetic band</A></B> 
  <DD>One of the subregions of a <A HREF="glossary.html#chromosome">chromosome</A> 
    visible microscopically after special staining. 
    <P> 
  <DT><B><A NAME="cytogenetic_map">cytogenetic map</A></B> 
  <DD>A type of <A HREF="glossary.html#genetic_map">genetic map</A> relating 
    <A HREF="glossary.html#gene">gene</A> positions to <A HREF="glossary.html#chromosome">chromosomal</A> 
    banding patterns. The maps are built from relating the positions of genes 
    to <A HREF="glossary.html#cytogenetic_marker">cytogenetic markers</A> or 
    by <A HREF="glossary.html#in_situ"><I>in situ</I> hybridization</A>. 
    <P> 
  <DT><B><A NAME="cytogenetic_marker">cytogenetic marker</A></B> 
  <DD>1. A structure within a <A HREF="glossary.html#chromosome">chromosome</A> 
    that is visible by microscopic examination, possibly after special staining 
    methods are used.<BR>
    2. A <A HREF="glossary.html#chromosome_rearrangement">chromosome rearrangement</A> 
    that is visible by microscopic examination. 
    <P> 
  <DT><B><A NAME="cytoplasm">cytoplasm</A></B> 
  <DD>That part of a eukaryotic cell that is not the <A HREF="glossary.html#nucleus">nucleus</A>. 
    <!-- DDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDD -->
    <A NAME="d"></A> 
    <P> 
  <DT><B><A NAME="degenerate">degenerate</A></B> 
  <DD>A term describing one of the qualities of the <A HREF="glossary.html#genetic_code">genetic 
    code</A>, specifically, that some amino acids may be specified by more than 
    one <A HREF="glossary.html#codon">codon</A>. 
    <P> 
  <DT><B><A NAME="deficiency">deficiency</A></B> 
  <DD>A type of <A HREF="glossary.html#mutation">mutation</A> caused by loss 
    of one or more <A HREF="glossary.html#nucleotide">nucleotides</A> from a 
    <A HREF="glossary.html#DNA">DNA</A> segment. Deficiencies can be very large, 
    encompassing many <A HREF="glossary.html#gene">genes</A> and <A HREF="glossary.html#megabase">megabases</A> 
    of DNA, to the point of producing a visible cytological abnormality in a <A HREF="glossary.html#chromosome">chromosome</A>. 
    Small deficiencies within a gene can alter the <A HREF="glossary.html#reading_frame">reading 
    frame</A>, and thus the <A HREF="glossary.html#amino_acid">amino acid</A> 
    sequence of the encoded <A HREF="glossary.html#protein">protein</A>. See 
    the <A HREF="http://www.nhgri.nih.gov/DIR/VIP/Glossary/Illustration/deletion.html">Figure</A> 
    at NHGRI. 
    <P> 
  <DT><B><A NAME="denaturation">denaturation</A></B> 
  <DD>1. The separation of the two strands of a double-stranded <A HREF="glossary.html#nucleic_acid">nucleic 
    acid</A> caused by treatments that overcome hydrogen bonding, e.g. heat. <BR>
    2. A usually irreversible change in the conformation of a <A HREF="glossary.html#protein">protein</A> 
    caused by treatments that overcome hydrogen bonding, <A HREF="glossary.html#hydrophobic">hydrophobic</A> 
    interactions, or other chemical forces that maintain the structure of proteins, 
    e.g. heat. 
    <P> 
  <DT><B><A NAME="dimorphic">dimorphic</A></B> 
  <DD>Having two forms. 
    <P> 
  <DT><B><A NAME="diploid">diploid</A></B> 
  <DD>Having twice the <A HREF="glossary.html#chromosome">chromosome</A> number 
    normally found in a <A HREF="glossary.html#gamete">gamete</A>. Normal zebrafish 
    are diploid, having a chromosome set from the maternal gamete (the egg) and 
    a chromosome set from the paternal gamete (the sperm). See also <A HREF="glossary.html#haploid">haploid</A>. 
    <P> 
  <DT><B><A NAME="DNA">DNA</A></B> 
  <DD>Deoxyribonucleic acid. The <A HREF="glossary.html#nucleic_acid">nucleic 
    acid</A> of which <A HREF="glossary.html#gene">genes</A> are made. See the 
    <A HREF="http://www.nhgri.nih.gov/DIR/VIP/Glossary/Illustration/dna.html">Figure</A> 
    at <A HREF="glossary.html#NHGRI">NHGRI</A>. See also <A HREF="glossary.html#central_dogma">central 
    dogma</A>, <A HREF="glossary.html#nucleic_acid">nucleic acid</A>, and <A HREF="glossary.html#RNA">RNA</A>. 
    <P> 
  <DT><B><A NAME="DNA_construct">DNA construct</A></B> 
  <DD>An assembly of <A HREF="glossary.html#DNA">DNA</A> sequences made <A HREF="glossary.html#in_vitro"><I>in 
    vitro</I></A> to serve an experimental purpose. 
    <P> 
  <DT><B><A NAME="DNA_mapping_panel">DNA mapping panel</A></B> 
  <DD>A data set obtained by DNA typing of <A HREF="glossary.html#polymorphism">polymorphic</A> 
    <A HREF="glossary.html#marker">markers</A> in <a href="glossary.html#hybrid">hybrid</a> 
    <A HREF="glossary.html#backcross">crosses</A> of zebrafish. See a list of 
    zebrafish Mapping Panels at <A HREF="glossary.html#ZFIN">ZFIN</A>. 
    <P> 
  <DT><B><A NAME="DNA_polymerase">DNA polymerase</A></B> 
  <DD>An <A HREF="glossary.html#enzyme">enzyme</A> that carries out <A HREF="glossary.html#replication">replication</A>. 
    <P> 
  <DT><B><A NAME="DNA_segment">DNA segment</A></B> 
  <DD>1. A length of <A HREF="glossary.html#DNA">DNA</A>. <BR>
    2. In <A HREF="glossary.html#ZFIN">ZFIN</A>, a DNA segment is a genomic feature 
    recognized by <A HREF="glossary.html#anonymous_DNA">anonymous DNA</A> probes. 
    <A HREF="glossary.html#gene_symbol">Symbols</A> for such segments most commonly 
    represent <A HREF="glossary.html#intergenic">intergenic</A> <A HREF="glossary.html#marker">markers</A> 
    used in <A HREF="glossary.html#genetic_map">genetic mapping</A>. 
    <P> 
  <DT><B><A NAME="dNTP">dNTP</A></B> 
  <DD>Deoxyribonucleotide triphosphate. A generic term referring to the four deoxyribonucleotides: 
    dATP, dCTP, dGTP and dTTP. See the <A HREF="http://www.nhgri.nih.gov/DIR/VIP/Glossary/Illustration/rna.html">Figure</A> 
    at <A HREF="glossary.html#NHGRI">NHGRI</A>. 
    <!--addition of x-refs 1/19 to Dominant  -->
    <P> 
  <DT><B><A NAME="dominant">dominant</A></B> 
  <DD>One of a series of terms applied to the <A HREF="glossary.html#phenotype">phenotypic</A> 
    effect of a particular <A HREF="glossary.html#allele">allele</A> in reference 
    to another allele (usually the standard <A HREF="glossary.html#wild_type">wild-type</A> 
    allele) with respect to a given <A HREF="glossary.html#trait">trait</A>. 
    An allele "A" is said to be dominant with respect to the allele "a" if the 
    A/A <A HREF="glossary.html#homozygote">homozygote</A> and the A/a <A HREF="glossary.html#heterozygote">heterozygote</A> 
    are phenotypically identical and different from the a/a homozygote. See also 
    <A HREF="glossary.html#codominant">codominant</A>, <A HREF="glossary.html#recessive">recessive</A>, 
    <A HREF="glossary.html#semidominant">semidominant</A>. 
    <P> 
  <DT><B><A NAME="duplication">duplication</A></B> 
  <DD>An additional copy of a <A HREF="glossary.html#DNA">DNA</A> segment present 
    in the <A HREF="glossary.html#genome">genome</A>. <A HREF="glossary.html#gene">Gene</A> 
    duplication is the source of <A HREF="glossary.html#paralog">paralogous</A> 
    genes. See the <A HREF="http://www.nhgri.nih.gov/DIR/VIP/Glossary/Illustration/duplication.html">Figure</A> 
    at NHGRI. 
    <!-- EEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE -->
    <A NAME="e"></A> 
    <P> 
  <DT><B><A NAME="EC_number">EC number</A></B> 
  <DD>A number assigned to a type of <A HREF="glossary.html#enzyme">enzyme</A> 
    according to a scheme of standardized enzyme nomenclature developed by the 
    Enzyme Commission of the Nomenclature Committee of the International Union 
    of Biochemistry and Molecular Biology (IUBMB). EC numbers may be found in 
    ENZYME, the Enzyme nomenclature database, maintained at the <A HREF="http://www.expasy.ch/enzyme/">ExPASy</A> 
    molecular biology server of the Geneva University Hospital and the University 
    of Geneva, Switzerland. 
    <P> 
  <DT><B><A NAME="early_pressure">early pressure (EP)</A></B> 
  <DD><a href="zfbook/chapt7/7.5.html">Early pressure</a> method to produce <a href="glossary.html#homozygous">homozygous</a> 
    <a href="glossary.html#diploid">diploid</a> offspring. 
    <P> 
  <DT><B><A NAME="electroporation">electroporation</A></B> 
  <DD>The use of strong, brief pulses of electric current to create temporary 
    holes in cell <A HREF="glossary.html#membrane">membranes</A>, allowing the 
    introduction of <A HREF="glossary.html#DNA">DNA</A> or other molecules. 
    <P> 
  <DT><B><A NAME="electrophoresis">electrophoresis</A></B> 
  <DD>The separation of charged molecules ( <A HREF="glossary.html#DNA">DNA</A>, 
    <A HREF="glossary.html#RNA">RNA</A> or <A HREF="glossary.html#protein">protein</A>) 
    in an electrical field, usually in a supporting medium such as an agarose 
    or polyacrylamide gel. 
    <P> 
  <DT><B><A NAME="EMS">EMS</A></B> 
  <DD>Ethyl methanesulfonate (methanesulfonic acid ethyl ester). A chemical <A HREF="glossary.html#mutagen">mutagen</A>. 
    <P> 
  <DT><B><A NAME="endogenous">endogenous</A></B> 
  <DD>Contained within. In genetics, endogenous <A HREF="glossary.html#virus">viruses</A> 
    are those that are integrated into a host <A HREF="glossary.html#genome">genome</A> 
    and transmitted to progeny as <A HREF="glossary.html#chromosome">chromosomal</A> 
    elements. 
    <P> 
  <DT><B><A NAME="endonuclease">endonuclease</A></B> 
  <DD>A <A HREF="glossary.html#protein">protein</A> that cleaves the phosphodiester 
    backbone of a <A HREF="glossary.html#nucleic_acid">nucleic acid</A> <A HREF="glossary.html#enzyme">enzymatically</A>, 
    for example a <A HREF="glossary.html#restriction_enzyme">restriction enzyme</A>. 
    <P> 
  <DT><B><A NAME="enhancer">enhancer</A></B> 
  <DD>One of the necessary <A HREF="glossary.html#regulatory_element">regulatory 
    elements</A> of a <A HREF="glossary.html#gene">gene</A>. An enhancer is a 
    site on <A HREF="glossary.html#DNA">DNA</A> to which a complex of transcription 
    factors bind to affect the availability of the <A HREF="glossary.html#promoter">promoter</A> 
    to <A HREF="glossary.html#RNA_polymerase">RNA polymerase</A>. A gene may 
    have multiple enhancers. 
    <P> 
  <DT><B><A NAME="enhancer_trap">enhancer trap</A></B> 
  <DD>A type of <A HREF="glossary.html#DNA_construct">DNA construct</A> containing 
    a <A HREF="glossary.html#reporter_gene">reporter gene</A> sequence downstream 
    of a <A HREF="glossary.html#promoter">promoter</A> that is capable of integrating 
    into random <A HREF="glossary.html#chromosome">chromosomal</A> locations. 
    Integration of the enhancer trap near an <A HREF="glossary.html#enhancer">enhancer</A> 
    allows the expression of a new mRNA encoding the reporter gene. The reporter 
    gene is therefore expressed in the cells and developmental stages where the 
    enhancer is active. See also <A HREF="glossary.html#gene_trap">gene trap</A>. 
    <P> 
  <DT><B><A NAME="ENU">ENU</A></B> 
  <DD>Ethylnitrosourea. A chemical <A HREF="glossary.html#mutagen">mutagen</A>. 
    In zebrafish, the <A HREF="glossary.html#mutation">mutation</A> rate caused 
    by ENU can be as high as one mutation/ <A HREF="glossary.html#locus">locus</A>/500-1000 
    <A HREF="glossary.html#gamete">gametes</A>. 
    <P> 
  <DT><B><A NAME="enzyme">enzyme</A></B> 
  <DD>A <A HREF="glossary.html#protein">protein</A> (or rarely, <A HREF="glossary.html#RNA">RNA</A>) 
    that catalyzes a chemical reaction. 
    <P> 
  <DT><B><A NAME="EP">EP</A></B> 
  <DD>See early pressure. 
    <P> 
  <DT><B><A NAME="epigenetic">epigenetic</A></B> 
  <DD>Refers to factors affecting the development or function of an organism other 
    than the primary sequence of the target <A HREF="glossary.html#gene">genes</A>. 
    <P> 
  <DT><B><A NAME="epistasis">epistasis</A></B> 
  <DD>Masking of a <A HREF="glossary.html#phenotype">phenotypic</A> <A HREF="glossary.html#trait">trait</A> 
    through the action of a <A HREF="glossary.html#mutation">mutant</A> <A HREF="glossary.html#allele">allele</A>. 
    For example, albino (absence of pigment) is epistatic to melanin pigment <A HREF="glossary.html#gene">genes</A> 
    that determine the dark color of zebrafish eyes and stripes. 
    <P> 
  <DT><B><A NAME="EST">EST</A></B> 
  <DD>Expressed Sequence Tag. A partial sequence of a randomly chosen <A HREF="glossary.html#cDNA">cDNA</A>, 
    obtained from the results of a single DNA sequencing reaction. ESTs are used 
    to identify <A HREF="glossary.html#transcription">transcribed</A> regions 
    in <A HREF="glossary.html#genome">genomic</A> sequence, to characterize patterns 
    of <A HREF="glossary.html#gene_expression">gene expression</A> in the tissue 
    that was the source of the <A HREF="glossary.html#cDNA">cDNA</A> and as markers 
    for genetic mapping. 
    <P> 
  <DT><B><A NAME="ethidium_bromide">ethidium bromide</A></B> 
  <DD>A fluorescent dye that intercalates between <A HREF="glossary.html#base_pair">base 
    pairs</A> in double-stranded <A HREF="glossary.html#nucleic_acid">nucleic 
    acids</A> or between <A HREF="glossary.html#base">bases</A> in single-stranded 
    nucleic acids. Ethidium bromide is commonly used to visualize <A HREF="glossary.html#DNA">DNA</A> 
    on agarose gels. 
    <P> 
  <DT><B><A NAME="euchromatin">euchromatin</A></B> 
  <DD>The part of the <A HREF="glossary.html#genome">genome</A> characterized 
    by relatively high <A HREF="glossary.html#gene">gene</A> density and relative 
    absence of highly repetitive sequences. See also <A HREF="glossary.html#heterochromatin">heterochromatin</A>. 
    <P> 
  <DT><B><A NAME="euploid">euploid</A></B> 
  <DD>Having a <A HREF="glossary.html#chromosome">chromosome</A> number that 
    is an integral multiple of the <A HREF="glossary.html#haploid">haploid</A> 
    number without segmental <A HREF="glossary.html#duplication">duplications</A> 
    or <A HREF="glossary.html#deficiency">deficiencies</A>. 
    <P> 
  <DT><B><A NAME="evolution">evolution</A></B> 
  <DD>Change of the <A HREF="glossary.html#gene">genes</A> of a population over 
    time, resulting in new species. 
    <P> 
  <DT><B><A NAME="evolutionary_conservation">evolutionary conservation</A></B> 
  <DD>The presence of <A HREF="glossary.html#similarity">similar</A> <A HREF="glossary.html#gene">genes</A>, 
    portions of genes, or <A HREF="glossary.html#chromosome">chromosome</A> segments 
    in different species, reflecting both the common origin of species and an 
    important functional property of the conserved element. 
    <P> 
  <DT><B><A NAME="exon">exon</A></B> 
  <DD>Part of a <A HREF="glossary.html#gene">gene</A> whose sequence is present 
    in a mature <A HREF="glossary.html#mRNA">mRNA</A> after <A HREF="glossary.html#splicing">splicing</A>. 
    See also <A HREF="glossary.html#intron">Intron</A>. 
    <P> 
  <DT><B><A NAME="expressivity">expressivity</A></B> 
  <DD>The relative constancy of the <A HREF="glossary.html#phenotype">phenotype</A> 
    of individuals of a given <A HREF="glossary.html#genotype">genotype</A>. 
    <A HREF="glossary.html#mutation">Mutations</A> said to have variable expressivity 
    show a relatively large amount of phenotypic variation among individuals having 
    the same genotype. See also <A HREF="glossary.html#penetrance">penetrance</A>. 
    <!-- FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF -->
    <A NAME="f"></A> 
    <P> 
  <DT><B><A NAME="FISH">FISH</A></B> 
  <DD>Fluorescent <I>in situ</I> hybridization. A method for determining the <A HREF="glossary.html#cytogenetic">cytogenetic</A> 
    location of a cloned segment of <A HREF="glossary.html#DNA">DNA</A>. The 
    DNA is labeled with a fluorescent dye and <A HREF="glossary.html#hybridization">hybridized</A> 
    to a cytological preparation of <A HREF="glossary.html#chromosome">chromosomes</A> 
    that has been <A HREF="glossary.html#denaturation">denatured</A> to allow 
    <A HREF="glossary.html#nucleic_acid">nucleic acid</A> hybridization between 
    chromosomal DNA and the <A HREF="glossary.html#probe">probe</A>. The site 
    of hybridization is determined by fluorescence microscopy. See the <A 

HREF="http://www.nhgri.nih.gov/DIR/VIP/Glossary/Illustration/fish.html">Figure</A> 
    at <A HREF="glossary.html#NHGRI">NHGRI</A>. See also <A HREF="glossary.html#in_situ"><I>in 
    situ</I> hybridization</A>. 
    <P> 
  <DT><B><A NAME="fluorography">fluorography</A></B> 
  <DD>The detection of <A HREF="glossary.html#radiation">radiation</A> or of 
    a fluorescent compound by secondary light that was generated by the excitation 
    of a "fluor" or a screen by light, a beta particle or a gamma ray. 
    <P> 
  <DT><B><A NAME="floxed">floxed</A></B> 
  <DD>Refers to a <A HREF="glossary.html#DNA_construct">DNA construct</A> in 
    which a <A HREF="glossary.html#gene">gene</A> or gene segment is flanked 
    by <A HREF="glossary.html#loxP">loxP</A> sites in the same orientation; <A HREF="glossary.html#Cre">Cre 
    recombinase</A> excises the segment between the loxP sites. 
    <P> 
  <DT><B><A NAME="ftp">FTP</A></B> 
  <DD>File Transfer Protocol. A method for transferring files to and from remote 
    computer systems. 
    <P> 
  <DT><B><A NAME="frameshift">frameshift</A></B> 
  <DD>A type of <A HREF="glossary.html#mutation">mutation</A> in which there 
    is an <A HREF="glossary.html#insertion">insertion</A> or <A HREF="glossary.html#deficiency">deficiency</A> 
    that changes the <A HREF="glossary.html#reading_frame">reading frame</A>. 
    <!-- GGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGG -->
    <A NAME="g"></A> 
    <P> 
  <DT><B><A NAME="gain-of-function">gain-of-function mutation</A></B> 
  <DD>A type of <A HREF="glossary.html#mutation">mutation</A> in which the altered 
    <A HREF="glossary.html#gene_product">gene product</A> possesses a new <A HREF="glossary.html#molecular_function">molecular 
    function</A> or a new pattern of <A HREF="glossary.html#gene_expression">gene 
    expression</A>. Gain-of-function mutations are almost always <A HREF="glossary.html#dominant">dominant</A> 
    or <A HREF="glossary.html#semidominant">semidominant</A>. See also: 
    <UL>
      <LI><A HREF="glossary.html#amorphic">amorphic mutation</A></LI>
      <LI><A HREF="glossary.html#antimorphic">antimorphic mutation</A></LI>
      <LI><A HREF="glossary.html#hypermorphic">hypermorphic mutation</A></LI>
      <LI><A HREF="glossary.html#hypomorphic">hypomorphic mutation</A></LI>
      <LI><A HREF="glossary.html#loss-of-function">loss-of-function mutation</A></LI>
      <LI><A HREF="glossary.html#neomorphic">neomorphic mutation</A></LI>
      <LI><A HREF="glossary.html#null_mutation">null mutation</A></LI>
    </UL>
    <P> 
  <DT><B><A NAME="gamete">gamete</A></B> 
  <DD>One of the differentiated cells that is a product of <A HREF="glossary.html#meiosis">meiosis</A>. 
    In animals, sperm or egg cells.<BR>
    <P> 
  <DT><B><A NAME="GDB">GDB</A></B> 
  <DD>Genome Database, a database of human genes. Links to GDB are returned when 
    human genes are displayed, for example in the &quot;orthologues&quot; section 
    of a gene record<A 

HREF="http://www.informatics.jax.org/searches/homology_form.shtml"></A>. See the 
    <A HREF="http://gdbwww.gdb.org/gdb/">central node</A> at GDB by clicking this 
    link. 
    <P> 
  <DT><B><A NAME="genbank">GenBank</A></B> 
  <DD>The database of <A HREF="glossary.html#nucleic_acid">nucleic acid</A> sequences 
    at <A HREF="glossary.html#NCBI">NCBI</A>.<BR>
    <P> 
  <DT><B><A NAME="gene">gene</A></B> 
  <DD>The basic unit of heredity; a portion of <A HREF="glossary.html#DNA">DNA</A> 
    that (usually) codes for a <A HREF="glossary.html#protein">protein</A> product. 
    See also <A HREF="glossary.html#central_dogma">central dogma</A>.<BR>
    In <A HREF="glossary.html#ZFIN">ZFIN</A>, "gene" may be used to refer to 
    any of the following: 
    <OL>
      <LI> A <A HREF="glossary.html#locus">locus</A> in the nuclear <A HREF="glossary.html#genome">genome</A> 
        characterized by an altered <A HREF="glossary.html#phenotype">phenotype</A> 
        or by an effect on an inserted <A HREF="glossary.html#reporter_gene">reporter 
        gene</A>, such as a <A HREF="glossary.html#gene_trap">gene trap</A> or 
        <A HREF="glossary.html#enhancer_trap">enhancer trap</A>. 
      <LI> A locus in the nuclear genome that is necessary and sufficient to express 
        the complete complement of functional products derived from a unit of 
        <A HREF="glossary.html#transcription">transcription</A>. 
      <LI> A locus in the zebrafish nuclear genome identified by <A HREF="glossary.html#hybridization">hybridization</A> 
        to a <A HREF="glossary.html#nucleic_acid">nucleic acid</A> segment derived 
        from a non-zebrafish species, where the segment used as <A HREF="glossary.html#probe">probe</A> 
        represents some portion of a functional unit of transcription in the nuclear 
        genome of the non-zebrafish species. 
      <LI> An <A HREF="glossary.html#exon">exon</A>-encoding segment of the <A HREF="glossary.html#germline">germ-line</A> 
        nuclear genome located within a region that undergoes <A HREF="glossary.html#somatic">somatic</A> 
        rearrangement. 
      <LI> A locus in the nuclear genome that is within an <A HREF="glossary.html#intron">intron</A> 
        of (but not, itself, an exon of) a unit of transcription, which gives 
        rise to a functional product upon <A HREF="glossary.html#transcript">transcript</A> 
        <A HREF="glossary.html#RNA_processing">processing</A> of the host unit. 
    </OL>
    <P> 
  <DT><B><A NAME="gene_complex">gene complex</A></B> 
  <DD>A number of apparently functionally or <A HREF="glossary.html#evolution">evolutionarily</A> 
    related <A HREF="glossary.html#locus">loci</A> that are genetically closely 
    <A HREF="glossary.html#linkage">linked</A>. Alternative states of complexes 
    are referred to as <A HREF="glossary.html#haplotype">haplotypes</A> rather 
    than <A HREF="glossary.html#allele">alleles</A>. 
    <P> 
  <DT><B><A NAME="gene_conversion">gene conversion</A></B> 
  <DD>A type of <A HREF="glossary.html#recombination">nonreciprocal recombination</A> 
    event in which a recipient strand of <A HREF="glossary.html#DNA">DNA</A> 
    receives information from another strand having an <A HREF="glossary.html#allele">allelic</A> 
    difference. The recipient strand has its original allele "converted" to the 
    new allele as a consequence of the event. 
    <P> 
  <DT><B><a name="gene_duplication">gene duplication</a></b> 
  <DD>See <a href="glossary.html#duplication">Duplication</a>. 
    <P> 
  <DT><B><A NAME="gene_expression">gene expression</A></B> 
  <DD> <A HREF="glossary.html#transcription">Transcriptional</A> activity of 
    a <A HREF="glossary.html#gene">gene</A> resulting in one or more <A HREF="glossary.html#RNA">RNA</A> 
    products and, usually, following <A HREF="glossary.html#translation">translation</A>, 
    one or more <A HREF="glossary.html#protein">protein</A> products. 
    <P> 
  <DT><B><A NAME="gene_family">gene family</A></B> 
  <DD>A group of <A HREF="glossary.html#paralog">paralogous</A> <A HREF="glossary.html#gene">genes.</A> 
    <P> 
  <DT><B><A NAME="gene_name">gene name</A></B> 
  <DD>In <A HREF="glossary.html#ZFIN">ZFIN</A>, a "gene name" is a word or phrase 
    that uniquely identifies a <A HREF="glossary.html#gene">gene</A>. The gene 
    name has an abbreviation that is the <A HREF="glossary.html#gene_symbol">gene 
    symbol</A>. See also <A HREF="glossary.html#allele_designation">Allele Designation</A>, 
    <A HREF="glossary.html#gene_symbol">Gene Symbol</A>. 
    <P> 
  <DT><B><A NAME="gene_ontology">Gene Ontology (GO)</A></B> 
  <DD>A set of <A HREF="glossary.html#controlled_vocabulary">controlled vocabularies</A> 
    used to describe biological features within a specified domain of biological 
    knowledge. See the <A HREF="http://www.geneontology.org/">GO Consortium</A> 
    site for further information. 
    <P> 
  <DT><B><A NAME="gene_product">gene product</A></B> 
  <DD>1. A <A HREF="glossary.html#protein">protein</A> molecule that is the product 
    of the <A HREF="glossary.html#gene_expression">expression</A> of a <A HREF="glossary.html#gene">gene</A>, 
    through which the gene influences development or metabolism. <BR>
    2. An <A HREF="glossary.html#RNA">RNA</A> molecule that is the product of 
    the expression of a gene, especially those cases in which the RNA molecule 
    is not <A HREF="glossary.html#translation">translated</A> (see <A HREF="glossary.html#tRNA">tRNA</A>, 
    <A HREF="glossary.html#rRNA">rRNA</A>). 
    <P> 
  <DT><B><A NAME="gene_symbol">gene symbol</A></B> 
  <DD>As used in <A HREF="glossary.html#ZFIN">ZFIN</A>, a "gene symbol" is a 
    unique abbreviation for the <A HREF="glossary.html#gene_name">gene name</A>. 
    See also <A HREF="glossary.html#allele_designation">allele designation</A>, 
    <A HREF="glossary.html#gene_name">gene name</A>. 
    <P> 
  <DT><B><A NAME="gene_trap">gene trap</A></B> 
  <DD>A type of <A HREF="glossary.html#DNA_construct">DNA construct</A> containing 
    a <A HREF="glossary.html#reporter_gene">reporter gene</A> sequence downstream 
    of a <A HREF="glossary.html#splice_acceptor">splice acceptor site</A> that 
    is capable of integrating into random chromosomal locations. Integration of 
    the gene trap into an <A HREF="glossary.html#intron">intron</A> allows the 
    expression of a new <A HREF="glossary.html#mRNA">mRNA</A> containing one 
    or more upstream <A HREF="glossary.html#exon">exons</A> followed by the reporter 
    gene. The reporter gene is thus <A HREF="glossary.html#gene_expression">expressed</A> 
    in the same cells and developmental stages as the <A HREF="glossary.html#gene">gene</A> 
    into which the gene trap has inserted. See also <A HREF="glossary.html#enhancer_trap">enhancer 
    trap</A> 
    <P> 
  <DT><B><A NAME="genetic_code">genetic code</A></B> 
  <DD>The relationship of the sixty-four <A HREF="glossary.html#nucleic_acid">nucleic 
    acid</A> <A HREF="glossary.html#codon">codons</A> to the twenty primary <A HREF="glossary.html#amino_acid">amino 
    acids</A>. See the <A 

HREF="http://www.accessexcellence.org/AB/GG/genetic.html">Figure</A> for the standard 
    genetic code. See also <A HREF="glossary.html#central_dogma">central dogma</A>. 
    <P> 
  <DT><B><A NAME="genetic_map">genetic map</A></B> 
  <DD>A map showing the position of <A HREF="glossary.html#gene">genes</A> or 
    <A HREF="glossary.html#marker">markers</A> on a <A HREF="glossary.html#chromosome">chromosome</A>. 
    See <A HREF="glossary.html#physical_map">physical map</A>,l<A HREF="glossary.html#linkage_map">inkage 
    map</A>, and <A HREF="glossary.html#cytogenetic_map">cytogenetic map</A>. 
    <P> 
  <DT><B><A NAME="genome">genome</A></B> 
  <DD>The total genetic information of a cell or <A HREF="glossary.html#organelle">organelle</A>. 
    In eukaryotes, "genome" usually refers to nuclear <A HREF="glossary.html#DNA">DNA</A> 
    rather than to <A HREF="glossary.html#mitochondria">mitochondrial</A> or 
    chloroplast DNA. 
    <P> 
  <DT><B><A NAME="genotype">genotype</A></B> 
  <DD>A description of the genetic information carried by an organism. In the 
    simplest case, "genotype" may refer to the information carried at a single 
    <A HREF="glossary.html#locus">locus</A>, as in A/A, A/a, or a/a. 
    <P> 
  <DT><B><A NAME="GFP">GFP</A></B> 
  <DD>Green Fluorescent Protein. A fluorescent marker used to label cells expressing 
    <a href="glossary.html#transgene">transgenes</a>. 
    <P> 
  <DT><B><A NAME="germline">germ line</A></B> 
  <DD>Cells of an animal that give rise to <A HREF="glossary.html#gamete">gametes</A>. 
    <!-- HHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHH -->
    <A NAME="h"></A> 
    <P> 
  <DT><B><A NAME="haploid">haploid</A></B> 
  <DD>Having the <A HREF="glossary.html#chromosome">chromosome</A> number normally 
    found in a <A HREF="glossary.html#gamete">gamete</A>. See also <A HREF="glossary.html#diploid">diploid</A>. 
    <P> 
  <DT><B><A NAME="haploinsufficient">haploinsufficient</A></B> 
  <DD>A description applied to a <A HREF="glossary.html#gene">gene</A> that produces 
    a <A HREF="glossary.html#mutant">mutant</A> <A HREF="glossary.html#phenotype">phenotype</A> 
    when present in a <A HREF="glossary.html#diploid">diploid</A> individual 
    <A HREF="glossary.html#heterozygous">heterozygous</A> for an <A HREF="glossary.html#amorphic">amorphic</A> 
    <A HREF="glossary.html#allele">allele</A>. 
    <P> 
  <DT><B><A NAME="haplotype">haplotype</A></B> 
  <DD>One of the alternative forms of the <A HREF="glossary.html#genotype">genotype</A> 
    of a <A HREF="glossary.html#gene_complex">gene complex</A>. This term is 
    applied to gene complexes rather than the term <A HREF="glossary.html#allele">allele</A>, 
    which refers to one of the forms of a single <A HREF="glossary.html#gene">gene</A>. 
    <P> 
  <DT><B><A NAME="heat_shock">heat shock (HS)</A></B> 
  <DD>1. <a href="zfbook/chapt7/7.5.html">Method</a> to produce <a href="glossary.html#homozygous">homozygous</a> 
    <a href="glossary.html#diploid">diploid</a> offspring.<BR>
    2. Method to induce expression of <a href="glossary.html#transgene">transgenes</a> 
    under the control of a heat shock <a href="glossary.html#promoter">promoter</a>. 
    <P> 
  <DT><B><A NAME="hemizygous">hemizygous</A></B> 
  <DD>The state of a <A HREF="glossary.html#gene">gene</A> present in only one 
    copy in a <A HREF="glossary.html#diploid">diploid</A> cell, such as a gene 
    on the <A HREF="glossary.html#X_chromosome">X chromosome</A> in a male mammal, 
    or a gene whose <A HREF="glossary.html#homolog">homologue</A> has been deleted. 
    <P> 
  <DT><B><A NAME="heterochromatin">heterochromatin</A></B> 
  <DD>1. The part of the <A HREF="glossary.html#genome">genome</A> characterized 
    by relatively low <A HREF="glossary.html#gene">gene</A> density and the presence 
    of highly repetitive sequences. Heterochromatin is more highly condensed than 
    <A HREF="glossary.html#euchromatin">euchromatin</A>.<BR>
    2. The <A HREF="glossary.html#X_chromosome">X chromosome</A> that is highly 
    condensed in a mammalian cell that has undergone <A HREF="glossary.html#X_inactivation">X 
    inactivation</A>. The inactive X chromosome resembles heterochromatin as defined 
    above with respect to their state of condensation and genetic inactivity, 
    although there is no change in the <A HREF="glossary.html#DNA">DNA</A> sequence 
    as a consequence of inactivation.<BR>
    See also <A HREF="glossary.html#euchromatin">euchromatin</A>. 
    <P> 
  <DT><B><A NAME="heterogametic">heterogametic</A></B> 
  <dd>Producing two types of <A HREF="glossary.html#euploid">euploid</A> <A HREF="glossary.html#gamete">gametes</A> 
    with respect to <A HREF="glossary.html#chromosome">chromosomal</A> content. 
    This term is applied to one of the sexes in species with chromosomal sex determination; 
    in mammals, males are heterogametic. See also <A HREF="glossary.html#homogametic">Homogametic</A>, 
    <A HREF="glossary.html#X_chromosome">X chromosome</A>, <A HREF="glossary.html#Y_chromosome">Y 
    chromosome</A>. 
    <P> 
  <DT><B><A NAME="heteropolymer">heteropolymer</A></B> 
  <DD>A polymer composed of different subunits. Some multimeric <A HREF="glossary.html#protein">proteins</A> 
    are normally heteropolymers. Heteropolymers can also be made experimentally, 
    using subunits derived from different species, as a test of <A HREF="glossary.html#homology">homology</A>. 
    Formation of a functional multimeric protein product using subunits from different 
    species is a demonstration of homology. 
    <P> 
  <DT><B><A NAME="heterozygote">heterozygote</A></B> 
  <DD>An individual that is <A HREF="glossary.html#heterozygous">heterozygous</A>. 
    <P> 
  <DT><B><A NAME="heterozygous">heterozygous</A></B> 
  <DD>Producing two <A HREF="glossary.html#gamete">gamete</A> types with respect 
    to at least one <A HREF="glossary.html#gene">gene</A> (A/a). 
    <P> 
  <DT><B><A NAME="heterozygosity">heterozygosity</A></B> 
  <DD>The state of being <A HREF="glossary.html#heterozygous">heterozygous</A>. 
    <P> 
  <DT><B><A NAME="homogametic">homogametic</A></B> 
  <DD>Producing a single type of <A HREF="glossary.html#euploid">euploid</A> 
    <A HREF="glossary.html#gamete">gametes</A> with respect to <A HREF="glossary.html#chromosome">chromosomal</A> 
    content. This term is applied to one of the sexes in species with chromosomal 
    sex determination; in mammals, females are homogametic. See also <A HREF="glossary.html#heterogametic">heterogametic</A>, 
    <A HREF="glossary.html#X_chromosome">X chromosome</A>, <A HREF="glossary.html#Y_chromosome">Y 
    chromosome</A>. 
    <P> 
  <DT><B><A NAME="homolog">homologue</A></B> 
  <DD> 1. >One of a pair of <A HREF="glossary.html#chromosome">chromosomes</A> 
    that segregate from one another during the first <A HREF="glossary.html#meiosis">meiotic</A> 
    division.<BR>
    2. A <A HREF="glossary.html#gene">gene</A> related to a second gene by descent 
    from a common ancestral <A HREF="glossary.html#DNA">DNA</A> sequence. The 
    term, homologue, may apply to the relationship between genes separated by the 
    event of speciation (see <A HREF="glossary.html#ortholog">orthologue</A>) or 
    to the relationship betwen genes separated by the event of genetic duplication 
    (see <A HREF="glossary.html#paralog">paralog</A>). <BR>
    3. A morphological structure in one species related to that in a second species 
    by descent from a common ancestral structure. 
    <P> 
  <DT><B><A NAME="homologous_recombination">homologous recombination</A></B> 
  <DD> 1. Reciprocal <A HREF="glossary.html#recombination">recombination</A> 
    between <A HREF="glossary.html#DNA">DNA</A> sequences that have a high degree 
    of <A HREF="glossary.html#similarity">similarity</A>.<BR>
    2. Reciprocal recombination between DNA sequences that have a high degree 
    of similarity and that are located at corresponding positions on <A HREF="glossary.html#homolog">homologous</A> 
    <A HREF="glossary.html#chromosome">chromosomes</A>. 
    <P> 
  <DT><B><A NAME="homology">homology</A></B> 
  <dd> 1. The relationship of any two characters that have descended from a common 
    <A HREF="glossary.html#ancestor">ancestor</A>. This term can apply to a morphological 
    structure, a <A HREF="glossary.html#chromosome">chromosome</a> or an individual 
    <A HREF="glossary.html#gene">gene</a> or <A HREF="glossary.html#DNA_segment">DNA 
    segment</a>. <BR>
    2. In ZFIN, assertions of mammalian homology imply presumed orthology even though there may be <a href="glossary.html#duplication">duplicates</a> 
    of the homologous gene in zebrafish.&nbsp; See the <A HREF="http://www.ncbi.nlm.nih.gov/Education/BLASTinfo/Orthology.html">Figure</A> 
    at NCBI. See also: <A HREF="glossary.html#homolog">homologue</A>, <A HREF="glossary.html#orthology">orthology</A>, 
    <A HREF="glossary.html#paralogy">paralogy</A>. 
    <P> 
  <DT><B><A NAME="homozygote">homozygote</A></B> 
  <DD>An individual that is <A HREF="glossary.html#homozygous">homozygous</A>. 
    <P> 
  <DT><B><A NAME="homozygous">homozygous</A></B> 
  <DD>Producing only a single <A HREF="glossary.html#gamete">gamete</A> type 
    with respect to one or more <A HREF="glossary.html#gene">genes</A> (A/A). 
    <P> 
  <DT><B><A NAME="homozygosity">homozygosity</A></B> 
  <DD>The state of being <A HREF="glossary.html#homozygous">homozygous</A>. 
    <P> 
  <DT><B><A NAME="horseradish_peroxidase">horseradish peroxidase</A></B> 
  <DD>An <A HREF="glossary.html#enzyme">enzyme</A> for which a <A HREF="glossary.html#chromogenic">chromogenic</A> 
    <A HREF="glossary.html#substrate">substrate</A> exists, commonly used as 
    a label for <A HREF="glossary.html#antibody">antibodies</A>. 
    <P> 
  <DT><B><A NAME="HTML">HTML</A></B> 
  <DD>Hypertext Markup Language. An authoring language for creating and sharing 
    electronic documents over the Internet. This document is written in HTML.<BR>
    You can view the HTML source code for this document by: 
    <UL>
      <LI>selecting <B>Page Source</B> from the <B>View</B> menu in Netscape Navigator</LI>
      <LI>selecting <B>Source</B> from the <B>View</B> menu in Microsoft Internet 
        Explorer.</LI>
    </UL>
    <P> 
  <DT><B><A NAME="hybrid">hybrid</A></B> 
  <DD>1. The offspring of two sexually-reproducing <A HREF="glossary.html#homozygous">homozygous</A> 
    individuals of different <A HREF="glossary.html#genotype">genotypes</A>.<BR>
    2. As a <A HREF="glossary.html#DNA">ZFIN</A> Mapping Data type, a <A HREF="glossary.html#radiation%20hybrid%20mapping">radiation 
    (somatic cell) hybrid</A> experiment. 
    <P> 
  <DT><B><A NAME="hybridization">hybridization</A></B> 
  <DD>With respect to <A HREF="glossary.html#nucleic_acid">nucleic acids</A>, 
    "hybridization" refers to the formation of double-stranded <A HREF="glossary.html#DNA">DNA</A>, 
    <A HREF="glossary.html#RNA">RNA</A>, or DNA/RNA hybrids by complementary 
    <A HREF="glossary.html#base_pair">base pairing</A>. 
    <P> 
  <DT><B><A NAME="hydrophilic">hydrophilic</A></B> 
  <DD>Literally, "water-loving"; polar or charged compounds that are soluble in 
    water. 
    <P> 
  <DT><B><A NAME="hydrophobic">hydrophobic</A></B> 
  <DD>Literally, "water-fearing"; nonpolar compounds that are immiscible with 
    water. The side chains of some <A HREF="glossary.html#amino_acid">amino acids</A> 
    are nonpolar, and hence <A HREF="glossary.html#protein">protein</A> sequences 
    rich in these amino acids tend to locate to the interior of the protein in 
    its native state, away from the solvent. 
    <P> 
  <DT><B><A NAME="hypertext">hypertext</A></B> 
  <DD>Text displayed electronically with embedded links to other text or to images, 
    sounds, movies or other multimedia content. This document is an example of 
    hypertext. 
    <P> 
  <DT><B><A NAME="hypermorphic">hypermorphic mutation</A></B> 
  <DD>A type of <A HREF="glossary.html#mutation">mutation</A> in which the altered 
    <A HREF="glossary.html#gene_product">gene product</A> possesses an increased 
    level of activity, or in which the <A HREF="glossary.html#wild_type">wild-type</A> 
    gene product is <A HREF="glossary.html#gene_expression">expressed</A> at 
    an increased level. <BR>
    See also: 
    <UL>
      <LI><A HREF="glossary.html#amorphic">amorphic mutation</A></LI>
      <LI><A HREF="glossary.html#antimorphic">antimorphic mutation</A></LI>
      <LI><A HREF="glossary.html#gain-of-function">gain-of-function mutation</A></LI>
      <LI><A HREF="glossary.html#hypomorphic">hypomorphic mutation</A></LI>
      <LI><A HREF="glossary.html#loss-of-function">loss-of-function mutation</A></LI>
      <LI><A HREF="glossary.html#neomorphic">neomorphic mutation</A></LI>
      <LI><A HREF="glossary.html#null_mutation">null mutation</A></LI>
    </UL>
    <P> 
  <DT><B><A NAME="hypomorphic">hypomorphic mutation</A></B> 
  <DD>A type of <A HREF="glossary.html#mutation">mutation</A> in which the altered 
    <A HREF="glossary.html#gene_product">gene product</A> possesses a reduced 
    level of activity, or in which the <A HREF="glossary.html#wild_type">wild-type</A> 
    gene product is <A HREF="glossary.html#gene_expression">expressed</A> at 
    a reduced level.<BR>
    See also: 
    <UL>
      <LI><A HREF="glossary.html#amorphic">amorphic mutation</A></LI>
      <LI><A HREF="glossary.html#antimorphic">antimorphic mutation</A></LI>
      <LI><A HREF="glossary.html#gain-of-function">gain-of-function mutation</A></LI>
      <LI><A HREF="glossary.html#hypermorphic">hypermorphic mutation</A></LI>
      <LI><A HREF="glossary.html#loss-of-function">loss-of-function mutation</A></LI>
      <LI><A HREF="glossary.html#neomorphic">neomorphic mutation</A></LI>
      <LI><A HREF="glossary.html#null_mutation">null mutation</A></LI>
    </UL>
    <!-- IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII -->
    <A NAME="i"></A> 
    <P> 
  <DT><B><A NAME="identity">identity</A></B> 
  <DD>In comparison of <A HREF="glossary.html#nucleic_acid">nucleic acid</A> 
    or <A HREF="glossary.html#protein">protein</A> sequences, the extent to which 
    two sequences have the same <A HREF="glossary.html#nucleotide">nucleotide</A> 
    or <A HREF="glossary.html#amino_acid">amino acid</A> at equivalent positions, 
    usually expressed as a percentage. See also <A HREF="glossary.html#similarity">similarity</A>. 
    <P> 
  <DT><B><A NAME="idiogram">idiogram</A></B> 
  <DD>An idealized drawing. 
    <P> 
  <DT><B><A NAME="IMAGE">I.M.A.G.E. Consortium</A></B> 
  <DD>Integrated Molecular Analysis of Genome Expression Consortium. A collection 
    of a large number of partially sequenced <a href="glossary.html#EST">ESTs</a> 
    or <A HREF="glossary.html#cDNA">cDNAs</A>. See the homepage of the <A HREF="http://image.llnl.gov/">I.M.A.G.E. 
    Consortium</A> for further information. 
    <P> 
  <DT><B><A NAME="immunofluorescence">immunofluorescence</A></B> 
  <DD>The detection of an <A HREF="glossary.html#antigen">antigen</A> in cytological 
    preparations by using a fluorescently-labeled <A HREF="glossary.html#antibody">antibody</A>. 
    <P> 
  <DT><B><A NAME="immunohistochemistry">immunohistochemistry</A></B> 
  <DD>A method of detecting the presence of specific <A HREF="glossary.html#protein">proteins</A> 
    in cells or tissues. Fixed cells or tissue on a microscope slide, made permeable 
    if necessary with a detergent, are reacted with a primary <A HREF="glossary.html#antibody">antibody</A> 
    to the specific protein to be assayed. The preparation is then treated with 
    a secondary antibody that has been coupled to an <A HREF="glossary.html#enzyme">enzyme</A> 
    and which is directed against the primary antibody (e.g. goat anti-rabbit 
    antibody). The preparation is then treated with a <A HREF="glossary.html#chromogenic">chromogenic</A> 
    <A HREF="glossary.html#substrate">substrate</A>. Alternatively, the secondary 
    antibody can be directly coupled to a fluorofluor. Microscopic examination 
    reveals the presence of labeling, and hence of the specific protein to be 
    detected. 
    <P> 
  <DT><B><A NAME="immunological_cross-reaction">immunological cross-reaction</A></B> 
  <DD>The binding of an <A HREF="glossary.html#antibody">antibody</A> to a <A HREF="glossary.html#protein">protein</A> 
    that is different from the protein against which the antibody was raised. 
    This result demonstrates sequence or structural similarity between the two 
    proteins and can be evidence of <A HREF="glossary.html#homology">homology</A>. 
    <P> 
  <DT><B><A NAME="imprinting">imprinting</A></B> 
  <DD>An <A HREF="glossary.html#epigenetic">epigenetic</A> modification of <A HREF="glossary.html#gene">genes</A> 
    that identifies a given gene as having been inherited from the maternal or 
    paternal parent. In mammals, some genes are <A HREF="glossary.html#gene_expression">expressed</A> 
    primarily from the maternally-inherited or paternally-inherited <A HREF="glossary.html#allele">alleles</A> 
    as a consequence of imprinting. 
    <P> 
  <DT><B><A NAME="in_silico"><I>in silico</I></A></B> 
  <DD>Computationally, as opposed to <A HREF="glossary.html#in_vitro"><I>in vitro</I></A> 
    or <A HREF="glossary.html#in_vivo"><I>in vivo</I></A>. 
    <P> 
  <DT><B><A NAME="in_situ"><I>in situ </I>hybridization</A></B> 
  <DD>A method of detecting the presence of specific <A HREF="glossary.html#nucleic_acid">nucleic 
    acid</A> sequences within a cytological preparation. A <A HREF="glossary.html#DNA">DNA</A> 
    or <A HREF="glossary.html#RNA">RNA</A> <A HREF="glossary.html#probe">probe</A> 
    is labeled radioactively or chemically and <A HREF="glossary.html#hybridization">hybridized</A> 
    to a cytological preparation to detect RNA or to a <A HREF="glossary.html#denaturation">denatured</A> 
    cytological preparation to detect DNA. The hybridization is detected by <A HREF="glossary.html#autoradiography">autoradiography</A> 
    (for radioactive probes) or by <A HREF="glossary.html#chromogenic">chromogenic</A> 
    reactions or fluorescence (for chemically-labeled probes). See also <A HREF="glossary.html#FISH">FISH</A>. 
    <P> 
  <DT><B><A NAME="in_vitro"><I>in vitro</I></A></B> 
  <DD>Literally, "in glass", meaning a reaction, process or experiment in a metaphorical 
    test tube rather than in a living organism. In <A HREF="glossary.html#ZFIN">ZFIN</A> 
    this term also applies to <A HREF="glossary.html#cDNA">cDNA</A> clones originating 
    from tissue culture cells rather than from tissues from whole organisms. See 
    also <A HREF="glossary.html#in_vivo"><I>in vivo</I></A>, <A HREF="glossary.html#in_silico"><I>in 
    silico</I></A>. 
    <P> 
  <DT><B><A NAME="in_vivo"><I>in vivo</I></A></B> 
  <DD>Literally, "in life", meaning a reaction, process or experiment in a living 
    organism rather than in a metaphorical test tube. See also <A HREF="glossary.html#in_vitro"><I>in 
    vitro</I></A>, <A HREF="glossary.html#in_silico"><I>in silico</I></A>. 
    <P> 
  <DT><B><A NAME="inbred_strain">inbred strain</A></B> 
  <DD>A strain that is essentially <A HREF="glossary.html#homozygous">homozygous</A> 
    at all loci due to brother-sister matings for at least 20 sequential generations. 
    <P> 
  <DT><B><A NAME="incross">incross</A></B> 
  <DD>A <A HREF="glossary.html#cross">cross</A> between two identically <A HREF="glossary.html#homozygous">homozygous</A> 
    individuals (A/A X A/A). <BR>
    See also: 
    <UL>
      <LI><A HREF="glossary.html#backcross">backcross</A></LI>
      <LI><A HREF="glossary.html#intercross">intercross</A></LI>
      <LI><A HREF="glossary.html#outcross">outcross</A></LI>
      <LI><A HREF="glossary.html#testcross">testcross</A></LI>
    </UL>
    <P> 
  <DT><B><A NAME="inhibitor">inhibitor</A></B> 
  <DD>A chemical compound that has the effect of blocking or slowing an enzymatic 
    reaction. 
    <P> 
  <DT><B><A NAME="insertion">insertion</A></B> 
  <DD>A type of <A HREF="glossary.html#mutation">mutation</A> in which one or 
    more <A HREF="glossary.html#nucleotide">nucleotides</A> is inserted into 
    a <A HREF="glossary.html#DNA">DNA</A> sequence. Small insertions within a 
    <A HREF="glossary.html#gene">gene</A> can alter the <A HREF="glossary.html#reading_frame">reading 
    frame</A>, and thus the <A HREF="glossary.html#amino_acid">amino acid</A> 
    sequence of the encoded <A HREF="glossary.html#protein">protein</A>. 
    <P> 
  <DT><B><A NAME="intercross">intercross</A></B> 
  <DD>A <A HREF="glossary.html#cross">cross</A> between two identically <A HREF="glossary.html#hybrid">hybrid</A> 
    individuals (A/a X A/a). <BR>
    See also: 
    <UL>
      <LI><A HREF="glossary.html#backcross">backcross</A></LI>
      <LI><A HREF="glossary.html#incross">incross</A></LI>
      <LI><A HREF="glossary.html#outcross">outcross</A></LI>
      <LI><A HREF="glossary.html#testcross">testcross</A></LI>
    </UL>
    <P> 
  <DT><B><A NAME="intergenic">intergenic</A></B> 
  <DD>Between <A HREF="glossary.html#gene">genes</A>. 
    <P> 
  <DT><B><A NAME="intron">intron</A></B> 
  <DD>Part of a <A HREF="glossary.html#gene">gene</A> whose sequence is <A HREF="glossary.html#transcription">transcribed</A> 
    but not present in a mature <A HREF="glossary.html#mRNA">mRNA</A> after <A HREF="glossary.html#splicing">splicing</A>. 
    See also <A HREF="glossary.html#exon">Exon</A>. 
    <P> 
  <DT><B><A NAME="inversion">inversion</A></B> 
  <DD>A type of <A HREF="glossary.html#mutation">mutation</A> in which a length 
    of <A HREF="glossary.html#DNA">DNA</A> is broken at two positions and repaired 
    in such a way that the medial segment is now present in reverse order. Inversions 
    range in size from those large enough to be visible <A HREF="glossary.html#cytogenetic">cytogenetically</A> 
    to those involving only a few base pairs. 
    <!-- JJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJ -->
    <A NAME="j"></A> 
    <!-- KKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKK -->
    <A NAME="k"></A> 
    <P> 
  <DT><B><A NAME="karyotype">karyotype</A></B> 
  <DD>A description of the condensed <A HREF="glossary.html#chromosome">chromosomes</A> 
    of a eukaryote as they are seen at metaphase. Additional details are revealed 
    by a variety of staining techniques that produce banded chromosomes. 
    <P> 
  <DT><B><A NAME="kb">kb</A></B> 
  <DD><a href="glossary.html#kilobase">Kilobase</a>. 
    <P> 
  <DT><B><A NAME="kilobase">kilobase</A></B> 
  <DD>Unit of <A HREF="glossary.html#DNA">DNA</A> or <A HREF="glossary.html#RNA">RNA</A> 
    sequence equal to 1000 <A HREF="glossary.html#nucleotide">nucleotides</A>. 
    <P> 
  <DT><B><A NAME="kinetochore">kinetochore</A></B> 
  <DD>A stucture formed adjacent to the <A HREF="glossary.html#centromere">centromere</A> 
    of a condensed <A HREF="glossary.html#chromosome">chromosome</A> that allows 
    the chromosome to attach to <A HREF="glossary.html#microtubule">microtubules</A> 
    of the <A HREF="glossary.html#meiosis">meiotic</A> or <A HREF="glossary.html#mitosis">mitotic</A> 
    <A HREF="glossary.html#spindle">spindle</A>. 
    <P> 
  <DT><B><A NAME="knockdown">knock-down</A></B> 
  <DD>A casual term for the reduction of gene function by injection of <a href="glossary.html#morpholino">morpholinos</a>. 
    <P> 
  <DT><B><A NAME="knockin">knock-in</A></B> 
  <DD>A casual term for a type of <A HREF="glossary.html#targeted_mutation">targeted 
    mutation</A> in which an alteration in gene function other than a <A HREF="glossary.html#loss-of-function">loss-of-function</A> 
    allele is produced. See also <A HREF="glossary.html#knockout">knock-out</A>. 
    <P> 
  <DT><B><A NAME="knockout">knock-out</A></B> 
  <DD>A casual term for a type of <A HREF="glossary.html#targeted_mutation">targeted 
    mutation</A> in which an <A HREF="glossary.html#amorphic">amorphic</A> ( 
    <A HREF="glossary.html#loss-of-function">loss-of-function</A>) allele is 
    produced. See also <A HREF="glossary.html#knockin">knock-in</A>. 
    <!-- LLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLL -->
    <A NAME="l"></A> 
    <P> 
  <DT><B><A NAME="library">library</A></B> 
  <DD>In molecular biology, a "library" is a complex mixture of recombinant DNA 
    molecules in a suitable <A HREF="glossary.html#cloning_vector">cloning vector</A> 
    representing either the entire <A HREF="glossary.html#genome">genome</A> 
    of an organism (a genomic library) or the <A HREF="glossary.html#mRNA">messenger 
    RNA</A> population of a whole organism, cell type, or tissue type (a <A HREF="glossary.html#cDNA">cDNA</A> 
    library). 
    <P> 
  <DT><B><A NAME="ligate">ligate</A></B> 
  <DD>In molecular biology, to join two separate <A HREF="glossary.html#DNA">DNA</A> 
    or <A HREF="glossary.html#RNA">RNA</A> segments to form a single DNA or RNA 
    molecule <A HREF="glossary.html#enzyme">enzymatically</A>. 
    <P> 
  <DT><B><A NAME="ligand">ligand</A></B> 
  <DD>A molecule that binds to a receptor <A HREF="glossary.html#protein">protein</A>. 
    <P> 
  <DT><B><A NAME="linkage">linkage</A></B> 
  <DD>The property displayed by two <A HREF="glossary.html#gene">genes</A> that 
    do not <A HREF="glossary.html#segregation">segregate</A> independently of 
    each other. Genes that are linked are on the same <A HREF="glossary.html#chromosome">chromosome</A>. 
    <P> 
  <DT><B><A NAME="linkage_analysis">linkage analysis</A></B> 
  <DD>The construction of a <A HREF="glossary.html#linkage_map">linkage map</A> 
    through the analysis of <A HREF="glossary.html#meiosis">meiotic</A> <A HREF="glossary.html#recombination">recombination</A> 
    frequencies between pairs of <A HREF="glossary.html#gene">genes</A>. 
    <P> 
  <DT><B><A NAME="linkage_map">linkage map</A></B> 
  <DD>A type of <A HREF="glossary.html#genetic_map">genetic map</A> showing relative 
    <A HREF="glossary.html#gene">gene</A> positions based on <A HREF="glossary.html#meiosis">meiotic</A> 
    <A HREF="glossary.html#recombination">recombination</A> frequencies. The 
    unit of measurement is the <A HREF="glossary.html#centimorgan">centimorgan</A>. 
    <P> 
  <DT><B><A NAME="locus">locus</A></B> 
  <DD>Literally, "place". The location of a <A HREF="glossary.html#gene">gene</A> 
    or set of genes on a <A HREF="glossary.html#chromosome">chromosome</A>. 
    <P> 
  <DT><B><A NAME="LOH2">LOH</A></B> 
  <DD>See <a href="#LOH">loss of heterozygosity</a>. 
    <P> 
  <DT><B><A NAME="loss-of-function">loss-of-function mutation</A></B> 
  <DD>A type of <A HREF="glossary.html#mutation">mutation</A> in which the altered 
    <A HREF="glossary.html#gene_product">gene product</A> lacks the <A HREF="glossary.html#molecular_function">molecular 
    function</A> of the <A HREF="glossary.html#wild_type">wild-type</A> <A HREF="glossary.html#gene">gene</A>. 
    Synonyms: <A HREF="glossary.html#amorphic">amorphic mutation</A>, <A HREF="glossary.html#null_mutation">null 
    mutation</A>. <BR>
    See also: 
    <UL>
      <LI><A HREF="glossary.html#antimorphic">antimorphic mutation</A></LI>
      <LI><A HREF="glossary.html#gain-of-function">gain-of-function mutation</A></LI>
      <LI><A HREF="glossary.html#hypermorphic">hypermorphic mutation</A></LI>
      <LI><A HREF="glossary.html#hypomorphic">hypomorphic mutation</A></LI>
      <LI><A HREF="glossary.html#neomorphic">neomorphic mutation</A></LI>
    </UL>
    <P> 
  <DT><B><A NAME="LOH">loss of heterozygosity</A></B> 
  <DD>A genetic event that can occur in the dividing cells of a <A HREF="glossary.html#diploid">diploid</A> 
    organism <A HREF="glossary.html#heterozygous">heterozygous</A> for one or 
    more <A HREF="glossary.html#marker">markers</A>, in which a daughter cell 
    becomes <A HREF="glossary.html#homozygous">homozygous</A> or <A HREF="glossary.html#hemizygous">hemizygous</A> 
    for one or more <A HREF="glossary.html#allele">alleles</A> through mitotic 
    <A HREF="glossary.html#recombination">recombination</A>, <A HREF="glossary.html#deficiency">deficiency</A>, 
    or <A HREF="glossary.html#gene_conversion">gene conversion</A>. "Loss of 
    heterozygosity (LOH)" events are often important steps in tumor progression. 
    See the <A HREF="http://www.people.virginia.edu/~rjh9u/lohanim.html">animation</A> 
    of this concept by Robert J. Huskey at the Univeristy of Virginia. 
    <P> 
  <DT><B><A NAME="loxP">loxP Sequence</A></B> 
  <DD>A 34 base pair <A HREF="glossary.html#DNA">DNA</A> sequence recognized 
    by the <A HREF="glossary.html#site-specific_recombination">site-specific 
    recombination</A> <A HREF="glossary.html#enzyme">enzyme</A> <A HREF="glossary.html#Cre">Cre</A>. 
    <!-- MMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMM -->
    <A NAME="m"></A> 
    <P> 
  <DT><B><A NAME="marker">marker</A></B> 
  <DD> 1. Any biological feature that can be positioned with respect to other 
    features on a chromosome, by genetic, physical or other mapping methods. For 
    example, a gene, anonymous DNA segment, mutation, or phenotype.<BR>
    2. A feature that distinguishes a particular biological state. For example, 
    an expression profile of natural or engineered genes, or a characteristic 
    morphology. <BR>
    3. In ZFIN, a Marker is an object for which unique official nomenclature must 
    be assigned. Markers in ZFIN can be of type: <A HREF="glossary.html#gene">gene</a>, 
    <a href="glossary.html#mutant">mutant</a>, <a href="glossary.html#BAC">BAC</a>/<a href="glossary.html#YAC">YAC</a>,<a href="glossary.html#cDNA"> 
    cDNA</a>, <a href="glossary.html#EST">EST</a>, <a href="glossary.html#SSLP">SSLP</a>, 
    <a href="glossary.html#SSR">SSR</a>, <a href="glossary.html#STS">STS</a>, 
    <a href="glossary.html#RAPD">RAPD</a>, <a href="glossary.html#RFLP">RFLP</a>. 
  <dd> 
    <p> 
  <dt><b><a name="mb">Mb</a></b> 
  <dd><a href="glossary.html#megabase">megabase</a>. 
  <dt>&nbsp;</dt>
  <DD> 
    <P> 
  <DT><B><A NAME="megabase">megabase (Mb)</A></B> 
  <DD>Unit of <A HREF="glossary.html#DNA">DNA</A> or <A HREF="glossary.html#RNA">RNA</A> 
    sequence equal to one million <A HREF="glossary.html#nucleotide">nucleotides</A>. 
    Abbreviated <a href="glossary.html#mb">Mb</a>. 
    <P> 
  <DT><B><A NAME="meiosis">meiosis</A></B> 
  <DD>A pair of nuclear divisions forming <A HREF="glossary.html#gamete">gametes</A> 
    wherein the number of <A HREF="glossary.html#chromosome">chromosomes</A> 
    is reduced from the diploid to the haploid number; resulting cells normally 
    contain one member of each pair of <A HREF="glossary.html#homolog">homologous</A> 
    chromosomes. 
    <P> 
  <DT><B><A NAME="meiotic_map">meiotic map</A></B> 
  <DD>See <a href="glossary.html#linkage_map">linkage map</a>. 
    <P> 
  <DT><B><A NAME="meiotic_mapping_panel">meiotic maping panel</A></B> 
  <DD>A set of <a href="glossary.html#DNA">DNAs</a> used to generate a <a href="glossary.html#linkage_map">linkage 
    map</a>. 
    <P> 
  <DT><B><A NAME="membrane">membrane</A></B> 
  <DD> 1. A phospholipid bilayer that forms a <A HREF="glossary.html#hydrophobic">hydrophobic</A> 
    barrier around and within cells.<BR>
    2. A sheet of nylon, nitrocellulose, or similar material that is used to create 
    a replica of a gel for <A HREF="glossary.html#southern_blot">Southern blots</A>, 
    <A HREF="glossary.html#northern_blot">Northern blots</A>, or <A HREF="glossary.html#western_blot">Western 
    blots</A>. 
    <P> 
  <DT><B><A NAME="mendelian">mendelian</A></B> 
  <DD> 1. That type of inheritance in which a specific <A HREF="glossary.html#trait">trait</A> 
    is affected by a set of <A HREF="glossary.html#allele">alleles</A> of a single 
    <A HREF="glossary.html#gene">gene</A>.<BR>
    2. That type of inheritance in which genetic information is transmitted by 
    one or more <A HREF="glossary.html#nucleus">nuclear</A> genes, as opposed 
    to cytoplasmic or <A HREF="glossary.html#epigenetic">epigenetic</A> mechanisms. 
    <P> 
  <DT><B><A NAME="MGI">MGI</A></B> 
  <DD><A HREF="http://www.informatics.jax.org/mgihome">Mouse Genome Informatics</A>. 
    The collection of bioinformatics projects at The Jackson Laboratory. 
    <P> 
  <DT><B><A NAME="microarray">microarray</A></B> 
  <DD>An array of <a href="glossary.html#DNA">DNA</a> or <a href="glossary.html#protein">protein</a> 
    samples that can be hybridized with <a href="glossary.html#probe">probes</a> 
    to study patterns of <a href="glossary.html#gene_expression">gene expression</a>. 
    <P> 
  <DT><B><A NAME="microsatellite">microsatellite marker</A></B> 
  <DD>See <a href="glossary.html#SSLP">SSLP</a>. 
    <P> 
  <DT><B><A NAME="microtubule">microtubule</A></B> 
  <DD>A cytoskeletal element of eukaryotic cells that is a long, generally straight, 
    hollow tube with an external diameter of 24 nm, consisting of polymerized 
    monomers of tubulin. Microtubules make up the bulk of the <A HREF="glossary.html#spindle">spindle</A>. 
    <P> 
  <DT><B><A NAME="mitochondria">mitochondria</A></B> 
  <DD>The <A HREF="glossary.html#organelle">organelles</A> that generate energy 
    in eukaryotic cells. Mitochondria have their own <A HREF="glossary.html#genome">genome</A> 
    encoding a subset of the <A HREF="glossary.html#protein">proteins</A> found 
    in mitochondria; the mitochondrial genome uses an alternate <A HREF="glossary.html#genetic_code">genetic 
    code</A>. 
    <P> 
  <DT><B><A NAME="mitochondrial">mitochondrial gene</A></B> 
  <DD>A <A HREF="glossary.html#gene">gene</A> contained within the mitochondrial 
    <A HREF="glossary.html#genome">genome</A> of a eukaryote, transmitted independently 
    of the nuclear genome. The mitochondrial genome is transmitted maternally 
    (from the female parent). 
    <P> 
  <DT><B><A NAME="mitosis">mitosis</A></B> 
  <DD>The division of the replicated <A HREF="glossary.html#chromosome">chromosomes</A> 
    of a eukaryotic cell into two daughter <A HREF="glossary.html#nucleus">nuclei</A> 
    that are genetically identical to that of the original cell. See the <A 

HREF="http://www.nhgri.nih.gov/DIR/VIP/Glossary/Illustration/metaphase.html">Figure</A> 
    at <A HREF="glossary.html#NHGRI">NHGRI</A>. 
    <P> 
  <DT><B><A NAME="mobile_genetic">mobile genetic element</A></B> 
  <DD>A <A HREF="glossary.html#DNA">DNA</A> segment carried within the <A HREF="glossary.html#chromosome">chromosomes</A> 
    that is capable of moving to new sites in the <A HREF="glossary.html#genome">genome</A> 
    other than by <A HREF="glossary.html#mutation">mutation</A>. See also <A HREF="glossary.html#retrotransposon">Retrotransposon</A>. 
    <P> 
  <DT><B><A NAME="molecular_function">molecular function</A></B> 
  <DD>Refers to the tasks or activities characteristic of particular <A HREF="glossary.html#gene">gene</A> 
    products. For example, <B>transcription factor</B> refers to one of a number 
    of <A HREF="glossary.html#protein">proteins</A> performing similar tasks. 
    In the GO Project vocabularies, Molecular Function is a primary class of terms. 
    See the <A HREF="http://www.geneontology.org/">GO Consortium</A> site for 
    further information. 
    <P> 
  <DT><B><A NAME="monoclonal">monoclonal antibody</A></B> 
  <DD>An <A HREF="glossary.html#antibody">antibody</A> produced by cultured cells 
    that have their origin in a single antibody-producing cell, and which is therefore 
    of a single molecular type, in contrast to the <B>polyclonal</B> antibodies 
    normally found in the serum of an immunized animal. 
    <P> 
  <DT><B><A NAME="monosomy">monosomy</A></B> 
  <DD>The condition of having a single <A HREF="glossary.html#chromosome">chromosome</A> 
    of a particular type; lacking a <A HREF="glossary.html#homolog">homologous</A> 
    chromosome. See also <A HREF="glossary.html#trisomy">Trisomy</A>. 
    <P> 
  <DT><B><A NAME="morpholino">morpholino</A></B> 
  <DD>An antisense oligonucleotide modified to make them more stable than RNA. 
    Morpholinos are used to inhibit <a href="glossary.html#translation">translation</a> 
    or <a href="glossary.html#splicing">splicing</a> of particular <a href="glossary.html#mRNA">mRNAs</a>. 
    The resulting reduction in gene function is sometimes casually referred to 
    as a <a href="glossary.html#knockdown">knock-down</a>. 
    <P> 
  <DT><B><A NAME="mosaic">mosaic</A></B> 
  <DD>An individual consisting of cells of two or more <A HREF="glossary.html#genotype">genotypes</A>. 
    One example is a wild-type host into which mutant cells have been transplanted. 
    <P> 
  <DT><B><A NAME="mRNA">mRNA</A></B> 
  <DD>Messenger RNA. An <A HREF="glossary.html#RNA">RNA</A> molecule that is 
    the product of <A HREF="glossary.html#transcription">transcription</A> of 
    a <A HREF="glossary.html#gene">gene</A>, after that molecule has been <A HREF="glossary.html#splicing">spliced</A> 
    and <A HREF="glossary.html#polyadenylation">polyadenylated</A>, that can 
    be <A HREF="glossary.html#translation">translated</A> into a <A HREF="glossary.html#protein">protein</A> 
    product. See the <A 

HREF="http://www.nhgri.nih.gov/DIR/VIP/Glossary/Illustration/mrna.html">Figure</A> 
    at <A HREF="glossary.html#NHGRI">NHGRI</A>. See also <A HREF="glossary.html#central_dogma">central dogma</A>. 
    <P> 
  <DT><B><A NAME="mutagen">mutagen</A></B> 
  <DD>An agent that causes <A HREF="glossary.html#mutation">mutations</A>. 
    <P> 
  <DT><B><A NAME="mutant">mutant</A></B> 
  <DD>1. A term applied to a <A HREF="glossary.html#gene">gene</a> or <A HREF="glossary.html#phenotype">phenotype</A> 
    altered by <A HREF="glossary.html#mutation">mutation</A>.<BR>
    2. An individual carrying a mutation. 
    <P> 
  <DT><B><A NAME="mutation">mutation</A></B> 
  <DD>1. The process through which <A 

HREF="glossary.html#gene">genes</A> undergo a structural change.<BR>
    2. Any permanent change in <A HREF="glossary.html#DNA">DNA</A>, i.e., in 
    its <A HREF="glossary.html#nucleotide">nucleotide</A> sequence. Examples 
    include <A HREF="glossary.html#chromosome_rearrangement">chromosome rearrangements</A> 
    and <A HREF="glossary.html#point_mutation">point mutations</A>. See also: 
    <UL>
      <LI><A HREF="glossary.html#chromosome_rearrangement">chromosome rearrangement</A></LI>
      <LI><A HREF="glossary.html#deficiency">deficiency</A></LI>
      <LI><A HREF="glossary.html#insertion">insertion</A></LI>
      <LI><A HREF="glossary.html#inversion">inversion</A></LI>
      <LI><A HREF="glossary.html#nucleotide_repeat">nucleotide repeat expansion</A></LI>
      <LI><A HREF="glossary.html#nucleotide_substitution">nucleotide substitution</A></LI>
      <LI><A HREF="glossary.html#transition">point mutation (transition)</A></LI>
      <LI><A HREF="glossary.html#transversion">point mutation (transversion)</A></LI>
      <LI><A HREF="glossary.html#translocation">translocation</A></LI>
      <LI><A HREF="glossary.html#transposition">transposition</A></LI>
      . 
    </UL>
    <!-- NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN -->
    <A NAME="n"></A> 
    <P> 
  <DT><B><A NAME="name">name</A></B> 
  <DD>1. As used in <A HREF="glossary.html#ZFIN">ZFIN</A>, a <A HREF="glossary.html#gene">gene</A> 
    "name" is a word or phrase that uniquely identifies a gene. The gene name 
    has an abbreviation that is the <A HREF="glossary.html#gene_symbol">gene 
    symbol</A>.<BR>
    2. As used in ZFIN, an <A HREF="glossary.html#allele">allele</A> "name" is 
    a set of letters and numbers that uniquely identifies a particular allele 
    of a gene. The allele name is the <A HREF="glossary.html#allele_designation">allele 
    designation</A>. 
    <P> 
  <DT><B><A NAME="NCBI">NCBI</A></B> 
  <DD><A HREF="http://www.ncbi.nlm.nih.gov/">National Center for Biotechnology 
    Information</A>. 
    <P> 
  <DT><B><A NAME="neomorphic">neomorphic mutation</A></B> 
  <DD>A type of <A HREF="glossary.html#mutation">mutation</A> in which the altered 
    <A HREF="glossary.html#gene_product">gene product</A> possesses a novel <A HREF="glossary.html#molecular_function">molecular 
    function</A> or a novel pattern of <A HREF="glossary.html#gene_expression">gene 
    expression</A>. Antimorphic mutations are usually <A HREF="glossary.html#dominant">dominant</A> 
    or <A HREF="glossary.html#semidominant">semidominant</A>. <BR>
    See also: 
    <UL>
      <LI><A HREF="glossary.html#amorphic">amorphic mutation</A></LI>
      <LI><A HREF="glossary.html#antimorphic">antimorphic mutation</A></LI>
      <LI><A HREF="glossary.html#gain-of-function">gain-of-function mutation</A></LI>
      <LI><A HREF="glossary.html#hypermorphic">hypermorphic mutation</A></LI>
      <LI><A HREF="glossary.html#hypomorphic">hypomorphic mutation</A></LI>
      <LI><A HREF="glossary.html#loss-of-function">loss-of-function mutation</A></LI>
      <LI><A HREF="glossary.html#null_mutation">null mutation</A></LI>
    </UL>
    <P> 
  <DT><B><A NAME="NHGRI">NHGRI</A></B> 
  <DD><A HREF="http://www.nhgri.nih.gov/">National Human Genome Research Institute</A>. 
    <P> 
  <DT><B><A NAME="noncoding_RNA">noncoding RNA</A></B> 
  <DD>An <A HREF="glossary.html#RNA">RNA</A> molecule that functions structurally 
    or catalytically (see <A HREF="glossary.html#ribozyme">ribozyme</A>) without 
    being <A HREF="glossary.html#translation">translated</A>. Noncoding RNAs 
    lack conserved <A HREF="glossary.html#open_reading_frame">open reading frames</A>. 
    <P> 
  <DT><B><A NAME="non-mendelian">non-Mendelian</A></B> 
  <DD>1. That type of inheritance in which a specific <A HREF="glossary.html#trait">trait</A> 
    is affected by a set of <A HREF="glossary.html#allele">alleles</A> of multiple 
    <A HREF="glossary.html#gene">genes</A>. Synonym: <A HREF="glossary.html#polygenic">polygenic</A>.<BR>
    2. That type of inheritance in which genetic information is transmitted other 
    than by <A HREF="glossary.html#nucleus">nuclear</A> genes. See <A HREF="glossary.html#epigenetic">epigenetic</A>, 
    <A HREF="glossary.html#mitochondria">mitochondria</A>. 
    <P> 
  <DT><B><A NAME="northern_blot">Northern blot</A></B> 
  <DD>An assay that detects specific <A HREF="glossary.html#RNA">RNA</A> molecules 
    using a <A HREF="glossary.html#DNA">DNA</A> or RNA <A HREF="glossary.html#probe">probe</A> 
    with sequence similarity. Samples are subjected to <A HREF="glossary.html#electrophoresis">electrophoresis</A> 
    on a slab gel. A replica of the gel is then made on a membrane by capillary 
    transfer. Specific RNA sequences are then detected on the membrane with a 
    radioactively- or chemically-labeled probe. See also <A HREF="glossary.html#southern_blot">Southern 
    blot</A> and <A HREF="glossary.html#western_blot">Western blot</A>. 
    <P> 
  <DT><B><A NAME="nucleic_acid">nucleic Acid</A></B> 
  <DD> <A HREF="glossary.html#DNA">DNA</A> or <A HREF="glossary.html#RNA">RNA</A>. 
    Each of these compounds consists of a backbone of sugar molecules (ribose 
    for RNA and deoxyribose for DNA) linked by single phosphate groups. Attached 
    to the sugars of the backbone are any of four nitrogenous <A HREF="glossary.html#base">bases</A>, 
    A, T, C or G for DNA and A, U, C or G for RNA. See the <A 

HREF="http://www.nhgri.nih.gov/DIR/VIP/Glossary/Illustration/rna.html">Figure</A> 
    at <A HREF="glossary.html#NHGRI">NHGRI</A>. 
    <P> 
  <DT><B><A NAME="nucleotide">nucleotide</A></B> 
  <DD>A monomer unit of <A HREF="glossary.html#nucleic_acid">nucleic acid</A>, 
    consisting of a <A HREF="glossary.html#purine">purine</A> or <A HREF="glossary.html#pyrimidine">pyrimidine</A> 
    base, a sugar molecule (ribose or deoxyribose), and phosphate group(s). 
    <P> 
  <DT><B><A NAME="nucleotide_repeat">nucleotide repeat expansion</A></B> 
  <DD>A type of <A HREF="glossary.html#mutation">mutation</A> in which a set 
    of tandemly repeated sequences <A HREF="glossary.html#replication">replicates</A> 
    inaccurately to increase the number of repeats. An example of this kind of 
    mutation in humans is the <A 

HREF="http://www3.ncbi.nlm.nih.gov/htbin-post/Omim/dispmim?309550#MOLECULAR 

GENETICS">FMR1</A> gene. See also <A HREF="glossary.html#microsatellite">microsatellite</A>. 
    <P> 
  <DT><B><A NAME="nucleotide_substitution">nucleotide substitution</A></B> 
  <DD>A <A HREF="glossary.html#point_mutation">point mutation</A>. 
    <P> 
  <DT><B><A NAME="nucleus">nucleus</A></B> 
  <DD>The <A HREF="glossary.html#organelle">organelle</A> in a eukaryotic cell 
    that contains the <A HREF="glossary.html#chromosome">chromosomes</A>. In 
    most types of eukaryotic cells, the nucleus breaks down as chromosomes condense 
    during cell division. See the <A 

HREF="http://www.nhgri.nih.gov/DIR/VIP/Glossary/Illustration/nucleus.html">Figure</A> 
    at <A HREF="glossary.html#NHGRI">NHGRI</A>. 
    <P> 
  <DT><B><A NAME="null_mutation">null mutation</A></B> 
  <DD>A type of <A HREF="glossary.html#mutation">mutation</A> in which the altered 
    <A HREF="glossary.html#gene_product">gene product</A> lacks the <A HREF="glossary.html#molecular_function">molecular 
    function</A> of the <A HREF="glossary.html#wild_type">wild-type</A> <A HREF="glossary.html#gene">gene</A>. 
    Synonyms: <A HREF="glossary.html#amorphic">amorphic mutation</A>, <A HREF="glossary.html#loss-of-function">loss-of-function 
    mutation</A>. See also: 
    <UL>
      <LI><A HREF="glossary.html#antimorphic">antimorphic mutation</A></LI>
      <LI><A HREF="glossary.html#gain-of-function">gain-of-function mutation</A></LI>
      <LI><A HREF="glossary.html#hypermorphic">hypermorphic mutation</A></LI>
      <LI><A HREF="glossary.html#hypomorphic">hypomorphic mutation</A></LI>
      <LI><A HREF="glossary.html#neomorphic">neomorphic mutation</A></LI>
    </UL>
    <P> 
  <DT><B><A NAME="numerical_aberration">numerical aberration</A>n</B> 
  <DD>A change in the number of <A HREF="glossary.html#chromosome">chromosomes</A> 
    from the <A HREF="glossary.html#wild_type">wild-type</A> number in the absence 
    of any <A HREF="glossary.html#chromosome_rearrangement">chromosome rearrangement</A>. 
    See also <A HREF="glossary.html#monosomy">monosomy</A>, <A HREF="glossary.html#trisomy">trisomy</A>. 
    <!-- OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO -->
    <A NAME="o"></A> 
    <P> 
  <DT><B><A NAME="offset">offset</A></B> 
  <DD>In <A HREF="glossary.html#ZFIN">ZFIN</A>, the limits of the position of 
    a <A HREF="glossary.html#gene">gene</A> on a <A HREF="glossary.html#chromosome">chromosome</A> 
    on the <A HREF="glossary.html#cytogenetic_map">cytogenetic map</A> or the 
    <A HREF="glossary.html#linkage_map">linkage map</A>. 
    <P> 
  <DT><B><A NAME="oligonucleotide">oligonucleotide</A></B> 
  <DD>A short length of <A HREF="glossary.html#nucleic_acid">nucleic acid</A>. 
    <P> 
  <DT><B><A NAME="OMIM">OMIM</A></B> 
  <DD>Online Mendelian Inheritance in Man. A <a href="http://www3.ncbi.nlm.nih.gov/Omim/">database</a> 
    of human inherited diseases and genes. 
    <P> 
  <DT><B><A NAME="ontology">ontology</A></B> 
  <DD>As used by researchers interested in the representation of biological knowledge 
    by computer programs and databases, "ontology" refers to a controlled vocabulary, 
    or set of such vocabularies, used to describe biological features within a 
    specified domain of biological knowledge. 
    <P> 
  <DT><B><A NAME="open_reading_frame">open reading frame (ORF)</A></B> 
  <DD>A length of <A HREF="glossary.html#nucleotide">nucleotide</A> sequence 
    that lacks <A HREF="glossary.html#termination_codon">termination codons</A> 
    in a given <A HREF="glossary.html#reading_frame">reading frame</A>. 
    <P> 
  <DT><B><A NAME="organelle">organelle</A></B> 
  <DD>One of a number of different kinds of <A HREF="glossary.html#membrane">membrane</A>-bound 
    substructures within a eukaryotic cell. Examples include the <A HREF="glossary.html#nucleus">nucleus</A>, 
    <A HREF="glossary.html#mitochondria">mitochondria</A>, and <B>chloroplasts</B>. 
    <P> 
  <DT><B><A NAME="ortholog">orthologue</A></B> 
  <DD>One of a set of homologous <A HREF="glossary.html#gene">genes</a> that 
    have diverged from each other as a consequence of speciation. For example, 
    the alpha globin genes of mouse and chick are orthologues. See the <A HREF="http://www.ncbi.nlm.nih.gov/Education/BLASTinfo/Orthology.html">Figure</A> 
    at NCBI. Some mammalian genes have two orthologues in zebrafish due to <a href="glossary.html#duplication">gene 
    duplication</a>. See also <A HREF="glossary.html#homolog">homologue</A>, <A HREF="glossary.html#paralog">paralog</A>, 
    <A HREF="glossary.html#orthology">orthology</A>. 
    <P> 
  <DT><B><A NAME="orthology">orthology</A></B> 
  <DD>The relationship of any two homologous characters whose common ancestor 
    lies in the most recent common ancestor of the taxa being considered. In ZFIN, 
    assertions of mammalian homology imply presumed orthology.&nbsp; See also: 
    <A HREF="glossary.html#homology">homology</A>, <A HREF="glossary.html#ortholog">orthologue</A>, 
    <A HREF="glossary.html#paralogy">paralogy</A>. 
    <P> 
  <DT><B><A NAME="feature">other genome feature</A></B> 
  <DD>In <A HREF="glossary.html#ZFIN">ZFIN</A>, "other genome feature" refers 
    to any feature of the <A HREF="glossary.html#genome">genome</A> that is considered 
    to have biological significance but cannot be classified with defined <A HREF="glossary.html#marker">marker</A> 
    types. Major classes of "other genome features" include <A HREF="glossary.html#endogenous">endogenous</A> 
    <A HREF="glossary.html#virus">viruses</A> and <A HREF="glossary.html#retrotransposon">retrotransposons</A>, 
    Integration Sites, and Repetitive Elements. An additional class of such features 
    includes genomic segments that function or are biologically significant as 
    <A HREF="glossary.html#DNA">DNA</A> elements. 
    <P> 
  <DT><B><A NAME="outcross">outcross</A></B> 
  <DD>A type of genetic <A HREF="glossary.html#cross">cross</A> in which an organism 
    is crossed to a strain from which it was not recently derived. See also: 
    <UL>
      <LI><A HREF="glossary.html#backcross">backcross</A></LI>
      <LI><A HREF="glossary.html#incross">incross</A></LI>
      <LI><A HREF="glossary.html#intercross">intercross</A></LI>
      <LI><A HREF="glossary.html#testcross">testcross</A></LI>
    </UL>
    <!-- PPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPPP -->
    <A NAME="p"></A> 
    <P> 
  <DT><B><A NAME="P1">P1</A></B> 
  <DD>A <A HREF="glossary.html#bacteriophage">bacteriophage</A> with a genome 
    size of over 100 kb that has been used as a <A HREF="glossary.html#cloning_vector">cloning 
    vector</A>. 
    <P> 
  <DT><B><A NAME="PAC">PAC</A></B> 
  <DD>P1 Artificial Chromosome. A type of <A HREF="glossary.html#cloning_vector">cloning 
    vector</A> derived from <A HREF="glossary.html#bacteriophage">bacteriophage</A> 
    P1 that allows foreign <A HREF="glossary.html#DNA">DNA</A> segments to be 
    cloned in bacteria. The capacity of a PAC is up to 100 kb of foreign DNA. 
    <P> 
  <DT><B><A NAME="paralog">paralog</A></B> 
  <dd>One of a set of homologous <A HREF="glossary.html#gene">genes</a> that 
    have diverged from each other as a consequence of genetic duplication. For 
    example, the human alpha globin and beta globin genes are paralogs. The relationship 
    between human alpha globin and mouse beta globin is also considered paralogous. 
    See the <A HREF="http://www.ncbi.nlm.nih.gov/Education/BLASTinfo/Orthology.html">Figure</A> 
    at NCBI. See also <A HREF="glossary.html#homolog">homologue</A>, <A HREF="glossary.html#ortholog">orthologue</A>, 
    and <A HREF="glossary.html#paralogy">paralogy</A>. 
    <P> 
  <DT><B><A NAME="paralogy">paralogy</A></B> 
  <dd>The relationship of any two homologous characters that arose by a genetic 
    duplication. See also <A HREF="glossary.html#homology">homology</A>, <A HREF="glossary.html#orthology">orthology</A>, 
    and <A HREF="glossary.html#paralog">paralog</A>. 
    <P> 
  <DT><B><A NAME="parent">parent</A></B> 
  <DD>1. Mother or father of an individual or a set of progeny.<br>
    2. In <A HREF="glossary.html#ZFIN">ZFIN</A>, this term refers to terms in 
    a hierarchical <A HREF="glossary.html#controlled_vocabulary">controlled vocabulary</A> 
    such as those containing <A HREF="glossary.html#gene_ontology">Gene Ontology 
    (GO)</A> terms. A "parent" of a term is a one any number of levels above it 
    in the hierarchy from which it is descended. For example, the GO term <A HREF="glossary.html#enzyme">enzyme</A> 
    [GO:0003824] is a <B>parent</B> to the GO term <B>alcohol dehydrogenase</B> 
    [GO:0004022]. See also: <A HREF="glossary.html#children">children</A>, <A HREF="glossary.html#ancestor">ancestor</A>, 
    <A HREF="glossary.html#sibling">sibling</A>. 
    <P> 
  <DT><B><A NAME="pcr">PCR</A></B> 
  <DD>Polymerase Chain Reaction. A method of amplifying specific <A HREF="glossary.html#DNA">DNA</A> 
    segments based on hybridization to a <A HREF="glossary.html#primer">primer</A> 
    pair. A DNA sample is <A HREF="glossary.html#denaturation">denatured</A> 
    by heating in the presence of a vast molar excess of short single-stranded 
    DNA primers (around 20 <A HREF="glossary.html#nucleotide">nucleotides</A>) 
    whose sequence is chosen based on the target sequence. The reaction mixture 
    also contains a <A HREF="glossary.html#thermostable">thermostable</A> <A HREF="glossary.html#DNA_polymerase">DNA 
    polymerase</A>, <A HREF="glossary.html#dNTP">dNTPs</A>, and buffer. The primer 
    sequences are selected so that they: 1) are derived from opposite strands 
    of the target sequence, 2) have their <A HREF="glossary.html#3'">3' ends</A> 
    facing each other, and 3) are separated by a length of DNA that can be reliably 
    synthesized <A HREF="glossary.html#in_vitro">in vitro</A>. The sample is 
    then cooled to a temperature that allows primer annealing and in vitro <A HREF="glossary.html#replication">replication</A>. 
    The sample is subjected to multiple cycles of denaturation and cooling to 
    allow multiple rounds of replication. The quantity of the target sequence 
    doubles during each cycle, causing the target sequence to be amplified, while 
    other <A HREF="glossary.html#DNA">DNA</A> sequences in the sample remain 
    unamplified. See the <A 

HREF="http://www.accessexcellence.org/AB/GG/polymerase.html">Figure</A> at Access 
    Excellence. 
    <P> 
  <DT><B><A NAME="penetrance">penetrance</A></B> 
  <DD>The fraction of individuals of a given <A HREF="glossary.html#genotype">genotype</A> 
    that show a particular <A HREF="glossary.html#phenotype">phenotype</A>, usually 
    expressed as a percentage. See also <A HREF="glossary.html#expressivity">expressivity</A>. 
    <P> 
  <DT><B><A NAME="phage">phage</A></B> 
  <DD>1. A <A HREF="glossary.html#bacteriophage">bacteriophage</A>, a virus capable 
    of infecting bacteria. <BR>
    2. A type of <A HREF="glossary.html#cloning_vector">cloning vector</A> derived 
    from a bacteriophage, usually capable of carrying an amount of foreign <A HREF="glossary.html#DNA">DNA</A> 
    that is at the upper range of that carried by a <A HREF="glossary.html#plasmid">plasmid</A>. 
    <P> 
  <DT><B><A NAME="phagemid">phagemid</A></B> 
  <DD>A type of <A HREF="glossary.html#cloning_vector">cloning vector</A> derived 
    from a <A HREF= "glossary.html#phage">phage</A> and a <A HREF= "glossary.html#plasmid">plasmid</A>. 
    Phagemids are capable of carrying an amount of foreign <A HREF="glossary.html#DNA">DNA</A> 
    comparable to a plasmid, but have some special feature such as the ability 
    to produce single-stranded DNA. 
    <P>
  <DT><B><A NAME="phenocopy">phenocopy</A></B>
  <DD>The condition of an individual resembling that of a <A HREF="glossary.html#phenotype">phenotype</A> 
    produced by a particular <A HREF="glossary.html#mutation">mutation</A> by 
    some experimental treatment other than the presence of that mutation, e.g., 
    drug treatment, <a href="glossary.html#morpholino">morpholino</a> injection. 
    <P> 
  <DT><B><A NAME="phenotype">phenotype</A></B> 
  <DD>A description of the observable state of an individual with respect to some 
    inherited characteristic. Often, individuals with different <A HREF="glossary.html#genotype">genotypes</A> 
    display the same phenotype. See <A HREF="glossary.html#dominant">dominant</A> 
    and <A HREF="glossary.html#recessive">recessive</A>. 
    <P> 
  <DT><B><A NAME="phosphorimaging">phosphorimaging</A></B> 
  <DD>The detection of <A HREF="glossary.html#radiation">radioactivity</A> using 
    "phosphor" compounds that emit visible light when exposed to radiation. Phosphorimaging 
    instruments produce images of, for example, <A HREF="glossary.html#southern_blot">Southern 
    blots</A> and <A HREF="glossary.html#northern_blot">Northern blots</A>, that 
    are comparable to those produced by <A HREF="glossary.html#autoradiography">autoradiography</A>, 
    with superior quantitation. 
    <P> 
  <DT><B><A NAME="phycoerythrin">phycoerythrin</A></B> 
  <DD>A fluorescent dye that can be coupled to <A HREF="glossary.html#antibody">antibodies</A> 
    for the detection of <A HREF="glossary.html#protein">proteins</A> on <A HREF="glossary.html#western_blot">Western 
    blots</A> using <A HREF="glossary.html#fluorography">fluorography</A>. 
    <P> 
  <DT><B><A NAME="physical_map">physical map</A></B> 
  <DD>A map of <A HREF="glossary.html#DNA">DNA</A> showing distances between 
    and within <A HREF="glossary.html#gene">genes</A> or specified <A HREF="glossary.html#marker">markers</A> 
    measured in <A HREF="glossary.html#base_pair">base pairs</A> of DNA. It is 
    based on the direct measurement of DNA. 
    <P> 
  <DT><B><A NAME="plasmid">plasmid</A></B> 
  <DD>A type of <A HREF="glossary.html#cloning_vector">cloning vector</A> derived 
    from autonomously-replicating extrachromosomal circular <A HREF="glossary.html#DNA">DNAs</A> 
    in bacteria. The amount of foreign DNA that can be carried in a plasmid is 
    small, ranging up to about 20 <A HREF="glossary.html#kb">kb</A>. 
    <P> 
  <DT><B><A NAME="pleiotropy">pleiotropy</A></B> 
  <DD>The production of a <A HREF="glossary.html#phenotype">phenotype</A> affecting 
    multiple <A HREF="glossary.html#trait">traits</A> by a single <A HREF="glossary.html#mutation">mutation</A>. 
    <P> 
  <DT><B><A NAME="point_mutation">point mutation</A></B> 
  <DD>A type of <A HREF="glossary.html#mutation">mutation</A> in which a single 
    <A HREF="glossary.html#nucleotide">nucleotide</A> is changed to one of the 
    other three possible nucleotides. See also <A HREF="glossary.html#nucleotide_substitution">nucleotide 
    substitution</A> <A HREF="glossary.html#transition">yransition</A>, <A HREF="glossary.html#transversion">transversion</A>. 
    <P> 
  <DT><B><A NAME="polyadenylation">polyadenylation</A></B> 
  <DD>The process by which a series of adenosine (A) ribonucleotides is added 
    to the <A HREF="glossary.html#3'">3'</A> end of a <A HREF="glossary.html#splicing">spliced</A> 
    <A HREF="glossary.html#RNA">RNA</A> to make a mature <A HREF="glossary.html#mRNA">mRNA</A>. 
    This addition to the RNA is sometimes referred to as a poly-A tail, and commonly 
    contains several hundred bases. 
    <P> 
  <DT><B><A NAME="polygenic">polygenic</A></B> 
  <DD>A <A HREF="glossary.html#trait">trait</A> determined by multiple <A HREF="glossary.html#gene">genes</A>. 
    <P> 
  <DT><b><a name="polymorphic_strains">polymorphic strains</a></b> 
  <DD>In zebrafish, polymorphic strains have differences in DNA sequence at many 
    loci. A panel of recombinant offspring derived from a cross between two polymorphic 
    parental strains can be used to establish <a href="glossary.html#linkage">linkage 
    </a> between any <a href="glossary.html#marker">marker</a> that is <a href="glossary.html#polymorphism">polymorphic</a> 
    between the parental strains and other polymorphic markers that have been 
    typed in each strain in the panel. 
    <P> 
  <DT><B><A NAME="polymorphism">polymorphism</A></B> 
  <DD>An instance of <A HREF="glossary.html#genotype">genotypic</A> variation 
    within a population. 
    <P> 
  <DT><B><A NAME="primer">primer</A></B> 
  <DD>A single-stranded <A HREF="glossary.html#nucleic_acid">nucleic acid</A> 
    that can "prime" <A HREF="glossary.html#replication">replication</A> of a 
    <A HREF="glossary.html#template">template</A>. More specifically, a single-stranded 
    nucleic acid capable of <A HREF="glossary.html#hybridization">hybridizing</A> 
    to a template single-stranded nucleic acid in such a way as to leave part 
    of the template to the <A HREF="glossary.html#3'">3' end</A> of the primer 
    single-stranded. <A HREF="glossary.html#DNA_polymerase">DNA polymerase</A> 
    can then synthesize a new strand starting from the 3' end of the primer, adding 
    <A HREF="glossary.html#nucleotide">nucleotides</A> to the growing strand 
    by base complementarity to the template. See also <A HREF="glossary.html#pcr">PCR</A>. 
    <P> 
  <DT><B><A NAME="probe">probe</A></B> 
  <DD>1. In molecular biology, a <A HREF="glossary.html#nucleic_acid">nucleic 
    acid</A> that has been labeled either radioactively or chemically that allows 
    the detection of nucleic acids with sequence similarity in a sample by hybridization. 
    Probes are used to detect <A HREF="glossary.html#DNA">DNA</A> on membranes 
    in <A HREF="glossary.html#southern_blot">Southern blots</A>, to detect <A HREF="glossary.html#RNA">RNA</A> 
    on membranes in <A HREF="glossary.html#northern_blot">Northern blots</A>, 
    and either DNA or RNA in cytological preparations for <A HREF="glossary.html#in_situ"><I>in 
    situ</I></A> hybridization.<BR>
    2. In ZFIN, the term "probe" applies not only to nucleic acid probes detected 
    as described above, but also to <a href="glossary.html#antibody">antibody</a> 
    probes used to detect proteins by <a href="glossary.html#immunohistochemistry">immunohistochemistry</a>. 
    <P> 
  <DT><B><A NAME="promoter">promoter</A></B> 
  <DD>One of the necessary <A HREF="glossary.html#regulatory_element">regulatory 
    elements</A> of a <A HREF="glossary.html#gene">gene</A>. The promoter is 
    the <A HREF="glossary.html#DNA">DNA</A> sequence to which <A HREF="glossary.html#RNA_polymerase">RNA 
    polymerase</A> binds and initiates <A HREF="glossary.html#transcription">transcription</A>. 
    See also <A HREF="glossary.html#enhancer">enhancer</A>. 
    <P> 
  <DT><B><A NAME="protein">protein</A></B> 
  <DD>A polymer of <A HREF="glossary.html#amino_acid">amino acids</A>. See the 
    <A 

HREF="http://www.nhgri.nih.gov/DIR/VIP/Glossary/Illustration/amino_acid.html">Figure</A> 
    at <A HREF="glossary.html#NHGRI">NHGRI</A>. 
    <P> 
  <DT><B><A NAME="protein_domain">protein domain</A></B> 
  <DD>A region of a <A HREF="glossary.html#protein">protein</A> responsible for 
    a particular function, as recognized experimentally and by the occurrence 
    of <A HREF="glossary.html#similarity">similar</A> segments in other proteins 
    sharing that function, e.g., a <A HREF="glossary.html#DNA">DNA</A> binding 
    domain. 
    <P> 
  <DT><B><A NAME="protein_histochemistry">protein histochemistry</A></B> 
  <DD>1. A method of detecting a particular <A HREF="glossary.html#enzyme">enzyme</A> 
    in a cell or tissue sample. A sample of cells or tissue is fixed, then treated 
    with a <A HREF="glossary.html#chromogenic">chromogenic</A> <A HREF="glossary.html#substrate">substrate</A> 
    for the enzyme to be detected. Microscopic examination reveals the presence 
    of staining, and hence of the specific <A HREF="glossary.html#protein">protein</A> 
    to be detected.<BR>
    2. <A HREF="glossary.html#immunohistochemistry">Immunohistochemistry</A>. 
    <P> 
  <DT><B><A NAME="proteome">proteome</A></B> 
  <DD>The complete collection of all <A HREF="glossary.html#protein">proteins</A> 
    encoded by the <A HREF="glossary.html#genome">genome</A> of an organism. 
    <P> 
  <DT><B><A NAME="pseudoautosomal">pseudoautosomal</A></B> 
  <DD>The small region of <A HREF="glossary.html#homolog">homology</A> shared 
    between the <A HREF="glossary.html#X_chromosome">X chromosome</A> and the 
    <A HREF="glossary.html#Y_chromosome">Y chromosome</A> in mammals. All crossovers 
    between the X and Y chromosomes occur in this region. 
    <P> 
  <DT><B><A NAME="pseudogene">pseudogene</A></B> 
  <DD>A non-functional <A HREF="glossary.html#locus">locus</A> derived from a 
    functional locus either by 1) replicative transfer, such as <A HREF="glossary.html#transposition">transposition</A>, 
    <A HREF="glossary.html#retrotransposon">retrotransposition</A> or <A HREF="glossary.html#duplication">duplication</A> 
    or by 2) <A HREF="glossary.html#mutation">mutation</A>, where the non-functional 
    locus is not considered an <A HREF="glossary.html#allele">allele</A> of an 
    existing functional locus. 
    <P> 
  <DT><B><A NAME="purine">purine</A></B> 
  <DD>One of the <A HREF="glossary.html#base">bases</A> of <A HREF="glossary.html#nucleic_acid">nucleic 
    acids</A>, either adenine (A) or guanine (G). See the <A 

HREF="http://www.nhgri.nih.gov/DIR/VIP/Glossary/Illustration/rna.html">Figure</A> 
    at <A HREF="glossary.html#NHGRI">NHGRI</A>. 
    <P> 
  <DT><B><A NAME="pyrimidine">pyrimidine</A></B> 
  <DD>One of the <A HREF="glossary.html#base">bases</A> of <A HREF="glossary.html#nucleic_acid">nucleic 
    acids</A>, cytosine (C), thymine (T) or uracil (U). See the <A 

HREF="http://www.nhgri.nih.gov/DIR/VIP/Glossary/Illustration/rna.html">Figure</A> 
    at <A HREF="glossary.html#NHGRI">NHGRI</A>. 
    <!-- QQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQ -->
    <A NAME="q"></A> 
    <P> 
  <DT><B><A NAME="QTL">quantitative trait locus (QTL)</A></B> 
  <DD>The type of <A HREF="glossary.html#marker">marker</A> described by statistical 
    association to quantitative variation in a particular <A HREF="glossary.html#phenotype">phenotypic</A> 
    trait that is thought to be controlled by the cumulative action of <A HREF="glossary.html#allele">alleles</A> 
    at multiple <A HREF="glossary.html#locus">loci</A>. 
    <P> 
  <DT><B><A NAME="query">query</A></B> 
  <DD>A request for information submitted to a computerized database. 
    <P> 
  <DT><B><A NAME="query_form">query form</A></B> 
  <DD>As used in <A HREF="glossary.html#ZFIN">ZFIN</A>, a Query Form is a web 
    page allowing users to retrieve information from the ZFIN database. 
    <P> 
  <DT><B><A NAME="query_sequence">query sequence</A></B> 
  <DD>A <A HREF="glossary.html#DNA">DNA</A> or <A HREF="glossary.html#protein">protein</A> 
    sequence submitted to a computerized database for comparison, e.g., a <A HREF="glossary.html#BLAST">BLAST</A> 
    search. 
    <!-- RRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRR -->
    <A NAME="r"></A> 
    <P> 
  <DT><B><A NAME="radiation">radiation</A></B> 
  <DD>1. Electromagnetic energy: gamma rays, X rays, ultraviolet light, visible 
    light, infrared light, microwaves and radio waves. In zebrafish genetics, 
    this term generally refers to gamma rays and X rays.<BR>
    2. Subatomic particles emitted by the decay of unstable isotopes: electrons 
    (beta particles) and helium nuclei (alpha particles). Common unstable isotopes 
    in molecular biology are tritium (<SUP>3</SUP>H), which emits low-energy beta 
    particles, <SUP>35</SUP>S, which emits beta particles of moderate energy, 
    and <SUP>32</SUP>P, which emits high-energy beta particles.<BR>
    3. Subatomic particles from a particle accelerator, such as protons, neutrons, 
    and electrons. 
    <P> 
  <DT><B><A NAME="radiation_hybrid_mapping">radiation hybrid (RH) mapping</A></B> 
  <DD>A type of <A HREF="glossary.html#genetic_map">genetic mapping</A> providing 
    resolution between relatively low-resolution <A HREF="glossary.html#linkage_analysis">linkage 
    analysis</A> and high-resolution <A HREF="glossary.html#physical_map">physical 
    mapping</A> by the assembly of contiguous cloned DNA segments. The method 
    consists of fusing <A HREF="glossary.html#radiation">irradiated</A> cultured 
    cells of one species with cultured cells of a different species. A panel of 
    hybrid cells is then tested for the occurrence of pairs of <A HREF="glossary.html#marker">markers</A>. 
    The closer two markers are to each other, the more likely that both are present 
    in an individual hybrid cell. Radiation hybrid mapping data for the zebrafish 
    are available a at <A HREF="glossary.html#ZFIN">ZFIN</A>. 
    <P> 
  <DT><B><A NAME="radiation_hybrid_panel">radiation hybrid (RH) panel</A></B> 
  <DD>A set of <a href="glossary.html#DNA">DNAs</a> used to generate a <a href="glossary.html#radiation_hybrid_mapping"> 
    radiation hybrid map</a>. 
    <P> 
  <DT><B><A NAME="RAPD">RAPD</A></B> 
  <DD>Random Amplified Polymorphic DNA. Segments of DNA amplified by <a href="glossary.html#PCR">PCR</a> 
    using short primers with sequences chosen at random that are used as <a href="glossary.html#polymorphism">polymorphic</a> 
    <a href="glossary.html#marker">markers</a> for mapping. 
    <P> 
  <DT><B><A NAME="reading_frame">reading frame</A></B> 
  <DD>One of three ways of reading a single strand of <A HREF="glossary.html#nucleic_acid">nucleic 
    acid</A> sequence as <A HREF="glossary.html#codon">codons</A>. 
    <!-- addition of xrefs to Recessive -->
    <P> 
  <DT><B><A NAME="recessive">recessive</A></B> 
  <DD>One of a series of terms applied to the <A HREF="glossary.html#phenotype">phenotypic</A> 
    effect of a particular <A HREF="glossary.html#allele">allele</A> in reference 
    to another allele (usually the standard <A HREF="glossary.html#wild_type">wild-type</A> 
    allele) with respect to a given <A HREF="glossary.html#trait">trait</A>. 
    An allele "a" is said to be recessive with respect to the allele "A" if the 
    A/A <A HREF="glossary.html#homozygote">homozygote</A> and the A/a <A HREF="glossary.html#heterozygote">heterozygote</A> 
    are phenotypically identical and different from the a/a homozygote. See also 
    <A HREF="glossary.html#codominant">Codominant</A>, <A HREF="glossary.html#dominant">Dominant</A>, 
    <A HREF="glossary.html#semidominant">Semidomint</A>. 
    <P> 
  <DT><B><A NAME="recombination">recombination</A></B> 
  <DD>Transfer of information from one <A HREF="glossary.html#DNA">DNA</A> molecule 
    to another. Recombination may be <B>reciprocal</B>, in which case the products 
    are equivalent to breakage of the two DNA molecules and rejoining of the broken 
    ends to form new molecules. Recombination may also be <B>nonreciprocal</B>, 
    in which case the product is equivalent to transfer of information from the 
    donor DNA molecule to the recipient DNA molecule, with no change in the donor 
    DNA molecule. Reciprocal recombination events are also called <A HREF="glossary.html#crossover">crossovers</A>. 
    <P> 
  <DT><B><A NAME="regulatory_element">regulatory element</A></B> 
  <DD>A <A HREF="glossary.html#DNA">DNA</A> sequence that is required for a <A HREF="glossary.html#gene">gene</A> 
    on the same DNA molecule to be <A HREF="glossary.html#transcription">transcribed</A>, 
    or to be transcribed in the proper cell type(s) and developmental stage(s). 
    See also <A HREF="glossary.html#enhancer">enhancer</A>, <A HREF="glossary.html#promoter">promoter</A>. 
    <P> 
  <DT><B><A NAME="regulatory_gene">regulatory gene</A></B> 
  <DD>A <A HREF="glossary.html#gene">gene</A> whose function is to regulate the 
    <A HREF="glossary.html#gene_expression">expression</A> of a <A HREF="glossary.html#structural_gene">structural 
    gene</A>. 
    <P> 
  <DT><B><A NAME="replication">replication</A></B> 
  <DD>The process of synthesizing a copy of a <A HREF="glossary.html#DNA">DNA</A> 
    molecule from <A HREF="glossary.html#nucleotide">nucleotides</A> using information 
    contained within one strand of a <A HREF="glossary.html#template">template</A> 
    DNA molecule. The new strand of DNA is synthesized from the <A HREF="glossary.html#5'">5' 
    end</A> to the <A HREF="glossary.html#3'">3' end</A>. See the <A 

HREF="http://www.nhgri.nih.gov/DIR/VIP/Glossary/Illustration/dna_replication.html">Figure</A> 
    at <A HREF="glossary.html#NHGRI">NHGRI</A>. 
    <P> 
  <DT><B><A NAME="reporter_gene">reporter gene</A></B> 
  <DD>A <A HREF="glossary.html#gene">gene</A> whose product is easily detected 
    and not ordinarily present in an organism or cell type under study that is 
    expressed as part of a <A HREF="glossary.html#DNA_construct">DNA construct</A> 
    introduced experimentally. Bacterial beta-galactosidase, whose activity can 
    be detected using a staining reaction, is a commonly used reporter gene. See 
    also <A HREF="glossary.html#enhancer_trap">enhancer trap</A>, <A HREF="glossary.html#gene_trap">gene 
    trap</A>. 
    <P> 
  <DT><B><A NAME="restriction_enzyme">restriction enzyme</A></B> 
  <DD>A <A HREF="glossary.html#protein">protein</A> that recognizes specific, 
    short <A HREF="glossary.html#nucleotide">nucleotide</A> sequences and cuts 
    <A HREF="glossary.html#DNA">DNA</A> at those sites. 
    <P> 
  <DT><B><A NAME="restriction_fragment">restriction fragment</A></B>
<DD>A length of <A HREF="glossary.html#DNA">DNA</A> whose ends are the result 
    of cutting by a <A HREF="glossary.html#restriction_enzyme">restriction enzyme</A>. 
    <P> 
  <DT><B><A NAME="retrotransposon">retrotransposon</A></B> 
  <DD>A type of <A HREF="glossary.html#mobile_genetic">mobile genetic element</A> 
    that utilizes an <A HREF="glossary.html#RNA">RNA</A> intermediate and <A HREF="glossary.html#reverse_transcriptase">reverse 
    transcriptase</A> to transpose. 
    <P> 
  <DT><B><A NAME="retrovirus">retrovirus</A></B> 
  <DD>A <A HREF="glossary.html#virus">virus</A> whose primary genetic material 
    is <A HREF="glossary.html#RNA">RNA</A> instead of <A HREF="glossary.html#DNA">DNA</A>. 
    <A HREF="glossary.html#replication">Replication</A> of the <A HREF="glossary.html#genome">genome</A> 
    of such a virus requires the RNA to be copied into DNA using <A HREF="glossary.html#reverse_transcriptase">reverse 
    transcriptase</A>. This group of viruses includes HIV (AIDS virus). 
    <P> 
  <DT><B><A NAME="reverse_transcriptase">reverse transcriptase</A></B> 
  <DD>An <A HREF="glossary.html#enzyme">enzyme</A> that is able to synthesize 
    <A HREF="glossary.html#DNA">DNA</A> from information in <A HREF="glossary.html#RNA">RNA</A>. 
    It requires an RNA <A HREF="glossary.html#template">template</A> and a DNA 
    or RNA <A HREF="glossary.html#primer">primer</A>. See also <A HREF="glossary.html#cDNA">cDNA</A>. 
    <!-- new 1/19 --Reversion and Revertant -->
    <P> 
  <DT><B><A NAME="reversion">reversion</A></B> 
  <DD>A <A HREF="glossary.html#mutation">mutation</A> event that alters an <A HREF="glossary.html#allele">allele</A> 
    conferring a <A HREF="glossary.html#mutant">mutant</A> <A HREF="glossary.html#phenotype">phenotype</A> 
    into one conferring a <A HREF="glossary.html#wild_type">wild-type</A> phenotype. 
    The mutation need not restore the <A HREF="glossary.html#gene">gene</A> to 
    its original <A HREF="glossary.html#allele">nucleotide</A> sequence to be 
    considered a reversion event. 
    <P> 
  <DT><B><A NAME="revertant">revertant</A></B> 
  <DD>An individual carrying an <A HREF="glossary.html#allele">allele</A> of 
    a given <A HREF="glossary.html#allele">gene</A> that at one time produced 
    a <A HREF="glossary.html#allele">mutant</A> <A HREF="glossary.html#allele">phenotype</A>, 
    but which since has undergone a subsequent <A HREF="glossary.html#allele">mutation</A> 
    that restored a <A HREF="glossary.html#allele">wild-type</A> phenotype. The 
    mutation need not restore the gene to its original <A HREF="glossary.html#allele">nucleotide</A> 
    sequence to be considered a <A HREF="glossary.html#reversion">reversion</A> 
    event. 
    <P> 
  <DT><B><A NAME="RFLP">RFLP</A></B> 
  <DD>Restriction Fragment Length Polymorphism. A genetic <A HREF="glossary.html#polymorphism">polymorphism</A> 
    with respect to the observed length of a <A HREF="glossary.html#restriction_fragment">restriction 
    fragment</A>. RFLPs can result from <A HREF="glossary.html#SNP">single nucleotide 
    polymorphisms</A> as well as from <A HREF="glossary.html#insertion">insertions</A>, 
    <A HREF="glossary.html#deficiency">deficiencies</A>, or <A HREF="glossary.html#microsatellite">microsatellite</A> 
    expansions. 
    <P> 
  <DT><B><A NAME="ribosome">ribosome</A></B> 
  <DD>A complex of <A HREF="glossary.html#protein">proteins</A> and <A HREF="glossary.html#RNA">RNA</A> 
    within which <A HREF="glossary.html#translation">translation</A> is carried 
    out. 
    <P> 
  <DT><B><A NAME="ribozyme">ribozyme</A></B> 
  <DD>An <A HREF="glossary.html#RNA">RNA</A> molecule with catalytic activity. 
    <P> 
  <DT><B><A NAME="RNA">RNA</A></B> 
  <DD>Ribonucleic acid. A <A HREF="glossary.html#nucleic_acid">nucleic acid</A> 
    that is the primary product of <A HREF="glossary.html#gene_expression">gene 
    expression</A>. Chemically, it differs from <A HREF="glossary.html#DNA">DNA</A> 
    by the substitution of ribose for deoxyribose in the sugar-phosphate backbone 
    and by the substitution of the base uracil for thymine. See the <A 

HREF="http://www.nhgri.nih.gov/DIR/VIP/Glossary/Illustration/rna.html">Figure</A> 
    at <A HREF="glossary.html#NHGRI">NHGRI</A>. See also <A HREF="glossary.html#central_dogma">central dogma</A> and <a href="glossary.html#DNA">DNA</a>. 
    <P> 
  <DT><B><A NAME="RNA_editing">RNA editing</A></B> 
  <DD>The alteration of the sequence of an <A HREF="glossary.html#RNA">RNA</A> 
    molecule by <A HREF="glossary.html#enzyme">enyzmatic</A> modification of 
    individual <A HREF="glossary.html#base">bases</A> without <A HREF="glossary.html#splicing">splicing</A>. 
    <P> 
  <DT><B><A NAME="RNA_polymerase">RNA polymerase</A></B> 
  <DD>An <A HREF="glossary.html#enzyme">enzyme</A> that carries out <A HREF="glossary.html#transcription">transcription</A>. 
    <P> 
  <DT><B><A NAME="RNA_processing">RNA processing</A></B> 
  <DD>Modifications to an <A HREF="glossary.html#RNA">RNA</A> molecule after 
    <A HREF="glossary.html#transcription">transcription</A>, including <A HREF="glossary.html#splicing">splicing</A>, 
    <A HREF="glossary.html#polyadenylation">polyadenylation</A>, and <A HREF="glossary.html#RNA_editing">editing</A>. 
    <P> 
  <DT><B><A NAME="RNAse">RNAse</A></B> 
  <DD>Ribonuclease. A <A HREF="glossary.html#protein">protein</A> that cleaves 
    the phosphodiester backbone of <A HREF="glossary.html#RNA">RNA</A> <A HREF="glossary.html#enzyme">enzymatically</A>. 
    <P> 
  <DT><B><A NAME="RNAse_protection">RNAse protection</A></B> 
  <DD>A method of detecting the presence of a specific <A HREF="glossary.html#RNA">RNA</A> 
    in a sample. A radioactively-labeled RNA <A HREF="glossary.html#probe">probe</A> 
    is prepared by <A HREF="glossary.html#transcription">transcribing</A> the 
    <A HREF="glossary.html#antisense">antisense</A> strand of a <A HREF="glossary.html#DNA_construct">DNA 
    construct</A>. The labeled <A HREF="glossary.html#probe">probe</A> is hybridized 
    to the sample. The sample is then treated with <A HREF="glossary.html#RNAse">RNAse</A>, 
    which is specific to single-stranded RNA. The sample is then subjected to 
    <A HREF="glossary.html#electrophoresis">electrophoresis</A> and <A HREF="glossary.html#autoradiography">autoradiography</A>. 
    The presence of full-length probe that has not been cleaved by RNAse indicates 
    the presence of the <A HREF="glossary.html#sense">sense</A> strand, and hence 
    <A HREF="glossary.html#gene_expression">gene expression</A>, in the sample. 
    <P> 
  <DT><B><A NAME="rRNA">rRNA</A></B> 
  <DD>Ribosomal RNA. The <A HREF="glossary.html#RNA">RNA</A> molecules that are 
    a structural and catalytic component of the <A HREF="glossary.html#ribosome">ribosome</A>. 
    <P> 
  <DT><B><A NAME="RT-PCR">RT-PCR</A></B> 
  <DD>Reverse-Transcription PCR. A method of amplifying <A HREF="glossary.html#mRNA">mRNA</A> 
    by first synthesizing <A HREF="glossary.html#cDNA">cDNA</A> with <A HREF="glossary.html#reverse_transcriptase">reverse 
    transcriptase</A>, then amplifying the cDNA using <A HREF="glossary.html#pcr">PCR</A>. 
    A positive result is evidence of a particular mRNA, and hence of <A HREF="glossary.html#gene_expression">gene 
    expression</A>, in a sample. 
    <!-- SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS -->
    <A NAME="s"></A> 
    <P> 
  <DT><B><A NAME="segregation">segregation</A></B> 
  <DD>1. The separation of <A HREF="glossary.html#homolog">homologous</A> chromosomes 
    during <A HREF="glossary.html#meiosis">meiosis</A>.<BR>
    2. The separation of different <A HREF="glossary.html#allele">alleles</A> 
    of the same <A HREF="glossary.html#gene">gene</A> during meiosis. 
    <!--addition of xrefs to Semidominant 1/19  -->
    <P> 
  <DT><B><A NAME="semidominant">semidominant</A></B> 
  <DD>One of a series of terms applied to the <A HREF="glossary.html#phenotype">phenotypic</A> 
    effect of a particular <A HREF="glossary.html#allele">allele</A> in reference 
    to another allele (usually the standard <A HREF="glossary.html#wild_type">wild-type</A> 
    allele) with respect to a given <A HREF="glossary.html#trait">trait</A>. 
    An allele "A" is said to be semidominant with respect to the allele "a" if 
    the A/A <A HREF="glossary.html#homozygote">homozygote</A> has a mutant phenotype, 
    the A/a <A HREF="glossary.html#heterozygote">heterozygote</A> has a less 
    severe phenotype, while the a/a <A HREF="glossary.html#homozygote">homozygote</A> 
    is wild-type. See also <A HREF="glossary.html#codominant">codominant</A>, 
    <A HREF="glossary.html#dominant">dominant</A>, <A HREF="glossary.html#recessive">recessive</A>. 
    <P> 
  <DT><B><A NAME="sense">sense</A></B> 
  <DD>1. In molecular biology, that strand of a <A HREF="glossary.html#DNA">DNA</A> 
    molecule whose sequence is represented in <A HREF="glossary.html#mRNA">mRNA</A>.<BR>
    2. In molecular biology, an <A HREF="glossary.html#RNA">RNA</A> molecule 
    normally processed into <A HREF="glossary.html#mRNA">mRNA</A> and <A HREF="glossary.html#translation">translated</A> 
    (rather than the <A HREF="glossary.html#complementary">complementary sequence</A>). 
    <P> 
  <DT><B><A NAME="sequence_annotation">sequence annotation</A></B> 
  <DD>Additional information added to <A HREF="glossary.html#genome">genomic</A> 
    sequence to identify <A HREF="glossary.html#gene">genes</A>, delimit the 
    <A HREF="glossary.html#intron">intron</A> and <A HREF="glossary.html#exon">exon</A> 
    structures of those <A HREF="glossary.html#gene">genes</A>, identify <A HREF="glossary.html#regulatory_element">regulatory 
    elements</A>, note the positions of <A HREF="glossary.html#allele">allelic</A> 
    variation, etc. 
    <!--new 1/19 Sex Chromosome and Sex Linked -->
    <P> 
  <DT><B><A NAME="sex_chromosome">sex chromosome</A></B> 
  <DD>Either of two <A HREF="glossary.html#chromosome">chromosomes</A> that are 
    sexually <A HREF="glossary.html#dimorphic">dimorphic</A> in species with 
    chromosomal (as opposed to genic) sex determination. In mammals, males are 
    the <A HREF="glossary.html#heterogametic">heterogametic</A> sex, having an 
    <A HREF="glossary.html#X_chromosome">X chromosome</A> and a <A HREF="glossary.html#Y_chromosome">Y 
    chromosome</A>, while females are the <A HREF="glossary.html#homogametic">homogametic</A> 
    sex, having two X chromosomes. 
    <P> 
  <DT><B><A NAME="sex_linked">sex linked</A></B> 
  <DD>Located on the <A HREF="glossary.html#X_chromosome">X chromosome</A> or 
    the <A HREF="glossary.html#Y_chromosome">Y chromosome</A>. 
    <P> 
  <DT><B><A NAME="sibling">sibling</A></B> 
  <DD>1. A brother or sister sharing the same parents.<br>
    2. In <A HREF="glossary.html#ZFIN">ZFIN</A>, this term refers to terms in 
    a hierarchical <A HREF="glossary.html#controlled_vocabulary">controlled vocabulary</A> 
    like ones containing <A HREF="glossary.html#gene_ontology">Gene Ontology 
    (GO)</A> terms. A "sibling" of a term is a term at the same level of the hierarchy 
    sharing at least one <A HREF="glossary.html#ancestor">ancestor</A>. For example, 
    the GO term <B>alcohol dehydrogenase</B> [GO:0004022] is a <B>sibling</B> 
    to the GO term <B>aldehyde oxidase</B> [GO:0004031]; they share the <A HREF="glossary.html#ancestor">ancestor</A> 
    term <A HREF="glossary.html#enzyme">enzyme</A> [GO:0003824]. See also: <A HREF="glossary.html#ancestor">ancestor</A>, 
    <A HREF="glossary.html#children">children</A>. 
    <P> 
  <DT><B><A NAME="similarity">similarity</A></B> 
  <DD>1. In comparison of <A HREF="glossary.html#nucleic_acid">nucleic acid</A> 
    sequences, the extent to which two <A HREF="glossary.html#nucleic_acid">nucleic 
    acid</A> sequences have identical <A HREF="glossary.html#base">bases</A> 
    at equivalent positions, usually expressed as a percentage. <BR>
    2. In comparison of <A HREF="glossary.html#protein">protein</A> sequences, 
    the extent to which the <A HREF="glossary.html#amino_acid">amino acid</A> 
    sequences of two <A HREF="glossary.html#protein">proteins</A> have identical 
    or functionally similar <A HREF="glossary.html#amino_acid">amino acids</A> 
    at equivalent positions, usually expressed as a percentage. See also <A HREF="glossary.html#identity">Identity</A>. 
    <P> 
  <DT><B><A NAME="site-specific_recombination">site-specific recombination</A></B> 
  <DD>Reciprocal <A HREF="glossary.html#recombination">recombination</A> between 
    specific target sequences catalyzed by a specific <A HREF="glossary.html#recombination">recombination</A> 
    <A HREF="glossary.html#enzyme">enzyme</A>, as opposed to general <A HREF="glossary.html#homologous_recombination">homologous 
    recombination</A>. One example is recombination at <A HREF="glossary.html#loxP">loxP</A> 
    sites catalyzed by <A HREF="glossary.html#Cre">Cre recombinase</A>. 
    <P> 
  <DT><B><A NAME="SNP">SNP</A></B> 
  <DD>Single Nucleotide Polymorphism. A type of <A HREF="glossary.html#polymorphism">polymorphism</A> 
    in which two <A HREF="glossary.html#chromosome">chromosomes</A> differ in 
    a given segment by the identity of a single <A HREF="glossary.html#base_pair">base 
    pair</A>. 
    <P> 
  <DT><B><A NAME="southern_blot">Southern blot</A></B> 
  <DD>An assay that detects specific <A HREF="glossary.html#DNA">DNA</A> molecules 
    using a <A HREF="glossary.html#DNA">DNA</A> or <A HREF="glossary.html#RNA">RNA</A> 
    <A HREF="glossary.html#probe">probe</A> with sequence similarity. Samples 
    are subjected to <A HREF="glossary.html#electrophoresis">electrophoresis</A> 
    on a slab gel. A replica of the gel is then made on a membrane by capillary 
    transfer following <A HREF="glossary.html#denaturation">denaturation</A>. 
    Specific <A HREF="glossary.html#DNA">DNA</A> sequences are then detected 
    on the membrane with a radioactively- or chemically-labeled <A HREF="glossary.html#probe">probe</A>. 
    See the <A 

HREF="http://www.accessexcellence.org/AB/GG/southBlotg.html">Figure</A> from Alberts, 
    et al., Molecular Biology of the Cell. See also <A HREF="glossary.html#northern_blot">Northern 
    blot</A> and <A HREF="glossary.html#western_blot">Western blot</A>. 
    <P> 
  <DT><B><A NAME="somatic">somatic</A></B> 
  <DD>Cells of an animal other than those that constitute the <A HREF="glossary.html#germline">germ 
    line</A>. 
    <P> 
  <DT><B><A NAME="somatic_cell_hybrid">somatic cell hybrid</A></B> 
  <DD>A type of mapping experiment permitting the assignment of <A HREF="glossary.html#marker">markers</A> 
    to <A HREF="glossary.html#chromosome">chromosomes</A>. The method consists 
    of fusing cultured cells of one species with cultured cells of a different 
    species. The hybrid cells are unstable in <A HREF="glossary.html#karyotype">karyotype</A> 
    during growth, with most chromosomes from one species typically being lost. 
    Among clonal populations of hybrid cells following growth, different chromosomes 
    are retained from one species. A panel of hybrid cell cultures can be assayed 
    for which zebrafish chromosomes (for example) are retained, and simultaneously 
    assayed for the presence of particular markers. The correlation of the presence 
    of a particular marker across the panel with the presence of a particular 
    zebrafish chromosome allows that marker to be assigned to that chromosome. 
    See also: <A HREF="glossary.html#radiation_hybrid_mapping">Radiation Hybrid 
    Mapping</A>. 
    <P> 
  <DT><B><A NAME="spindle">spindle</A></B> 
  <DD>The cellular apparatus that directs <A HREF="glossary.html#chromosome">chromosome</A> 
    movement during cell division in <A HREF="glossary.html#mitosis">mitosis</A> 
    or <A HREF="glossary.html#meiosis">meiosis</A>. The spindle is largely composed 
    of <A HREF="glossary.html#microtubule">microtubules</A>. See the <A 

HREF="http://www.nhgri.nih.gov/DIR/VIP/Glossary/Illustration/metaphase.html">Figure</A> 
    at <A HREF="glossary.html#NHGRI">NHGRI</A>. 
    <P> 
  <DT><B><A NAME="splice_acceptor">splice acceptor site</A></B> 
  <DD>In the <A HREF="glossary.html#splicing">splicing</A> of <A HREF="glossary.html#RNA">RNA</A>, 
    the site at the <A HREF="glossary.html#3'">3' end</A> of an <A HREF="glossary.html#intron">intron</A>. 
    See also <A HREF="glossary.html#splice_donor">splice donor site</A>. 
    <P> 
  <DT><B><A NAME="splice_donor">splice donor site</A></B> 
  <DD>In the <A HREF="glossary.html#splicing">splicing</A> of <A HREF="glossary.html#RNA">RNA</A>, 
    the site at the <A HREF="glossary.html#5'">5' end</A> of an <A HREF="glossary.html#intron">intron</A>. 
    See also <A HREF="glossary.html#splice_acceptor">splice acceptor site</A>. 
    <P> 
  <DT><B><A NAME="splice_junction">splice junction</A></B> 
  <DD>In the <A HREF="glossary.html#splicing">splicing</A> of <A HREF="glossary.html#RNA">RNA</A>, 
    the site of a former <A HREF="glossary.html#intron">intron</A> in a mature 
    <A HREF="glossary.html#mRNA">mRNA</A>. 
    <P> 
  <DT><B><A NAME="splicing">splicing</A></B> 
  <DD>Part of the processing of an <A HREF="glossary.html#RNA">RNA</A> <A HREF="glossary.html#transcription">transcript</A> 
    into <A HREF="glossary.html#mRNA">mRNA</A>, in which <A HREF="glossary.html#intron">introns</A> 
    are removed <A HREF="glossary.html#enzyme">enzymatically</A>. 
    <P> 
  <DT><B><A NAME="spontaneous">spontaneous</A></B> 
  <DD>As a type of <A HREF="glossary.html#mutation">mutation</A>, a mutation 
    that has occurred in the absence of any experimental <A HREF="glossary.html#mutagen">mutagenic</A> 
    treatment, such as irradiation or treatment with chemical mutagens. 
    <P> 
  <DT><B><A NAME="SQL">SQL</A></B> 
  <DD>Structured Query Language. A programming language for extracting information 
    from relational databases. The <A HREF="glossary.html#query_form">query forms</A> 
    at <A HREF="glossary.html#ZFIN">ZFIN</A> operate by generating instructions 
    in SQL. 
    <P> 
  <DT><B><A NAME="SSLP">SSLP</A></B> 
  <DD>A short (up to several hundred <a href="glossary.html#base_pair">base pairs</a>) 
    segment of <a href="glossary.html#DNA">DNA</a> that consists of multiple 
    tandem repeats of a two or three base-pair sequence. SSLPs expand and contract 
    (that is, add or remove repeat units) with a frequency much higher than other 
    types of <a href="glossary.html#mutation">mutations</a>, making them useful 
    as <a href="glossary.html#polymorphism">polymorphic</a> <a href="glossary.html#marker">markers</a> 
    in closely related zebrafish strains. SSLPs are also sometimes called <a href="glossary.html#microsatellite">microsatellite 
    markers</a>. 
    <p> 
  <DT><b><a name="SSR">SSR</a></b> 
  <DD>See <a href="glossary.html#SSLP">SSLP</a>. 
    <P> 
  <DT><B><A NAME="stop_codon">stop codon</A></B> 
  <DD>One of three <A HREF="glossary.html#codon">codons</A> that signal that 
    <A HREF="glossary.html#translation">translation</A> of an <A HREF="glossary.html#RNA">RNA</A> 
    sequence should cease. 
    <P> 
  <DT><B><A NAME="structural_gene">structural gene</A></B> 
  <DD>A <A HREF="glossary.html#gene">gene</A> that encodes an <A HREF="glossary.html#enzyme">enzyme</A> 
    or <A HREF="glossary.html#structural_protein">structural protein</A>, in 
    contrast to a <A HREF="glossary.html#regulatory_gene">regulatory gene</A>. 
    <P> 
  <DT><B><A NAME="structural_protein">structural protein</A></B> 
  <DD>A <A HREF="glossary.html#protein">protein</A> that functions as a structural 
    element of cells rather than as an <A HREF="glossary.html#enzyme">enzyme</A>, 
    for example, <B>collagen</B>. 
    <P> 
  <DT><B><A NAME="STS">STS</A></B> 
  <DD>Sequence Tagged Site. A short segment of unique sequence derived from <A HREF="glossary.html#genome">genomic</A> 
    <A HREF="glossary.html#DNA">DNA</A>. A large collection of STSs can be used 
    to assemble a <A HREF="glossary.html#physical_map">physical map</A> of the 
    genome from a collection of genomic clones (e.g. <A HREF="glossary.html#BAC">BACs</A> 
    or <A HREF="glossary.html#YAC">YACs</A>) by testing each clone for the presence 
    of each STS. Two clones that contain one or more STSs in common must overlap. 
    <P> 
  <DT><B><A NAME="substrate">substrate</A></B> 
  <DD>A molecule acted upon by an <A HREF="glossary.html#enzyme">enzyme</A>. 
    <P> 
  <DT><B><A NAME="symbol">symbol</A></B> 
  <DD>As used in <A HREF="glossary.html#ZFIN">ZFIN</A>, a <A HREF="glossary.html#gene">gene</A> 
    symbol is a unique abbreviation for the <A HREF="glossary.html#gene_name">gene 
    name</A>. 
    <P> 
  <DT><B><A NAME="synteny">synteny</A></B> 
  <DD>The state of being on the same <A HREF="glossary.html#chromosome">chromosome</A>. 
    A <A HREF="glossary.html#gene">gene</A> is also said to be syntenic to a 
    particular chromosome if it is known to be located on that chromosome but 
    is otherwise unmapped. See also <A HREF="glossary.html#conserved_synteny">conserved 
    synteny</A> 
    <P> 
  <DT><B><A NAME="synonym">synonym</A></B> 
  <DD>In <A HREF="glossary.html#ZFIN">ZFIN</A>, a synonym is a <A HREF="glossary.html#gene_symbol">gene 
    symbol</A> or <A HREF="glossary.html#gene_name">name</A> that has appeared 
    in the scientific literature, is not pending approval, and has never been 
    an <A HREF="glossary.html#approved">approved</A> symbol or name. 
    <!-- TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT -->
    <A NAME="t"></A> 
    <P> 
  <DT><B><A NAME="tab-delimited">tab-delimited</A></B> 
  <DD>A text file with data fields separated by "tab" characters. Such files can 
    be converted to spreadsheet files, such as those used by Microsoft Excel. 
    <P> 
  <DT><B><A NAME="targeted_mutation">targeted mutation</A></B> 
  <DD>A type of <A HREF="glossary.html#mutation">mutation</A> in which a chromosomal 
    <A HREF="glossary.html#gene">gene</A> is altered by the substitution of a 
    <A HREF="glossary.html#DNA_construct">DNA construct</A> assembled <A HREF="glossary.html#in_vitro">in 
    vitro</A>. The constructs are usually designed to eliminate gene function; 
    such targeted mutations are often casually referred to as <A HREF="glossary.html#knockout">knock-outs</A>. 
    Some DNA constructs are designed to alter gene function; such targeted mutations 
    are often casually referred to as <A HREF="glossary.html#knockin">knock-ins</A>. 
    <P> 
  <DT><B><A NAME="telomere">telomere</A></B>
<DD>A specialized structure at the ends of linear <A HREF="glossary.html#chromosome">chromosomes</A> 
    in eukaryotes. Telomeres confer stability on chromosome ends. Chromosome ends 
    lacking telomeres, such as those generated from interstitial sites by chromosome 
    breaks, are reactive, often fusing with other broken ends to generate <A HREF="glossary.html#chromosome_rearrangement">chromosome 
    rearrangements</A>. Telomeres also permit the ends of linear chromosomes to 
    <A HREF="glossary.html#replication">replicate</A> fully. See the <A 

HREF="http://www.nhgri.nih.gov/DIR/VIP/Glossary/Illustration/chromosome.html">Figure</A> 
    at <A HREF="glossary.html#NHGRI">NHGRI</A>. 
    <P> 
  <DT><B><A NAME="template">template</A></B> 
  <DD>In the process of <A HREF="glossary.html#replication">replication</A> or 
    <A HREF="glossary.html#transcription">transcription</A>, the strand of <A HREF="glossary.html#DNA">DNA</A> 
    that serves as the source of information. 
    <P> 
  <DT><B><A NAME="termination_codon">termination codon</A></B> 
  <DD>One of three <A HREF="glossary.html#codon">codons</A> that signal that 
    <A HREF="glossary.html#translation">translation</A> of an <A HREF="glossary.html#RNA">RNA</A> 
    sequence should cease. 
    <P> 
  <DT><B><A NAME="testcross">testcross</A></B> 
  <DD>A type of <A HREF="glossary.html#cross">cross</A> in which individuals 
    whose <A HREF="glossary.html#genotype">genotype</A> with respect to one or 
    more <A HREF="glossary.html#gene">genes</A> is unknown are crossed to a test 
    strain <A HREF="glossary.html#homozygous">homozygous</A> or <a href="glossary.html#heterozygous">heterozygous</a> 
    for a <A HREF="glossary.html#recessive">recessive</A> allele at the genes 
    under study. For example, a cross of an individual that was A/A or A/a (identical 
    in <A HREF="glossary.html#phenotype">phenotype</A>) to a/a would reveal the 
    genotype of the individual being tested, because if the individual being tested 
    were A/A, all of the progeny would show the <A HREF="glossary.html#dominant">dominant</A> 
    phenotype, while if the individual being tested were A/a, half of the progeny 
    would show the dominant phenotype and half would show the <A HREF="glossary.html#recessive">recessive</A> 
    phenotype.<BR>
    See also: 
    <UL>
      <LI><A HREF="glossary.html#backcross">backcross</A></LI>
      <LI><A HREF="glossary.html#incross">incross</A></LI>
      <LI><A HREF="glossary.html#intercross">intercross</A></LI>
      <LI><A HREF="glossary.html#outcross">outcross</A></LI>
    </UL>
    <P> 
  <DT><B><A NAME="texas_red">Texas red</A></B> 
  <DD>A fluorescent dye used to label <A HREF="glossary.html#antibody">antibodies</A> 
    for <A HREF="glossary.html#immunofluorescence">immunofluorescence</A> or 
    <A HREF="glossary.html#western_blot">Western blots</A>. 
    <P> 
  <DT><B><A NAME="thermostable">thermostable</A></B> 
  <DD>Used to describe an <A HREF="glossary.html#enzyme">enzyme</A> or other 
    <A HREF="glossary.html#protein">protein</A> that is not <A HREF="glossary.html#denaturation">denatured</A> 
    at temperatures that denature most other proteins.  
    <P> 
  <DT><B><A NAME="trait">trait</A></B> 
  <DD>A particular aspect of the <A HREF="glossary.html#phenotype">phenotype</A> 
    that can be measured or observed directly, e.g. blood flow or body color. 
    <P> 
  <DT><B><A NAME="transcript">transcript</A></B> 
  <DD>An <A HREF="glossary.html#RNA">RNA</A> molecule (or species of RNA molecule) 
    that is the product of <A HREF="glossary.html#transcription">transcription</A>. 
    <P> 
  <DT><B><A NAME="transcription">transcription</A></B> 
  <DD>The <A HREF="glossary.html#enzyme">enzymatic</A> synthesis of an <A HREF="glossary.html#RNA">RNA</A> 
    molecule directed by information in a <A HREF="glossary.html#DNA">DNA</A> 
    molecule. See the <A 

HREF="http://www.nhgri.nih.gov/DIR/VIP/Glossary/Illustration/mrna.html">Figure</A> 
    at <A HREF="glossary.html#NHGRI">NHGRI</A>. See also <A HREF="glossary.html#central_dogma">central 
    dogma</A>. 
    <P> 
  <DT><B><A NAME="transgene">transgene</A></B> 
  <DD>A <A HREF="glossary.html#gene">gene</A> in a living organism that is derived 
    from another organism and introduced experimentally. 
    <P> 
  <DT><B><A NAME="transition">transition</A></B> 
  <DD>A type of <A HREF="glossary.html#point_mutation">point mutation</A> in 
    which a <A HREF="glossary.html#purine">purine</A> is substituted for another 
    purine or a <A HREF="glossary.html#pyrimidine">pyrimidine</A> for another 
    pyrimidine. These substitutions include A for G, G for A, C for T, or T for 
    C. See also <A HREF="glossary.html#transversion">transversion</A>. 
    <P> 
  <DT><B><A NAME="translation">translation</A></B> 
  <DD>The <A HREF="glossary.html#enzyme">enzymatic</A> synthesis of a <A HREF="glossary.html#protein">protein</A> 
    molecule directed by the information in an <A HREF="glossary.html#mRNA">mRNA</A> 
    molecule. The <A HREF="glossary.html#mRNA">mRNA</A> is read from the <A HREF="glossary.html#5'">5' 
    end</A> to the <A HREF="glossary.html#3'">3' end</A>, with the <A HREF="glossary.html#protein">protein</A> 
    being synthesized from the <A HREF="glossary.html#amino_terminus">amino terminus</A> 
    to the <A HREF="glossary.html#carboxyl_terminus">carboxyl terminus</A>. See 
    the <A 

HREF="http://www.nhgri.nih.gov/DIR/VIP/Glossary/Illustration/mrna.html">Figure</A> 
    at <A HREF="glossary.html#NHGRI">NHGRI</A>. See also <A HREF="glossary.html#central_dogma">central 
    dogma</A>. 
    <P> 
  <DT><B><A NAME="translocation">translocation</A></B> 
  <DD>A type of <A HREF="glossary.html#mutation">mutation</A> in which two <A HREF="glossary.html#homolog">nonhomologous</A> 
    chromosomes are each broken and then repaired in such a way that: 
    <OL>
      <LI>the resulting <A HREF="glossary.html#chromosome">chromosomes</A> each 
        contain material from the other chromosome (a <B>reciprocal</B> translocation; 
        see the <A 

HREF="http://www.nhgri.nih.gov/DIR/VIP/Glossary/Illustration/translocation.html">Figure</A> 
        at <A HREF="glossary.html#NHGRI">NHGRI</A>),</LI>
      <LI>one of the chromosomes contains an <A HREF="glossary.html#insertion">insertion</A> 
        of material from the other chromosome, with the other chromosome containing 
        a <A HREF="glossary.html#deficiency">deficiency</A> (an <B>insertional</B> 
        translocation; see the <A 

HREF="http://www.nhgri.nih.gov/DIR/VIP/Glossary/Illustration/insertion.html">Figure</A> 
        at <A HREF="glossary.html#NHGRI">NHGRI</A>), or 
      <LI>the two chromosomes, each with breaks near the <A HREF="glossary.html#centromere">centromere</A>, 
        fuse to form a single chromosome with a single centromere (a <B>Robertsonian</B> 
        translocation). 
    </OL>
    <P> 
  <DT><B><A NAME="transposition">transposition</A></B> 
  <DD>1. A type of <A HREF="glossary.html#chromosome_rearrangement">chromosome 
    rearrangement</A> in which a segment of a <A HREF="glossary.html#chromosome">chromosome</A> 
    is moved to a different location on the same chromosome, resembling an <A HREF="glossary.html#translocation">insertional 
    translocation</A> involving a single chromosome.<BR>
    2. The movement of a <A HREF="glossary.html#mobile_genetic">mobile genetic 
    element</A> to a new location. 
    <P> 
  <DT><B><A NAME="transposon">transposon</A></B> 
  <DD>A type of <A HREF="glossary.html#mobile_genetic">mobile genetic element</A> 
    consisting of <A HREF="glossary.html#DNA">DNA</A> that moves to new <A HREF="glossary.html#genome">genomic</A> 
    locations <B>conservatively</B> (without replicating itself) or <B>replicatively</B> 
    (moving a copy of itself). 
    <P> 
  <DT><B><A NAME="transversion">transversion</A></B> 
  <DD>A type of <A HREF="glossary.html#point_mutation">point mutation</A> in 
    which a <A HREF="glossary.html#purine">purine</A> is substituted for a <A HREF="glossary.html#pyrimidine">pyrimidine</A> 
    or a pyrimidine for a purine. These substitutions include C or T for A, C 
    or T for G, A or G for C, and A or G for T. See also <A HREF="glossary.html#transition">transition</A>. 
    <P> 
  <DT><B><A NAME="trisomy">trisomy</A></B> 
  <DD>The condition of having three <A HREF="glossary.html#chromosome">chromosomes</A> 
    of a particular type. Down Syndrome in humans is a trisomy for chromosome 
    21. See also <A HREF="glossary.html#monosomy">monosomy</A>. 
    <P> 
  <DT><B><A NAME="tRNA">tRNA</A></B> 
  <DD>Transfer RNA. Small <A HREF="glossary.html#RNA">RNA</A> molecules that 
    bind to the <A HREF="glossary.html#codon">codons</A> of <A HREF="glossary.html#mRNA">mRNA</A> 
    in the <A HREF="glossary.html#ribosome">ribosome</A> after being "charged" 
    with <A HREF="glossary.html#amino_acid">amino acids</A>. 
    <!-- UUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUU -->
    <A NAME="u"></A> 
    <P> 
  <DT><B><A NAME="uniparental_disomy">uniparental disomy</A></B> 
  <DD>The inheritance, in a <A HREF="glossary.html#diploid">diploid</A> organism, 
    of both copies of a single <A HREF="glossary.html#chromosome">chromosome</A> 
    from one parent. This may result from the union of a <A HREF="glossary.html#gamete">gamete</A> 
    bearing two copies of one chromosome with a gamete bearing no copy of that 
    chromosome, or from the union of a gamete bearing two copies of one chromosome 
    with a normal gamete, followed by the loss of one chromosome through an error 
    in <A HREF="glossary.html#mitosis">mitosis</A>. Because of <A HREF="glossary.html#imprinting">imprinting</A>, 
    uniparental disomy can have <A HREF="glossary.html#phenotype">phenotypic</A> 
    consequences in mammals. See, for example, <A 

HREF="http://www3.ncbi.nlm.nih.gov/htbin-post/Omim/dispmim?176270#INHERITANCE">Prader-Willi 
    Syndrome</A>. 
    <P> 
  <DT><B><A NAME="unknown">unknown</A></B> 
  <DD>In <A HREF="glossary.html#ZFIN">ZFIN</A>, a map location of "unknown" means 
    that the <A HREF="glossary.html#marker">marker</A> has not yet been assigned 
    to a <A HREF="glossary.html#chromosome">chromosome</A>. 
    <P> 
  <DT><B><A NAME="URL">URL</A></B> 
  <DD>Uniform Resource Locator. An Internet address giving the protocol to be 
    used for obtaining resources on the Internet such as "ftp:" for an <A HREF="glossary.html#ftp">FTP</A> 
    site or "http:" for a World Wide Web page. It also includes the server name 
    and sometimes the path to the resource. The URL for <A HREF="glossary.html#ZFIN">ZFIN</A> 
    is "http://zfin.org". 
    <!-- VVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVV -->
    <A NAME="v"></A> 
    <P> 
  <DT><B><A NAME="vector">Vector</A></B> 
  <DD>See <A HREF="glossary.html#cloning_vector">cloning vector</A>. 
    <P> 
  <DT><B><A NAME="virus">virus</A></B> 
  <DD>A noncellular biological entity that requires a host cell for reproduction. 
    Viruses consist of a <A HREF="glossary.html#nucleic_acid">nucleic acid</A> 
    <A HREF="glossary.html#genome">genome</A> that is either <A HREF="glossary.html#DNA">DNA</A> 
    or, in the case of <A HREF="glossary.html#retrovirus">retroviruses</A>, <A HREF="glossary.html#RNA">RNA</A>. 
    The viral genome is covered with a <A HREF="glossary.html#protein">protein</A> 
    coat; some viruses have a host-derived membrane over the protein coat. 
    <!-- WWWWWWWWWWWWWWWWWWWWWWWWWWWWWWWWWWWWWWWWWWWWW -->
    <A NAME="w"></A> 
    <P> 
  <DT><B><A NAME="western_blot">Western blot</A></B> 
  <DD>An assay that detects specific <A HREF="glossary.html#protein">proteins</A> 
    within a protein mixture. Samples are subjected to <A HREF="glossary.html#electrophoresis">electrophoresis</A> 
    on a slab gel. A replica of the gel is then made on a membrane by electrophorectic 
    transfer. Specific proteins are then detected on the membrane using <A HREF="glossary.html#antibody">antibody</A> 
    staining. See <A HREF="glossary.html#southern_blot">Southern blot</A> and 
    <A HREF="glossary.html#northern_blot">Northern blot</A>. 
    <P> 
  <DT><B><A NAME="wild_type">wild type</A></B> 
  <DD>1. The <A HREF="glossary.html#phenotype">phenotype</A> with respect to 
    a given inherited characteristic that is considered to be the "normal" type 
    commonly found in natural populations.<BR>
    2. The <A HREF="glossary.html#allele">allele</A> of a particular <A HREF="glossary.html#gene">gene</A> 
    that confers the phenotype considered to be the "normal" type commonly found 
    in natural populations. N.B.: Because some <A HREF="glossary.html#DNA">DNA</A> 
    sequence <A HREF="glossary.html#polymorphism">polymorphisms</A> do not produce 
    different phenotypes, there can be multiple "wild-type" alleles of a gene. 
    <P> 
  <DT><B><A NAME="withdrawn">withdrawn</A></B> 
  <DD>With respect to <A HREF="glossary.html#gene">gene</A> nomenclature, a withdrawn 
    <A HREF="glossary.html#gene_symbol">symbol</A> or <A HREF="glossary.html#gene_name">name</A> 
    was once the <A HREF="glossary.html#approved">approved</A> symbol or name 
    for a <A HREF="glossary.html#marker">marker</A>; there is currently a different 
    approved symbol or name for that marker. 
    <!-- XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX -->
    <A NAME="x"></A> 
    <P> 
  <DT><B><A NAME="X_chromosome">X chromosome</A></B> 
  <DD>One of pair of <A HREF="glossary.html#chromosome">chromosomes</A> that 
    is sexually <A HREF="glossary.html#dimorphic">dimorphic</A> in mammals. Normal 
    female mammals have two X chromosomes, while normal male mammals have an X 
    chromosome and a Y chromosome. 
    <P> 
  <DT><B><A NAME="X_inactivation">X inactivation</A></B> 
  <DD>The condensation of all but one of the <A HREF="glossary.html#X_chromosome">X 
    chromosomes</A> of a mammal into a <A HREF="glossary.html#heterochromatin">heterochromatic</A> 
    state, eliminating <A HREF="glossary.html#gene_expression">gene expression</A> 
    from all but the active X chromosome. This process ensures that male and female 
    mammals have the same level of gene activity of X-chromosome genes. 
    <!-- YYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYY -->
    <A NAME="y"></A> 
    <P> 
  <DT><B><A NAME="Y_chromosome">Y chromosome</A></B> 
  <DD>One of pair of <A HREF="glossary.html#chromosome">chromosomes</A> that 
    is sexually <A HREF="glossary.html#dimorphic">dimorphic</A> in mammals. Normal 
    female mammals have two <A HREF="glossary.html#X_chromosome">X chromosomes</A>, 
    while normal male mammals have an X chromosome and a Y chromosome. 
    <P> 
  <DT><B><A NAME="YAC">YAC</A></B> 
  <DD>Yeast Artificial Chromosome. A type of <A HREF="glossary.html#cloning_vector">cloning 
    vector</A> containing a yeast <A HREF="glossary.html#centromere">centromere</A> 
    and <A HREF="glossary.html#telomere">telomeres</A> that allows large <A HREF="glossary.html#DNA">DNA</A> 
    segments to be cloned in yeast. A YAC can carry 200 - 1000 <A HREF="glossary.html#kb">kb</A> 
    of foreign <A HREF="glossary.html#DNA">DNA</A>. 
    <!-- ZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZ -->
    <A NAME="z"></A> 
    <p> 
  <dt><b><a name="ZFIN">ZFIN</a></b> 
  <dd>The <a href="http://zfin.org/">Zebrafish Information Network</a>. The on-line 
    zebrafish model organism database. 
    <p> 
  <dt><b><a name="ZIRC">ZIRC</a></b> 
  <dd>The <a href="stckctr/stckctr.html">Zebrafish International Resource Center</a>. 
    The resource that provides strains of <a href="glossary.html#wild_type">wild-type</a> 
    and <a href="glossary.html#mutant">mutant</a> zebrafish and research materials 
    used for their study. 
    <p> 
  <dt><b><a name="Z_marker">Z marker</a></b> 
  <dd>One of a large series of <a href="glossary.html#SSLP">SSLP</a> markers 
    in the zebrafish, developed at the Massachusetts General Hospital. These markers 
    have been used to map many <a href="glossary.html#marker">markers</a> and 
    to align the <a href="glossary.html#physical_map">physical</a> and <a href="glossary.html#linkage_map">linkage 
    maps</a> in zebrafish. 
    <!-- Markup for a definition

<P><DT><B><A NAME=""></A></B>

<DD>

End of markup -->
</DL>


<hr>
<script language="JavaScript" src="http://<!--|DOMAIN_NAME|-->/footer.js"></script>