<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html>
	<head>
		<title>
		Nomenclature Tutorial
		</title>
	</head>

<script language="JavaScript" src="/javascript/header.js"></script>



<h1 align="center">A Tutorial for Proposing Zebrafish Gene Nomenclature</h1>


<p align="center">
By the <a href="/zf_info/nomen_comm.html"> Zebrafish Nomenclature Committee</a><br>
Feel free to contact us at:  <a href="mailto:nomenclature@zfin.org">nomenclature@zfin.org</a></font></p>

<hr width="75%"><br>

<p>Finding an appropriate gene name and symbol to use for your newly discovered gene can be a frustrating endeavor! There are so many factors to consider, like whether homologs or orthologs are known in other species, whether it is a member of a gene family, and whether it has a duplicate. We hope the following tips will guide you and make the experience of gene naming more pleasant.</p>

<p><b>How to propose a good gene name and symbol:</b></p>
<ol>
<li> Become familiar with the zebrafish nomenclature
guidelines: <a href="/zf_info/nomen.html">http://zfin.org/zf_info/nomen.html</a>.</li><br><br>

<li>Search <a href="http://zfin.org/">ZFIN</a> (http://zfin.org/) to ensure that your gene hasn't already been named. A good way to do this is to use  <a href="/cgi-bin/webdriver?MIval=aa-blast.apg">ZFIN BLAST</a>.  Use the Nucleotide -Nucleotide BLAST.

<p>If there is no identical  match check to see if your gene has a closely related gene such as genes created by the genome duplication or a tandemly duplicated gene.

<p>Gene duplicates, resulting from the whole genome duplication, often have differences in the third nucleotide or "wobble" position in a codon. Closely related sequences with mismatches separated by stretches of 2, 5, or 8 identical nucleotides may indicate that the genes are duplicates.</li><br><br>

<li>If you believe your gene is the duplicate of a named zebrafish gene based on a high sequence identity and conserved synteny, propose the gene symbol comprised of the existing, named gene symbol with an "a" or "b" attached to the end.  The existing gene will be renamed by adding an "a" or "b" to the original symbol depending on chromosomal location.<br><br>

If you are reporting one or both duplicates and have mapped them to duplicate chromosome segments, make an effort to keep the "a" and "b" terminology consistent with other duplicates already mapped on the chromosomal segment.

<blockquote>
  <blockquote>
    <blockquote>
      <b>Example: </b>You identify the two zebrafish duplicates of
      the human gene <i>ZZZ</i>. You have mapped one duplicate to LG3 near the hoxb "a" genes and the other duplicate you have mapped to LG12 near the hoxb "b" genes.  Propose <i>zzza</i> for the gene on LG3 and <i>zzzb</i> for the gene on LG12 to conform to the nomenclature of hoxb genes already mapped to those chromosomes.
    </blockquote>
  </blockquote>
</blockquote></li><br>

<li>If the gene is new to zebrafish but has a human or mouse
ortholog, check the Human Nomenclature Database or the Mouse Genome Informatics resources for the name and symbol used in those species:

<blockquote>
  <blockquote>
    <blockquote>
        <b>Human:</b> <a href="http://www.gene.ucl.ac.uk/cgi-bin/nomenclature/searchgenes.pl">http://www.gene.ucl.ac.uk/cgi-bin/nomenclature/searchgenes.pl</a><br><br>

        <b>Mouse:</b> <a href="http://www.informatics.jax.org/searches/marker_form.shtml">http://www.informatics.jax.org/searches/marker_form.shtml  </a>
    </blockquote>
  </blockquote>
</blockquote>


<p>Human and mouse gene names in these two resources are approved by nomenclature committees.</p></li><br>

<li>If there is an established gene name and symbol in human for the ortholog, use the same symbol for zebrafish (of course following zebrafish nomenclature guidelines like all lower case italics, no "z" prefix, etc.).  If there is no orthologous human gene, but there is an orthologous mouse gene, use the same name and symbol used for the mouse ortholog.

<blockquote>
  <blockquote>
    <blockquote>
      <b>Example:</b> Suppose you isolate the zebrafish ortholog of the human <i>"dihydrolipoamide branched chain transacylase E2"</i>.  Searching the Human Nomenclature Database returns the gene symbol <i>DBT</i>.  Searching the Mouse Genome Informatics resources returns the gene symbol <i>Dbt</i>.  Therefore, an appropriate symbol adhering to the zebrafish nomenclature guidelines would be <i>dbt</i>.

    </blockquote>
  </blockquote>
</blockquote>

<p>This is only a starting point for proposing a good symbol. You aren't done yet!</p></li><br>

<li>If the gene you are naming is a member of a gene family,
check to see what is used for the "root" symbol of the gene family in zebrafish.
If there is no established gene family in zebrafish but there is in human or mouse, use
the same "root" symbol established in human and mouse.

<blockquote>
  <blockquote>
    <blockquote>
      <b>Example:</b> You isolate a new calcium channel, voltage-dependent, alpha 1 subunit gene.  Searching ZFIN using the gene query form, you find that all the known calcium channel, voltage-dependent, alpha 1 subunits begin with "cacna1".  The last member in the cacna1 series is <i>cacna1f</i>.  Therefore, a good starting point would be to propose <i>cacna1g</i>.
    </blockquote>
  </blockquote>
</blockquote></li><br>

<li>We've already seen how to propose a zebrafish gene symbol
in cases where it is a duplicate, an ortholog of a known human or mouse gene, or when it
is a member of a gene family. What if your gene is none of these?

<blockquote>
  <blockquote>
    <blockquote>
      <b>Example:</b> You isolate a novel gene that is an apoptosis determining protein expressed in neurons.  You want to call it <i>"apoptosis determining protein, neuronal"</i>.  <i>adpn</i> would be a starting point.
    </blockquote>
  </blockquote>
</blockquote></li><br>

<li>Check ZFIN to ensure that the symbol you want to use is
not already in use for a different zebrafish gene or mutant. Also, check the Human Nomenclature
Database and the Mouse Genome Informatics resources to ensure the symbol you want to use is not
already in use for a different gene in human or mouse.

<blockquote>
  <blockquote>
    <blockquote>
      <p><b>Example from above:</b> <i>dihydrolipoamide branched chain transacylase E2; dbt</i>.  You search ZFIN for a gene already named <i>dbt</i> as well as for a mutant whose abbreviation is <i>dbt</i>.  A quick search shows that dbt is not already used for a zebrafish gene or mutant.  We know <i>DBT</i> and <i>Dbt</i> are dihydrolipoamide branched chain transacylase E2 in human and mouse.  The symbol <i>dbt</i> is still OK.</p>

      <p><b>Example from above:</b> <i>calcium channel, voltage-dependent, alpha 1G subunit; cacna1g</i>.  You know <i>cacna1g</i> isn't already used for a zebrafish gene or mutant.  However, when you search the Human Nomenclature Database, you find a <i>CACNA1G</i> gene in humans and when you search the Mouse Genome Informatics site, you find <i>Cacna1g</i>.  If your zebrafish cacna1 gene is not the ortholog of the human <i>CACNA1G</i> and mouse <i>Cacna1g</i>, you need to propose a different symbol.</p>

      <p><b>Example from above:</b> <i>apoptosis determining protein, neuronal; adpn</i>.  You find no zebrafish gene or mutant called <i>adpn</i>.  However, when you search the Human Nomenclature Database and the Mouse Genome Informatics resources, you find that <i>ADPN/Adpn</i> is already used for the <i>adiponutrin</i> gene.  You need to come up with a different symbol for your apoptosis determining protein gene expressed in neurons. </p>
    </blockquote>
  </blockquote>
</blockquote></li><br>

<li>Make sure the symbol you select does not interrupt an existing gene family in zebrafish, human, or mouse by entering a truncated form of your proposed symbol in a gene query form and searching the databases using the "begins with" operator.

<blockquote>
  <blockquote>
    <blockquote>
      <b>Example:</b> You isolate a heart expressed cytokeratin
      and you want to propose <i>hexc</i>. <i>hexc</i> is not used for zebrafish, human, or
      mouse genes. However, you search ZFIN, the Human Nomenclature Database and the Mouse Genome
      Informatics resources for "hex" using the "begins with"
      operator. You find that in both human and mouse, there are <i>HEXA/Hexa</i> and <i>HEXB/Hexb</i>
      genes for <i>hexosaminidase A</i> and <i>hexosaminidase B</i>. <i>hexc</i> for your
      zebrafish gene is not a good symbol because if another human or mouse <i>hexosaminidase</i>
      gene is found, it will be assigned <i>HEXC/Hexc</i>.
    </blockquote>
  </blockquote>
</blockquote></li><br>

<li>Check other resources to ensure that the symbol you
select is not a very commonly used, unapproved symbol. Useful resources to check are:

<blockquote>
  <blockquote>
    <blockquote>
      <p><b>Entrez Gene:</b> <a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?db=gene">http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?db=gene</a></p>
	  <p><b>HUGO Gene Nomenclature Committee:</b>  <a href="http://www.genenames.org/">http://www.genenames.org/</a></p>
      <p><b>The Genome Database:</b> <a href="http://gdbwww.gdb.org/gdb/">http://gdbwww.gdb.org/gdb/</a></p>
      <p><b>OMIM:</b> <a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?db=OMIM">http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?db=OMIM</a></p>
      <p><b>PubMed:</b> <a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?db=PubMed">http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?db=PubMed</a> </p>
    </blockquote>
  </blockquote>
</blockquote></li><br>

<li><p>When you have found a suitable symbol, send it to the
zebrafish nomenclature committee for confirmation and final approval (<a href="mailto:nomenclature@zfin.org">nomenclature@zfin.org</a>). We will then assign and reserve this symbol for your novel gene.</p>

<p>Although we typically have a quick turn around time, please
don't wait until the last minute to request approval of a gene name and symbol. Often a
seemingly simple request is more complicated than anticipated and may require
correspondence with other nomenclature committees or independent analyses on our part.</p>

<p>To help us provide quick turn around time for your requests,
the following information is greatly appreciated and will be held in strict confidence:

<ul>
  <li>Provide a sequence accession number if the sequence is public in an on-line sequence database.  It is very helpful if you include the amino acid sequence, nucleotide sequence, or both with your request if they are not publicly available. </li>
  <li>If you have mapped the gene on a mapping panel, provide the map position and on
    which mapping panel the gene is mapped. Preferably indicate its location with respect to
    the two nearest flanking "z" markers (MGH SSLPs - <a href="http://zebrafish.mgh.harvard.edu/">http://zebrafish.mgh.harvard.edu/</a>).  If you have placed the gene on the Sanger Institute's Ensembl contig (<a href="http://www.ensembl.org/Danio_rerio/">http://www.ensembl.org/Danio_rerio/</a>), please provide the coordinates. </li>
  <li>If you believe the zebrafish gene is the ortholog of a gene
    in another species, provide the evidence (amino acid sequence comparison, nucleotide
    sequence comparison, conserved synteny, probe cross hybridization, etc.).</li>
  <li>If possible, please provide a phylogenetic tree or description of the relationships between your zebrafish gene and orthologous or related genes in other species. </li>
  <li>Include any other information or commentary you view as
    pertinent in assigning proper gene nomenclature.</li>
  </ul>
</li>
</ol>

<p>If you need help with any of this analysis, contact us (<a href="mailto:nomenclature@zfin.org">nomenclature@zfin.org</a>). </p>



<hr>


<script language="JavaScript" src="/javascript/footer.js"></script>
