<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html>
	<head>
		<title>
			ZFIN Zebrafish Nomenclature
		</title>

<style type="text/css">
<!--
h3 {
   background-color: #cccccc
   }
-->
</style>

	</head>

<script language="JavaScript" src="/header.js"></script>


<H1 align="center">Zebrafish Nomenclature Guidelines</H1>
  
<center>Based on <i>Trends in Genetics</i> Genetic Nomenclature Guide (1998)
</CENTER>
<br>

<p><b><i>(Updated July 25, 2005)</i></b></p>

<table width="100%" cellspacing=20>

<tr><td width="50%" valign="top" style="border: medium ridge #999999; padding: 2%">
<P> <STRONG>NOMENCLATURE INFORMATION</STRONG><BR><BR>
  The current nomenclature guidelines are updates to rules established during 
  a discussion session at a meeting in Ringberg, Germany, in March 1992, and are 
  widely accepted by most zebrafish labs.<br>
   <BR>
 <a name="Approval"></a> <a href="/zf_info/nomen_comm.html">Zebrafish Nomenclature Committee</a><B> (ZNC).</B> For questions, contact <a 
href="mailto:nomenclature@zfin.org">nomenclature@zfin.org</a>
   (also see <a href="#11.">Contributors</a> below).</p><br>

<a name="Approval"></a>

<p style="border: medium ridge #999999; padding: 1%; background-color: #cccccc"><b>Obtaining approval for gene names and symbols.  </b>Investigators should contact the committee for advice about gene symbols at: <a href="mailto:nomenclature@zfin.org">nomenclature@zfin.org<a/>.  Registered ZFIN users can receive automatic approval for new mutant locus names by logging in at: <a href="/">http://zfin.org/</a> and selecting the "Register Locus/Allele Designation" link next to "Mutants / Transgenics".<br></p>

<p><a href ="/zf_info/zfbook/lab_desig.html">Laboratory Designations for Mutant Alleles</a></p>

<p><a href ="/zf_info/nomen_tutorial.html">A Tutorial for Proposing Zebrafish Gene Nomenclature</a></p><br>

<P><b>Links to other nomenclature guidelines. </b><a href="http://www.gene.ucl.ac.uk/cgi-bin/nomenclature/searchgenes.pl">Human</a>, 
  <a href="http://www.informatics.jax.org/mgihome/nomen/">Mouse</a>, <a href="http://flybase.bio.indiana.edu/docs/nomenclature/lk/nomenclature.html/">Fly 
  (Drosophila)</a>, <a href="http://www.yeastgenome.org/gene_guidelines.shtml">Yeast 
  (Saccharomyces)</a>, <a href="http://www.informatics.jax.org/mgihome/nomen/genefamilies/index.shtml">Gene 
  families</a> 
</td>



<td style="border: medium ridge #999999; padding: 2%" valign="top">
<b>CONTENTS:</b>
<ul>
  <li><a href="#1.">1. Gene names and symbols</a><br>
  <ul>
    <li><a href="#1.1.">1.1. Genes (loci) identified by mutation</a><br>
    <li><a href="#1.2.">1.2. Genes identified by cloning</a><br>
    <ul>
	    <li><a href="#Duplicate">Duplicated genes</a><br>
    </ul>
    <li><a href="#1.3.">1.3. Genes identified only by genomic sequencing projects</a><br>
    <li><a href="#1.4.">1.4. Transcriptional variants</a><br>
  </ul>
    <li><a href="#2.">2. Alleles</a><br>
    
  <li><a href="#3.">3. Proteins</a><br>
    
  <li><a href="#4.">4. Genotypes</a><br>
    <ul>
	 <li><a href="#4.1.">4.1. Unlinked loci</a><br>
	 <li><a href="#4.2.">4.2. Linked loci</a><br>
	 <li><a href="#4.3.">4.3. Unmapped loci</a><br>
	 <li><a href="#4.4.">4.4. Poorly resolved loci</a><br>
    </ul>
    <li><a href="#5.">5. Phenotypes</a><br>
    <li><a href="#6.">6. Chromosomes and aberrations</a><br>
  <ul>
    <li><a href="#6.1.">6.1. Deficiencies</a> <br>
    <li><a href="#6.2.">6.2. Translocations</a><br>
	  <ul>
	  <li><a href="#6.2.1.">6.2.1. Reciprocal translocations</a><br>
	  <li><a href="#6.2.2.">6.2.2. Resolved translocations</a><br>
	  </ul>
    <li><a href="#6.3.">6.3. Transgenes</a><br>
  </ul>
    
  <li><a href="#7.">7. Priority in Names</a><br>
    
  <li><a href="#8.">8. Website</a><br>
    
  <li><a href="#9.">9. Genome project</a> <br>
    
  <li><a href="#10.">10. Contributors</a><sup></sup><br>
    
  <li><a href="#11.">11. References</a><br>
  </li>
</ul>
</p>

</td>

</table><br><br>


<hr width=90%><br><br>


<h3 align="left"><b><a name="1.">1. GENE NAMES AND SYMBOLS</a></b></h3>
<blockquote> 
  <p><strong><a name="1.1."></a>1.1. Genes (loci) identified by mutation</strong> 
  </p>
  <blockquote> 
    <p><b>1.1.1.</b> The name of a gene identified by mutation should refer to 
      the ostensible phenotype of the mutant without any inherent interpretation 
      of the phenotype. It should not be too specific. Determining the molecular 
      mechanism by which a phenotype is produced can be a formidable task and 
      should not be inferred by the name of the gene. The name should be a mnemonic 
      for the mutant phenotype. For example, <i>cyclops</i> is an excellent name 
      for the mutation isolated in Eugene that produces a cyclopic phenotype. 
    </p>
    <p><b>1.1.2.</b> The name should be unique, one or two words, and italicized. 
      Longer names tend to be impractical to use. Avoid choosing names identical 
      to those used in other species. </p>
    <p><b>1.1.3.</b> For mutations in different genes that all have similar phenotypes, 
      a single name followed by consecutive numbers should not be used. This nomenclature 
      makes the field less accessible to the outsider. A mutation in a second 
      gene resulting in cyclopia could be called <i>fused eyes</i>, but should 
      not be called <i>cyclops2</i>. </p>
    <p><b>1.1.4.</b> If the mutated gene has not yet been named or described in 
      a publication and is known to be the orthologue of a mammalian gene, the 
      approved mammalian gene name and symbol should be used, following the guidelines 
      described below for naming genes identified by cloning. </p>
  </blockquote>
  <p><b>Specific guidelines:</b> Genes identified by mutation are typically named 
    to reflect the mutant phenotype. Full gene names are lowercase italic, no 
    hyphens. Gene symbols consist of three lowercase letters. The letters should 
    be unique with respect to other named zebrafish mutants and genes and, if 
    possible, gene symbols in mammalian species. The symbol should be derived 
    from the full name. For example, the abbreviated name for <i>cyclops</i> is 
    <i>cyc</i>. Numbers should generally not be used in naming a gene identified 
    only by mutation. </p>
  <blockquote> 
    <p> <b>Examples:</b><br>
      &nbsp;&nbsp;&nbsp;&nbsp;Names - <i>cyclops, brass, no tail</i><br>
      &nbsp;&nbsp;&nbsp;&nbsp;Symbols - <i>cyc, brs, ntl</i> </p>
  </blockquote>
  <p>Locus and allele names should be registered at <a 
href="/">ZFIN</a>.</p>
</blockquote>
<hr width=75%>
<blockquote> 
  <p> <strong><a name="1.2."></a>1.2. Genes identified by cloning</strong><br>
    Zebrafish gene designations should not be preceded by 'Z' or 'Zf'. Gene names 
    should be lower case, italic. Punctuation such as hyphens and full stops (periods) 
    in gene names are discouraged. When mammalian orthologues are known, the same 
    name and abbreviation should be used if possible, except all letters are lower 
    case, italic. When orthologous relationships are unclear, lower case letters 
    may be appended. Different members of a gene family are sequentially numbered. 
    Gene symbols should be lower case italicized letters or lower case letters 
    and an italicized number or letter (no hyphen). Genes identified as homologues 
    to human or mouse genes should use the same symbol, if possible, and thus 
    are not limited to 3 letters. </p>
  <blockquote> 
    <p> <b>Examples:</b><br>
      &nbsp;&nbsp;&nbsp;&nbsp;Names - <i>engrailed 1a, engrailed 2b</i><br>
      &nbsp;&nbsp;&nbsp;&nbsp;Symbols - <i>eng1a, eng2b, msxa, msxb</i> </p>
  </blockquote>
  <p>In cases where the mammalian orthologue of a zebrafish gene has a different 
    name that has been approved by the appropriate nomenclature committee, the 
    approved mammalian name may be appended to the zebrafish name, lower case 
    in parentheses.</p>
  <blockquote> 
    <p><b>Examples:</b><i> syu (shh), swr (bmp2b)</i></p>
  </blockquote>
  <p><b><a name="Duplicate">Duplicated genes:</a></b> Most of the zebrafish genome 
    contains duplicated segments that probably result from an extra genome-wide 
    duplication. This duplication occurred in the ray fin fish lineage after it 
    diverged from the lobe fin lineage (that include avian or mammalian species). 
    For this reason, zebrafish often have two copies of a gene that is present 
    as a single copy in mammals. </p>
  <p>In these cases, symbols for the two zebrafish genes should be the same as 
    the approved symbol of the human orthologue followed by "<i>a</i>" or "<i>b</i>" 
    to indicate that they are duplicate copies. Before these symbols are assigned, 
    it is important to provide evidence by mapping that the two copies reside 
    on duplicated chromosome segments. It is preferable that all copies in one 
    of the duplicate chromosome segments use the same "<i>a</i>" or "<i>b</i>" 
    suffix, although this will not always be possible for historical reasons. 
  </p>
  <p>This terminology should not be used for tandem duplicates or duplicates that 
    resulted prior to the divergence of ray fin and lobe fin fish. </p>
  <blockquote> 
    <p><b>Examples:</b> <i>hoxa13a, hoxa13b</i></p>
  </blockquote>
  <p>Special cases sometimes arise when naming gene duplicates that resulted from 
    the ray-fin fish whole-genome duplication. When there is a unique mammalian 
    orthologue, but addition of the <i>a, b</i> suffixes would conflict with a 
    different mammalian gene symbol, then numerical suffixes <i>.1, .2</i> should 
    be appended to the orthologous mammalian gene symbol instead of <i>a, b</i>.</p>
  <p>When mammalian gene duplications prevent identification of a unique mammalian 
    orthologue, then an alternate gene symbol should be chosen. A possible choice 
    would be an approved gene symbol from a unique non-mammalian orthologue. 
    If necessary to avoid conflicts with existing mammalian gene symbols, the 
    two zebrafish genes should then be distinguished from each other by addition 
    of the terminal<i>.1, .2</i> instead of <i>a, b</i>.</p>
  <blockquote> 
    <p><b>Examples:</b><i> stat5.1, stat5.2</i></p>
  </blockquote>
</blockquote>
<hr width=75%>
<blockquote> 
  <p><b><a name="1.3.">1.3. Genes identified only by genomic sequencing projects</a></b><br>
    Large-scale genome sequencing projects use a variety of prediction methods 
    to identify both open reading frames and genes. Some of these genes are already 
    known, others are new. Novel genes identified by these means that can not 
    be named because they encode known motifs or because they are not members 
    of a gene family are assigned a name comprised of a prefix, a clone name, 
    and an integer. The prefix is used to specify the research institution that 
    identified the gene (e.g., &quot;<i>si</i>&quot; for the Sanger Institute). 
    A colon separates the prefix from the clone name. The clone name is the name 
    of the clone from which the gene was first identified. If part of a gene is 
    found in more than one clone, the name of the first clone in which the 5' 
    portion of the gene is found takes precedence. A full stop (period) separates 
    the clone name from the integer. Integers are assigned serially to genes found 
    in the clone and do not indicate the order of genes.</p>
  <blockquote> 
    <p><b>Examples:</b> <i>si:bz3c13.1, si:bz3c13.2, si:bz3c13.3 </i></p>
  </blockquote>
  <p> When genes initially identified by genomic sequencing projects are better 
    diagnosed, the gene nomenclature will be changed to maintain consistency within 
    gene families or to indicate orthology relationships with genes in other species. 
  </p>
</blockquote>
<hr width=75%>
<blockquote> 
  <p><b><a name="1.4.">1.4. Transcript variants</a></b><br>
    Transcript variants that originate from the same gene are not normally given 
    different gene symbols and names. However, variants from a single locus can 
    be distinguished by adding to the end of the full name a comma, "<i>transcript 
    variant</i>", and a serial number; and by adding to the end of the symbol 
    an underscore, "<i>tv</i>", and a serial number.</p>
  <blockquote> <b>Examples:</b><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Names -<br>
    &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<i>myosin VIa, transcript 
    variant 1</i><br>
    &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<i>myosin VIa, transcript 
    variant 2</i><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Symbols -<br>
    &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<i>myo6a_tv1</i><br>
    &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<i>myo6a_tv2</i> </blockquote>
</blockquote>
<hr width=90%>
<br>
<br>
<a name="2."> 
<h3>2. ALLELES</h3>
</a> 
<p>Wild-type and mutant alleles are indicated by superscript plus and mutant allele 
  designation, respectively. Specific mutant alleles are designated by a superscript 
  following the gene symbol. Alleles are given a laboratory-specific designation, 
  e.g. <i>b</i> is Eugene; <i>m</i> is MGH, Boston; <i>t</i> is Tuebingen, Germany. 
  (The full list can be found on <a href="zfbook/lab_desig.html">ZFIN</a>.) Other 
  letters should not immediately follow the laboratory designation but may be 
  appended to the end of the allele name. Dominant alleles have a <i>d</i> in 
  the first position of the superscript to distinguish them from recessive alleles.</p>
<blockquote> 
  <p> <b>Examples:</b><br>
    &nbsp;&nbsp;&nbsp;&nbsp;<i>cyc<sup>+</sup>, brs<sup>+</sup></i><br>
    &nbsp;&nbsp;&nbsp;&nbsp;<i>lof<sup>dt2</sup>, cyc<sup>b16</sup>, cyc<sup>m101</sup>, 
    cyc<sup>t219</sup></i> </p>
</blockquote>
<p>To distinguish heterozygotes and homozygotes and multiple mutations on a single 
  chromosome, each allele should be separated by a slash &quot;/&quot;.</p>
<blockquote> 
  <p> <b>Examples:</b><br>
    &nbsp;&nbsp;&nbsp;&nbsp;<i>rse<sup>b140</sup>/ rse<sup>+</sup></i> (heterozygote, 
    can be abbreviated <i>rse<sup>b140/+</sup></i>)<br>
    &nbsp;&nbsp;&nbsp;&nbsp;<i>rse<sup>b140</sup>/ rse<sup>b140</sup></i> (homozygote, 
    can be abbreviated <i>rse<sup>b140/b140</sup></i> or <i>rse<sup>b140</sup></i>)<br>
    &nbsp;&nbsp;&nbsp;&nbsp;<i>rse<sup>b140</sup> dla<sup>+</sup>/ rse<sup>+</sup> 
    dla<sup>dx2</sup></i> (transheterozygote for <i>rse</i> and <i>dla</i> alleles 
    on same chromosome)</p>
</blockquote>
<hr width=90%>
<br>
<br>
<a name="3."></a> 
<h3>3. PROTEINS</h3></a>
<p>The protein symbol is the same as the gene symbol, but non-italic and the first 
  letter is uppercase. </p>
<blockquote> 
  <p> <b>Examples:</b> Cyc, Brs, Eng1a, Eng2b, Ntl </p>
</blockquote>
<p>In cases where the mammalian orthologue of a zebrafish protein has a different 
  name that has been approved by the appropriate nomenclature committee, the approved 
  mammalian name may be appended to the zebrafish name, non-italic, first letter 
  uppercase<sup>2</sup>.</p>
<blockquote> 
  <p> <strong>Examples</strong>: Syu (Shh), Swr (Bmp2b)</p>
</blockquote>
<p>Note the differences between zebrafish and mammalian naming conventions:</p>
<blockquote> 
  <p><strong>species / gene / protein</strong><br>
    zebrafish / <i>shh</i>/ Shh<br>
    human / <i>SHH</i> / SHH<br>
    mouse / <i>Shh</i> / SHH<br>
</blockquote>
<hr width=90%>
<br>
<br>
<a name="4."> 
<h3>4. GENOTYPES</h3>
</a> 
<blockquote> <a name="4.1."></a><b>4.1. Unlinked loci</b> 
  <blockquote> 
    <p><a name="4.1.1."></a><b>4.1.1. Homozygous</b> genotypes, the genotype at 
      each locus is listed in order according to Linkage Group number, from 1 
      to 25, with a semicolon to separate loci on different chromosomes. </p>
    <blockquote> 
      <p> <b>Examples:</b> <i>rse<sup>b140</sup></i>; <i>gol<sup>b16</sup></i></p>
    </blockquote>
    <p> <a name="4.1.2."></a><b>4.1.2. Heterozygous</b> genotypes, loci on homologous 
      chromosomes are separated by a slash. </p>
    <blockquote> 
      <p> <b>Examples:</b> <i>rse<sup>b140</sup>/rse<sup>+</sup></i>; <i>gol<sup>b16</sup>/gol<sup>m592</sup></i></p>
    </blockquote>
    <p>Another case is a new mutation induced in a line expressing a transgene 
      as a marker, for example, using a GFP line to find interesting mutations 
      that alter GFP expression. </p>
    <blockquote> 
      <p> <b>Example:</b> <i>Tg(NBT:MAPT-GFP)<sup>zc1</sup></i>; <i>abc<sup>b000</sup></i> 
      </p>
    </blockquote>
  </blockquote>
  <p> <a name="4.2."></a><b>4.2. Linked loci</b> - the haplotype on each chromosome 
    is written as a sequence, with a space separating syntenic loci, and loci 
    are placed in the order they appear on the Linkage Group, top to bottom. Homologous 
    chromosomes are separated by a slash, and non-homologous chromosomes are separated 
    by semicolons. </p>
  <blockquote> 
    <p> <b>Examples:</b> <i>rse<sup>b140</sup> leo<sup>t1</sup></i>; <i>gol<sup>b16</sup></i> 
    </p>
  </blockquote>
  <p><a name="4.3."></a><b>4.3. Unmapped loci</b> - genotypes of unmapped loci 
    are listed alphabetically within braces following genotypes of mapped loci 
    on different chromosomes. </p>
  <blockquote> 
    <p><b>Examples:</b> <i>rse<sup>b140</sup></i>; {<i>edi<sup>tz253</sup></i>;<i> 
      esr<sup>tj236</sup></i>} (<i>edi</i> and <i>esr</i> are unmapped, all three 
      loci are on different chromosomes)</p>
  </blockquote>
  <p><a name="4.4."></a><b>4.4. Poorly resolved loci on same chromosome</b> - 
    genotypes of loci on the same chromosome at locations that are undetermined 
    or poorly resolved are listed alphabetically within braces.</p>
  <blockquote> 
    <p><b>Examples:</b> 
    <blockquote> {<i>abc<sup>b000</sup> def<sup>m000</sup></i>} (poorly resolved 
      loci on same chromosome)<i><br>
      rse<sup>b140</sup></i> {<i>abc<sup>b000</sup> def<sup>m000</sup></i>} <i>leo<sup>t1</sup></i> 
      (poorly resolved loci in a known interval between mapped loci, all on same 
      chromosome)</blockquote>
  </blockquote>
</blockquote>
<hr width=90%>
<br>
<br>
<a name="5."> 
<h3>5. PHENOTYPES</h3>
</a> 
<p>Phenotypes are designated by the name or symbol of the mutation in italics. 
</p>
<blockquote> 
  <p> <b>Examples:</b> <i>cyclops, cyc</i> </p>
</blockquote>
<hr width=90%>
<br>
<br>
<a name="6."> 
<h3>6. CHROMOSOMES AND ABERRATIONS</h3>
</a> 
<p>Chromosomes are not yet cytologically defined. Linkage groups are designated 
  by non-italic numerals, each linkage group corresponds to one of the 25 haploid 
  chromosomes. Chromosome differences have not been observed between males and 
  females.</p>
<blockquote> <b>Examples:</b><br>
  LG1 to LG25 
  <p> <b>Chromosome rearrangements:</b> Common prefixes include:<br>
    <i>Cr</i>, complex rearrangement<br>
    <i>Df</i>, deficiency<br>
    <i>Dp</i>, duplication<br>
    <i>In</i>, inversion<br>
    <i>Is</i>, insertion<br>
    <i>T</i>, translocation<br>
    <i>Tg</i>, transgene 
  <p> 
  <hr width=75%>
</blockquote>
<blockquote> 
  <p><a name="6.1."><b>6.1. Deficiencies:</b></a> The general format for naming 
    a deficiency is: </p>
  <p><i>Df(LG##)xxx<sup>allele</sup></i></p>
  <p><i>Df</i> indicated deficiency. The term <i>xxx</i> should describe the salient 
    features of the deficiency, as determined by the investigator. In cases where 
    the deficiency removes sequences from a named gene, it should contain the 
    standard symbol for that <a href="#1.">gene</a>. The <a 
href="#2.">allele name</a> should follow standard nomenclature conventions (laboratory 
    designation followed by allele name).</p>
  <p> The linkage group where the deficiency maps should be specified by its number 
    (##) using two digits (i.e., 03 for LG 3) so that computers will order them 
    properly. </p>
  <blockquote> 
    <p><b> Example: </b><i>Df(LG12)dlx3<sup>b380</sup></i> </p>
  </blockquote>
  <hr width=75%>
  <p><a name="6.2."></a><b>6.2. Translocations: </b>The general format for naming 
    translocations depends upon the type of translocation:</p>
  <blockquote> 
    <p><a name="6.2.1."></a><b>6.2.1. Reciprocal translocations:</b> </p>
    <p>These rearrangements have two separate chromosomal elements, and each element 
      has a distinct name:</p>
    <p><i>T(LG##;LG##)xxx<sup>allele</sup>,##U.##L </i>and <i>T(LG##;LG##)xxx<sup>allele</sup>,##U.##L</i> 
    </p>
    <p><i>T</i> indicates translocation. The elements in the parentheses are the 
      linkage groups involved, the lower numbered linkage group is listed first, 
      and the linkage groups are separated by a semicolon. The linkage groups 
      should be specified by their numbers (##) using two digits (i.e., 03 for 
      LG 3) so that computers will order them properly. </p>
    <p>The term <i>xxx</i> should describe some salient feature of the translocation, 
      as determined by the investigator. In cases where the translocation moves 
      a named gene primarily studied by the investigator, <i>xxx</i> would usually 
      be the standard <a href="#1.">symbol</a> for that gene. Alternatively, <i>xxx</i> 
      could just be an experimental series number. </p>
    <p>The <a href="#2.">allele name</a> should follow standard nomenclature conventions 
      (laboratory designation followed by allele name, superscripted). After the 
      allele name comes a comma, and then a phrase that indicates the new order 
      of the chromosome, starting from the top of the chromosome as displayed 
      by convention. The first number (##) is the LG number, followed by upper 
      case <i>U</i> to indicate the upper arm of a chromosome or by upper case 
      <i>L</i> to indicate the lower arm of a chromosome. The location of the 
      centromere is indicated by a period. No spaces. </p>
    <blockquote> 
      <p><b>Example:</b><br>
        &nbsp;&nbsp;&nbsp;&nbsp;<i>T(LG02;LG12)cyc<sup>b213</sup>,02U.12L02L</i> 
        and <i>T(LG02;LG12)cyc<sup>b213</sup>,12U.12L02L</i></p>
      <p>&nbsp;&nbsp;&nbsp;&nbsp;<img src="translocation.jpeg"></p>
    </blockquote>
    <p>This example illustrates a reciprocal translocation where a portion of 
      the lower arm of LG12 was translocated interstitially into the proximal 
      lower arm of LG2 and a portion of the lower arm of LG2 was translocated 
      to the distal lower arm of LG12.</p>
    <p><a name="6.2.2."></a><b>6.2.2. Resolved translocations:</b></p>
    <p>In some cases, the two elements of the translocation can separate, a mutant 
      line would have just one of the elements. This results in the animal being 
      monosomic for some chromosome regions and trisomic for others. In these 
      cases, the mutant line would be designated with just one of the elements 
      rather than two as in the reciprocal designation above. The allele name 
      would remain the same to indicate their common origin and common breakpoint. 
    </p>
    <blockquote> 
      <p><b>Example:</b> <i>T(LG02;LG12)cyc<sup>b213</sup>,02U.12L02L</i></p>
    </blockquote>
  </blockquote>
  <hr width=75%>
  <p><a name="6.3."><b>6.3. Transgenes</b></a>: The general format for naming 
    a transgene is: </p>
  <p><i>Tg(xxx)<sup>allele</sup></i></p>
  <p><i>Tg</i> indicates transgene. The term <i>xxx</i> should describe the most 
    salient features of the transgene, as determined by the investigator. Brevity 
    and clarity in the transgene name are favored, in general, over exhaustive 
    detail. In cases where the insertion utilizes sequences from a named gene, 
    it should contain the standard symbol for that gene. The allele name should 
    follow standard nomenclature conventions (laboratory designation followed 
    by allele name).</p>
  <p>If the insertion is a compound construct consisting of regulatory sequence linked 
    to coding region, the insertion may be designated using the symbol of the 
    regulatory region, separated by a colon from the symbol for the coding region. 
    Regulatory sequence could be derived from either an enhancer or promoter, 
    and may be denoted by the symbol of the regulated gene.</p>
  <blockquote> 
    <p><b>Example:</b> <i>Tg(promoter:gene)<sup>allele</sup></i></p>
  </blockquote>
  <p>In cases where the map location is known and where this information is useful 
    (i.e., when the location of the transgene affects its expression or when the 
    transgene is used as a map marker) the linkage group number (##) should also 
    be included: <i>Tg(LG##,promoter:gene)<sup>allele</sup></i></p>
  <blockquote> 
    <p><b>Example:</b> <i>Tg(LG03,ERE:nic1)<sup>b512</sup></i></p>
  </blockquote>
  <p>Each transgenic name should have no more than one colon. Regardless of how 
    they are ordered in the construct, all regulatory sequences should be listed 
    to the left of the colon, and all coding sequences should be listed to the 
    right of the colon. Multiple regulatory or coding sequences should be separated 
    by hyphens.</p>
  <blockquote> 
    <p><b>Example:</b> <i>Tg(-3.5hhex-ERE:sptb-GFP)<sup>bz1</sup></i></p>
  </blockquote>
  <p> Here, a fusion protein of <i>spectrin beta (sptb)</i> and GFP is driven 
    by a combination of an estrogen response element and an upstream enhancer 
    from the <i>hhex</i> gene.</p>
</blockquote>
<hr width=90%>
<br>
<br>
<a name="7."> 
<h3>7. PRIORITY IN NAMES</h3>
</a> 
<blockquote> 
  <p><b>7.1.</b> When mutations in the same gene are found independently in more 
    than one laboratory and given different names, then the name that appears 
    first in the literature will be given priority. Sometimes different mutant 
    alleles of the same gene can have very different phenotypes and it may take 
    a while before it is shown that the mutations are in the same gene. The first 
    published name would get priority and all other alleles would adopt this designation.</p>
  <p><b>7.2.</b> When a mutation is found in a previously cloned zebrafish gene 
    (which has been published), then the mutant will take the cloned gene's name. 
    For example, a mutation was induced in the <i>msxb</i> gene. The mutation 
    in the gene is also be referred to as <i>msxb</i> (in this case, <i>Df(LG01)msxb,</i> 
    because the mutation is a deficiency). </p>
  <p><b>7.3.</b> If both the cloned gene and the mutation are already known by 
    different names and at some later point they are found to be the same gene, 
    then the name of the mutation usually takes priority. The unique name of the 
    mutation may be more relevant to the function of the gene than the cloned 
    gene's name, and this unique mutant name may be easier to remember or use 
    than the gene name. For example, an author may find it more convenient to 
    refer to <i>cyclops (cyc)</i>, rather than <i>nodal-related 2 (ndr2)</i>. 
    Exceptions would include cases where orthologous relationships with genes 
    in other species are well established and these orthologues have a commonly 
    used name. For example, an author may find it preferable to refer to <i>sonic 
    hedgehog (shh)</i>, rather than the <i>sonic you (syu)</i> gene.</p>
</blockquote>
<hr width=90%>
<br>
<br>
<a name="8."> 
<h3>8. WEBSITE</h3>
</a> <b>The Zebrafish Information Network</b> (<a href="/">ZFIN</a>) provides 
access to information about zebrafish genetics, genomics, and development as well 
as links to researchers, laboratories, and publications.</p> 
<hr width=90%>
<br>
<br>
<a name="9."> 
<h3>9. GENOME PROJECT</h3>
</a> The <a href="/zf_info/catch/catch.html">genome project</a> began in 1994, 
and by 1996 the genetic map was closed. NIH funded major <a href="http://www.nih.gov/science/models/zebrafish/">programs</a> 
to develop a doubled haploid meiotic mapping panel, deficiency strains and expressed 
sequence tags (ESTs), The ESTs and anonymous markers have been mapped on two radiation-hybrid 
panels. The <a href="http://www.sanger.ac.uk/">Sanger Institute</a> began <a href="http://www.sanger.ac.uk/Projects/D_rerio/">full 
genome squencing</a> in 2001. A physical map is being constructed from the <a href="/zf_info/zfprbs.html">BAC 
libraries</a> used for sequencing. Genomic information is updated regularly on 
<a href="/ZFIN">ZFIN</a>. 
<hr width=90%>
<br>
<br>
<a name="10."> 
<h3>10. CONTRIBUTORS</h3>
</a> Marc Ekker <a href="mailto:marc.ekker@science.uottawa.ca">(marc.ekker@science.uottawa.ca</a>), 
Center for Advanced Research in Environmental Genomics, University of Ottawa, 
Ontario, Canada<br>
Mary Mullins (<a href="mailto:mullins@mail.med.upenn.edu">mullins@mail.med.upenn.edu</a>), 
Department of Cell and Developmental Biology, University of Pennsylvania, USA<br>
John Postlethwait (<a href="mailto:jpostle@oregon.uoregon.edu">jpostle@oregon.uoregon.edu</a>), 
Institute of Neuroscience, University of Oregon, USA <br>
Sridhar Ramachandran (<a 
href="mailto:sramacha@uoneuro.uoregon.edu">sramacha@uoneuro.uoregon.edu</a>), 
Zebrafish Information Network, University of Oregon, USA<br>
Monte Westerfield (<a href="mailto:monte@uoneuro.uoregon.edu">monte@uoneuro.uoregon.edu</a>), 
Institute of Neuroscience, University of Oregon, USA 
<hr width=90%>
<br>
<br>
<a name="11."> 
<h3>11. REFERENCES</h3>
</a> 1. <a href="/zf_info/monitor/vol2.1/vol2.1.html">The Zebrafish Science Monitor 
(1992) Sept. 21</a>.<br>
2. Mullins, M. (1995) Genetic methods: conventions for naming zebrafish genes 
in The Zebrafish Book (3rd edition, Westerfield, M., ed.), pp 7.1-7.4, University 
of Oregon Press.<br>
3. Genetic Nomenclature Guide, Trends in Genetics (1998). 
<hr width=90%>
<br>
<br>

</BODY>
<script language="JavaScript" src="/footer.js"></script>
