<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html>
	<head>
		<title>
			ZFIN Zebrafish Nomenclature
		</title>
	</head>

<script language="JavaScript" src="http://<!--|DOMAIN_NAME|-->/header.js"></script>

<CENTER>
<H1>Zebrafish Nomenclature Guidelines</H1>
   <B>Updated November 12, 2001</B><BR>
   Based on <i>Trends in Genetics</i> Genetic Nomenclature Guide (1998)
</CENTER>
<P> <STRONG>NOMENCLATURE INFORMATION</STRONG><BR>
  The current nomenclature guidelines are updates to rules established during 
  a discussion session at a meeting in Ringberg, Germany, in March 1992, and are 
  widely accepted by most zebrafish labs.<br>
   <BR>
   <B>Zebrafish Nomenclature Committee.</B> <a 
href="/zf_info/nomen_comm.html">Information</a>
   about the committee. For questions, contact <a 
href="mailto:nomenclature@zfin.org">nomenclature@zfin.org</a>
   (also see <a href="#11.">Contributors</a> below).
<P><a name="Approval"></a><b>Obtaining approval for gene names and symbols. </b>Investigators 
  should contact the committee for advice about gene symbols at: <a 
href="mailto:nomenclature@zfin.org">nomenclature@zfin.org</a> and registered ZFIN 
  users can receive automatic approval for new mutant locus names at: <a href="/ZFIN">http://zfin.org/ZFIN</a> 
<P><b>Links to other nomenclature guidelines. </b><a href="http://www.gene.ucl.ac.uk/cgi-bin/nomenclature/searchgenes.pl">Human</a>, 
  <a href="http://www.informatics.jax.org/mgihome/nomen/">Mouse</a>, <a href="http://flybase.bio.indiana.edu/">Fly 
  (Drosophila)</a>, <a href="http://genome-www.stanford.edu/Saccharomyces/">Yeast 
  (Saccharomyces</a>, <a href="http://www.informatics.jax.org/mgihome/nomen/genefamilies/index.shtml">Gene 
  families</a> 
<HR WIDTH=90%>
<P><b>CONTENTS:</b>
<ul>
  <li><a href="#1.">1. Gene names and symbols</a><br>
  <ul>
    <li><a href="#1.1.">1.1. Genes (loci) identified by mutation</a><br>
    <li><a href="#1.2.">1.2. Genes identified by cloning</a><br>
    <ul>
	    <li><a href="#Duplicate">Duplicated genes</a><br>
    </ul>
    <li><a href="#1.3.">1.3. Genes identified only by genomic sequencing projects</a><br>
  </ul>
    <li><a href="#2.">2. Alleles</a><br>
    
  <li><a href="#3.">3. Proteins</a><br>
    
  <li><a href="#4.">4. Genotypes</a><br>
    <ul>
	 <li><a href="#4.1.">4.1. Unlinked loci</a><br>
	 <li><a href="#4.2.">4.2. Linked loci</a><br>
	 <li><a href="#4.3.">4.3. Unmapped loci</a><br>
	 <li><a href="#4.4.">4.4. Poorly resolved loci</a><br>
    </ul>
    <li><a href="#5.">5. Phenotypes</a><br>
    <li><a href="#6.">6. Chromosomes and aberrations</a><br>
  <ul>
    <li><a href="#6.1.">6.1. Deficiencies</a> <br>
    <li><a href="#6.2.">6.2. Translocations</a><br>
	  <ul>
	  <li><a href="#6.2.1.">6.2.1. Reciprocal translocations</a><br>
	  <li><a href="#6.2.2.">6.2.2. Resolved translocations</a><br>
	  </ul>
    <li><a href="#6.3.">6.3. Transgenes</a><br>
  </ul>
    
  <li><a href="#7.">7. Strains</a><br>
    
  <li><a href="#8.">8. Priority in Names</a><br>
    
  <li><a href="#9.">9. Website</a><br>
    
  <li><a href="#10.">10. Genome project</a> <br>
    
  <li><a href="#11.">11. Contributors</a><sup></sup><br>
    
  <li><a href="#12.">12. References</a><br>
  </li>
</ul>
<hr width=90%>
<h3 align="left"><b><a name="1.">1. GENE NAMES AND SYMBOLS</a></b></h3>
<blockquote>
  <p><STRONG><a name="1.1."></a>1.1. GENES (LOCI) identified by mutation</STRONG><br>
    <BR>
  </p>
  <blockquote>
    <p><b>1.1.1.</b> The name of a gene identified by mutation should refer to 
      the ostensible phenotype of the mutant without any inherent interpretation 
      of the phenotype. It should not be too specific. Determining the molecular 
      mechanism by which a phenotype is produced can be a formidable task and 
      should not be inferred by the name of the gene. The name should be a mnemonic 
      for the mutant phenotype. For example, <i>cyclops </i>is an excellent name 
      for the mutation isolated in Eugene that produces a cyclopic phenotype. 
    </p>
    <p><b>1.1.2.</b> The name should be unique, one or two words, and italicized. 
      Longer names tend to be impractical to use. Avoid choosing names identical 
      to those used in other species. </p>
    <p><b>1.1.3.</b> For mutations in different genes that all have similar phenotypes, 
      a single name followed by consecutive numbers should not be used. This nomenclature 
      makes the field less accessible to the outsider. A mutation in a second 
      gene resulting in cyclopia could be called <i>fused eyes,</i> but should 
      not be called <i>cyclops2.</i> </p>
    <p><b>1.1.4.</b> If the mutated gene has not yet been named or described in 
      a publication and is known to be the orthologue of a mammalian gene, the 
      approved mammalian gene name and symbol should be used, following the guidelines 
      described below for naming genes identified by cloning. </p>
  </blockquote>
  <p><b>Specific guidelines:</b>. Genes identified by mutation are typically named 
    to reflect the mutant phenotype. Full gene names are lowercase italic, no 
    hyphens. Gene symbols consist of three lowercase letters. The letters should 
    be unique with respect to other named zebrafish mutants and genes and, if 
    possible, gene symbols in mammalian species. The symbol should be derived 
    from the full name. For example, the abbreviated name for <i>cyclops </i>is 
    <i>cyc.</i> Numbers should generally not be used in naming a gene identified 
    only by mutation. </p>
  <blockquote>
    <p> <B>Examples:</B><BR>
      &nbsp;&nbsp;&nbsp;&nbsp;Names - <I>cyclops, brass, no tail</I><BR>
      &nbsp;&nbsp;&nbsp;&nbsp;Symbols - <I>cyc, brs, ntl</I> </p>
  </blockquote>
  <p>Locus and allele names should be registered at <a 
href="/">ZFIN</a> </p>
  <HR WIDTH=90%>
  <p> <STRONG><a name="1.2."></a>1.2. GENES identified by cloning</STRONG><BR>
    Zebrafish gene designations should not be preceded by 'Z' or 'Zf'. Gene names 
    should be lower case, italic. Punctuation such as hyphens and full stops (periods) 
    in gene names are discouraged. When mammalian orthologues are known, the same 
    name and abbreviation should be used if possible, except all letters are lower 
    case, italic. When orthologous relationships are unclear, lower case letters 
    may be appended. Different members of a gene family are sequentially numbered. 
    Gene symbols should be lower case italicized letters or lower case letters 
    and an italicized number or letter (no hyphen). Genes identified as homologues 
    to human or mouse genes should use the same symbol, if possible, and thus 
    are not limited to 3 letters. </p>
  <blockquote>
    <p> <B>Examples:</B><BR>
      &nbsp;&nbsp;&nbsp;&nbsp;Names - <I>engrailed1a, engrailed2</I>b<BR>
      &nbsp;&nbsp;&nbsp;&nbsp;Symbols - <I>eng1a, eng2b, msxa, msxb</I> </p>
  </blockquote>
  <p>In cases where the mammalian orthologue of a zebrafish 
    gene has a different name that has been approved 
    by the appropriate nomenclature committee, the approved 
    mammalian name may be appended to the zebrafish 
    name, lower case in parentheses.</p>
  <blockquote>
    <p><b>Examples:</b><i> syu (shh); swr (bmp2b)</i></p>
  </blockquote>
  <p><b><a name="Duplicate">Duplicated genes:</a></b> Most of the zebrafish genome 
    contains duplicated segments that probably result from an extra genome-wide 
    duplication. This duplication occurred in the ray fin fish lineage after it 
    diverged from the lobe fin lineage (that include avian or mammalian species). 
    For this reason, zebrafish often have two copies of a gene that is present 
    as a single copy in mammals. </p>
  <p>In these cases, symbols for the two zebrafish genes should be the same as 
    the approved symbol of the human orthologue followed by "a" or "b" to indicate 
    that they are duplicate copies. Before these symbols are assigned, it is important 
    to provide evidence by mapping that the two copies reside on duplicated chromosome 
    segments. It is preferable that all copies in one of the duplicate chromosome 
    segments use the same "a" or "b" suffix, although this will not always be 
    possible for historical reasons. </p>
  <p>This terminology should not be used for tandem duplicates or duplicates that 
    resulted prior to the divergence of ray fin and lobe fin fish. </p>
  <blockquote>
    <p><b>Examples:</b><br>
      &nbsp;&nbsp;&nbsp;&nbsp;Symbols - <i>hoxa13a, hoxa13b</i> </p>
  </blockquote>
  <hr>
  <p><b><a name="1.3.">1.3. GENES identified only by genomic sequencing projects</a></b><br>
    Large-scale genome sequencing projects use a variety of prediction methods 
    to identify both open reading frames and genes. Some of these genes are already 
    known, others are new. Novel genes identified by these means that can not 
    be named because they encode known motifs or because they are not members 
    of a gene family are assigned a name comprised of a prefix, a clone name, 
    and an integer. The prefix is used to specify the research institution that 
    identified the gene ('SC' for Sanger Centre). A colon separates the prefix 
    from the clone name. The clone name is the name of the clone from which the 
    gene was first identified. If part of a gene is found in more than one clone, 
    the name of the first clone in which the 5' portion of the gene is found takes 
    precedence. A full stop (period) separates the clone name from the integer. 
    Integers are assigned serially to genes found in the clone and do not indicate 
    the order of genes. </p>
  <blockquote>
    <p><b>Examples:</b><br>
      &nbsp;&nbsp;&nbsp;&nbsp;<i>SC:bz3c13.1, SC:bz3c13.2, SC:bz3c13.3 </i></p>
  </blockquote>
  <p> When genes initially identified by genomic sequencing projects are better 
    diagnosed, the gene nomenclature will be changed to maintain consistency within 
    gene families or to indicate orthology relationships with genes in other species. 
  </p>
</blockquote>
<HR WIDTH=90%>
<P align="left"> <a name="2."><STRONG>2. ALLELES</STRONG></a><br>
  <BR>
  Wild-type and mutant alleles are indicated by superscript plus and mutant allele 
  designation, respectively. Specific mutant alleles are designated by a superscript 
  following the gene symbol. Alleles are given a laboratory-specific designation, 
  e.g. <I>b </I>is Eugene; <I>m </I>is MGH, Boston; <I>t </I>is Tuebingen, Germany. 
  (The full list can be found on <A HREF="zfbook/lab_desig.html">ZFIN</A>). Other 
  letters should not immediately follow the laboratory designation but may be 
  appended to the end of the allele name. Dominant alleles have a <I>d</I> in 
  the first position of the superscript to distinguish them from recessive alleles. 
<blockquote>
   <p> <B>Examples:</B><BR>
     &nbsp;&nbsp;&nbsp;&nbsp;<I>cyc<SUP>+</SUP> brs<SUP>+</SUP></I><BR>
     &nbsp;&nbsp;&nbsp;&nbsp;<I>lof<SUP>dt2</SUP> cyc<SUP>b16</SUP> 
cyc<SUP>m101</SUP>
     cyc<SUP>t219</SUP></I>
   </p>
</blockquote>
 
<p>To distinguish heterozygotes and homozygotes and multiple mutations on a single 
  chromosome, each allele should be separated by a slash &quot;/&quot;. 
<blockquote>
   <p> <b>Examples:</b><br>
    &nbsp;&nbsp;&nbsp;&nbsp;<I>msxb<SUP>b220</SUP>/ msxb<SUP>+</SUP></I> (heterozygote, 
    can be abbreviated <I>msx<SUP>b220/+</SUP></I>)<br>
     &nbsp;&nbsp;&nbsp;&nbsp;<I>msxb<SUP>b220</SUP>/ 
msxb<SUP>b220</SUP></I> (homozygote, can be abbreviated 
<I>msx<SUP>b220/b220</SUP></I> or <I>msx<SUP>b220</SUP></I>)<br>
     &nbsp;&nbsp;&nbsp;&nbsp;<I>msxb<SUP>b220</SUP> dla<SUP>+</SUP>/ 
msxb<SUP>+</SUP>
     dla<SUP>dx2</SUP></I> (transheterozygote for <I>msxb</I> and 
<I>dla</I> alleles
     on same chromosome)</p>
</blockquote>
<HR WIDTH=90%>
<P align="left"> <STRONG><a name="3."></a>3. PROTEINS<br>
  </STRONG><BR>
  The protein symbol is the same as the gene symbol, but non-italic and the first 
  letter is uppercase. 
<blockquote> 
  <p> <B>Examples:</B> Cyc Brs Eng1a Eng2b Ntl </p>
</blockquote>
<p>In cases where the mammalian orthologue of a zebrafish protein has a different 
  name that has been approved by the appropriate nomenclature committee, the approved 
  mammalian name may be appended to the zebrafish name, non-italic, first letter 
  uppercase².<br>
<blockquote> 
  <p> <strong>Examples</strong>: Syu (Shh); Swr (Bmp2b)</p>
</blockquote>
<HR WIDTH=90%>
  <p> <STRONG><a name="4."></a>4. GENOTYPES</STRONG><br>
    <BR>
  
<blockquote> <a name="4.1."></a><B>4.1. Unlinked loci</B> 
  <blockquote> 
    <p><a name="4.1.1."></a><b>4.1.1. Homozygous</b> genotypes, the genotype at each locus is listed 
      in order according to Linkage Group number, from 1 to 25, with a semicolon 
      to separate loci on different chromosomes. </p>
    <blockquote> 
      <p> <B>Examples:</B> <I>rse<SUP>b140</SUP></I>; <I>gol<SUP>b16</SUP></I></p>
    </blockquote>
    <p> <a name="4.1.2."></a><b>4.1.2. Heterozygous</b> genotypes, loci on homologous chromosomes are 
      separated by a slash. </p>
    <blockquote> 
      <p> <B>Examples:</B> <I>rse<SUP>b140</SUP>/rse<SUP>+</SUP></I>;<I> gol<SUP>b16</SUP>/gol<SUP>m592</SUP></I></p>
    </blockquote>
    <p>Another case is a new mutation induced in a line expressing a transgene 
      as a marker, for example, using a GFP line to find interesting mutations 
      that alter GFP expression. </p>
    <blockquote> 
      <p> <b>Example:</b> <i>Tg(NBT:MAPT-GFP)<sup>zc1</sup></i>; <i>abc<sup>b000</sup></i> 
      </p>
    </blockquote>
  </blockquote>
  <p> <a name="4.2."></a><B>4.2. Linked loci</B> - the haplotype on each chromosome is written as 
    a sequence, with a space separating syntenic loci, and loci are placed in 
    the order they appear on the Linkage Group, top to bottom. Homologous chromosomes 
    are separated by a slash, and non-homologous chromosomes are separated by 
    semicolons. </p>
  <blockquote> 
    <p> <B>Examples:</B> <I>rse<SUP>b140</SUP> leo<SUP>t1</SUP></I>; <I>gol<SUP>b16</SUP></I> 
    </p>
  </blockquote>
  <p><a name="4.3."></a><b>4.3. Unmapped loci -</b> genotypes of unmapped loci are listed alphabetically 
    within braces following genotypes of mapped loci on different chromosomes 
  </p>
  <blockquote> 
    <p><b>Examples:</b><i> rse<sup>b140</sup></i>; {<i>edit<sup>z253</sup></i>;<i> 
      esr<sup>tj236</sup></i>} (<i>edi</i> and <i>esr</i> are unmapped, all three 
      loci are on different chromosomes)</p>
  </blockquote>
  <p><a name="4.4."></a><b>4.4. Poorly resolved loci on same chromosome -</b> genotypes of loci on 
    the same chromosome at locations that are undetermined or poorly resolved 
    are listed alphabetically within braces.</p>
  <blockquote> 
    
      <p><b>Examples:</b><br>
    <blockquote>
	    {<i>abc<sup>b000</sup> def<sup>m000</sup></i>} (poorly resolved loci on 
        same chromosome)<i><br>
      rse<sup>b140</sup></i> {<i>abc<sup>b000</sup> def<sup>m000</sup></i>} <i>leo<sup>t1</sup></i> 
      (poorly resolved loci in a known interval between mapped loci, all on same 
      chromosome.</blockquote>
  </blockquote>
</blockquote>
<hr width=90%>
<br>
<strong><a name="5."></a>5. PHENOTYPES</strong><br>
<br>
Phenotypes are designated by the name or symbol of the mutation in italics.
<blockquote>
   <p> <b>Examples:</b> <i>cyclops cyc</i> </p>
</blockquote>
<hr width=90%>
<P align="left"> <a name="6."><STRONG>6. CHROMOSOMES AND ABERRATIONS</STRONG></a><br>
  <BR>
  Chromosomes are not yet cytologically defined. Linkage groups are designated 
  by non-italic numerals, each linkage group corresponds to one of the 25 haploid 
  chromosomes. Chromosome differences have not been observed between males and 
  females. 
<P>
<B>Examples:</B><BR>
LG1 to LG25
<P> <B>Chromosome rearrangements</B>: Common prefixes include:<BR>
   Cr, complex rearrangement<BR>
   Df, deficiency<BR>
   Dp, duplication<BR>
   In, inversion<BR>
   Is, insertion<BR>
   T, translocation<BR>
   Tg, transgene
<P><BR>
<blockquote>
   
  <p><a name="6.1."><b>6.1. Deficiencies</b></a>: The general format for 
    naming a deficiency is: </p>
   <p><i>Df(LG##)xxx<sup>allele</sup></i></p>
   <p><i>Df</i> indicated deficiency. The term xxx should describe the salient
     features of the deficiency, as determined by the investigator. In 
cases where
     the deficiency removes sequences from a named gene, it should contain the
     standard symbol for that <a href="#1.">gene</a>. The <a 
href="#2.">allele
     name</a> should follow standard nomenclature conventions 
(laboratory designation
     followed by allele name).</p>
   <p> The linkage group where the deficiency maps should be specified 
by its number
     (##) using two digits (i.e. 03 for LG 3) so that computers will order them
     properly. </p>
   <blockquote>
     <p><b> Example: </b><i>Df(LG12)dlx3<sup>b380</sup></i><br>
     </p>
   </blockquote>
   <br>
   
  <p><a name="6.2."></a><b>6.2. Translocations: </b>The general format 
    for naming translocations depends upon the type of translocation:</p>
   <blockquote>
     
    <p><a name="6.2.1."></a><b>6.2.1. Reciprocal translocations:</b> </p>
   <p>These rearrangements have two separate chromosomal elements, and 
each element
     has a distinct name:</p>
     <p><i>T(LG##;LG##)xxx<sup>allele</sup>,##U.##L </i>and 
<i>T(LG##;LG##)xxx<sup>allele</sup>,##U.##L</i>
     </p>
     <p>T indicates translocation. The elements in the parentheses are 
the linkage
       groups involved, the lower numbered linkage group is listed 
first, and the
       linkage groups are separated by a semicolon. The linkage groups should be
       specified by their numbers (##) using two digits (i.e. 03 for 
LG 3) so that
       computers will order them properly. </p>
   <p>The term xxx should describe some salient feature of the translocation, as
     determined by the investigator. In cases where the translocation 
moves a named
     gene primarily studied by the investigator, xxx would usually be 
the standard
     <a href="#1.">symbol</a> for that gene. Alternatively, xxx could just be
     an experimental series number. </p>
     
    <p>The <a href="#2.">allele name</a> should follow standard nomenclature 
      conventions (laboratory designation followed by allele name, superscripted). 
      After the allele name comes a comma, and then a phrase that indicates the 
      new order of the chromosome, starting from the top of the chromosome as 
      displayed by convention. The first number (##) is the LG number, followed 
      by upper case Auto indicate the upper arm of a chromosome or by upper case 
      L to indicate the lower arm of a chromosome. The location of the centromere 
      is indicated by a period. No spaces. </p>
   <blockquote>
       <p><b>Example:</b><br>
        &nbsp;&nbsp;&nbsp;&nbsp;<i>T(LG02;LG12)cc<sup>b213</sup>,02U.12L02L</i> 
        and <i>T(LG02;LG12)cyc<sup>b213</sup>,12U.12L02L</i></p>
     <p>&nbsp;&nbsp;&nbsp;&nbsp;<img src="translocation.jpeg"></p>
   </blockquote>
     
    <p>This example illustrates a reciprocal translocation where a portion of 
      the lower arm of LG12 was translocated interstitially into the proximal 
      lower arm of LG2 and a portion of the lower arm of LG2 was translocated 
      to the distal lower arm of LG12.</p>
     
    <p><a name="6.2.2."></a><b>6.2.2. Resolved translocations:</b></p>
   <p>In some cases, the two elements of the translocation can 
separate, a mutant
     line would have just one of the elements. This results in the animal being
     monosomic for some chromosome regions and trisomic for others. In 
these cases,
     the mutant line would be designated with just one of the elements 
rather than
     two as in the reciprocal designation above. The allele name would 
remain the
     same to indicate their common origin and common breakpoint. </p>
   <blockquote>
       <p><b>Example:</b> <i>T(LG02;LG12)cyc<sup>b213</sup>,02U.12L02L</i></p>
   </blockquote>
   </blockquote>
   <br>
   
  <p><a name="6.3."><b>6.3. Transgenes</b></a>: The general format for naming 
    a transgene is: </p>
   
  <p><i>Tg(xxx)<sup>allele</sup></i></p>
   
  <p><i>Tg</i> indicates transgene. The term xxx should describe the salient features 
    of the transgene, as determined by the investigator. It is always contained 
    within parentheses. In cases where the insertion utilizes sequences from a 
    named gene, it should contain the standard symbol for that . The <a href="#2.">allele 
    name</a> should follow standard nomenclature conventions (laboratory designation 
    followed by allele name). If the insertion is a compound construct consisting 
    of the promoter of one gene linked to the coding region of another, the insertion 
    may be designated using the symbol of the first, separated by a colon from 
    the symbol of the second: </p>
   
  <p><i>Tg(promoter:gene)<sup>allele</sup></i></p>
   <p>In cases where the map location is known and where this 
information is useful
     (i.e., when the location of the transgene affects its expression 
or when the
     transgene is used as a map marker) the linkage group number (##) 
should also
     be included: </p>
   
  <p><i>Tg(Lg##,promoter:gene)<sup>allele</sup></i></p>
   <blockquote>
     
    <p><b>Example:</b> <i>Tg(Lg03,ERE:nic1)<sup>b512</sup></i></p>
   </blockquote>
</blockquote>
<HR WIDTH=90%>
<P> <STRONG><a name="7."></a>7. STRAINS</STRONG><br>
  <BR>
  Strain names are non-italic, first letter upper case. Abbreviations are non-italic, 
  all uppercase.
<P>
<blockquote> 
  <p> <B>Examples:</B> Darjeeling DAR Indonesia INDO</p>
</blockquote>
<HR WIDTH=90%>
  <p><a name="8."></a><STRONG>8. PRIORITY IN NAMES</Strong><br><BR>
 
<blockquote><b>8.1</b> When mutations in the same gene are found independently 
  in more than one laboratory and given different names, then the name that appears 
  first in the literature will be given priority. Sometimes different mutant alleles 
  of the same gene can have very different phenotypes and it may take a while 
  before it is shown that the mutations are in the same gene. The first published 
  name would get priority and all other alleles would adopt this designation. 
  <p><b>8.2</b> When a mutation is found in a previously cloned zebrafish gene 
    (which has been published), then the mutant will take the cloned gene's name. 
    For example, a mutation was induced in the <i>msxb</i> gene. The mutation 
    in the gene is also be referred to as <i>msxb</i> (in this case, <i>Df1(msxb),</i> 
    because the mutation is a deficiency). </p>
  <p><b>8.3</b> If both the cloned gene and the mutation are already known by 
    different names and at some later point they are found to be the same gene, 
    then the name of the mutation usually takes priority. The unique name of the 
    mutation may be more relevant to the function of the gene than the cloned 
    gene's name, and this unique mutant name may be easier to remember or use 
    than the gene name. For example, an author may find it more convenient to 
    refer to cyclops <i>(cyc),</i> rather than nodal-related-gene-two<i> (ndr2).</i> 
    Exceptions would include cases where orthologous relationships with genes 
    in other species are well established and these orthologues have a commonly 
    used name. For example, an author may find it preferable to refer to sonic 
    hedgehog <i>(shh),</i> rather than the sonic you <i>(syu)</i> gene.</p>
</blockquote>
<hr width=90%>
<p><STRONG><a name="9."></a>9. WEBSITE</STRONG><br>
  <BR>
   <B>The Zebrafish Information Network</B> (<A 
HREF="/">ZFIN</A>)
   provides access to information about zebrafish genetics, genomics, 
and development
   as well as links to researchers, laboratories, and publications.</p>
<HR WIDTH=90%>
<P> <STRONG><a name="10."></a>10. GENOME PROJECT</STRONG><br>
  <BR>
  The <a 
href="/zf_info/catch/catch.html">genome project</a> began in 1994, and by 1996 
  the genetic map was closed. NIH funded major <a href="http://www.nih.gov/science/models/zebrafish/">programs</a> 
  to develop a doubled haploid meiotic mapping panel, deficiency strains and expressed 
  sequence tags (ESTs), The ESTs and anonymous markers have been mapped on two 
  radiation-hybrid panels. The <a href="http://www.sanger.ac.uk/">Sanger Centre</a> 
  began <a href="http://www.sanger.ac.uk/Projects/D_rerio/">full genome squencing</a> 
  in 2001. A physical map is being constructed from the <a href="http://zfin.org/zf_info/zfprbs.html">BAC 
  libraries</a> used for sequencing. Genomic information is updated regularly 
  on <a href="/ZFIN">ZFIN</a>. 
<HR WIDTH=90%>
<P> <STRONG><a name="11."></a>11. CONTRIBUTORS</STRONG><BR>
   Marc Ekker <a href="mailto:mekker@lri.caa">(mekker@lri.ca</a>) Loeb Institute
   for Medical Research, Ottawa Civic Hospital, Canada<br>
   Ken Frazer <a 
href="mailto:ksf@cs.uoregon.edu">(ksf@cs.uoregon.edu)</a> Zebrafish
   International Resource Center, University of Oregon, USA<br>
   Mary Mullins <A HREF="mailto:mullins@mail.med.upenn.edu">(mullins@mail.med.upenn.edu)</A>
   Department of Cell and Developmental Biology, University of 
Pennsylvania, USA<BR>
   John Postlethwait<A HREF="mailto:jpostle@oregon.uoregon.edu">(jpostle@oregon.uoregon.edu)</A>
   Institute of Neuroscience, University of Oregon, USA <BR>
   Monte Westerfield <A HREF="mailto:monte@uoneuro.uoregon.edu">(monte@uoneuro.uoregon.edu)</A>
   Institute of Neuroscience, University of Oregon, USA
<HR WIDTH=90%>
<P> <STRONG><a name="12."></a>12. REFERENCES</STRONG><BR>
  1. <a href="/zf_info/monitor/vol2.1/vol2.1.html">The Zebrafish Science Monitor 
  (1992) Sept 21</a><BR>
  2. Mullins, M. (1995) Genetic methods: conventions for naming zebrafish genes 
  in The Zebrafish Book (3rd edition, Westerfield, M., ed.), pp 7.1-7.4, University 
  of Oregon Press.<br>
   3. Genetic Nomenclature Guide, Trends Genetics (1998).<BR>

</BODY>
<script language="JavaScript" src="http://<!--|DOMAIN_NAME|-->/footer.js"></script>
