<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html>
	<head>
		<title>
			ZFIN Zebrafish Nomenclature
		</title>

<style type="text/css">
<!--
h3 {
   background-color: #cccccc
   }
-->
</style>

	</head>

<script language="JavaScript" src="/javascript/header.js"></script>

<body>
<H1 align="center">Zebrafish Nomenclature Guidelines</H1>
  
<center>Based on <i>Trends in Genetics</i> Genetic Nomenclature Guide (1998)
</CENTER>
<br>

<table width="100%" cellspacing=20>

<tr><td width="50%" valign="top" style="border: medium ridge #999999; padding: 2%">
<P> <STRONG>NOMENCLATURE INFORMATION</STRONG><BR><BR>
  The current nomenclature guidelines are updates to rules established during 
  a discussion session at a meeting in Ringberg, Germany, in March 1992, and are 
  widely accepted by most zebrafish labs.<br>
   <BR>
  <B>Zebrafish Nomenclature Committee(ZNC)</B>
  <ul>
  <li><a href="/zf_info/nomen_comm.html">Our Activities</a>, <a href="#7">Contributors</a>
  <li>For questions and advice about appropriate nomenclature, contact us at <a 
href="mailto:nomenclature@zfin.org">nomenclature@zfin.org</a>
  </ul>

<a name="Approval"></a>

<P> <STRONG>APPROVAL FOR GENE AND MUTANT NAMES</STRONG><BR><BR>
Please use one of our submission forms to propose a new name for a gene or mutant and to provide supporting information. Your submission will be sent to the ZFIN nomenclature coordinator for review and will be treated in confidence
<p>

<input type="button" value="Submit a Proposed GENE Name" onclick="javascript:window.open('/<!--|WEBDRIVER_PATH_FROM_ROOT|-->?MIval=aa-new_gene.apg');">
<p>
      <input type="button" value="Submit a Proposed LOCUS/LINE Name" onclick="javascript:window.open('/<!--|WEBDRIVER_PATH_FROM_ROOT|-->?MIval=aa-new_line.apg');">

<P> <STRONG>ADDITIONAL RESOURCES</STRONG>
<p><a href ="/zf_info/nomen_tutorial.html">A Tutorial for Proposing Zebrafish Gene Nomenclature</a></p>
<p><a href ="/zf_info/zfbook/lab_desig.html">Laboratory Line Designations</a></p>

<P><b>Other nomenclature guidelines: </b><a href="http://www.genenames.org/guidelines.html">Human</a>, 
  <a href="http://www.informatics.jax.org/mgihome/nomen/">Mouse</a>, <a href="http://flybase.bio.indiana.edu/docs/nomenclature/lk/nomenclature.html/">Fly 
  (Drosophila)</a>, <a href="http://www.yeastgenome.org/gene_guidelines.shtml">Yeast 
  (Saccharomyces)</a>, <a href="http://www.informatics.jax.org/mgihome/nomen/genefamilies/index.shtml">Gene 
  families</a> 
</td>



<td style="border: medium ridge #999999; padding: 2%" valign="top">
<br>
<b>CONTENTS:</b>



  <li><a href="#1">1. Gene names and Symbols</a><br>
  <ul>
    <li><a href="#1.1">1.1. Genes identified by cloning</a><br>
    <li><a href="#1.1.2">1.2. Duplicated genes</a><br>
    <li><a href="#1.1.3">1.3. Genes (loci) identified by mutation</a><br>
    <li><a href="#1.1.4">1.4. Genes identified by genomic sequencing projects</a><br>
      <li><a href="#1.1.5">1.5. Genes identified only by other large scale projects</a><br>
      <li><a href="#1.1.6">1.6. Transcript variants</a><br>
  </ul>
  
  
    <li><a href="#2">2. Proteins</a><br>
    
  <li><a href="#3">3. Alleles and Genotypes</a><br>
  <ul>
      <li><a href="#3.1">3.1 Line designations</a><br>
      <li><a href="#3.2">3.2 Genotype nomenclature for publications</a><br>
      <li><a href="#3.3">3.3 Genotype display at ZFIN</a><br>
      </ul>
  

  <li><a href="#4">4. Chromosomes and aberrations</a><br>
    <ul>
	 <li><a href="#4.1">4.1.  Deficiencies</a><br>
	 <li><a href="#4.2">4.2. Translocations</a><br>
	 <li><a href="#4.3">4.3. Transgenic lines and constructs</a><br>
    </ul>
    
    <li><a href="#5">5. Priority in Names</a><br>
    <li><a href="#6">6. Mapping and Sequencing information</a><br>

    
  <li><a href="#7">7. Contributors</a><br>
    
  <li><a href="#8">8. References</a><br>
    
  </li>



</td>

</table><br><br>


<br><br>


<h3 align="left"><b><a name="1">1. GENE NAMES AND SYMBOLS</a></b></h3>
<p>Full gene names are lowercase italic, and gene symbols are three or more lowercase letters and are also italicized. The letters should be unique with respect to other named zebrafish mutants and genes and except in cases of established orthology, where the gene symbol should match that of the orthologue. Zebrafish gene designations should not be preceded by 'Z' or 'Zf'.  The use of punctuation such as period and hyphens in gene names or symbols is discouraged, except under specific circumstances described below.<br>
Gene names should be registered at <a href="/<!--|WEBDRIVER_PATH_FROM_ROOT|-->?MIval=aa-new_gene.apg">zfin</a>

<blockquote> 
  <p><strong><a name="1.1">1.1. Genes identified by cloning</a></strong> 
  </p>
  <blockquote> 
    <p>Genes should be named after the mammalian orthologue whenever possible. When mammalian orthologues are known, the same name and abbreviation should be used, except all letters are italicized and lower case.  Members of a gene family are sequentially numbered. 

<p><b>Examples:</b>
    &nbsp;&nbsp;&nbsp;&nbsp;Names - <i>engrailed 1a, engrailed 2b</i>
    &nbsp;&nbsp;&nbsp;&nbsp;Symbols - <i>eng1a, eng2b</i>
    
<p>In some cases when a zebrafish gene has been renamed to the mammalian orthologue from an older zebrafish name, it is still preferable within a publication to refer to the previous name.  Refer to the previous name by appending the previous name in parentheses. Previous names are searchable at ZFIN.

<p><b>Examples:</b> <i>shha (syu), bmp2b (swr)</i>

     </blockquote> 
    <p><b><a name="1.1.2">1.2. Duplicated genes</a> </b>The zebrafish genome contains duplicated segments that resulted from a genome-wide duplication in the ray fin fish lineage after it diverged from the lobe fin lineage (that included avian and mammalian species). For this reason, zebrafish often have two copies of a gene that is present as a single copy in mammals. 

 <p>In these cases, symbols for the two zebrafish genes should be the same as the approved symbol of the human or mouse orthologue followed by "a" or "b" to indicate that they are duplicate copies. Before these symbols are assigned, it is important to provide evidence by mapping that the two copies reside on duplicated chromosome segments. It is preferable that all copies in one of the duplicate chromosome segments use the same "a" or "b" suffix, although this will not always be possible for historical reasons. This terminology should not be used for duplicates that resulted prior to the divergence of ray fin and lobe fin fish. In these cases it is preferable to use terminology that is most consistent with the mammalian nomenclature.

 <p><b>Examples:</b> <i>hoxa13a, hoxa13b</i>

 <p>In some cases when there is a unique mammalian orthologue, but addition of the a, b suffixes would conflict with a different mammalian gene symbol, then numerical suffixes .1, .2 should be appended to the orthologous mammalian gene symbol instead of a, b. Tandem duplicates with a single mammalian orthologue may also be appended with a .1, .2, using the same symbol as the mammalian orthologue. 

 <p><b>Examples:</b> <i>stat5.1, stat5.2</i>

 <p>When mammalian gene duplications prevent identification of a unique mammalian orthologue, then an alternate gene symbol should be chosen. A possible choice would be an approved gene symbol from a unique non-mammalian orthologue. When a gene is homologous to a human gene, but orthology is ambiguous, the gene should be named after the closest mammalian homologue with the word 'like' appended to the name of the homologue.  In some cases, a gene family described in zebrafish is homologous to a mammalian gene family but the evolution of the gene family is ambiguous. Under these circumstances the zebrafish gene family should be named with the same stem as the mammalian gene family with the gene number beginning after the end of the mammalian numbering and continuing sequentially throughout the gene family.  If the members of the gene family are on the same chromosome, the adjacent genes should be given sequential numbers.  
 </p>
    <p><b><a name="1.1.3">1.3. Mutant loci with unidentified genes </a></b> Mutant loci for which the gene has not yet been identified are given placeholder gene names.  When the gene is identified, it is renamed following standard nomenclature guidelines as described above. Genes identified by mutation are typically named to reflect the mutant phenotype. The symbol should be derived from the full name. Numbers should generally not be used in naming a mutant. 

 <p><b>Example:</b> <i>touchy feely, tuf</i>

 <p>Mutant names should be registered at <a href="/<!--|WEBDRIVER_PATH_FROM_ROOT|-->?MIval=aa-new_line.apg">zfin</a>. 
</p>
    <p><b><a name="1.1.4">1.4. Genes identified only by genomic sequencing projects</a></b>Large-scale genome sequencing projects use a variety of prediction methods to identify both open reading frames and genes. Some of these genes are already known, while others are new. Novel genes identified by these means often cannot be identified and are assigned a name comprised of a prefix, a clone name, and an integer. The prefix is used to specify the research institution that identified the gene (e.g., "si" for the Sanger Institute). A colon separates the prefix from the clone identifier. In many cases, there are multiple predicted reading frames in a single clone. These genes are distinguished with a full stop (period) between the clone name and an integer. Integers are assigned to genes in the clone as they are identified and do not indicate the order of genes. If part of a gene is found in more than one clone, the name of the first clone in which the 5' portion of the gene is found takes precedence.

 <p><b>Examples:</b> <i>si:bz3c13.1, si:bz3c13.2, si:bz3c13.3</i>

 <p>Genes initially identified by genomic sequencing projects are renamed using standard nomenclature guidelines (described above) as more information about them becomes available.  
 
     <p><b><a name="1.1.5">1.5. Genes identified only by other large scale projects</a></b> Large-scale sequencing of ESTs or full length cDNA clone sets often result in large numbers of unidentified genes. These are given placeholder names with the project prefix, a colon and a clone number, similar to genes identified by genomic sequencing projects.  In these cases, the clones usually contain only one or a fragment of a single gene. 
     
      <p><b>Examples:</b> <i>im:7044540</i>, zgc:165514
 
  
    <p><b><a name="1.1.6">1.6. Transcript variants</a></b> Transcript variants that originate from the same gene are not normally given different gene symbols and names. However, variants from a single gene can be distinguished in publications by adding to the end of the full name a comma, "transcript variant", and a serial number; and by adding to the end of the symbol an underscore, "tv", and a serial number.

 <p><b>Examples:</b>
    &nbsp;&nbsp;&nbsp;&nbsp;Names -
        <i>myosin VIa, transcript variant 1
        myosin VIa, transcript variant 2</i>
    &nbsp;&nbsp;&nbsp;&nbsp;Symbols -
        <i>myo6a_tv1
        myo6a_tv2</i>



     </blockquote> 

<h3 align="left"><a name="2">2. PROTEINS</a></h3>

<p>The protein symbol is the same as the gene symbol, but non-italic and the first letter is uppercase. 

<p><b>Examples:</b> Ndrw, Brs, Eng1a, Eng2b, Ntl 

<p>Note the differences between zebrafish and mammalian naming conventions:

<p>species / gene / protein<br>
zebrafish / <i>shha</i>/ Shha<br>
human / <i>SHH</i> / SHH<br>
mouse / <i>Shh</i> / SHH <br>

<p>In publications, it is sometimes convenient to refer to a protein which has been renamed based on orthology using the more commonly known name in parentheses following the current name. 

<p><b>Examples:</b> Shha (Syu), Bmp2b (Swr)
</p>

<h3><a name="3">3. ALLELES and GENOTYPES</a></h3>

 <blockquote>  <p><b><a name="3.1">3.1 Line designations</a></b>
  

<p>When describing genes wild-type alleles are indicated using a superscript "+", while mutant alleles are indicated using a superscript line designation. Line designations are composed of a institution-specific designation followed by a number.  The full list of institution designations can be found at <a href="http://zfin.org/zf_info/zfbook/lab_desig.html">ZFIN</a>.

<p>Institute specific line designations should be two or three letters in length, preferably  two letters.  These designations should not be the same as a gene name in mouse or human.   The institution designation should be followed by a unique number specific to a particular line.  Other letters should not immediately follow the institution designation but may be appended to the end of the line designation to make it unique.  Line designations should only contain alphanumeric characters.    Dominant alleles have a d in the first position of the line designation to distinguish them from recessive alleles.   This means that the letter 'd' cannot begin an institution designation.   Line designations for transgenic lines follow these same rules, so the same number cannot be give to both a transgenic line and a mutant allele.

 <p><b>Examples:</b> <i>b</i> is the Eugene designation; <i>m</i> is  for MGH, Boston; <i>t</i> is Tuebingen, Germany<br>
   &nbsp;&nbsp;&nbsp;&nbsp; wild type:  <i>lof<sup>+</sup>, ndr2<sup>+</sup>, brs<sup>+</sup></i><br>
   &nbsp;&nbsp;&nbsp;&nbsp; mutant: <i>lof<sup>dt2</sup>, ndr2<sup>b16</sup>, ndr2<sup>m101</sup>, ndr2<sup>t219</sup></i>

  <p><b><a name="3.2">3.2 Genotype nomenclature for publications</a></b>
  
  <p>For unlinked loci, heterozygotes and homozygotes are distinguished by having each allele separated by a slash "/".

  <p><b>Examples:</b><br>
      &nbsp;&nbsp;&nbsp;&nbsp;<i>ednrb1<sup>b140</sup>/ ednrb1<sup>+</sup></i> (heterozygote, can be abbreviated <i>ednrb1<sup>b140/+</sup></i>)
      &nbsp;&nbsp;&nbsp;&nbsp;<i>ednrb1<sup>b140</sup>/ ednrb1<sup>b140</sup></i> (homozygote, can be abbreviated <i>ednrb1<sup>b140/b140</sup></i> or <i>ednrb1<sup>b140</sup></i>) 

  <p>For homozygous genotypes, the genotype at each locus is listed in order according to Chromosome number, from 1 to 25, with a semicolon to separate loci on different chromosomes. 

  <p><b>Examples:</b> <i>ednrb1<sup>b140</sup>; slc24a5<sup>b16</sup></i>

  <p>For heterozygous genotypes, loci on homologous chromosomes are separated by a slash. 

  <p><b>Examples:</b> <i>ednrb1<sup>b140</sup>/ednrb1<sup>+</sup>; slc24a5<sup>b16</sup>/slc24a5<sup>m592</sup></i>

  <p>For linked loci, the haplotype on each chromosome is written as a sequence, with a space separating syntenic loci, and loci are placed in the order they appear on the Linkage Group, top to bottom. Homologous chromosomes are separated by a slash, and non-homologous chromosomes are 
separated by semicolons. 

  <p><b>Examples:</b> <i>ednrb1<sup>b140</sup> cx41.8<sup>t1</sup>; slc24a5<sup>b16</sup></i>

  <p>For unmapped loci, genotypes of unmapped loci are listed alphabetically within braces following genotypes of mapped loci on different chromosomes. 

  <p><b>Examples:</b> <i>ednrb1<sup>b140</sup>; mycbp2<sup>tj236</sup> {edit<sup>z253</sup>}</i> (edi is unmapped, all three loci are on different chromosomes)

  <p>Poorly resolved loci on same chromosome are listed alphabetically within braces. 

  <p><b>Examples:</b><br>
&nbsp;&nbsp;&nbsp;&nbsp;{<i>abcb<sup>000</sup> def<sup>m000</sup>}</i> (poorly resolved loci on same chromosome)
&nbsp;&nbsp;&nbsp;&nbsp;<i>ednrb1<sup>b140</sup> {abcb<sup>000</sup> def<sup>m000</sup>} cx41.8<sup>t1</sup></i> (poorly resolved loci in a known interval between mapped loci, all on same chromosome)

  <p><b><a name="3.3">3.3 Genotype displays in ZFIN</a></b>
  
  <p>Due to technical constraints, genotypes at ZFIN are shown in alphabetical order by gene, and then by allele designation. See below for display of complex genotypes involving transgenic or chromosomal rearrangements. </p>

     </blockquote> 

<h3><a name="4">
4. CHROMOSOMES AND ABERRATIONS</a></h3>

<p>The chromosome numbering system corresponds to the old Linkage Group designations with what was LG 1 now named Chr 1. Chromosomes are designated by non-italic numerals, 1 to 25.  Chromosome differences have not been observed between males and females.   

<p><b>Examples:</b><br>
&nbsp;&nbsp;&nbsp;&nbsp;Chr1 to Chr25 

<p>Chromosome rearrangements are indicated with the following prefixes, followed by the details within parentheses. See below for specific examples. Common prefixes include:<br>
<I>Df</I>, deficiency<br>
<I>Dp</I>, duplication<br>
<I>In</I>, inversion<br>
<I>Is</I>, insertion<br>
<I>T</I>, translocation<br>
<I>Tg</I>, transgene <br>



<blockquote> <a name="4.1"></a><b>4.1. Deficiencies</b> <br>
The general format for naming a deficiency is: <br>
<I>Df(Chr<sup>##</sup>)xxx<sup>allele</sup></I>

<p><I>Df</I> indicated deficiency. The term <I>xxx</I> should describe the salient features of the deficiency, as determined by the investigator. In cases where the deficiency removes sequences from a named gene, it should contain the standard symbol for that <a href="http://zfin.org/zf_info/nomen.html#1.">gene</a>. The <a href="http://zfin.org/zf_info/nomen.html#2.">allele name</a> should follow standard nomenclature conventions (laboratory designation followed by allele name).

<p>The chromosome where the deficiency maps should be specified by its number (##) using two digits (i.e., 03 for Chr 3) so that computers will order them properly. 

<p><b>Example:</b> <i>Df(Chr12)dlx3<sup>b380</sup></i>

<p>Deficiencies are written as an allele of a gene when the gene is disrupted at one of the two breakpoints of the deficiency.   The omitted genes are listed as missing.

<p>Example:
<i>ndr2<sup>b16</sup></i> is a deficiency in which a portion of ndr2 is missing, causing the deficiency to be an allele of ndr2. 

  
<p> <a name="4.2"></a><b>4.2. Translocations</b> <br>
The general format for naming translocations depends upon the type of translocation:

<p>Reciprocal translocations have two separate chromosomal elements, and each element has a distinct name:
<i>T(Chr##;Chr##</sup>)xxx<sup>allele</sup>,##U.##L</i> and <i>T(Chr##;Chr##)xxx<sup>allele</sup>,##U.##L</i>

<p><i>T</i> indicates translocation. The elements in the parentheses are the chromosomes involved, the lower numbered chromosome  is listed first, and the chromosomes are separated by a semicolon. The chromosomes should be specified by their numbers (##) using two digits (i.e., 03 for Chr 3) so that computers will order them properly. 

<p>The term <i>xxx</i> should describe some salient feature of the translocation, as determined by the investigator. In cases where the translocation moves a named gene primarily studied by the investigator, <i>xxx</i> would usually be the standard <a href="http://zfin.org/zf_info/nomen.html#1.">symbol</a> for that gene. Alternatively, <i>xxx</i> could just be an experimental series number. 

<p>The <a href="http://zfin.org/zf_info/nomen.html#2.">allele name</a> should follow standard nomenclature conventions (laboratory designation followed by allele name, superscripted). After the allele name comes a comma, and then a phrase that indicates the new order of the chromosomes, starting from the top of the chromosome as displayed by convention. The first number (##) is the Chr number, followed by upper case <i>U</i> to indicate the upper arm of a chromosome or by upper case <i>L</i> to indicate the lower arm of a chromosome. The location of the centromere is indicated by a period. No spaces.  Translocations are written as an allele of a gene when the gene is disrupted at one of the breakpoints of the translocation. There can be as many as four alleles of a translocation.

<p><b>Example:</b>
    &nbsp;&nbsp;&nbsp;&nbsp;<i>T(Chr02;Chr12)cyc<sup>b213</sup>,02U.12L02L</i> and <i>T(Chr02;Chr12)cyc<sup>b213</sup>,12U.12L02L</i>

<p>This example illustrates a reciprocal translocation where a portion of the lower arm of Chr12 was translocated interstitially into the proximal lower arm of Chr2 and a portion of the lower arm of Chr2 was translocated to the distal lower arm of Chr12.

<p>Resolved translocations are where the two elements of the translocation separate and a mutant line has just one of the elements. This results in the animal being monosomic for some chromosome regions and trisomic for others. In these cases, the mutant line would be designated with just one of the elements rather than two as in the reciprocal designation above. The allele name would remain the same to indicate their common origin and common breakpoint. 

<p><b>Example:</b> <i>T(Chr02;Chr12)cyc<sup>b213</sup>,02U.12L02L</i>

<p><a name="4.3"></a><b>4.3. Transgenic lines and constructs</b>
<p>Transgenic constructs now have their own pages in ZFIN. Transgenic construct names are important because the construct name is used in the transgenic line nomenclature when the insertion is NOT an allele of a gene (see below). 

<blockquote><p> <a name="4.3.1"></a><b>4.3.1 Transgenic constructs</b>

<p><i>Tg(promoter:gene)</i>

<p><i>Tg</i> indicates transgene. Within the parentheses, the most salient features of the transgene should be described. Brevity and clarity in the transgene name are favored, in general, over exhaustive detail. Regulatory sequences should be listed to the left of the colon, and coding sequences to the right of the colon. Not all transgenic constructs will have both promoter and coding elements, and in this case, the colon will not be used. In cases where a construct utilizes sequences from a named gene, it should contain the standard zebrafish lowercase symbol for that gene. Human genes may be listed in all capital letters, mouse genes with the first letter capitalized, and other non-zebrafish genes with the two letter species abbreviation (Ss for Salmo Salar) and the first letter of the gene capitalized. For commonly used reporter elements (i.e. GFP, DsRed) use standard naming conventions. 

<p><b>Example:</b> <i>Tg(SsNdr2:GFP)</i> 

<p>Regulatory sequence could be derived from either an enhancer or promoter, and are denoted by the symbol of the regulated gene. Regulatory or coding sequence fusions should be separated by hyphens. In cases where a number of constructs are generated with differing sizes of promoter elements, these may be specified within the parentheses as follows: 

<p><b>Example:</b> <i>Tg(-3.5hhex-ERE:sptb-GFP)</i> 

<br><i>Tg(-6.0hhex-ERE:sptb-GFP)</i> 

<p>Here, two constructs having a fusion protein of <i>spectrin beta (sptb)</i> and GFP driven by a combination of an estrogen response element and an upstream enhancer containing either 3.5kb or 6.0kb 5' to the <i>hhex</i> gene. 

<p>However, in a number of cases, the changes within the construct may be too small to change the number of kb. In this case, the constructs will be appended with a period and a number within the parentheses, referring to the element that has changed, instead of including further details in the name. Alternatively, if the .1, .2 nomenclature conflicts with a gene name, then a number may be placed at the beginning of the construct name. The numbering should start with a 1 and increment by one for each different construct. The details of construct differences will be available on the construct pages. 

<p><b>Example:</b> <i>Tg(-1.7shha.1:GFP)</i><br>
 &nbsp;&nbsp;&nbsp;&nbsp;<i>Tg(-1.7shha.2:GFP)</i><br>
 &nbsp;&nbsp;&nbsp;&nbsp;<i>Tg(-1.7shha.3:GFP)</i><br>

<p>Sometimes within a single construct, there are multiple cassettes, each containing regulatory and coding sequences. In this case, it is necessary to distinguish between what is coding in the first cassette with what is regulatory in the second. Multiple cassettes may be distinguished using a comma. In the following example, isl3 promoter and retinoic acid response element drive GAL4, and UAS drives GFP. 

<p><b>Example:</b> <i>Tg(isl3-RARE:GAL4,UAS:GFP)</i>

<p>Enhancer, promoter, and gene-trap constructs may use Et, Pt, or Gt, all of which are considered types of transgenic constructs. 


<p> <a name="4.3.2"></a><b>4.3.2 Enhancer trap, promoter trap, gene trap constructs</b>

<p>These all use the same nomenclature convention as described for transgenic constructs above, substituting Pt, Gt, Et as necessary. 



<p> <a name="4.3.3"></a><b>4.3.3 Transgenic lines</b>

<p>Transgenic lines are of two types, those that are known to create alleles of genes and those that are not known to create alleles of genes. 
For a line that does not create an allele of a gene, the feature name consists of the construct name appended with a unique line number with no superscript. The line number should begin with the laboratory designation followed by a unique number. 

<p><b>Example:</b> <i>Tg(hsp70l:GFP)mik6</i>

<p>For lines that do create alleles of a gene, a standard genetic representation is used, where the allele designation is superscripted above the gene, but is appended with a Tg to indicate that it is a transgenic insertion allele. Details regarding the construct used will be available on the genotype page. Gene traps and enhancer traps known to create alleles of a gene are handled in a similar fashion, appending Gt or Et to the allele designation. 

<p><b>Example:</b> <i>arnt<sup>hi2639cTg</sup></i><br>
<i>parg<sup>mn2Et</sup></i>

    
     
     <p> <a name="4.3.4"></a><b>4.3.4 Display of complex genotypes at ZFIN</b>
     
     <p>Genotypes at ZFIN are shown in alphabetical order with transgenic lines that are not alleles of genes first, then other alleles.

<p><b>Example:</b> <i>Tg(-0.7her5:EGFP)ne2067;hmgcrbs<sup>617/s617</sup></i>


</blockquote></blockquote>

<h3><b><a name="5">5. PRIORITY IN NAMES</a></b></h3>

<p>As described above, zebrafish genes are named based on orthology to a human or mouse gene. If an ortholog cannot be identified, then the name that appears first in the literature will be given priority assuming it follows other nomenclature guidelines. ZFIN recommends submission of proposed gene names via the ZFIN form or consultation with the zebrafish nomenclature committee (<a href=mailto:nomenclature@zfin.org>nomenclature@zfin.org</a>) for nomenclature assignment. 

<p>When a mutation is found in a previously cloned zebrafish gene, then the mutant will be referred to as an allele of the gene. If both the cloned gene and the mutation are known by different names and later found to be the same gene, then the name of the gene usually takes priority.  The exception to this rule is when the mammalian gene has a gene symbol that is less than two characters such as the mouse gene brachyury which has the symbol T.  In this case the zebrafish gene retained the original name <i>no tail, ntl.</i>  

</p>









<h3><b><a name="6">6. MAPPING AND SEQUENCING INFORMATION</a></b></h3>

<p>The <a href="http://zfin.org/zf_info/catch/catch.html">genome project</a> began in 1994, and by 1996 the genetic map was closed. NIH funded major <a href="http://www.nih.gov/science/models/zebrafish/">programs</a> to develop a doubled haploid meiotic mapping panel, deficiency strains and expressed sequence tags (ESTs), The ESTs and anonymous markers have been mapped on two radiation-hybrid panels. The <a href="http://www.sanger.ac.uk/">Sanger Institute</a> began <a href="http://www.sanger.ac.uk/Projects/D_rerio/">full genome squencing</a> in 2001. A physical map is being constructed from the <a href="http://zfin.org/zf_info/zfprbs.html">BAC libraries</a> used for sequencing. Genomic information is updated regularly on <a href="/<!--|WEBDRIVER_PATH_FROM_ROOT|-->?MIval=aa-ZDB_home.apg">ZFIN</a>. 
</p>



<br>
<br>
<h3><a name="7">7. CONTRIBUTORS</a></h3>
<p>Marc Ekker (<a href=mailto:mekker@uottawa.ca>mekker@uottawa.ca</a>), Center for Advanced Research in Environmental Genomics, University of Ottawa, Ontario, Canada<br>
Mary Mullins (<a href=mailto:mullins@mail.med.upenn.edu>mullins@mail.med.upenn.edu</a>), Department of Cell and Developmental Biology, University of Pennsylvania, USA<br>
John Postlethwait (<a href=mailto:jpostle@oregon.uoregon.edu>jpostle@oregon.uoregon.edu</a>), Institute of Neuroscience, University of Oregon, USA <br>
Monte Westerfield (<a href=mailto:monte@uoneuro.uoregon.edu>monte@uoneuro.uoregon.edu</a>), Institute of Neuroscience, University of Oregon, USA <br>
Erik Segerdell, XenBase, University of Calgary, Canada<br>
Melissa Haendel (<a href=mailto:mhaendel@uoneuro.uoregon.edu>mhaendel@uoneuro.uoregon.edu</a>), Zebrafish Information Network, University of Oregon, USA<br>
Ceri Van Slyke (<a href=mailto:van_slyke@uoneuro.uoregon.edu>van_slyke@uoneuro.uoregon.edu</a>), Zebrafish Information Network, University of Oregon, USA<br>

  <p> 



<br>
<br>
<h3><a name="8">8. REFERENCES</a> </h3>

<p>1. The Zebrafish Science Monitor (1992) Sept. 21.<br>
2. Mullins, M. (1995) Genetic methods: conventions for naming zebrafish genes in The Zebrafish Book (3rd edition, Westerfield, M., ed.), pp 7.1-7.4, University of Oregon Press.<br>
3. Genetic Nomenclature Guide, Trends in Genetics (1998). <br>

<br>
<br>

</BODY>
<script language="JavaScript" src="/javascript/footer.js"></script>
