<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html>
	<head>
		<title>
			ZFIN Genomic Resources
		</title>
	</head>

<script language="JavaScript" src="/header.js"></script>

<h1 align="center">Genomic Resources for Zebrafish</h1>

<p></p>
<b><font size="+1">Genome and Map Resources</font></b
<ul>
  <li><b><a href="http://www.sanger.ac.uk/Projects/D_rerio/">The Danio rerio Sequencing Project at Sanger</a>:</b><br>
The <a href="http://www.sanger.ac.uk/Projects/D_rerio/"><b>Sanger zebrafish homepage</b></a> hosts all of the efforts of the zebrafish genome sequencing project, including the <a href="http://www.sanger.ac.uk/Projects/D_rerio/wgs.shtml"><b>whole genome sequencing and assembly project</b></a> with automated annotation in <a href="http://www.ensembl.org/Danio_rerio/"><b>Ensembl</b></a> and the <a href="http://www.sanger.ac.uk/Projects/D_rerio/clones.shtml"><b>clone mapping and sequencing project</b></a> with manual annotation in <a href="http://vega.sanger.ac.uk/Danio_rerio/"><b>Vega</b></a>.
  </li>
  <li><b><a href="http://vega.sanger.ac.uk/Danio_rerio/">Zebrafish Genome Annotation Browser (Vega)</a>:</b><br>
This site allows browsing of zebrafish BAC and PAC clones that were sequenced, analysed and manually annotated within the Sanger zebrafish genome project. Vega includes
    <ul type=circle>
      <li> manually annotated genes with poly A features
      <li> protein feature and domain annotation
      <li> Blast searches against all zebrafish ESTs, cDNAs and the WashU and IMCB EST cluster</li>
      <li> Blast searches against all vertebrate proteins, cDNAs and ESTs</li>
      <li> mapping of unists markers</li>
      <li> ab initio gene predictions</li>
      <li> tRNA prediction</li>
      <li> CpG island prediction and repeat masking</li>
    </ul>
You can search Vega with <a href="http://vega.sanger.ac.uk/Multi/blastview?species=Danio_rerio"><b>Blast</b></a>.  For Blast searches against all finished and unfinished clones go <a href="http://www.sanger.ac.uk/cgi-bin/blast/submitblast/d_rerio"><b>here</b></a>.<br>
The manual annotation is a joined effort between the Sanger Institute (gene structure annotation) and ZFIN (nomenclature).
 </lI>
  <li><a href="http://www.ensembl.org/Danio_rerio/"><b>Zebrafish Genome Browser at Ensembl</b></a><b>:</b><br>
  This site allows browsing of the latest zebrafish whole genome assembly <a href="http://www.sanger.ac.uk/Projects/D_rerio/Zv5_assembly_information.shtml"><b>Zv5</b></a>
    <ul type="circle">
      <li> automatically annotated Ensembl genes</li>
      <li> automatically annotated Ensembl EST genes</li>
      <li> protein feature and domain annotation</li>
      <li> Blast searches against all zebrafish ESTs, cDNAs and the WashU and IMCB EST clusters</li>
      <li> Blast searches against all vertebrate proteins, Uniprot, all vertebrate cDNAs and ESTs</li>
      <li> mapping of unists markers</li>
      <li> mapping of the Affymetrix zebrafish probes</li>
      <li> mapping of SNps in dbSNP</li>
      <li> comparative analysis with other Ensembl species (human, mouse, rat, chicken, fugu, tetraodon)</li>
      <li> BAC end placement</li>
      <li> transcription start site predictions with Eponine</li>
      <li> ab initio gene predictions</li>
      <li> tRNA prediction</li>
      <li> CpG island prediction and repeat masking</li>
    </ul>
Additional DAS tracks with SNP and indel analysis derived from the comparison of the whole genome shotgun reads are available. Ensembl also offers completely customizable downloads with <a href ="http://www.ensembl.org/Multi/martview?species=Danio_rerio"><b>EnsMart</b></a><br>
You can search Ensembl with either <a href="http://www.ensembl.org/Multi/blastview?species=Danio_rerio"><b>Blast</b></a> or <a href="http://www.ensembl.org/Multi/blastview?species=Danio_rerio"><b>SSAHA</b></a>.
  </li>
  <li>Additional services available at the Sanger Institute
      <ul type="circle">
        <li> The <a href="http://www.sanger.ac.uk/Projects/D_rerio/WebFPC/zebrafish/small.shtml"><b>WebFPC database</b></a> hosts the physical map and allows searching and browsing for clone information.</li>
        <li> The <a href="http://www.sanger.ac.uk/cgi-bin/Projects/D_rerio/mapsearch"><b>mapsearch site</b></a> offers easy access to all mapping related data (mapped ESTs, markers, BAC ends) in the current assembly.</li>
        <li> The <a href="http://www.sanger.ac.uk/Projects/D_rerio/fishmask.shtml"><b>RepeatMasker</b></a> service masks your sequence data with a choice of repeat libraries.</li>
	<li> The <a href="http://microrna.sanger.ac.uk/"><b>miRBase</b></a> service is the new home for microRNA data, incorporating the database  and gene naming roles previously provided by the miRNA Registry, and including the new miRBase Target database.
       </ul>
   </li>

 <li><b><a href="http://genome.ucsc.edu/cgi-bin/hgGateway?hgsid=85282730&clade=vertebrate&org=Zebrafish&db=0">Zebrafish Genome Browser at UCSC</a>:</b><br>
    A joint collaboration between UCSC and the Zebrafish Genome Initiative at Children's Hospital Boston provides several useful tracks of genome data, including: 
      <br>
<ul type="circle"><li>Mapping & Sequencing:  Position / Conigs / Scaffolds / RH Map / BAC End Pairs & Singles / Gap / GC Percent</li>
     <li>Genes & Gene Prediction:  RefSeq Genes / Ensembl Genes</li>
     <li>mRNA & EST:  ZF mRNAs & ESTs / spliced ESTs / WZ EST Clusters / nonZF mRNAs</li>
     <li>Expression & Regulation:  Affy ZF</li>
     <li>Comparative Genomics:  Fugu Chain & Net / Human Proteins / HG Blast</li>
     <li>Variation & Repeats:  RepeatMasker / SimpleRepeats</li></ul>
     Custom annotation tracks for the zebrafish research community include: 
     <ul type="circle">
       <li> RH Map (about 9K sequences)</li>
       <li> BAC End Clones (over 288K pairs, 23K singletons)</li>
       <li> WZ ESTs (43K sequences, 76K alignments, 39K EST alignments)</li>
       <li> Affymetrix Zebrafish Microarray Probe Alignments (15K probes)
       <li> Human Protein (over 40K alignments)</li>
     </ul>

  <li><b><a href=" http://big.nyp.edu.sg/zpcr/servlet/index">Automatic zebrafish-specific PCR primer design service</a>:</b><br>
 This service provides "one stop shopping" for picking primers, BLAST of primer
sequences for specificity against the latest zebrafish genome sequence,
and graphical views of the locations of primers with different confidence
levels. All you need is a few clicks. No more cutting and pasting from one
website to another. The program flows automatically. Detailed information
for use of this program can be found in the "general info" and "help file".</li>  

<li><B><a href="http://134.174.23.160/zfaca/hash/master020106public.aspx">Zebrafish Affy Chip Annotation Project:</a></b><br>
The Trans-NIH Zebrafish Genome Initiative at Children's Hospital Boston is
pleased to provide a new web service for retrieving hand-annotated
Affymetrix zebrafish gene probes to the research community.  A total of
12092 probes from the complete set of 15617 (115 of which are controls) have
been annotated with human known genes (UCSC HG16):  7846 (from 15617 total)
probes have been manually annotated by our labs, plus 4246 from Affymetrix
annotations dated 2005 Sept. 17.    We will continue annotating as new
sequence data become available and by comparative genomics approaches. 
Researchers can choose to download the entire annotation as a text file, or
retrieve their specific gene probes of interest by submitting a list of
Affymetrix gene probe IDs of interest.</li>

<li><b><a href="http://www.niob.knaw.nl/researchpages/cuppen/zfmodels/">The ZF-MODELS consortium's knock-out service for zebrafish:</a></b><br>
The ZF-MODELS consortium would like to draw attention to its public  and free knock-out service for zebrafish, based on the resequencing  of mutagenized fish (TILLING). Up to 60 knock-outs will be performed  for researchers outside the consortium, whose requests are considered  on the basis of their scientific merit. The workload is distributed  between the laboratories of Sanger Institute (D. Stemple), Hubrecht  Laboratory (E. Cuppen), and TU Dresden (M. Brand). To request a  zebrafish KO mutant, please send a brief (approx. one page) summary  to Dr. Edwin Cuppen <mailto:ecuppen@niob.knaw.nl>. This summary  should contain the following information:<br>

Information on the gene to be knocked out: Name of the gene and  Ensembl ID. Please note that the ZF-MODELS consortium will only  accept requests for KOs based on Ensembl IDs. The requester should  also check the Ensembl annotation. In case the annotation is  incorrect, the requester has to make sure that it is changed. The ZF- MODELS consortium will not screen for a gene as long as the  annotation is not correct. Scientific relevance: Explain why is it important to knock this gene  out.
Zebrafish background of the lab requesting the KO mutant: Explain how  any work on the KO mutant provided would be embedded in an  environment in your lab where people can actually work with zebrafish  mutants.<br>

"ZF-MODELS - Zebrafish Models for Human Development and Disease" is  an Integrated Project funded by the Sixth Framework Programme of the  European Commission. For further information please see the project  website <a href="http://www.zf-models.org">http://www.zf-models.org.</a>



  <li><b><a href=" http://microarray.no">Zebrafish Oligonucleotide Arrays</a>:</b><br>
The Norwegian Microarray Consortium offers their array products for academic use to customers world wide. One limitation is oligonucleotide arrays, which cannot provided to users in the US and Japan, due to licensing conditions. A moderate license fee will be added to the base price for oligonucleotide arrays sold to users in the UK, France, Germany, Italy, Austria, Belgium, Sweden, Switzerland, Luxembourg, and The Netherlands.</li>

  <li><b><a href="http://danio.mgh.harvard.edu/blast/blast_grp.html">Zebrafish BLAST 
    server at MGH</a>:</b> <br>
  MGH site provides a BLAST search of a gene or marker of interest to the assembled sequence contigs of the zebrafish genomic shotgun sequence from the Sanger project. </li>

 <li><b><a href="http://danio.mgh.harvard.edu/markers/ssr.html">MGH web-based tool for automating primer selection</a>:</b><br>
    MGH site offers this tool which searches for a supercontig, scaffold, or finished BAC  clone by name or accession number, derives all di- and tri- nucleotide repeat sequences in the supercontig, masks the
sequence around the repeats for larger zebrafish repetitive sequences and then designs unique sequence primer pairs to amplify the simple sequence repeats. These are potentially polymorphic simple sequence repeats (SSRs) that can be used in genetic mapping/positional cloning projects. The site also offers a tool to <a href="http://danio.mgh.harvard.edu/userMarkers/usrInputSsr.html"> process your sequence for simple repeats/primer generation</a>  
  </li>
  
  <li><b><a href="http://giscompute.gis.a-star.edu.sg/~govind/zebrafish/version2/">Zebrafish Compugen microarray annotation database:</a></b><br></li>
  
Compugen array clones are annotated at this site.  The GenBank id of genes in the Compugen  oligo list can be used to access information from  Unigene, Entrez Gene, Gene Ontology, human and mouse HomoloGene, Unigene protein similarities of mouse and human, ZGI index.  The gene chromosomal location is also available.   The details of the microarray annotation database are published in PLoS Genetics  (2005).
  
 <li><b><a href="http://134.174.23.160/compGenomics/">BLAST 
    hits of Human Genes from Sanger sequence maintained by Zon lab</a>: </b><br>
    Each human gene annotated on the Santa Cruz site 
    was compared by BLAST against the Sanger Institute 
    zebrafish genome sequence. These data can be used 
    to help identify zebrafish orthologues of human 
    or Fugu genes. </li>
  <li>
  <b><a href="http://wwwmap.tuebingen.mpg.de/">T&uuml;bingen Map of the Zebrafish Genome</a> </b><br>
  </li>
  <li><b><a href="/<!--|CGI_BIN_DIR_NAME|-->/ZFIN_jump?record=JUMPTOREFCROSS">RH, 
    meiotic and integrated maps of the genome</a>: </b><br>
    Centralized site for all meiotic and radiation hybrid 
    maps of the zebrafish genome. </li>
   <li><b><a href="http://www.ncbi.nlm.nih.gov/genome/guide/zebrafish/index.html">Zebrafish Genome Resources at NCBI</a></b><br>
</ul>
<b><font size="+1">cDNA and EST Resources</font></b>
<p>
<ul>
  <li><a href="http://zgc.nci.nih.gov/"><strong>Zebrafish Gene
  Collection (ZGC)</strong></a><b>:</b> NIH initiative for production of
  full length sequences and cDNA clones.</li>
  <li><b><a href="http://compbio.dfci.harvard.edu/tgi/cgi-bin/tgi/gimain.pl?gudb=zfish">TIGR  Zebrafish Gene Index (ZGI)</a>:</b><br>
    The TIGR Zebrafish Gene Index (ZGI) integrates research data from international zebrafish gene research projects and represents a non-redundant view of all zebrafish genes and data on their expression patterns, cellular roles, functions, and evolutionary relationships.</li>
  <li><b><a href="http://www.genoscope.cns.fr/zie/">Zebrafish Inner Ear ESTs Database</a>:</b><br>
   This database consists of 4694 contigs or singletons obtained from ~18000 reads of cDNAs that were isolated from the microdissected otocysts of zebrafish embryos at 20-30 hours postfertilization, following subtraction with a pool of liver cDNAs from adult fishes.
  <li><b><a href="http://bioweb.imcb.nus.edu.sg/zest.htm">Zebrafish EST Database at Singapore</a></b><br>
</ul>

<b><font size="+1">Mutants and Stocks</font></b>
<p>
  <ul>
   <li><b><a href="/zirc/">Zebrafish International Resource Center (ZIRC)</a></b><br></li>

   <li><b><a href="
http://www.eb.tuebingen.mpg.de/core-facilities/zebrafish-stockcenter/tubingen-zebrafish-stockcenter/ ">T&uuml;bingen Zebrafish stockcenter</a></b><br></li>

  </ul>
 
  <b><font size="+1">Morpholinos</font></b>
  <p>
  <ul>
   <li><b><a href="moqc.pdf">Quality control testing of Gene Tools morpholinos.</a></b><br></li>

   </ul>

 
  <b><font size="+1">White Papers</font></b>
  <p>
  <ul>
   <li><b><a href="Zebrafish_disease_project.pdf">The Zebrafish & Disease Project: Zebrafish as a model system to study and cure human diseases</a></b><br></li>

   </ul>
<script language="JavaScript" src="/footer.js"></script>
