<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html>
	<head>
		<title>
			ZFIN Genomic Resources
		</title>
	</head>

<script language="JavaScript" src="/header.js"></script>





<h1 align="center">Genomic Resources for Zebrafish</h1>


<p></p>
<h4 align="left"><em>(Updated 30 September 2004)</em></h4>
<b><font size="+1">Genome and Map Resources</font></b
<ul>
  <li><b><a href="http://www.sanger.ac.uk/Projects/D_rerio/">The Danio rerio Sequencing Project at Sanger</a>:</b><br>
The <a href="http://www.sanger.ac.uk/Projects/D_rerio/"><b>Sanger zebrafish homepage</b></a> hosts all of the efforts of the zebrafish genome sequencing project, including the <a href="http://www.sanger.ac.uk/Projects/D_rerio/wgs.shtml"><b>whole genome sequencing and assembly project</b></a> with automated annotation in <a href="http://www.ensembl.org/Danio_rerio/"><b>Ensembl</b></a> and the <a href="http://www.sanger.ac.uk/Projects/D_rerio/clones.shtml"><b>clone mapping and sequencing project</b></a> with manual annotation in <a href="http://vega.sanger.ac.uk/Danio_rerio/"><b>Vega</b></a>.
  </LI>

  <li><b><a href="http://vega.sanger.ac.uk/Danio_rerio/">Zebrafish Genome Annotation Browser (Vega)</a>:</b><br>
This site allows browsing of zebrafish BAC and PAC clones that were sequenced, analysed and manually annotated within the Sanger zebrafish genome project. Vega includes
    <ul type=circle>
      <li> manually annotated genes with poly A features
      <li> protein feature and domain annotation
      <li> Blast searches against all zebrafish ESTs, cDNAs and the WashU and IMCB EST cluster</li>
      <li> Blast searches against all vertebrate proteins, cDNAs and ESTs</li>
      <li> mapping of unists markers</li>
      <li> ab initio gene predictions</li>
      <li> tRNA prediction</li>
      <li> CpG island prediction and repeat masking</li>
    </ul>
You can search Vega with <a href="http://vega.sanger.ac.uk/Multi/blastview?species=Danio_rerio"><b>Blast</b></a>.  For Blast searches against all finished and unfinished clones go <a href="http://www.sanger.ac.uk/cgi-bin/blast/submitblast/d_rerio"><b>here</b></a>.<br>
The manual annotation is a joined effort between the Sanger Institute (gene structure annotation) and ZFIN (nomenclature).
 </lI>
  <li><a href="http://www.ensembl.org/Danio_rerio/"><b>Zebrafish Genome Browser at Ensembl</b></a><b>:</b><br>
  This site allows browsing of the latest zebrafish whole genome assembly <a href="http://www.sanger.ac.uk/Projects/D_rerio/Zv4_assembly_information.shtml"><b>Zv4</b></a>
    <ul type=circle>
      <li> automatically annotated Ensembl genes</li>
      <li> automatically annotated Ensembl EST genes</li>
      <li> protein feature and domain annotation</li>
      <li> Blast searches against all zebrafish ESTs, cDNAs and the WashU and IMCB EST clusters</li>
      <li> Blast searches against all vertebrate proteins, Uniprot, all vertebrate cDNAs and ESTs</li>
      <li> mapping of unists markers</li>
      <li> mapping of the Affymetrix zebrafish probes</li>
      <li> mapping of SNps in dbSNP</li>
      <li> comparative analysis with other Ensembl species (human, mouse, rat, chicken, fugu, tetraodon)</li>
      <li> BAC end placement</li>
      <li> transcription start site predictions with Eponine</li>
      <li> ab initio gene predictions</li>
      <li> tRNA prediction</li>
      <li> CpG island prediction and repeat masking</li>
    </ul>
Additional DAS tracks with SNP and indel analysis derived from the comparison of the whole genome shotgun reads are available. Ensembl also offers completely customizable downloads with <a href ="http://www.ensembl.org/Multi/martview?species=Danio_rerio"><b>EnsMart</b></a><br>
You can search Ensembl with either <a href="http://www.ensembl.org/Multi/blastview?species=Danio_rerio"><b>Blast</b></a> or <a href="http://www.ensembl.org/Multi/blastview?species=Danio_rerio"><b>SSAHA</b></a>.
  </li>
  <li>Additional services available at the Sanger Institute
      <ul type=circle>
        <li> The <a href="http://www.sanger.ac.uk/Projects/D_rerio/WebFPC/zebrafish/small.shtml"><b>WebFPC database</b></a> hosts the physical map and allows searching and browsing for clone information.</li>
        <li> The <a href="http://www.sanger.ac.uk/cgi-bin/Projects/D_rerio/mapsearch"><b>mapsearch site</b></a> offers easy access to all mapping related data (mapped ESTs, markers, BAC ends) in the current assembly.</li>
        <li> The <a href="http://www.sanger.ac.uk/Projects/D_rerio/fishmask.shtml"><b>RepeatMasker</b></a> service masks your sequence data with a choice of repeat libraries.</li>
       </ul>
   </li>

 <li><b><a href="http://genome.ucsc.edu">Zebrafish Genome Browser at UCSC</a>:</b><br>
    A joint collaboration between UCSC and the Zebrafish Genome Initiative at Children's Hospital Boston provides several useful tracks of genome data, including: 
      <br>
<ul type=circle><li>Mapping & Sequencing:  Position / Conigs / Scaffolds / RH Map / BAC End Pairs & Singles / Gap / GC Percent</li>
     <li>Genes & Gene Prediction:  RefSeq Genes / Ensembl Genes</li>
     <li>mRNA & EST:  ZF mRNAs & ESTs / spliced ESTs / WZ EST Clusters / nonZF mRNAs</li>
     <li>Expression & Regulation:  Affy ZF</li>
     <li>Comparative Genomics:  Fugu Chain & Net / Human Proteins / HG Blast</li>
     <li>Variation & Repeats:  RepeatMasker / SimpleRepeats</li></ul>
     Custom annotation tracks for the zebrafish research community include: 
     <ul type=circle>
       <li> RH Map (about 9K sequences)</li>
       <li> BAC End Clones (over 288K pairs, 23K singletons)</li>
       <li> WZ ESTs (43K sequences, 76K alignments, 39K EST alignments)</li>
       <li> Affymetrix Zebrafish Microarray Probe Alignments (15K probes)
       <li> Human Protein (over 40K alignments)</li>
     </ul>

  <li><b><a href="http://danio.mgh.harvard.edu/blast/blast_grp.html">Zebrafish BLAST 
    server at MGH</a>:</b> <br>
    MGH site provides a BLAST search of a gene or marker 
    of interest to the assembled sequence contigs of 
    the zebrafish genomic shotgun sequence from the 
    Sanger project. </li>

 <li><b><a href="http://danio.mgh.harvard.edu/markers/ssr.html">MGH web-based tool for automating primer selection</a>:</b><br>
    MGH site offers this tool which searches for a Zv3 supercontig or BAC clones by name, derives 
all di- and tri- nucleotide repeat sequences in the supercontig, masks the
sequence around the repeats for larger zebrafish repetitive sequences and then designs unique sequence primer pairs to amplify the simple sequence repeats. These are potentially polymorphic simple sequence repeats (SSRs) that can be used in genetic mapping/positional cloning projects.
  
  </li>
 <li><b><a href="http://134.174.23.160/HumanblastZebrafish/">BLAST 
    hits of Human Genes from Sanger sequence maintained by Zon lab</a>: </b><br>
    Each human gene annotated on the Santa Cruz site 
    was compared by BLAST against the Sanger Institute 
    zebrafish genome sequence. These data can be used 
    to help identify zebrafish orthologues of human 
    or Fugu genes. </li>
  <li>
  <b><a href="http://wwwmap.tuebingen.mpg.de/">Tübingen Map of the Zebrafish Genome</a> </b><br>
  </li>
  <li><b><a href="http://zfin.org/cgi-bin/ZFIN_jump?record=JUMPTOREFCROSS">RH, 
    meiotic and integrated maps of the genome</a>: </b><br>
    Centralized site for all meiotic and radiation hybrid 
    maps of the zebrafish genome. </li>
   <li><b><a href="http://www.ncbi.nlm.nih.gov/genome/guide/zebrafish/index.html">Zebrafish Genome Resources at NCBI</a></b><br>
</ul>
<b><font size="+1">cDNA and EST Resources</font></b>
<p>
<ul>
  <li><a href="http://zgc.nci.nih.gov/"><strong>Zebrafish Gene
  Collection (ZGC)</strong></a><b>:</b> NIH initiative for production of
  full length sequences and cDNA clones.</li>
 <li><a href="http://www.genetics.wustl.edu/fish_lab/frank/cgi-bin/fish/"><b>Washington 
    University EST sequencing project</b></a><b>: </b><br>
    This site hosts the project in which over 200,000 
    ESTs have been sequenced. It offers a unigene-like 
    set (&#145;wz&#146; ESTs) that can be found by key 
    word search or BLAST. </li>
  <li><b><a href="http://www.tigr.org/tdb/tgi/zgi/">TIGR  Zebrafish Gene Index (ZGI)</a>:</b><br>
    The TIGR Zebrafish Gene Index (ZGI) integrates research data from international zebrafish gene research projects and represents a non-redundant view of all zebrafish genes and data on their expression patterns, cellular roles, functions, and evolutionary relationships.</li>
  <li><b><a href="http://www.genoscope.cns.fr/zie/">Zebrafish Inner Ear ESTs Database</a>:</b><br>
   This database consists of 4694 contigs or singletons obtained from ~18000 reads of cDNAs that were isolated from the microdissected otocysts of zebrafish embryos at 20-30 hours postfertilization, following subtraction with a pool of liver cDNAs from adult fishes.
  <li><b><a href="http://bioweb.imcb.nus.edu.sg/zest.htm">Zebrafish EST Database at Singapore</a></b><br>
</ul>

<b><font size="+1">Mutants and Stocks</font></b>
<p>
  <ul>
   <li><b><a href="/zirc/home/guide.php">Zebrafish International Resource Center (ZIRC)</a></b><br></li>

   <li><b><a href="http://www.eb.tuebingen.mpg.de/services/stockcenter/zebraf_stockcenter.html">Tübingen Zebrafish stockcenter</a></b><br></li>

  </ul>


</body>

<script language="JavaScript" src="/footer.js"></script>
