<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html>
	<head>
		<title>
			ZFIN Genomic Resources
		</title>
	</head>

<script language="JavaScript" src="/javascript/header.js"></script>



<h1 align="center"><a name="map">Genomic Resources for Zebrafish</a></h1>


<ul>
<a href="#map">Genome and Map Resources</a> <br>
<a href="#primers">Primers</a> <br>
<a href="#micro">Microarrays</a> <br>
<a href="#cDNA">cDNA and EST Resources</a> <br>
<a href="#stock">Stock Centers</a> <br>
<a href="#muta">Mutagenesis Projects</a> <br>
<a href="#traps">Gene and Enhancer Traps</a> <br>
<a href="#morph">Morpholinos</a> <br>
<a href="#paths">Pathways</a> <br>
<a href="#paths">Neurobehavioral / Physiological Models</a> <br>
<a href="#papers">White Papers</a> <br>
</ul>


<p></p>
<b><font size="+1">Genome and Map Resources</font></b>
<ul>
  <li><b><a href="http://www.sanger.ac.uk/Projects/D_rerio/">The Danio rerio Sequencing Project at Sanger</a>:</b><br>
The <a href="http://www.sanger.ac.uk/Projects/D_rerio/"><b>Sanger zebrafish homepage</b></a> hosts all of the efforts of the zebrafish genome sequencing project, including the <a href="http://www.sanger.ac.uk/Projects/D_rerio/wgs.shtml"><b>whole genome sequencing and assembly project</b></a> with automated annotation in <a href="http://www.ensembl.org/Danio_rerio/"><b>Ensembl</b></a> and the <a href="http://www.sanger.ac.uk/Projects/D_rerio/clones.shtml"><b>clone mapping and sequencing project</b></a> with manual annotation in <a href="http://vega.sanger.ac.uk/Danio_rerio/"><b>Vega</b></a>.
  </li>
  <br>
  <li><b><a href="http://vega.sanger.ac.uk/Danio_rerio/">Zebrafish Genome Annotation Browser (Vega)</a>:</b><br>
This site allows browsing of zebrafish BAC and PAC clones that were sequenced, analysed and manually annotated within the Sanger zebrafish genome project. Vega includes
    <ul type=circle>
      <li> manually annotated genes with poly A features
      <li> protein feature and domain annotation
      <li> Blast searches against all zebrafish ESTs, cDNAs and the WashU and IMCB EST cluster</li>
      <li> Blast searches against all vertebrate proteins, cDNAs and ESTs</li>
      <li> mapping of unists markers</li>
      <li> ab initio gene predictions</li>
      <li> tRNA prediction</li>
      <li> CpG island prediction and repeat masking</li>
    </ul>
You can search Vega with <a href="http://vega.sanger.ac.uk/Multi/blastview?species=Danio_rerio"><b>BLAT</b></a>.  For Blast searches against all finished and unfinished clones go <a href="http://www.sanger.ac.uk/cgi-bin/blast/submitblast/d_rerio"><b>here</b></a>.<br>
The manual annotation is a joined effort between the Sanger Institute (gene structure annotation) and ZFIN (nomenclature).
 </lI>
 <br> <li><a href="http://www.ensembl.org/Danio_rerio/"><b>Zebrafish Genome Browser at Ensembl</a>:</b><br>
  This site allows browsing of the latest zebrafish whole genome assembly <b><a href="http://www.ensembl.org/Danio_rerio/Info/Index">Zv9</a></b>
  <ul type=circle>
      <li> automatically annotated Ensembl genes</li>
      <li> automatically annotated Ensembl EST genes</li>
      <li> protein feature and domain annotation</li>
      <li> Blast searches against all zebrafish ESTs, cDNAs and the WashU and IMCB EST clusters</li>
      <li> Blast searches against all vertebrate proteins, Uniprot, all vertebrate cDNAs and ESTs</li>
      <li> mapping of unists markers</li>
      <li> mapping of the Affymetrix zebrafish probes</li>
      <li> mapping of SNps in dbSNP</li>
      <li> comparative analysis with other Ensembl species (human, mouse, rat, chicken, fugu, tetraodon)</li>
      <li> BAC end placement</li>
      <li> transcription start site predictions with Eponine</li>
      <li> ab initio gene predictions</li>
      <li> tRNA prediction</li>
      <li> CpG island prediction and repeat masking</li>
      </ul>

 <br>Additional DAS tracks with SNP and indel analysis derived from the comparison of the whole genome shotgun reads are available. Ensembl also offers completely customizable downloads with <a href ="http://www.ensembl.org/biomart/martview?species=Danio_rerio"><b>EnsMart</b></a><br>
You can search Ensembl with <a href="http://www.ensembl.org/Multi/blastview?species=Danio_rerio"><b>BLAT</b></a>.
  </li>
  <br> <li>Additional services available at the Sanger Institute
      <ul type="circle">
        <li> The <a href="http://www.sanger.ac.uk/Projects/D_rerio/WebFPC/zebrafish/small.shtml"><b>WebFPC database</b></a> hosts the physical map and allows searching and browsing for clone information. Please note that this database is not updated anymore. For up to date information on the clone path please check <a href="http://www.sanger.ac.uk/cgi-bin/humpub/chromoview">chromoview</a></li>

        <li> The <a href="http://www.sanger.ac.uk/Projects/D_rerio/fishmask.shtml"><b>RepeatMasker</b></a> service masks your sequence data with a choice of repeat libraries.</li>
	<li> The <a href="http://microrna.sanger.ac.uk/"><b>miRBase</b></a> service is the new home for microRNA data, incorporating the database  and gene naming roles previously provided by the miRNA Registry, and including the new miRBase Target database.
       </ul>
   </li>

  <br><li><b><a href="http://genome.ucsc.edu/cgi-bin/hgGateway?hgsid=85282730&clade=vertebrate&org=Zebrafish&db=0">Zebrafish Genome Browser at UCSC</a>:</b><br>
    A joint collaboration between UCSC and the Zebrafish Genome Initiative at Children's Hospital Boston provides several useful tracks of genome data, including:
      <br>
<ul type="circle"><li>Mapping & Sequencing:  Position / Conigs / Scaffolds / RH Map / BAC End Pairs & Singles / Gap / GC Percent</li>
     <li>Genes & Gene Prediction:  RefSeq Genes / Ensembl Genes</li>
     <li>mRNA & EST:  ZF mRNAs & ESTs / spliced ESTs / WZ EST Clusters / nonZF mRNAs</li>
     <li>Expression & Regulation:  Affy ZF</li>
     <li>Comparative Genomics:  Fugu Chain & Net / Human Proteins / HG Blast</li>
     <li>Variation & Repeats:  RepeatMasker / SimpleRepeats</li></ul>
      <br>Custom annotation tracks for the zebrafish research community include:
     <ul type="circle">
       <li> RH Map (about 9K sequences)</li>
       <li> BAC End Clones (over 288K pairs, 23K singletons)</li>
       <li> WZ ESTs (43K sequences, 76K alignments, 39K EST alignments)</li>
       <li> Affymetrix Zebrafish Microarray Probe Alignments (15K probes)
       <li> Human Protein (over 40K alignments)</li>
     </ul>



   <br><li><b><a href="http://www.ncbi.nlm.nih.gov/genome/guide/zebrafish/index.html">Zebrafish Genome Resources at NCBI</a></b><br>

This page provides information about genomic resources available from NCBI and the zebrafish research community, including sequence, mapping, clone information, and NCBI's independent gene annotation.<br><br>


    <li><b><a href="http://fishmap.igib.res.in/">FishMap Zebrafish Genomics Knowledgebase:</a></b><br>

    A community resource for Zebrafish Genomics.
    <br> <br>
    Bhartiya D, Maini J, Sharma M, Joshi P, Laddha SV, Jalali S, Patowary
A, Purkanti R, Lalwani M, Singh AR, Chauhan R, Singh N, Bhardwaj A,
Scaria V, Sivasubbu S.
FishMap Zv8 update--a genomic regulatory map of zebrafish.
Zebrafish. 2010 Jun;7(2):179-80.
<br><br>
Meli R, Prasad A, Patowary A, Lalwani MK, Maini J, Sharma M, Singh AR,
Kumar G, Jadhav V, Scaria V, Sivasubbu S.
FishMap: a community resource for zebrafish genomics.
Zebrafish. 2008 Summer;5(2):125-30.
 <br> <br>



</li>


    <li><b><a href="http://zfblast1.danio.tchlab.org/compGenomics.htm">BLAST
    hits of Human Genes from Sanger sequence maintained by Zon lab</a>:</b><br>
    Each human gene annotated on the Santa Cruz site
    was compared by BLAST against the Sanger Institute
    zebrafish genome sequence. These data can be used
    to help identify zebrafish orthologues of human
    or Fugu genes. </li>

     <br> <li><b><a href="http://danio.mgh.harvard.edu/blast/blast_grp.html">Zebrafish BLAST
    server at MGH</a>:</b> <br>
  MGH site provides a BLAST search of a gene or marker of interest to the assembled sequence contigs of the zebrafish genomic shotgun sequence from the Sanger project. </li><br>


  <li>
  <b><a href="http://wwwmap.tuebingen.mpg.de/">T&uuml;bingen Map of the Zebrafish Genome</a> </b><br>
  </li>

  <br> <li><b><a href="/<!--|CGI_BIN_DIR_NAME|-->/ZFIN_jump?record=JUMPTOREFCROSS">RH,
    meiotic and integrated maps of the genome</a>: </b><br>
    Centralized site for all meiotic and radiation hybrid
    maps of the zebrafish genome. </li>
</b><br>


      </ul>

   <b><a name="primers">Primers</a></b><br>
<ul>
<br><li><b><a href="http://www.molecularstation.com/forum/zebrafish-forum/21360-automatic-zebrafish-specific-pcr-primer-design-service.html">Automatic zebrafish-specific PCR primer design service</a>:</b><br>
 This service provides "one stop shopping" for picking primers, BLAST of primer
sequences for specificity against the latest zebrafish genome sequence,
and graphical views of the locations of primers with different confidence
levels. All you need is a few clicks. No more cutting and pasting from one
website to another. The program flows automatically. Detailed information
for use of this program can be found in the "general info" and "help file".</li>


<br> <li><b><a href="http://danio.mgh.harvard.edu/markers/ssr.html">MGH web-based tool for automating primer selection</a>:</b><br>
    MGH site offers this tool which searches for a supercontig, scaffold, or finished BAC  clone by name or accession number, derives all di- and tri- nucleotide repeat sequences in the supercontig, masks the
sequence around the repeats for larger zebrafish repetitive sequences and then designs unique sequence primer pairs to amplify the simple sequence repeats. These are potentially polymorphic simple sequence repeats (SSRs) that can be used in genetic mapping/positional cloning projects. The site also offers a tool to <a href="http://danio.mgh.harvard.edu/userMarkers/usrInputSsr.html"> process your sequence for simple repeats/primer generation</a>
  </li>
  </ul>




 <br><b><a name="micro">Microarrays</a></b><br>

   <ul>
<br><li><B><a href="http://134.174.23.160/zfaca/hash/master020106public.aspx">Zebrafish Affy Chip Annotation Project:</a></b><br>
The Trans-NIH Zebrafish Genome Initiative at Children's Hospital Boston aims to annotate zebrafish genes with human known genes for functional and comparative studies.  This website provides a new web service for retrieving hand-annotated zebrafish genes represented on the Affymetrix Zebrafish Genome Array to the research community.  A total of 11074 probesets from the complete set of 15617 (115 of which are controls) have been manually annotated with human known genes (UCSC HG16). We are currently manually annotating the remaining probesets which have putative Zv6 Genscan human-zebrafish orthologs. We will continue annotating as new sequence data become available and by comparative genomics approaches. Researchers can choose to download the entire annotation as a text file, or retrieve their specific gene probes of interest by submitting a list of Affymetrix probeset IDs of interest.

 <p>The latest release is now available at  <a href="http://134.174.23.160/zfACA/hash/cumulative_expanded.aspx">http://134.174.23.160/zfACA/hash/cumulative_expanded.aspx</a>.
</li>

  <li><b><a href="http://titlis.umassmed.edu/danio/">University of Massachusetts Medical School Zebrafish Microarray Database</a>: </b><br>

To access microarray data type username: "guest", password: "password"; once you have logged on, click on "Analysis" in the upper right hand corner.   <a href="http://zfin.org/cgi-bin/ZFIN_jump?record=ZDB-PUB-060927-11">Covassin, L., Amigo, J.D., Suzuki, K., Teplyuk, V., Straubhaar, J., and Lawson, N.D. (2006) Global analysis of hematopoietic and vascular endothelial gene expression by tissue specific microarray profiling in zebrafish. Dev. Biol. 299(2): 551-562</a>
  </li>

<br><li><b><a href="http://serine.umdnj.edu/~ouyangmi/cgi-bin/zebrafish/profile.htm">DNA Microarray Data from Amacher, Farber and Ho labs</a></b><br>
</li>

<br><li><b><a href="http://zf-espresso.tuebingen.mpg.de/search.php">Microarray expression at ZF-Espresso</a></b><br>
</li>

<br> <li><b><a href=" http://microarray.no">Zebrafish Oligonucleotide Arrays</a>:</b><br>
The Norwegian Microarray Consortium offers their array products for academic use to customers world wide. One limitation is oligonucleotide arrays, which cannot provided to users in the US and Japan, due to licensing conditions. A moderate license fee will be added to the base price for oligonucleotide arrays sold to users in the UK, France, Germany, Italy, Austria, Belgium, Sweden, Switzerland, Luxembourg, and The Netherlands.</li>

 <br><li><b><a href="http://giscompute.gis.a-star.edu.sg/~govind/zebrafish/version2/">Zebrafish Compugen microarray annotation database:</a></b><br>

Compugen array clones are annotated at this site.  The GenBank id of genes in the Compugen  oligo list can be used to access information from  Unigene, Entrez Gene, Gene Ontology, human and mouse HomoloGene, Unigene protein similarities of mouse and human, ZGI index.  The gene chromosomal location is also available.   The details of the microarray annotation database are published in PLoS Genetics  (2005).</li></ul>

</ul>



<br><b><a name="cDNA">cDNA and EST Resources</a></b>
<p>
<ul>
  <li><a href="http://zgc.nci.nih.gov/"><strong>Zebrafish Gene
  Collection (ZGC)</strong></a><b>:</b> NIH initiative for production of
  full length sequences and cDNA clones.</li>

 <b> 1819 ZGC clones recloned in pENTR223.1</b>

 In May 2008, ZGC has added 1819 full length ZGC clones recloned into a Gateway Entry vector (<a href="http://mgc.nci.nih.gov/Vectors/prot_pENTR223.1">pENTR223.1;</a>) for expression analysis and other applications.<br>

These clones can be found by searching Entrez nucleotide for "NIH_ZGC_36".<br>


<br>  <li><b><a href="http://compbio.dfci.harvard.edu/tgi/cgi-bin/tgi/gimain.pl?gudb=zfish">DFCI  Zebrafish Gene Index (ZGI)</a>:</b><br>
    The DFCI Zebrafish Gene Index (ZGI) integrates research data from international zebrafish gene research projects and represents a non-redundant view of all zebrafish genes and data on their expression patterns, cellular roles, functions, and evolutionary relationships.</li>

<br>  <li><b><a href="http://www.genoscope.cns.fr/zie/">Zebrafish Inner Ear ESTs Database</a>:</b><br>
   This database consists of 4694 contigs or singletons obtained from ~18000 reads of cDNAs that were isolated from the microdissected otocysts of zebrafish embryos at 20-30 hours postfertilization, following subtraction with a pool of liver cDNAs from adult fishes.

<br><br><li><b><a href="http://www.openbiosystems.com/GeneExpression/Non-Mammalian/Fish/ZebrafishIMCB/">Zebrafish EST Database at Singapore</a></b><br>
</ul>

<b><a name="stock">Stock Centers</a></b>
<p>
  <ul>
   <li><b><a href="/zirc/">Zebrafish International Resource Center (ZIRC)</a></b><br></li>

  <br> <li><b><a href="http://www.eb.tuebingen.mpg.de/core-facilities/zebrafish-stockcenter">T&uuml;bingen Zebrafish stockcenter</a></b><br></li>

  </ul>

  <b>
<a name="muta">Mutagenesis Projects</a></b>
<p>
<ul>

  <li><b>
   Targeted mutations using zinc-finger nucleases (ZFNs): </b>The <a href="http://lawsonlab.umassmed.edu/">Lawson</a> and <a href="http://labs.umassmed.edu/WolfeLab">Wolfe</a> labs have compiled resources relating to their recent <a href="http://zfin.org/cgi-bin/webdriver?MIval=aa-pubview2.apg&amp;OID=ZDB-PUB-080527-25">publication</a> on targeted gene inactivation using ZFNs:

<p>A list of plasmids and strains (available at <a href="http://www.addgene.org/">addgene):</a><br>
<a href="http://lawsonlab.umassmed.edu/PDFs/B1Hrec.pdf">http://lawsonlab.umassmed.edu/PDFs/B1Hrec.pdf</a>

<p>Help on getting started with <a href="http://www.ncbi.nlm.nih.gov/pubmed/18332042?dopt=Abstract">bacterial 1-hybrid screens</a>:<br>

<a href="http://lawsonlab.umassmed.edu/PDFs/B1HStart.pdf">http://lawsonlab.umassmed.edu/PDFs/B1HStart.pdf</a><br>

<p>An on-line search tool to identify and design library oligos:<br>
<a href="http://pgfe.umassmed.edu/ZFPsearch.html">http://pgfe.umassmed.edu/ZFPsearch.html</a><br>

<li><b>
   <a href="http://www.sanger.ac.uk/Projects/D_rerio/zmp/">Zebrafish Mutation Resource</a> (Sanger Institute):</b>
This project seeks to identify, phenotype and distribute a large number of mutant zebrafish lines. Mutations are identified from populations of chemically mutagenised animals using PCR of specific exons followed by re-sequencing. Mutants are analysed for morphological and molecular differences and distributed to the community. This is an open resource and is now open to requests from the community. The <a href="http://www.sanger.ac.uk/Projects/D_rerio/zmp/">Zebrafish Mutation Resource Tracking page</a> lists the mutants generated by the Sanger <a href="http://www.sanger.ac.uk/Projects/D_rerio/zmp/">Zebrafish Mutation Resource</a> and the  <a href="http://www.zf-health.org/workpackages/wp4.html">ZF-HEALTH consortium</a>.

  <br><br><li><b><a href="
http://labs.fhcrc.org/moens/Tilling_Project.html">Moens Zebrafish TILLING Project</a></b> The Moens lab has generated a cryopreserved library of 8,640 ENU-mutagenized zebrafish which we screen re-iteratively for induced mutations in genes of interest to the zebrafish community, since 2009 as part of the <a href="https://webapps.fhcrc.org/science/tilling/">Zebrafish TILLING Consortium</a>. These mutants are made available to the community from the Moens lab or through the Zebrafish International Resource Center (ZIRC).

   <br><br><li><b><a href="https://webapps.fhcrc.org/science/tilling/">Zebrafish TILLING Consortium</a></b> The Zebrafish TILLING Consortium is a joint effort of the Moens lab at the Fred Hutchinson Cancer Research Center, the Solnica-Krezel lab at Vanderbilt University, and the Postlethwait lab at the University of Oregon. The consortium-based approach will be to screen genes of interest sequentially at the three locations. Go to this site to request genes to be screened by the consortium.

  <br><br><li><b><a href="http://web.mit.edu/hopkins/">Hopkins lab large insertional mutagenesis screen database:</a></b><br></li>
   Detailed information on the mutants resulting from the large insertional mutagenesis screen described in Amsterdam, et al. (2004) Identification of 315 genes essential for early zebrafish development. <a href="/cgi-bin/ZFIN_jump?record=ZDB-PUB-040708-1">Proceeding of the National Academy of Sciences 101: 12792-12797</a> </b><br></li>


</ul>




  <b>
<a name="traps">Gene and Enhancer Traps</a></b>
<p>
   <ul>
   <li><b><a href="http://kawakami.lab.nig.ac.jp/pattern/">zTrap</a></b> The Zebrafish Gene Trap and Enhancer Trap Database at the Kawakami laboratory.
</ul>


 <b><a name="morph">Morpholinos</a></b>
  <p>
  <ul>

  <li><b><a href="http://www.morpholinodatabase.org/">Morpholino Database</a></b><br></li>
  <br>
   <li><b><a href="moqc.pdf">Quality control testing of Gene Tools morpholinos.</a></b><br></li>

   </ul>


<b><a name="paths">Pathways</a></b>

  <p>
  <ul>
   <li><b>
  <a href="http://www.wikipathways.org/index.php/WikiPathways">Zebrafish Pathways at WikiPathways</a>.</b> This site allows browsing, editing, downloading and publishing of zebrafish pathways that are annotated for computational analysis and illustrated for human consumption.

<ul>
  <li>includes metabolic, signaling, regulatory and disease pathways
  <li>browse and search for pathways by title or gene content
  <li>download pathways as images for presentations
  <li>download pathway for data visualization and analysis
  <li>edit pathways using a built-in drawing tool
  <li>publish new pathways based on your knowledge and research
  <li>add literature references to important pathway components or steps
  <li>discuss, share and collaborate on pathways with colleagues globally
  <li>compete with others to curate the highest ranking pathways
</ul>
 <p>WikiPathways is also accessible programmatically through a <a href="http://www.wikipathways.org/index.php/Help:WikiPathways_Webservice">web service API</a>, which supports search, download, and data mapping.


   </ul>

   <p></p>
<b><a name="papers">Neurobehavioral / Physiological Models</a></b>
<p>
<ul>
  <li><b><a href="http://www.tulane.edu/~znpindex/ ">Zebrafish Neurophenome Project (ZNP)</a>:</b><br>
ZNP is a comprehensive resource for neurobehavioral and physiological data of adult zebrafish models. ZNP incorporates validated and curated data from work published in this field, to improve the accessibility of current knowledge to researchers interested in using adult zebrafish models.
  </ul>



  <b><a name="papers">White Papers</a></b>
  <p>
  <ul>
   <li><b><a href="Zebrafish_disease_project.pdf">The Zebrafish & Disease Project: Zebrafish as a model system to study and cure human diseases</a></b><br></li>

   </ul>
<script language="JavaScript" src="/javascript/footer.js"></script>
