<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html>
	<head>
		<title>
			ZFIN Zebrafish Breakout Group
		</title>
	</head>

<script language="JavaScript" src="/javascript/header.js"></script>

&nbsp;
<table COLS=2 WIDTH="600" >
<tr>
<td WIDTH="210"><b><i><font size=+4>Catch of the day...</font></i></b>
<center><font face="Times New Roman,Times"><font size=+1>News from the
zebrafish research community</font></font></center>
</td>

<td WIDTH="100"><img SRC="catch.gif" BORDER=0 height=67 width=210></td>
</tr>
</table>

<hr WIDTH="100%"><b>Report from the NIH Non-Mammalian Models Meeting, February
16, 1999</b>
<br>
<hr size="5">
<center>
<h1>
<b>Zebrafish Breakout Group</b></h1></center>

<center>
<h1>
<b>Recommendations</b></h1></center>

<hr size="5">The zebrafish is a powerful model system for the genetic analysis
of vertebrate embryogenesis, organ development, and disease. Its unique
power is its tractable, phenotype driven mutation screens and readily accessible
transparent embryos. Because of its facile forward genetics, zebrafish
accelerates gene discovery; because of its accessible embryos, it promotes
deep understanding of gene function; because of its phylogenetic position,
it informs mechanisms of genome conservation. More than 2000 mutants are
currently available in more than 600 genes, and new phenotypic screens
are uncovering more mutations in many laboratories.
<p>The zebrafish system relies on the ability to clone and characterize
mutant genes rapidly. Positional cloning and insertional mutagenesis have
been successfully utilized to identify the molecular nature of mutations,
but candidate gene approaches have been the most successful to-date. The
infrastructure of the community and the recent Zebrafish Genome Initiative
have added value to the system by shortening the time to gene isolation,
and with the addition of a few more reagents and services, the cloning
of mutations could become routine. New investigators are becoming interested
in zebrafish and a mission of the current community is to make the system
more easily accessible and tractable. The current RFA and implementation
of the suggestions listed below for additional support will have significant
impact on investigator-driven research in the field.
<h2>
<b>A. Current RFA</b></h2>
In 1998, the NIH issued a Request for Applications to improve genomic resources
for zebrafish. Five projects were funded. These three year projects have
begun to vastly increase the number of expressed sequence tags and the
number of mapped expressed sequences and anonymous markers, and are providing
a collection of chromosomal deletions that cover the entire genome. The
funded projects are:
<ul>
<li>
100,000 ESTs 5' and 3' sequences</li>

<li>
10,000 ESTs mapped to RH panel</li>

<li>
3,000 ESTs mapped meiotically</li>

<li>
0.5 cM microsatellite map, 5000 microsatellites</li>

<li>
Deletion panel to cover the genome</li>
</ul>

<h2>
<b>B. More Infrastructure Needed for Genomics</b></h2>
<b><u>I. ESTs</u>. </b>A large collection of ESTs benefits all strategies
for the molecular identification of mutations. For positional cloning,
sample sequencing of clones in the critical interval can reveal sequence
identity to ESTs. For insertional mutagenesis, tags isolated from genomic
DNA adjacent to inserts can reveal overlap with ESTs, thereby reducing
cDNA library screening. For the candidate approach, the greater the number
of ESTs to serve as candidates, the greater the likelihood that a mutated
gene is represented by a known EST. Knowing the genetic location and expression
pattern further allows potential candidate genes to be ruled out rapidly,
thereby facilitating cloning of mutations. Therefore, as high priority,
our community needs more EST sequences, and more ESTs characterized by
genetic location and expression pattern.
<ul>
<li>
Make new, normalized cDNA libraries from multiple stages and tissues. ($150,000/year
x 3 years).</li>

<li>
Sequence 5' and 3' ends of 100,000 cDNAs. ($400,000/year x 3 years.)</li>

<li>
Map the ESTs to compare to mutant map positions. ($160/EST X 5,000/year
X 3 years = $2,400,000).</li>

<li>
Determine the expression pattern of the ESTs for comparison to mutant phenotypes.
$150,000/year X 3 years for 10,000 cDNAs. Make these ready for the Zebrafish
Gene Expression Atlas.</li>

<li>
Develop and make widely available microarrays of 10,000-20,000 ESTs. ($150,000/year
X 3 years.)</li>
</ul>
<b><u>II. Stock Center</u>. </b>The power of zebrafish is the collection
of mutations isolated on the basis of mutant phenotypes. Currently, laboratories
are unable to keep all the mutations they recover for lack of space and
funds for maintenance. Requests for the distribution of mutant stocks burdens
all laboratories that have generated substantial mutant collections. Most
mutations that are not of immediate interest must be maintained as frozen
sperm. This hinders research because it requires at least three months
to obtain homozygous embryos starting from frozen sperm, and stocks adequate
for experimentation can take up to 9 months from the thawing of sperm.
Clearly a stock center for the preservation and distribution of mutant
stocks is key to the rapid analysis of zebrafish mutants. NIH and State
of Oregon funds have been obtained to support the construction of a stock
center in Eugene and to partially fund the preservation and distribution
of zebrafish stocks. More funds are required to maintain all of the current
stocks, not to mention the many new mutations made by new, innovative and
directed mutations screens.
<ul>
<li>
Stock Center: Additional personnel needed. $250K/year additionally.</li>
</ul>
<b><u>III. Data Base.</u></b> Ready access of the entire community to genomic
and developmental data is required to turn data into understanding. The
zebrafish database ZFIN is positioned to become a tool for archiving and
retrieving data on zebrafish. Links must be made between the zebrafish
database and the database of other organisms. A substantial increase in
database support is needed to turn ZFIN into the tool the community requires.
<ul>
<li>
Data editors to annotate expression and phenotypic data from old and current
papers. ($500,000/year X 3 years.)</li>

<li>
Data base: Need bioinformaticists to expand capabilities to include conservation
of syntenies, histological atlas, and Zebrafish Gene Expression Atlas.
($500,000/year X 3 years.)</li>
</ul>
<b><u>IV. Mutant mapping.</u></b> One of the first steps to the molecular
characterization of a zebrafish mutation is its genetic mapping. This rules
out many potential EST candidates, and "rules in" others, as well as providing
an initial stage for positional cloning. It is important to map quickly
mutations in each gene mutated in the two large screens, 600 mutations,
by a mechanism that guarantees that mutant locations are rapidly made public.
Only about 150 of these have been mapped to date.
<ul>
<li>
Mapping mutations. ($2000/Mutation x 500 mutations = $1,000,000 over 3
years.)</li>
</ul>
<b><u>V. Physical Map of Genome.</u></b> DNA pooling methods and the huge
number of anonymous polymorphisms available by microsatellites, RAPDs,
and AFLPs permit the identification of DNA polymorphisms very closely linked
to zebrafish mutations. Chromosome walks from those linked markers requires
large insert DNA libraries, and is facilitated by a physical map of the
genome. A physical map that includes telomeres furthermore provides reagents
for genome sequencing.
<ul>
<li>
BAC libraries, 200 kb clones: ($250,000 for 10-genome coverage.)</li>

<li>
BACs contiged and anchored to the genetic map. ($1,000,000/year x 3 years.)</li>

<li>
YAC telomere library. Knowing the end of the chromosomes will aid positional
cloning of genes in these regions. ($150,000)</li>
</ul>
<b><u>VI. Genome Sequencing.</u></b> Complete knowledge of the sequence
of the zebrafish genome would provide a tool to revolutionize our efforts
to understand zebrafish biology.
<p>Phase 1, start sequencing of the genome. $3M first year. Initially science
driven as to regions to start sequencing, focussing on regions of highest
interest and regions orthologous to those already sequenced in other important
model systems such as human and pufferfish. This will define repeat structure
of the genome and will provide information on the structure of intron-exon
boundaries.
<p>Phase 2. Ramping up to sequence the entire genome beginning in 2000
with targeted completion 2008.
<p><b><u>VII. Technology development.</u></b> Although zebrafish has tremendous
advantages as a system for forward genetics, the development of several
additional methodologies would further improve the system. Key among these
needs is gene knockout technology and the construction of more transgenic
zebrafish. $1.5m/y
<ul>
<li>
ES cells</li>

<li>
transgenesis to provide more GFP lines</li>

<li>
Gal4 UAS system to investigate gene interactions</li>

<li>
RNAi to eliminate gene functions</li>
</ul>

<h2>
<b>C. More Infrastructure Needed for Cell Biology</b></h2>
<u>I. Imaging.</u> The beauty of zebrafish lies in its optically clear
embryos and their accessibility to experimental intervention. Some additional
descriptive studies exploiting these attributes would significantly enhance
our ability to understand mutant phenotypes. Chief among these would be
a four dimensional fish, that is, a time-lapse reconstruction of development
in real time. Substantial computer assisted imaging must accompany this
so that different organ systems, tissues, and cell types can be highlighted
and so that the images can be viewed from any orientation desired by the
viewer. Furthermore, the gene expression patterns of the ESTs generated
in item B1 above and other isolated genes must be visible on this 4D fish.
Furthermore, the images must be retrievable from a database in any of a
number of ways determined by the user (for example, the expression patterns
of all loci that begin expression within 1 hour of the appearance of sonic
hedgehog transcript in the floorplate).
<p>4D Fish. X,Y,Z images in real time, including gene expression patterns
<br>More fate maps for later stages. This would help in the interpretation
of gene expression patterns.
<p><b><u>II. Novel and improved phenotyping.</u></b> Members of the community
perceive that more effort needs to go into developing novel means for phenotyping
zebrafish, for example by electrophysiological methods or to assay behaviors,
or to detect changes in ions such as calcium. These goals are probably
best achieved by RO1's.
<p><b><u>III. Protein expression and biochemistry.</u></b> Although the
community feels at this point that infrastructure for genomics should take
priority to infrastructure for proteomics, the construction of monoclonal
antibody panels from various stages and tissues, and the establishment
of tissue-specific cell lines would be useful.
<ul>
<li>
Monoclonal antibodies for various stages</li>

<li>
Cell lines from various tissues</li>
</ul>
<b>D. Budget</b>
<p>Here is listed the budget for the genomics items in part B above. The
cell biology items are also needed, but we were less able to put dollar
figures on them. To accomplish the aims in part B would cost about $20
million spent over a three year period.
<br>&nbsp;
<table WIDTH="98%" >
<tr VALIGN=TOP>
<td WIDTH="23%"></td>

<td ALIGN=LEFT WIDTH="19%"><b><font color="#000000">Items</font></b></td>

<td ALIGN=RIGHT WIDTH="19%"><b>Per Year</b></td>

<td ALIGN=CENTER WIDTH="19%"><b>Years</b></td>

<td ALIGN=RIGHT WIDTH="19%"><b>Total</b></td>
</tr>

<tr ALIGN=RIGHT VALIGN=TOP>
<td></td>

<td></td>

<td ALIGN=RIGHT>($ in thousands)</td>

<td></td>

<td ALIGN=RIGHT>($ in thousands)</td>
</tr>

<tr VALIGN=TOP>
<td></td>

<td></td>

<td></td>

<td></td>

<td></td>
</tr>

<tr VALIGN=TOP>
<td><b>ESTs</b></td>

<td></td>

<td></td>

<td></td>

<td></td>
</tr>

<tr VALIGN=TOP>
<td></td>

<td>cDNA libraries</td>

<td ALIGN=RIGHT>150</td>

<td ALIGN=CENTER>3</td>

<td ALIGN=RIGHT>450</td>
</tr>

<tr VALIGN=TOP>
<td></td>

<td>End sequences</td>

<td ALIGN=RIGHT>400</td>

<td ALIGN=CENTER>3</td>

<td ALIGN=RIGHT>1,200</td>
</tr>

<tr VALIGN=TOP>
<td></td>

<td>Map ESTs</td>

<td ALIGN=RIGHT>800</td>

<td ALIGN=CENTER>3</td>

<td ALIGN=RIGHT>2,400</td>
</tr>

<tr VALIGN=TOP>
<td></td>

<td>Expression</td>

<td ALIGN=RIGHT>150</td>

<td ALIGN=CENTER>3</td>

<td ALIGN=RIGHT>450</td>
</tr>

<tr VALIGN=TOP>
<td></td>

<td>Microarrays</td>

<td ALIGN=RIGHT>150</td>

<td ALIGN=CENTER>3</td>

<td ALIGN=RIGHT>450</td>
</tr>

<tr VALIGN=TOP>
<td></td>

<td></td>

<td ALIGN=RIGHT></td>

<td ALIGN=CENTER></td>

<td ALIGN=RIGHT></td>
</tr>

<tr VALIGN=TOP>
<td><b>Stock Center</b></td>

<td></td>

<td ALIGN=RIGHT>250</td>

<td ALIGN=CENTER>3</td>

<td ALIGN=RIGHT>750</td>
</tr>

<tr VALIGN=TOP>
<td></td>

<td></td>

<td ALIGN=RIGHT></td>

<td ALIGN=CENTER></td>

<td ALIGN=RIGHT></td>
</tr>

<tr VALIGN=TOP>
<td><b>Data Base</b></td>

<td></td>

<td ALIGN=RIGHT></td>

<td ALIGN=CENTER></td>

<td ALIGN=RIGHT></td>
</tr>

<tr VALIGN=TOP>
<td></td>

<td>Data editors</td>

<td ALIGN=RIGHT>500</td>

<td ALIGN=CENTER>3</td>

<td ALIGN=RIGHT>1,500</td>
</tr>

<tr VALIGN=TOP>
<td></td>

<td>Bioinformaticists</td>

<td ALIGN=RIGHT>500</td>

<td ALIGN=CENTER>3</td>

<td ALIGN=RIGHT>1,500</td>
</tr>

<tr VALIGN=TOP>
<td></td>

<td></td>

<td ALIGN=RIGHT></td>

<td ALIGN=CENTER></td>

<td ALIGN=RIGHT></td>
</tr>

<tr VALIGN=TOP>
<td><b>Mutant Mapping</b></td>

<td></td>

<td ALIGN=RIGHT>333</td>

<td ALIGN=CENTER>3</td>

<td ALIGN=RIGHT>1000</td>
</tr>

<tr VALIGN=TOP>
<td></td>

<td></td>

<td ALIGN=RIGHT></td>

<td ALIGN=CENTER></td>

<td ALIGN=RIGHT></td>
</tr>

<tr VALIGN=TOP>
<td><b>Physical Map</b></td>

<td></td>

<td ALIGN=RIGHT></td>

<td ALIGN=CENTER></td>

<td ALIGN=RIGHT></td>
</tr>

<tr VALIGN=TOP>
<td></td>

<td>BAC libraries</td>

<td ALIGN=RIGHT>250</td>

<td ALIGN=CENTER>1</td>

<td ALIGN=RIGHT>250</td>
</tr>

<tr VALIGN=TOP>
<td></td>

<td>Contigs</td>

<td ALIGN=RIGHT>1,000</td>

<td ALIGN=CENTER>3</td>

<td ALIGN=RIGHT>3,000</td>
</tr>

<tr VALIGN=TOP>
<td></td>

<td>Telomeric YACs</td>

<td ALIGN=RIGHT>150</td>

<td ALIGN=CENTER>1</td>

<td ALIGN=RIGHT>150</td>
</tr>

<tr VALIGN=TOP>
<td></td>

<td></td>

<td ALIGN=RIGHT></td>

<td ALIGN=CENTER></td>

<td ALIGN=RIGHT></td>
</tr>

<tr VALIGN=TOP>
<td><b>Genome Sequence</b></td>

<td></td>

<td ALIGN=RIGHT></td>

<td ALIGN=CENTER></td>

<td ALIGN=RIGHT></td>
</tr>

<tr VALIGN=TOP>
<td></td>

<td>Phase 1</td>

<td ALIGN=RIGHT>3,000</td>

<td ALIGN=CENTER>1</td>

<td ALIGN=RIGHT>3,000</td>
</tr>

<tr VALIGN=TOP>
<td></td>

<td>Phase 2</td>

<td ALIGN=RIGHT>1,500</td>

<td ALIGN=CENTER>3</td>

<td ALIGN=RIGHT>4,500</td>
</tr>

<tr VALIGN=TOP>
<td></td>

<td></td>

<td ALIGN=RIGHT></td>

<td ALIGN=CENTER></td>

<td ALIGN=RIGHT></td>
</tr>

<tr VALIGN=TOP>
<td><b>Total</b></td>

<td></td>

<td ALIGN=RIGHT>9,133</td>

<td></td>

<td ALIGN=RIGHT>20,600</td>
</tr>
</table>


</body>

<script language="JavaScript" src="/javascript/footer.js"></script>
