<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html>
	<head>
		<title>
			ZFIN CONSTRUCTION OF A 
		</title>
	</head>

<script language="JavaScript" src="http://<!--|DOMAIN_NAME|-->/header.js"></script>

<h1 align="center"><strong>CONSTRUCTION OF A </strong></h1>
<h1 align="center"><strong>HIGH-RESOLUTION ZEBRAFISH</strong></h1>
<h1 align="center"><strong>GENETIC MAP</strong></h1>

<p align="center">By Mark C. Fishman, MD, Cardiovascular Research Center, Massachusetts General Hospital, 
149 13th Street , 4th Floor, Charlestown, MA  02129 USA

<br wp="br1"><br wp="br2"><h2><strong>Project goal</strong></h2>
Add markers to the existing microsatellite map to an average density of 1 marker per 0.3-0.5 cM. 
Based on the 2000 marker map, this will require approximately 7000 total markers or 500 new 
markers.

<br wp="br1"><br wp="br2"><h2><strong>Progress</strong></h2>

<table  width="93%">
<tr valign="top"><td width="25%"><em>Prior to RFA</em></td>
<td width="74%"><em>705 marker map (Knapik et al., 1998) </em> 

<p><em>2000 marker map (Shimoda et al., 1999, in press)</em></td></tr>
<tr valign="top"><td><h4><strong>To date</strong></h4>
</td>
<td></td></tr>
<tr valign="top"><td><strong></strong>February 1999</td>
<td>Funding awarded (RFA DK-98-006).</td></tr>
<tr valign="top"><td>May 1999</td>
<td>First quarterly update released:

<ul>
  <li>200 new markers added (total 2200)</li>
  <li>3300 new sequence-tagged sites submitted to NCBI</li>
</ul>

<p>Re-designed MGH web site to provide up-to-date, user-friendly 
access to map data.</td></tr>
<tr valign="top"><td><h4><strong>Projected</strong></h4>
</td>
<td></td></tr>
<tr valign="top"><td><strong></strong>August 1999</td>
<td>
<ul>
  <li>Genotype 600 additional STS's; expect 250 can be placed on 
map.</li>
  <li>Prepare additional libraries for identification of additional 
microsatellite sequences.</li>
</ul>
</td></tr>
<tr valign="top"><td>Subsequent quarters</td>
<td>
<ul>
  <li>Identify microsatellite sequences by hybridization to genomic 
libraries</li>
  <li>Sequence clones and develop STS's</li>
</ul>

<ul>
  <li>Genotype new STS's</li>
</ul>
</td></tr></table>

<br wp="br1"><br wp="br2">
<p>Achieving our goal of mapping 5200 markers over 3 years will require mapping an average of 
500 markers per quarter.  This will require a team of six technicians; to date we have hired and 
trained three.  Recruiting additional technicians is a top priority.  Certain aspects of the project, 
particularly screening genomic libraries, are most efficient when performed by the entire team. 
During quarters when this work is underway, the number of markers placed on the map may be 
reduced.

<br wp="br1"><br wp="br2"><h2><strong>Data release and availability</strong></h2>
<h4><strong>1.  Sequence Tagged Sites (STS's)</strong></h4>
STS data (genomic sequence and PCR primers) will be submitted to the dbSTS database at NCBI 
within 48 hours of analysis.  All STS's will be submitted prior to and regardless of genotyping 
and mapping.

<br wp="br1"><br wp="br2"><h4><strong>2.  Microsatellite markers</strong></h4>
New updates of the map will be calculated and released quarterly.  All markers genotyped during 
that quarter will be included in the update.  Calculating positions is a time-consuming process 
regardless of the number of markers being added, so it is not practical to calculate or release this 
data more frequently.  Map locations for new markers will be added to the existing map 
displayed on our local web site and submitted to ZFIN once the positions have been reviewed for 
possible errors, typically within 48-72h of analysis.

<br wp="br1"><br wp="br2"><h4><strong>3.  Data availability</strong></h4>
During the past quarter we have substantially enhanced our local resources for displaying map 
data.  We have set up a central database system to consolidate data on all aspects of marker 
generation including sequencing, PCR primer design, genotyping and mapping.  Our web site 
queries this database to generate up-to-date information for every request to ensure that the 
community has access to the latest data.  In addition, map positions are submitted to ZFIN and 
genomic sequences to NCBI.

<br wp="br1"><br wp="br2"><h2>Contact information</h2>

<table  width="100%">
<tr valign="top"><td width="49%">CVRC Zebrafish web server</td>
<td width="54%"><a href="http://zebrafish.mgh.harvard.edu/mapping/">http://zebrafish.mgh.harvard.edu/mapping/</a></td></tr>
<tr valign="top"><td>Contact address for mapping team

<p>(distributed to informatics coordinator, 
mapping team leader and system 
administrator plus others as needed)</td>
<td><a href="mailto: zfhelp@cvrc.mgh.harvard.edu">zfhelp@cvrc.mgh.harvard.edu</a></td></tr>
<tr valign="top"><td>Mark Fishman</td>
<td><a href="mailto: fishman@cvrc.mgh.harvard.edu">fishman@cvrc.mgh.harvard.edu</a></td></tr>
<tr valign="top"><td>Donald Jackson, Informatics Coordinator</td>
<td><a href="mailto: jackson@cvrc.mgh.harvard.edu">jackson@cvrc.mgh.harvard.edu</a></td></tr>
<tr valign="top"><td>John Ziniti, Mapping Team Leader</td>
<td><a href="mailto: ziniti@helix.mgh.harvard.edu">ziniti@helix.mgh.harvard.edu</a></td></tr>
<tr valign="top"><td>Lab telephone</td>
<td>(617) 726-6902</td></tr></table>

</body>

<script language="JavaScript" src="http://<!--|DOMAIN_NAME|-->/footer.js"></script>
