<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html>
	<head>
		<title>
			ZFIN Zebrafish Genomics Initiative
		</title>
	</head>

<script language="JavaScript" src="http://<!--|DOMAIN_NAME|-->/header.js"></script>

&nbsp;
<table COLS=2 WIDTH="600" >
<tr>
<td WIDTH="210"><b><i><font size=+4>Catch of the day...</font></i></b>
<center><font face="Times New Roman,Times"><font size=+1>News from the
zebrafish research community</font></font></center>
</td>

<td WIDTH="100"><img SRC="catch.gif" BORDER=0 height=67 width=210></td>
</tr>
</table>

<hr WIDTH="100%"><b>October 20, 1998</b>
<br><b>Trans-NIH Zebrafish Coordinating Committee</b>
<br>
<hr size="4">
<h1>
<b>ZEBRAFISH GENOMICS INITIATIVE GRANTEES MEETING</b></h1>

<hr size="4">Natcher Center, National Institutes of Health
<h2>
<b>MEETING REPORT</b></h2>

<hr><b>Background</b>
<br>The February 1997 Boston workshop, Current Advances in Defining the
Zebrafish Genome, precipitated recommendations promoting the use of zebrafish
as a model organism for the study of vertebrate development. Program staff
from the National Institute of Diabetes and Digestive and Kidney Diseases
(NIDDK), National Center for Research Resources (NCRR), and National Human
Genome Research Institute (NHGRI) reviewed the report. A report was presented
to the Director of the NIH, in spring, 1997, with the recommendation to
develop the zebrafish system for genetic studies of vertebrate embryogenesis
and disease. In response to these recommendations, the Director of the
NIH formed the Trans-NIH Zebrafish Coordinating Committee , which first
met in the fall of 1997. This working group is composed of representatives
from most of NIH's Institutes and Centers having an interest in promoting
zebrafish as a research model. The Committee is co-chaired by Dr. Josephine
Briggs of NIDDK and Dr. Tyl Hewitt of the National Institute for Child
Health and Human Development (NICHD).
<p>The initial effort of this group resulted in 13 Institutes sponsoring
a Request for Applications (RFA) entitled, "Genomic Resources for the Zebrafish"
(DK-98-006). The purpose of the RFA was to solicit applications aimed at
creating resources to facilitate the mapping and positional cloning of
genes in the zebrafish, specifically the (1) generation of a genetic map
with a resolution of 0.3 cM or better, (2) development of expressed sequence
tags (ESTs) from existing and new cDNA libraries featuring specific developmental
time points and tissues, and (3) creation of a physical or radiation hybrid
(RH) map. Currently, 18 Institutes support the initiative and will participate
in the oversight of funded projects. The principal awards were made through
NIDDK and will be managed by NIDDK, NICHD, NHGRI, and other participating
Institutes through the Trans-NIH Zebrafish Coordinating Committee. Four
awards were granted in summer, 1998 and a fifth award was made in fall,
1998.
<h3>
<b>Meeting Objectives</b></h3>
The meeting, chaired by Dr. Briggs and Dr. Hewitt, brought together the
five grantees, representatives from the Trans-NIH Zebrafish Coordinating
Committee, informatics experts, and the Chair of the Executive Zebrafish
Advisory Committee (EZAC). The meeting objectives included (1) an explanation
of the structure of the zebrafish genomics initiative, (2) presentation
of short-term and long-term goals for the five funded projects, (3) a review
of informatics resources and limitations, and (4) planning and policy discussions.
<h3>
<b>Structure of Initiative</b></h3>
The zebrafish genomics initiative is currently funded at $4.7 million per
year for 3 years. The principal investigators (PIs) comprise the Steering
Committee. During the meeting, Dr. Leonard Zon was elected Chair of the
Steering Committee. This committee will determine the agenda for annual
meetings. Additional meetings may occur by teleconference as needed. An
external advisory committee will advise the Trans-NIH Zebrafish Coordinating
Committee in setting goals for the zebrafish genome initiative and examining
research progress in relation to those goals. Until membership is finalized,
Dr. Nancy Hopkins has agreed to chair the committee. The meeting participants
were asked to review a tentative list of EZAC members and suggest additional
expertise. Dr. Richard Myers was recommended. Tascon, Inc. has been contracted
to provide logistical support for meetings and teleconferences.
<p>Dr. Briggs announced that the Trans-NIH Coordinating Committee will
sponsor a zebrafish workshop on May 10-13, 1999 at NIH. She suggested that
the Steering Committee hold its annual meeting at this time.
<h3>
<b>Aims of Individual-Funded Projects</b></h3>
<b>Dr. Stephen Johnson</b>, Washington University School of Medicine will
use an integrated approach to genomic resource development. EST sequencing
will be utilized to develop sequence tagged sites (STSs) that will be used
to identify single nucleotide polymorphisms (SNPs) between inbred strains.
SNPs will be mapped on RH and meiotic mapping panels. Genetic and physical
mapping of 10,000 zebrafish ESTs will provide candidate genes for zebrafish
mutations and a dense framework for initiating positional cloning projects.
Comparison of zebrafish gene map positions with human homologues will reveal
roles for human genes and assist in reconstructing the evolutionary history
of the vertebrate genome.
<p>Oligonucleotide hybridization fingerprint analysis of 278,000 independent
zebrafish cDNA clones from various tissues and developmental stages will
be analyzed to identify 50,000 clusters likely to represent each of the
zebrafish genes. Approximately 10,000 STS markers, generated from 3' ESTs,
will be used in collaboration with Dr. Zon, to generate a 7,500-marker
RH map. Sequencing and map data will be updated weekly on a publicly accessible
website.
<p>As the mapping strains SJD and C32 have residual polymorphism (less
than 5%), they will be bred to complete homozygosity. Polymerase chain
reaction (PCR)-based methods will be used to identify individuals, homozygous
at residual polymorphic loci, for propagation. As the SJD and C32 strains
are male-biased (typically less than 5% female), ingression of dominant
female-promoting genes will be achieved by serial backcrossing.
<p>Discussion focused on two topics -- stock centers and avoiding redundancy
in mapping efforts. To avoid mapping the same genes repeatedly, publicly
accessible websites will be monitored regularly, particularly between Dr.
Johnson and Dr. William Talbot. All the PIs agreed that maintaining large
fish stocks, due to imbalanced sex ratios, is problematic. Issues include
the demand for increased physical space for fish tanks and the possibility
of strain contamination. The need for a stock center that maintains test
strains was emphasized.
<p><b>Dr. Leonard Zon</b>, Children's Hospital Corporation, Boston, will
utilize amplified fragment length polymorphism (AFLP) technology to define
genetic markers within the zebrafish genome. A RH map will be developed
by (1) comparing four RH panels for retention frequency and resolution,
(2) constructing an anchored framework map of microsatellite markers and
cloned cDNAs on the RH panel with the most appropriate resolution, and
(3) positioning 5,000 to 10,000 EST markers on the selected RH panel. Data
will be posted on the Zebrafish Information Network (ZFIN) at regular intervals.
<p>Currently, two RH panels are available -- the Ekker panel and Goodfellow
panel (named after the investigators). Characterization of the panels involves
the following criteria: (1) retention rate -- the percentage of zebrafish
genome in each hybrid, (2) connectivity -- the ability to detect linkage
between distant markers, (3) resolution -- the ability to determine the
order of nearby loci, and (4) typical fragment size or average breakpoint
frequency. The two panels have similar retention rates, but the Goodfellow
panel produces a larger number of smaller fragments (approximately 66 kb),
while the Ekker panel produces a smaller number of larger fragments (100
kb to 150 kb). Resolution and connectivity comparisons are not yet completed.
<p>RH and genetic maps yield similar results, but the genetic map is more
accurate. The advantage of RH mapping is the ability to position many markers
rapidly. Remaining RH mapping issues for this project concern choosing
a panel for EST mapping and information technology coordination. The latter
issue involves choosing a software program for analysis (RH Mapper or SA
Mapper), data storage format, and Internet publishing format. Selection
of a RH panel for EST mapping will be based on resolution, connectivity,
and the cost and availability of the panel. An advisory panel will help
in the selection process.
<p><b>Dr. Mark Fishman</b>, Massachusetts General Hospital (MGH), plans
to increase the density of microsatellite markers on the zebrafish genetic
map by achieving an average intermarker distance of 0.3 cM. Currently,
a low-density map exists, consisting of 705 markers covering all 25 zebrafish
chromosomes at an average interval of 5 cM. Microsatellite markers are
based on simple sequence length polymorphism (SSLPs) repeats found throughout
the genome. These markers are the standard starting point for positional
cloning and serve as anchors for other types of maps.
<p>Microsatellite markers present several advantages over other types of
markers, including (1) their abundance and relatively uniform distribution
throughout the genome, (2) polymorphism between outbred strains, (3) their
reliability and ease in use, which makes them ideal tools for small laboratories
or high-throughput facilities, (4) a high incidence of codominance permits
tracking of all four alleles, (5) rapid progress in positional cloning
projects as they are single locus amplification products. The latter characteristic
also facilitates anchoring a RH-based EST map to the genetic map.
<p>Data will be posted on the local MGH website (<a href="http://zebrafish.mgh.harvard.edu">http://zebrafish.mgh.harvard.edu
</a>),
ZFIN (<a href="/ZFIN/">http://<!--|DOMAIN_NAME|-->/ZFIN/
</a>),
and the National Center for Biotechnology Information (NCBI) database for
STSs (<a href="http://www.ncbi.nlm.nih.gov/">http://www.ncbi.nlm.nih.gov/
</a>).
Data information will vary depending on the database, but will include
marker name, primer sequence, PCR conditions, and product sizes in multiple
strains.
<p>Subsequent discussion questioned the optimal number of markers, based
on cost-effectiveness. Participants agreed with Dr. Fishman's approach
to achieve an average intermarker distance of 0.3 cM as this resolution
facilitates chromosome walks. The added expense for increased resolution
produces minimal research gains. Instead, individual investigators with
a mutation or tag between two markers can expand the local map. Dr. Zon
suggested that a central broker, such as Dr. John Postlethwait at the University
of Oregon, notify investigators mapping the same region or bin. This process
would expedite determining the order of local markers. The participants
agreed that incentives are needed to encourage submission of prepublished
data to centralized databases.
<p><b>Dr. William Talbot</b>, Skirball Institute, New York University Medical
Center, will meiotically localize 3,000 zebrafish genes on an integrated
genetic map to create a resource that will enhance the comparative and
functional analysis of the vertebrate genome. The genes will be mapped
by scoring single-strand conformational polymorphisms (SSCPs) in 3' untranslated
regions (UTRs) and other nonconserved regions. As genes are uniquely suited
markers for comparative genomics, the project will allow construction of
a map of chromosomal segments conserved between zebrafish and human. This
will accelerate gene mapping in zebrafish by defining boundaries of conserved
segments and will facilitate comparison with the gene-rich maps of mouse
and human.
<p>A framework map will be constructed by assigning 500 publicly available
CA repeat markers in a homozygous diploid (HD) mapping panel. Genes will
be mapped to this panel and the data made available to the zebrafish community.
SSLP markers in the framework map will serve as a standard of comparison
to physical maps and genetic maps made with different mapping panels (e.g.,
mutation mapping crosses). The HD mapping panel is expected to achieve
a resolution of 2 cM on the genetic map. A World Wide Web interface will
allow rapid public access to map information and streamline data management.
<p>Discussion focused on DNA libraries. Dr. Marco Marra explained that
libraries are not equivalent. Libraries physically suitable for hybridization
may not perform well at the sequencing stage. Additionally, cell types
may be inappropriate for particular tasks. For example, lambda-based cDNA
libraries are not well suited for high-throughput sequencing efforts. Technologies
used with libraries must also be considered. Fingerprinting produces single
representatives of a particular gene and eliminates redundancy, but focuses
on a small subset of libraries. EST-based techniques support more complex
libraries, but target fewer genes.
<p><b>Dr. Marnie Halpern</b>, Carnegie Institute of Washington, will produce
a comprehensive deletion panel for the zebrafish genome. The panel will
integrate locations of specific deficiencies with the evolving zebrafish
genetic map. The project will involve collecting, preserving, and cataloging
existing deficiencies and translocation strains. Strains will be maintained
by freezing sperm and storing samples in a centralized repository. The
status and properties of each deficiency strain will be recorded in a database.
Additionally, recently isolated and newly gamma-ray-induced deficiency
strains will be recovered and characterized.
<p>Deficiency genome DNA samples will be typed using mapping primers derived
from cloned genes, ESTs, or anonymous DNA markers (e.g., SSLPs). High-quality
DNA samples from mutant embryos, preselected for the loss of specific chromosomal
regions, will be arrayed in multiwell plates for high-throughput mapping
and distribution to the zebrafish community. Lastly, expressed sequences
will be correlated to mutant phenotypes. Preliminary descriptions and video
images of mutant phenotypes corresponding to each deficiency will be stored
in an online database. This will enable researchers to immediately correlate
deleted gene sequences with a candidate mutant phenotype.
<p>Requested deficiency mutation strains will be thawed for <i>in vitro</i>
fertilization of wildtype eggs and shipped as live embryos. An estimated
200 embryos will be recoverable from each deletion strain, providing enough
DNA for 500 PCR reactions per embryo. Aliquots sufficient for 1,000 PCR
reactions will be provided upon request.
<p>Dr. Fishman stated that his team could help select microsatellite markers
to be used as anchors for the genetic map. Ideally, the anchors should
be evenly distributed, with a marker every 5 cM. Crude mapping would require
approximately 100 markers, with more refined projects requiring about 500
markers. The markers selected as anchors could be listed on ZFIN and refined
over time. His suggestion was well received.
<h3>
<b>Informatics</b></h3>
Dr. Monte Westerfield presented information about ZFIN. ZFIN began after
the 1994 Cold Spring Harbor meeting on zebrafish development and genetics.
At that time, meeting participants recognized that much of the generated
zebrafish data would not be published in traditional journals, and a centralized
database would provide access to that data. The ZFIN project, originally
funded by a National Science Foundation starter grant, is now funded by
NCRR.
<p>ZFIN offers a broad range of information, such as an annotated atlas
of embryonic and larval anatomy, meiotic and RH maps, descriptions of wildtype
and mutant strains, links to related websites, commercial sources for reagents,
and publications on research methods and technologies. Currently, ZFIN
contains descriptions of approximately 2000 mutants, with images for 85%
of the mutants. A simple dictionary of 30 to 40 terms allows a search for
mutants based on five criteria. Future plans call for an expanded dictionary
to include standardized descriptors for anatomical structures, physiological
processes, and developmental stages. Development of interfaces for user-friendly
access to data is also planned. A September 1998 meeting at the University
of Oregon on zebrafish mapping bioinformatics discussed the status and
future plans for ZFIN. A summary of that meeting was provided to the participants
of the Grantees Meeting.
<p>Discussion centered on the compatibility of mapping software, the compatibility
of the different types of maps, and the expeditious availability of new
data. As mapping laboratories need to develop and maintain their own in-house
databases for organizing and curating locally generated map data, Dr. Westerfield
is concerned that each laboratory will develop its own specialized software.
He recommended that mapping laboratories coordinate their local software
resources. A common software program (e.g., RH Mapper) would facilitate
data sharing and ease the work involved in exporting data to ZFIN.
<p>The second issue discussed at the meeting involves integrating data
from different mapping panels to allow direct comparisons of different
maps. Dr. Westerfield explained that a fully automated computer generation
of a single, definitive, integrated map is not practical due to ambiguities
and conflicts inherent in data from different mapping experiments. Producing
such a map would require extensive curation by a chromosome committee.
Dr. Fishman suggested the use of an anchored set of maps. While an anchored
map would not provide perfect colinear alignment, it would ease comparison
between different maps. On a related topic, participants discussed cross-species
genomic analysis. Dr. Westerfield explained that future plans call for
ZFIN links to records of orthologous genes in other species databases and
links to external databases (e.g., GenBank). A mechanism for displaying
and comparing syntenic relationships between zebrafish and other organisms
is also planned.
<p>Participants agreed that new data should be made publicly available
when it is generated. Dr. Westerfield suggested that unpublished data (e.g.,
mapped genes and mutations) be held for a waiting period after submission
to ZFIN, but before journal publication. The waiting period would allow
researchers to register and reserve mutant and gene names before publication.
Additionally, journals should be encouraged to require ZFIN accession numbers
for genes, mutations, and map markers by the time of publication. This
requirement would provide incentive for researchers to submit their data
to common databases.
<h3>
<b>Planning and Policy Discussion</b></h3>

<h4>
<u>Data Sharing and Release Policies</u></h4>
The investigators agreed that sharing sequence data is not problematic;
standards exist for other projects, and data can be updated on a timely
basis. Mapping data is problematic as it must be error-checked in a time-consuming
manner. Dr. Halpern stated that deletion data could be promptly posted
on ZFIN if researchers understand the considerable lag time in obtaining
live embryos. Dr. Briggs requested that the five grantees draft a written
policy for data release.
<h4>
<b><u>Dissemination of Proposed Data and Resource-Sharing Plans</u></b></h4>
The participants discussed mechanisms to encourage collaboration within
the zebrafish community and move the genome project forward. The following
suggestions were made: (1) list the names of the grantees and contact information
on ZFIN, (2) on ZFIN note how often updates will be posted, (3) allow investigators
to anonymously post map locations of unpublished genes and mutations for
a 6-month period prior to journal publication; this would allow researchers
to establish collaborations with one another while protecting their data
before publication, (4) make a large-scale mapping effort part of the initiative
project so that 100 to 200 mutations could be mapped quickly and made available,
(5) submit a letter or brief article to
<i>Science</i> or <i>Nature</i>
describing the dissemination process and spirit of collaboration.
<h4>
<u>Development and Oversight of Database</u></h4>
All the grantees are committed to using ZFIN as a centralized resource.
Dr. Westerfield reiterated the recommendations made at the September 1998
Zebrafish Mapping Bioinformatics meetingC(1) only mapped ESTs should be
stored in ZFIN with a link to access a complete listing in NCBI's dbEST,
(2) storing primer sequences in ZFIN would be useful, even though this
may duplicate data stored elsewhere, (3) information about allele size
should be included for each strain in which a marker has been mapped, (4)
information about the quality of each marker should be added, e.g., the
percent of successful amplification to indicate reliability, (5) a list
of linked markers to each mutant should be added, (6) information on deficiency
chromosomes should be added including the mutant phenotype, linkage group,
missing markers, and flanking markers that are present.
<h4>
<u>International Relations</u></h4>
Several participants expressed the need to offer funds for additional staffing
of the stock center in Tuebingen, Germany. This center provides fish globally,
but is understaffed. Dr. David Badman remarked that Dr. Harold Varmus,
NIH Director, is interested in this issue. Funds may be available through
the Fogerty International Center. On a related topic, the grantees voiced
concern on the restrictive nature of transfer agreements with the Max Planck
Institute. Dr. Briggs stated that these issues would be pursued.
<h4>
<u>Future Resource Needs</u></h4>
The grantees were asked to list and prioritize resources that would augment
their current projects should additional funds become available. The investigators
agreed to the following priorities: (1) informatics -- increase support
for development and eight additional personnel, (2) EST sequencing -- increase
the goal of 50,000 sequences to 100,000 and represent organogenesis through
36 hours of development, (3) EST mapping -- increase the number of mapped
ESTs from 10,000 to 25,000 over 3 years, (4) physical map -- add support
for fingerprinted contig maps, (5) new equipment, such as improved devices
for genotyping and robotics for 384-well PCR machines, (6) facilitatory
projects such as development of better vectors for transgenics, promoter
identification, marked chromosomes, and targeted mutation.
<br>&nbsp;
<h3>

<hr size="5"><b>Participants</b></h3>
&nbsp;
<table COLS=2 WIDTH="100%" >
<tr>
<td>
<dt>
<b>Dr. David Badman</b></dt>

<dt>
National Institute of Diabetes and Digestive and Kidney Diseases</dt>

<dt>
301 594-7717</dt>

<br><a href="mailto:David_Badman@nih.gov">David_Badman@nih.gov</a></td>

<td>
<dt>
<b>Dr. Josephine Briggs, Co-Chair</b></dt>

<dt>
National Institute of Diabetes and Digestive and Kidney Diseases</dt>

<dt>
301 496-6325</dt>

<br><a href="mailto:briggsj@extra.niddk.nih.gov">briggsj@extra.niddk.nih.gov</a></td>
</tr>

<tr>
<td>
<dt>
<b>Dr. Jill Carrington</b></dt>

<dt>
National Center for Research Resources</dt>

<dt>
301 435-0776</dt>

<br><a href="mailto:jillc@ep.ncrr.nih.gov">jillc@ep.ncrr.nih.gov</a></td>

<td>
<dt>
<b>Dr. Hemin Chin</b></dt>

<dt>
National Institute of Mental Health</dt>

<dt>
301 443-1706</dt>

<br><a href="mailto:hemin@codon.nih.gov">hemin@codon.nih.gov</a></td>
</tr>

<tr>
<td>
<dt>
<b>Dr. Mark Ekker</b></dt>

<dt>
Ottawa Civic Hospital</dt>

<dt>
613 798-5555 ext 6033</dt>

<dt>
<a href="mailto:ekker@civich.ottawa.on.ca">ekker@civich.ottawa.on.ca</a></dt>
</td>

<td><b>Dr. Elise Feingold</b>
<dt>
National Human Genome Research Institute</dt>

<dt>
301 496-7531</dt>

<dt>
<a href="mailto:feingole@odder.nhgri.nih.gov">feingole@odder.nhgri.nih.gov</a></dt>
</td>
</tr>

<tr>
<td>
<dt>
<b>Dr. Adam Felsenfeld</b></dt>

<dt>
National Human Genome Research Institute</dt>

<dt>
301 496-7531</dt>

<br><a href="mailto:adam_felsenfeld@nih.gov">adam_felsenfeld@nih.gov</a></td>

<td><b>Dr. Mark Fishman, Grantee</b>
<br>Massachusetts General Hospital
<br>617 726-3738
<br><a href="mailto:fishman@cvrc.mgh.harvard.edu">fishman@cvrc.mgh.harvard.edu&nbsp;</a></td>
</tr>

<tr>
<td>
<dt>
<b>Dr. Andreas Fritz</b></dt>

<dt>
Department of Biology</dt>

<dt>
Emory University</dt>

<dt>
404 727-9012</dt>

<br><a href="mailto:afritz@biology.emory.edu">afritz@biology.emory.edu</a></td>

<td>
<dt>
<b>Dr. Maria Giovanni</b></dt>

<dt>
National Eye Institute</dt>

<dt>
301 496-0484</dt>

<br><a href="mailto:myg@nei.nih.gov">myg@nei.nih.gov</a></td>
</tr>

<tr>
<td>
<dt>
<b>Dr. Judith Greenberg</b></dt>

<dt>
National Institute of General Medical Sciences</dt>

<dt>
301 594-0943</dt>

<br><a href="mailto:greenbej@nigms.nih.gov">greenbej@nigms.nih.gov</a></td>

<td>
<dt>
<b>Dr. Marnie Halpern, Grantee</b></dt>

<dt>
Carnegie Institute of Washington</dt>

<dt>
410 554-1218</dt>

<br><a href="mailto:halpern@mail1.ciwemb.edu">halpern@mail1.ciwemb.edu</a></td>
</tr>

<tr>
<td>
<dt>
<b>Dr. Deborah Henken</b></dt>

<dt>
National Institute of Child Health and Human Development</dt>

<dt>
301 496-5541</dt>

<br><a href="mailto:dh50g@nih.gov">dh50g@nih.gov</a></td>

<td>
<dt>
<b>Dr. Tyl Hewitt, Co-Chair</b></dt>

<dt>
National Institute of Child Health and Human Development</dt>

<dt>
301 496-5541</dt>

<br><a href="mailto:hewittt@hd01.nichd.nih.gov">hewittt@hd01.nichd.nih.gov</a></td>
</tr>

<tr>
<td>
<dt>
<b>Dr. Nancy Hopkins</b></dt>

<dt>
Massachusetts Institute of Technology</dt>

<dt>
617 253-6414</dt>

<br><a href="mailto:nhopkins@mit.edu">nhopkins@mit.edu</a></td>

<td>
<dt>
<b>Dr. Don Jackson</b></dt>

<dt>
Massachusetts General Hospital</dt>

<dt>
617 726-6902</dt>

<br><a href="mailto:jackson@cvrc-taco.mgh.harvard.edu">jackson@cvrc-taco.mgh.harvard.edu</a></td>
</tr>

<tr>
<td><b>Dr. Steve Johnson</b>
<br>Washington University
<br>314-362-0362
<br><a href="mailto:monte@uoneuro.uoregon.edu">sjohnson@genetics.wustl.edu</a></td>

<td><strong>Dr. Robert Karp</strong>
National Institute of Alcohol Abuse and Alcoholism
<br>301 443-2239
<br><a href="mailto:rkarp@willco.niaaa.nih.gov">rkarp@willco.niaaa.nih.gov</a></td>
</td></tr>

<tr><td><strong>Dr. Carol Letendre</strong>
<br>National Heart, Lung, and Blood Institute
<br>301 435-0080
<br><a href="mailto:letendrc@gwgate.nhlbi.nih.gov">letendrc@gwgate.nhlbi.nih.gov </a></td>

<td>
<dt>
<b>Dr. John McPherson</b></dt>
<dt>
Washington University</dt>
<dt>314 286-1848</dt>
<br><a href="mailto:jmcphers@watson.wustl.edu">jmcphers@watson.wustl.edu</a></td>
</tr>

<tr>
<td>
<dt>
<b>Dr. Grace Shen</b></dt>
<dt>
National Cancer Institute</dt>
<dt>
301 435-5226</dt>
<br><a href="mailto:gs35r@nih.gov">gs35r@nih.gov</a></td>

<td><strong>Dr. William Talbot, Grantee</strong> 
<br>New York University Medical Center
<br>212 263-1932
<br><a href="mailto:talbot@saturn.med.nyu.edu">talbot@saturn.med.nyu.edu</a></td></tr>

<tr><td><b>Dr. Monte Westerfield</b>
<br>University of Oregon
<br>541-346-4607
<br><a href="mailto:monte@uoneuro.uoregon.edu">monte@uoneuro.uoregon.edu </a></td>

<td><strong>Dr. Leonard Zon, Grantee</strong> 
<br>Children's Hospital Corporation, Boston
<br>617 355-7707
<br><a href="mailto:zon@rascal.med.harvard.edu">zon@rascal.med.harvard.edu</a></td>

</tr>
</table>


</body>

<script language="JavaScript" src="http://<!--|DOMAIN_NAME|-->/footer.js"></script>
