<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html>
	<head>
		<title>
			Zebrafish Resources
		</title>
	</head>

<body>
<table COLS=2 WIDTH="600" >
<tr>
<td WIDTH="210"><b><i><font size=+4>Catch of the day...</font></i></b>
<center><font face="Times New Roman,Times"><font size=+1>News from the
zebrafish research community</font></font></center>
</td>

<td WIDTH="100"><img SRC="catch.gif" BORDER=0 height=67 width=210></td>
</tr>
</table>

<hr WIDTH="100%">
<b>Fresh Caught - April 2002</b> <br>
<hr WIDTH="100%">
<b><font size="+2">Genomic Resources for Zebrafish</font></b>
<p><b><font size="+1">Links:</font></b></p>
<ul>
  <li><b><a href="http://www.sanger.ac.uk/Projects/D_rerio/">Sanger Institute 
    whole-genome sequence</a>: </b><br>
    This site hosts all of the efforts of the zebrafish genome sequencing project, 
    including whole genome shotgun sequence and selective BAC shotgun sequences. 
    In addition, the fingerprinting (physical map) of the zebrafish genome is 
    presented in a graphic display. </LI>
  <li><a href="http://www.ensembl.org/Danio_rerio/"><b>Ensembl database of annotated 
    zebrafish sequence</b></a><b>:<br>
    </b>Provides easy access to sequence data including predicted structure and 
    location of known genes in the genome sequence and predictions of novel genes, 
    with supporting evidence. </li>
  <li><a href="http://www.sanger.ac.uk/Projects/D_rerio/mapping.shtml"><b>Physical 
    map based on fingerprinted BACs</b></a><b>:</b><br>
    A bacterial clone physical map of the genome is being constructed using restriction 
    enzyme fingerprinting<b>.</b> </li>
  <li><b><a href="http://zfin.org/cgi-bin/ZFIN_jump?record=JUMPTOREFCROSS">RH, 
    meiotic and integrated maps of the genome</a>: </b><br>
    Centralized site for all meiotic and radiation hybrid maps of the zebrafish 
    genome. </li>
  <li><a href="http://www.genetics.wustl.edu/fish_lab/frank/cgi-bin/fish/"><b>Washington 
    University EST sequencing project</b></a><b>: </b><br>
    This site hosts the project in which over 200,000 ESTs have been sequenced. 
    It offers a unigene-like set (&#145;wz&#146; ESTs) that can be found by key 
    word search or BLAST. </li>
  <li><b><a href="http://134.174.23.160/HumanblastZebrafish/">BLAST hits of Human 
    Genes from Sanger sequence</a>: </b><br>
    Each human gene annotated on the Santa Cruz site was compared by BLAST against 
    the Sanger Institute zebrafish genome sequence. These data can be used to 
    help identify zebrafish orthologues of human or Fugu genes. </li>
  <li><b><a href="http://bahama.jgi-psf.org/fugu/bin/fugu_search">A BLAST server 
    for the FUGU genome</a>: </b><br>
    At this site, you can input amino acid or nucleotide sequence from any species 
    and do a BLASTN or TBLASTN search of the Fugu genome. The results point to 
    scaffolds from the Fugu genome of roughly 50,000 base pairs containing the 
    genes of interest. Synteny is highly conserved between zebrafish and Fugu. 
    Thus, this database is extremely useful for predicting the genomic locations 
    of zebrafish genes. </li>
  <li><b><a href="http://genome.ucsc.edu/">Santa Cruz Human Genome site</a>: </b><br>
    This site compiles the human genome as a large contig. It can be searched 
    using keywords or BLAST. </li>
  <li><b><a href="http://www.genome.wustl.edu/gsc/overgo/overgo.html">Server for 
    making overgoes</a>: </b> <br>
    This site helps design oligos utilized for hybridization to filters from BAC/PAC 
    libraries. These oligos give good signal to noise ratios for isolating genomic 
    clones. See the <a href="http://zon.tchlab.org/">Zon lab</a>&#146;s <a href="http://zon.tchlab.org/PositionalCloningGuide.doc">Positional 
    Cloning Guide</a> for details. </li>
  <li>The BAC end sequence data will soon be available. Two sites will host this 
    information, Stefan Schuster at Tuebingen Germany, and the Sanger Institute. 
  </li>
  <li><a href="http://zfin.org/zf_info/genome.html"><b>Links to genome databases 
    for other species</b></a></li>
</ul>
<hr WIDTH="100%">
<p><b><font size="+1">Contacts:</font></b></p>
<ul>
  <li>Information from Sanger about sequencing, sequence annotation: <a href="mailto:zface@sanger.ac.uk">Kerstin 
    Jekosh</a><br>
  </li>
  <li>Information about fingerprinting, mapping: <a href="mailto:sjh@sanger.ac.uk">Sean 
    Humphray</a></li>
  <li>BAC clones used in the physical map are available from the <a href="https://www.rzpd.de/">RZDP</a> 
</ul>

<hr>
Questions of comments: 
<address>Pat Edwards<br>
Institute of Neuroscience<br>
University of Oregon<br>
Eugene, OR 97403 USA<br>
fax: 503-346-4548<br>
<a href="mailto:edwards@uoneuro.uoregon.edu"> edwards@uoneuro.uoregon.edu</a> 
</address>
<hr>
<a href="/index.html">ZFIN <img src="fish_net.gif"></a> 

</body>
</html>