<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html>
	<head>
		<title>
			ZFIN Zon letter
		</title>
	</head>

<script language="JavaScript" src="/javascript/header.js"></script>

<table COLS=2 WIDTH="600" >
<tr>
<td WIDTH="210"><b><i><font size=+4>Catch of the day...</font></i></b>
<center><font face="Times New Roman,Times"><font size=+1>News from the
zebrafish research community</font></font></center>
</td>

<td WIDTH="100"><img SRC="catch.gif" BORDER=0 height=67 width=210></td>
</tr>
</table>

<hr WIDTH="100%"><b>March 25, 1999</b>
<br>
<hr WIDTH="100%">
<p>Dear Fellow Zebrafish Researchers,
<p>To help investigators in the field, I have decided to start a column
for ZFIN called 'Catch of the Day'.&nbsp; I will ask an expert in the field
to write a short paragraph defining major technical advances or new events
regarding zebrafish.&nbsp; Most of these reports will be focused on genetics,
genomics or supporting infrastructure.&nbsp; Periodically, experimental
techniques will be highlighted.&nbsp; I sincerely hope that you will find
the information useful.
<p><b>The Zebrafish Genome Initiative</b>
<p>In February 1997, several investigators of the zebrafish community met
at Children's Hospital of Boston for a meeting on the Zebrafish Genome.&nbsp;
As a result of this meeting, a plan was presented to the NIH director,
Harold Varmus, for a Zebrafish Genome Initiative.&nbsp; In May 1997, I
was asked to present the plan to the NIH Institute directors.&nbsp; Each
of the Institute directors warmly embraced the zebrafish system as an exciting
model, both relevant to understanding normal vertebrate development and
human disease.&nbsp; As a result, an NIH Institute-wide <a href="/zf_info/catch/oct98.html">Zebrafish
Genome Initiative</a> was formed, overseen by the <a href="/zf_info/catch/nih.html">Trans-NIH
Zebrafish Coordinating Committee</a> which is co-chaired by Josie Briggs
(National Institute of Diabetes and Digestive and Kidney Diseases [NIDDK])
and Tyl Hewitt (National Institute of Child Health and Human Development
[NICHD]).&nbsp; An RFA was funded which includes five individual projects
at a price tag of about $4,700,000.&nbsp; Thirteen Institutes contributed
funds, with the NIDDK responsible for managing the grants.&nbsp; Projects
led by Marnie Halpern, Steve Johnson, Will Talbot, and Len Zon began in
September 1998, and Mark Fishman joined in February 1999.
<p>The project developed by Mark Fishman will create a 0.5-1 cM microsatellite
map of the zebrafish genome.&nbsp; Steve Johnson and the group at the University
of Washington were funded to generate 100,000 expressed sequence tags (ESTs).&nbsp;
These ESTs can be utilized as candidates for genes identified in mutant
screens.&nbsp; Hans Lehrach has provided oligo-fingerprinted libraries
for this effort and these normalized libraries should contain unique genes.&nbsp;
Thus, sequencing 100,000 genes from these libraries is more efficient than
random sequencing of ESTs from different libraries.&nbsp; Several developmental
stages from zebrafish embryo libraries will be utilized as well as different
tissues.&nbsp; The Zon laboratory was funded to place 5,000 ESTs on zebrafish-hamster
or zebrafish-mouse radiation hybrid (RH) panels.&nbsp; We currently are
learning whether the Goodfellow RH Panel or the Ekker RH Panel will be
utilized for this purpose.&nbsp; Steve Johnson and John McPherson will
place an additional 5,000 ESTs on the radiation hybrid panel.&nbsp; Will
Talbot and John Postlethwait were funded to place another group of 3,000
ESTs on the meiotic map.&nbsp; A number of these genes will also be utilized
in the radiation hybrid panel so the meiotic maps and the radiation hybrid
maps can be compared.
<p>For meiotic mapping of a particular gene in your lab, it may be helpful
to contact Will or John.&nbsp; They can provide DNA samples for mapping.&nbsp;
The radiation hybrid panel by Marc Ekker is distributed throughout the
community and the Goodfellow Panel is available from Research Genetics.&nbsp;
Simple PCR reactions with oligonucleotides representing the 3' untranslated
regions of genes can be used to position them on the RH map.&nbsp; Once
a map position is known, it is possible to define whether mutants are localized
to that region by comparing the RH and meiotic maps.&nbsp; The large scale
mapping efforts in Tuebingen and Mass General have provided the map positions
of a number of mutants; by contacting those labs, it should be possible
to investigate your mutant candidate gene.&nbsp; The meiotic map data are
available on ZFIN and the RH map data will soon be released as well.&nbsp;
Using the radiation hybrid panels, my laboratory has found one candidate
that proved to be a mutant gene.
<p>Another project in the Genome Initiative is to create deletion mutants
of the entire genome.&nbsp; This is the work of Marnie Halpern and Andreas
Fritz.&nbsp; These mutants will be collected and placed in the Stock Center.&nbsp;
Deletion mutants provide a powerful resource for investigation of gene
mapping and gene function.&nbsp; Recently, an NIH grant was approved for
a <a href="/zirc/">Zebrafish
Stock Center</a> at the University of Oregon, Eugene.&nbsp; Monte Westerfield
is the PI of this grant and we understand that in the summer of 2000, the
Stock Center will make these deletion lines, as well as other mutants,
available.&nbsp; This resource is critical for our field.&nbsp; In addition,
there is a general feeling that ZFIN will be supported to develop further
informatics support for zebrafish mutants and genes and to serve as an
interface to other genome databases.&nbsp; For any questions regarding
the stock center or ZFIN, please refer to Monte Westerfield.&nbsp; He will
be submitting a 'Catch of the Day' column shortly.
<p>In February, there was another meeting of the NIH under the direction
of Dr. Varmus.&nbsp; This meeting was the <a href="/zf_info/catch/feb99.html">Non-Mammalian
Models Workshop</a>.&nbsp; At this meeting, we proposed significant increases
in funding for further efforts in zebrafish genomics.&nbsp; I have included
in this mailing, the latest report that John Postlethwait and I wrote,
regarding the resources that we hope to see developed for zebrafish research.&nbsp;
If the NIH budget does well in the year 2000, many of these studies could
be funded.&nbsp; The NIH is having another zebrafish meeting, Genomic &amp;
Genetic Tools for the Zebrafish, in May 1999.&nbsp; This will be an excellent
meeting and will result in a report to Dr. Varmus.&nbsp; Our hope is that
this will be an exciting period for zebrafish research.
<p>Please contact me with any comments.
<p>Sincerely,
<p>Leonard I. Zon, M.D.
<h3>
<b>Contacts:</b></h3>
Marc Ekker - <a href="mailto:ekker@civich.ottawa.on.ca">ekker@civich.ottawa.on.ca</a>
<br>Marnie Halpern - <a href="mailto:halpern@mail1.ciwemb.edu">halpern@mail1.ciwemb.edu</a>
<br>Steve Johnson - <a href="mailto:monte@uoneuro.uoregon.edu">sjohnson@genetics.wustl.edu</a>
<br>Will Talbot - <a href="mailto:talbot@saturn.med.nyu.edu">talbot@saturn.med.nyu.edu</a>
<br>John Postlethwait - <a href="mailto:jpostle@oregon.uoregon.edu">jpostle@oregon.uoregon.edu</a>
<br>Monte Westerfield - <a href="mailto:monte@uoneuro.uoregon.edu">monte@uoneuro.uoregon.edu</a>
<br>Len Zon - <a href="mailto:zon@rascal.med.harvard.edu">zon@rascal.med.harvard.edu</a>

<script language="JavaScript" src="/javascript/footer.js"></script>
