<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html>
	<head>
		<title>
			ZFIN ZF Monitor Vol 2(2)
		</title>
	</head>

<script language="JavaScript" src="/javascript/header.js"></script>
<b>ZEBRAFISH CODON USAGE TABLE</b><p>
<p>
By Anat Yarden<p>
<p>
Department of Chemical Immunology, The Weizmann Institute of Science, Rehovot
76100, Israel.<p>
<p>
	The pattern of choices between synonymous codons varies from one gene to
another according to the type of genome the gene occurs in (NAR 9, r43, 1981).
Thus, codon usage may be biased in a species specific manner. Therefor, the
codon usage table of an organism may prove to be useful for designing
degenerate oligonucleotide primers for PCR, cloning, etc. In addition,
knowledge of codon usage may yield a wide range of information about an open
reading frame, such as whether an open reading frame is indeed likely to be a
gene (<i>J. Heredity</i> <b>83</b>:239, 1992). Today codon usage tables are
available for many species through anonymous FTP site (132.183.190.10) from Dr.
Michael Cherry (CHERRY@FRODO.MGH.HARVARD.EDU)(see also <i>NAR 20 Suppl</i>.,
2111-2118, 1992). However, a zebrafish codon usage table has not yet been
assembled. I have pulled out of the genebank the available zebrafish sequences
(GenBank release 74.0, EMBL release 33.0), and sorted the sequence of exonic
fragments from them. Those exonic fragments were used to generate a zebrafish
codon usage table by using the "codonfrequency" program available through GCG.
Table 1 lists the genes which were included in this analysis (coding sequences
from additional gene fragments were also used, but are not listed). The
zebrafish codon usage table shown in Table 2 includes the total number of times
a specific codon appeared in all the fragments analyzed (appears in brackets)
and the preferred ratio by which the various codons were used for a specific
amino acid. Additional sequences, when available, can be added periodically to
the table in order to increase its statistical value.<p>
<p>
<b><u>Table 1. List of Genes Used for the Zebrafish Codon Usage
Analysis</u></b><u></u><p>
<u></u><p>
msh-D 	muscle segment homeobox <br>
zf-cad1	caudal homeobox<br>
epen		ependymin<br>
krox-20	zinc finger protein (krox-20)<br>
antp		antennapedia class zf'26'<br>
pax(zf-a)	drosophila paired homolog<br>
alpha-tropo	skeletal muscle alpha-tropomyosin<br>
hox(zf-22)	homeobox hox2.2<br>
hox(zf-61)	homeobox hox(zf-61)<br>
hox(zf-114)	homeobox hox (zf-114)<br>
krox-24	zinc finger protein (krox-24)<br>
hox(zf-21)	homeobox hox(zf-21)<br>
zf-54		homeotic zf-54<br>
eng		engrailed-like homeobox<br>
 zn-cad	N-cadherin (Bitzur, Kam &amp; Geiger, unpublished)
<p>
<b><u>Table 2. Zebrafish Codon Usage Table</u></b>
<p>
Gly<br>
	GGG	0.18 (41)<br>
	GGA	0.31 (70)<br>
	GGT	0.21 (49)<br>
	GGC	0.30 (68)<p>
Glu<br>
	GAG	0.71 (212)<br>
	GAA	0.29 (86)<p>
Asp<br>
	GAT	0.39 (75)<br>
	GAC	0.61 (115)<p>
Val<br>
	GTG	0.39 (67)<br>
	GTA	0.15 (26)<br>
	GTT	0.20 (35)<br>
	GTC	0.26 (46)<p>
Ala<br>
	GCG	0.16 (41)<br>
	GCA	0.22 (56)<br>
	GCT	0.31 (78)<br>
  GCC  0.31 (79)<p>
Arg<br>
	AGG	0.15 (49 )<br>
	AGA	0.24 (76)<p>
Ser<br>
	AGT	0.15 (57)<br>
	AGC	0.20 (76)<p>
Lys<br>
	AAG	0.47 (120)<br>
	AAA	0.53 (137)<p>
Asn<br>
	AAT	0.36 (73)<br>
	AAC	0.64 (132)<p>
Met<br>
	ATG	1.00 (101)<p>
Ile<br>
	ATA	0.23 (41)<br>
	ATT	0.28 (50)<br>
	ATC	0.49 (86)<p>
Thr<br>
	ACG	0.15 (42)<br>
	ACA	0.26 (71)<br>
	ACT	0.23 (62)<br>
	ACC	0.36 (98)<p>
Trp<br>
	TGG	1.00 (51)<p>
End<br>
	TGA	0.41 (9)<p>
Cys<br>
	TGT	0.47 (32)<br>
	TGC	0.53 (36)<p>
End<br>
	TAG	0.23 (5)<br>
	TAA	0.36 (8)<p>
Tyr<br>
	TAT	0.38 (54)<br>
	TAC	0.63 (90)<p>
Leu<br>
	TTG	0.13 (43)<br>
	TTA	0.09 (30)<p>
Phe<br>
	TTT	0.39 (59)<br>
	TTC	0.61 (91)<p>
Ser<br>
	TCG	0.08 (30)<br>
	TCA	0.20 (76)<br>
  TCT  0.18 (70)<br>
	TCC	0.20 (78)<p>
Arg<br>
	CGG	0.14 (43)<br>
	CGA	0.14 (45)<br>
	CGT	0.11 (35)<br>
	CGC	0.22 (70)<p>
Gln<br>
	CAG	0.64 (125)<br>
	CAA	0.36 (70)<p>
His<br>
	CAT	0.39 (44)<br>
	CAC	0.61 (69)<p>
Leu<br>
	CTG	0.34 (111)<br>
	CTA	0.08 (28)<br>
	CTT	0.15 (49)<br>
	CTC	0.21 (70)<p>
Pro<br>
	CCG	0.16 (39)<br>
	CCA	0.27 (66)<br>
	CCT	0.24 (59)<br>
	CCC	0.32 (77)<p>
<hr>
<A HREF="contents.html">Return to Contents</A><br>

</body>
<script language="JavaScript" src="/javascript/footer.js"></script>
