<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html>
	<head>
		<title>
			ZFIN ZFMonitor Vol 3(1)
		</title>
	</head>

<script language="JavaScript" src="/javascript/header.js"></script>

<H1><i>THE ZEBRAFISH SCIENCE MONITOR</i></H1>
<B><I>All the Zebrafish News That's Fit to Print</I><BR></B>
<img src="figures/logo.gif"><br>
<B>Volume 3, Issue 1, January 10, 1994, Eugene, Oregon</B><p><hr>
<B>Contents:</B><br>
<A HREF="#1">EXPRESSION OF THE NOVEL ZEBRAFISH GENE hlx-1 IN THE PRECHORDAL PLATE AND DURING DEVELOPMENT</A><p>
<A HREF="#2">BOSTON AREA ZEBRAFISH SYMPOSIUM</A><p>
<A HREF="#3">E-MAIL DISTRIBUTION LIST</A><p>
<A HREF="#4">AN INEXPENSIVE AND EASY MICROINJECTION EMBRYO-TRAY</A><p>
<A HREF="#5">UPDATE - ZEBRAFISH GENETICS AND DEVELOPMENT MEETING - UPDATE</A><p>
<A HREF="#6">SEGMENT AND CELL TYPE LINEAGE RESTRICTIONS DURING PHARYNGEAL ARCH DEVELOPMENT IN THE ZEBRAFISH EMBRYO</A><p>
<A HREF="#7">RESTRICTION OF NEURAL CREST CELL FATE IN THE TRUNK OF THE EMBRYONIC ZEBRAFISH</A><p>
<A HREF="#8">NON-ISOTOPIC <i>IN SITU</i> HYBRIDIZATION PROCEDURE FOR SECTIONED MATERIAL</A><p>
<A HREF="#9">ZF ADDRESS UPDATE</A><p>
<A HREF="#10">ZEBRAFISH REFERENCES</A><p>
<b><i><br><hr>
<H3><A NAME="1">EXPRESSION OF THE NOVEL ZEBRAFISH GENE hlx-1 IN THE PRECHORDAL PLATE AND DURING DEVELOPMENT</A></H3></i></b><br>
<br>
Anders Fjose<I>1</I>, Juan-Carlos Izpisua-Belmonte<I>2</I>, Catherine Fromental-Remain<I>3</I>, and Denis Duboule<I>2</I>. <br>
<I>1</I>Department of Biochemistry &amp; Molecular Biology, University of Bergen, &Aring;rstadveien 19, N-5009 Bergen, NORWAY.  <I>2</I>European Molecular Biology Laboratory, Meyerhofstrasse 1, D-6900 GERMANY.  <I>3</I>LGME du CNRS, Unite INSERM 184, Faculte de Medecine, 11 rue Humann, Strasbourg, FRANCE.<br>
	<br>
The zebrafish <i>hlx-1</i> gene belongs to the H2.0 subfamily of homeobox genes and is closely related to the mouse <i>Dbx</i> gene with respect to both homeodomain homology (96.7%) and neural expression during embryogenesis.  Analysis of <i>hlx-1</i> expression by <i>in situ</i> hybridization reveals several particularly interesting features.  In late gastrula embryos, <i>hlx-1</i> transcripts are detected within a circular area in the region of the presumptive rostral brain.  Subsequently, the expression domain becomes restricted to the hypoblast and undergoes dynamic changes involving conversion into a longitudinal stripe which elongates and retracts following a temporal sequence. The site of transient <i>hlx-1</i> expression along the ventral midline of the rostral neuroectoderm, which in part corresponds to the prechordal plate, suggests a role in the determination of head mesoderm as well as in patterning of the rostral brain.  As the midline stripe gradually disappears, the <i>hlx-1</i> gene becomes regionally expressed within the diencephalon and at a specific dorsoventral level along the hindbrain and spinal cord.  In the hindbrain, expression is initiated in dorsoventrally restricted transverse stripes which correlate with the segmental pattern of rhombomeres.  The stripes fuse into bilateral columns that are later converted to two series of paired transverse stripes at the rhombomere borders.  This pattern is consistent with the proposed subdivision of hindbrain segments into rhombomere centers separated by border regions.<br>
<br>
<b><br>
<H3><A NAME="2">BOSTON AREA ZEBRAFISH SYMPOSIUM</A></H3></b><br>
<br>
G. Heinrich, Department of Medicine, Section of Biomolecular Medicine, The University Hospital, 88 East Newton St., Boston, MA 02118-2393.<br>
	<br>
The implications of the study of zebrafish development for human health and disease were the subject of a one day symposium held on December 10, 1993, at Boston University Medical Center (BUMC).  The symposium was organized by G. Heinrich, S. Martin, and D. Strehlow, all from the Biomolecular Medicine Section at University Hospital, at BUMC.  Financial and staff support was provided by the Evans Medical Foundation.  	<p>
Gerhard Heinrich, from BUMC, opened the symposium with welcoming remarks and thanks to the Evans Foundation, the speakers, and his co-workers who helped organize the meeting.  	<p>
Mark Fishman, from the Massachusetts General Hospital (MGH), gave an introductory overview stressing that a unique combination of embryological and genetic techniques makes it possible to address organ development in the zebrafish.  The focus of his laboratory is in cardiovascular and intestinal systems. 	A central approach to a genetic analysis of development is the generation and study of mutants with developmental phenotypes.  Wolfgang Driever, from the MGH, discussed the generation of mutants using chemical mutagens and presented a scheme for the screening and identification of developmental mutants. Didier Stainier and Michael Pack, both from the Fishman laboratory, presented zebrafish mutants generated in the Driever laboratory that have defects in heart and gut development and related their phenotypes to human developmental diseases of the heart, aorta, and gastrointestinal systems.  	<p>
Roger Breitbart, from Children's Hospital, discussed a family of transcription factors, the MEF's, that are involved in muscle development.  Analysis of a spontaneous mutant, <i>bloodless</i>, and additional mutants including <i>spadetail</i>, by Leonard Zon, provided much new insight into hematopoiesis in the zebrafish. Jarema Malicki, from the Driever laboratory, presented eye and ear mutants obtained by the chemical method.  The naming of these mutants was imaginative and included <i>amadeus</i> and <i>plymouth rock</i>.  	<p>
The need to perfect genetic tools such as insertional mutagenesis, embryonic stem cell culture and transfection, and expression of exogenous genes in transgenic fish was discussed by Nancy Hopkins from MIT and illustrated with recent advances in her laboratory.  	<p>
David Strehlow, from the Heinrich laboratory at BUMC, talked about the fate map of blastomeres from the early embryo and presented data that demonstrate surprising effects on development of injections of retrovirus particles into the early embryo.  	<p>
The <i>trk</i> genes, a family of transmembrane tyrosine kinases that serve as high affinity receptors for neurotrophins, and play critical roles in nervous system development, was discussed by Stella Martin, also from the Heinrich laboratory of BUMC.  Surprisingly, the zebrafish <i>trk</i> gene family has five distinct members whereas only three <i>trk</i>s have been discovered in mammals so far.  	<p>
The next three presentations addressed eye development.  Julie Sandell, from BUMC, found early expression of GABA and GAD, and suggested that GABA might be a trophic factor in the developing eye. George Hyatt, from the John Dowling laboratories at Harvard, revealed that retinoic acid greatly affects eye development and the effects depend on the timing of retinoic acid treatment and include duplication of the retina.  The enzymes that synthesize retinoic acid were analyzed by Nick Marsh-Armstrong from the Walter Gilbert and John Dowling laboratories at Harvard.  He showed that at least two enzymes exist, and they are differentially regulated, possibly explaining a retinoic acid gradient in the eye but not in the body where additional mechanisms are at work.	<p>
Hazel Sive, from the Whitehead Institute at MIT, presented the last talk of the symposium.  She compared the utility of frog, chicken, fish, and mouse for genetic and embryonic studies of development. Although the zebrafish came out high on the list, a novel species which she called Fromashken, took the prize.  The Fromashken, as might be guessed from its name, combines the advantages of all four species as well as their anatomical features.  Hazel Sive illustrated that techniques that have led to rapid advances in our understanding of the molecular mechanisms of frog development can also be applied to the zebrafish embryo.<br><BR>
<b><br>
<H3><A NAME="3">E-MAIL DISTRIBUTION LIST</A></H3></b><br>
<br>
Pat Edwards, Institute of Neuroscience, University of Oregon, Eugene, OR 97403-1254<br>
	<br>
We have put together an electronic mail distribution list comprised of those on the regular zebrafish mailing list who have provided e-mail addresses.  This list can be a means of distributing information quickly and inexpensively among the zebrafish research community.  We realize, of course, that not everyone on the mailing list has access to e-mail and there are some of you who do not use it, but it can still benefit many.	<p>
Recently, a message was relayed via the distribution list and several "bounced back" because of incorrect addresses.  These are included in our "ZF Address Update" section.  If you would like to be included on our e-mail distribution list and are not, presently, please send me your current e-mail address. 	My e-mail address is edwards@uoneuro.uoregon.edu.  We also ask you to keep us apprised of your current address, phone number and FAX numbers and we will make every effort to keep our lists up to date and published regularly in the <i>Monitor</i>.<br>
<br>
<b><br>
<H3><A NAME="4">AN INEXPENSIVE AND EASY MICROINJECTION EMBRYO-TRAY</A></H3></b><br>
<br>
F. Argenton<I>1</I>, S. Bitzur<I>2</I>, and A. Yarden<I>2</I><br>
<I>1</I>Instituto Zooprofilattico delle Venezie, Via G. Orus 2. Dipartimento di Biologia, Via Trieste 75, 35121 Padova, Italy; <I>2</I>Department of Chemical Immunology, The Weizmann Institute of Science, Rehovot 76100, Israel.<br>
	<br>
We have developed a new inexpensive and easy method to position zebrafish embryos for microinjection.  Our technique is a simplification of the method published by Eric S. Weinberg (<i>Zebrafish Science Monitor</i> <b>2</b>(1):4-5) and of the method taught by Manfred Schartl in the course "Gene Expression and Regulation in Laboratory Fish", W&uuml;rzburg Germany, Feb. 1993.	<p>
Embryos in their chorions are pushed into depressions in an agarose disk which are formed with 1 mm capillaries.  The depressions are round and gently "hug" the embryos in such a way that the embryos remain in the depression during microinjections and while the micropipette is pulled out after injection.  There is, thus, no need to use a plastic cover  for impeding the movement of the embryos as described in the previous method.  In addition, once the embryos are positioned, the tray can be moved around without the embryos changing their positions.	<p>
1.  Place a few 1 mm x 5-6 cm glass capillaries (W.P.I.) on the bottom of a 90 x 15 mm petri dish. Pour 20 ml of warm 1.2% agarose (prepared in embryo medium containing 1 ppm methylene blue) over the capillaries (Fig. 1-a).  Once the capillaries start floating, push them back down to the bottom of the plate with fine forceps and arrange them in parallel rows (Fig. 1-b).	<p>
2.  When the agarose has solidified, reduce the diameter of the agarose disk by cutting around its perimeter with a razor blade (Fig. 1-c).  Turn the dish over and let the agarose disk containing the gel-coated capillaries fall onto another wet petri dish (Fig. 1-d,e).  Seal the dish onto the plate with more warm, liquid agarose (Fig. 1-f).	<p>
3.  Cover the capillaries with embryo medium and gently remove them from the agarose with fine forceps.  U-shaped grooves remain in the agarose.  The plate can be stored in the cold  and reused for several injections.	<p>
4.  Gently squeeze the embryos, still in their chorions, into the grooves and orient them  towards the injection pipette with fine forceps hooded with a plastic tip.  We prefer to orient the embryos with their animal poles towards the pipette (Fig. 1,  bottom).  More then 20 embryos can be lined up in each groove.	<p>
5.  For microinjection we move the petri dish by hand to position the embryo below the injection pipette.  We use a foot pedal controlled Eppendorf 5242 microinjector, position the micropipettes with a Leitz micromanipulator, and monitor the procedure with a Wild stereo-microscope at x32 magnification.	<p>
6.  Following injection, lift the embryos up out of the grooves. <br>
<br>
<b><br>
<H3><A NAME="5">UPDATE - ZEBRAFISH GENETICS AND DEVELOPMENT MEETING - UPDATE</A></H3></b><br>
from the organizers: W. Driever, J. Eisen, D. Grunwald, and C. Kimmel<br>
<b><br>
REMINDER</b><br>
<br>
The first open invitation meeting on Zebrafish Genetics and Development will take place at Cold Spring Harbor Laboratory April 27-May 1.  ABSTRACTS AND REGISTRATION FORMS ARE DUE FEBRUARY 9.  Forms can be obtained from Meetings Coordinator, Cold Spring Harbor Laboratory, 1 Bungtown Road, Cold Spring Harbor, New York 11724-2213. TEL: 516-367-8346; FAX: 516-367-8845<br>
   <b><br>
MEETING FORMAT</b><br>
<br>
Given the response we have already received, it looks like a large portion of the zebrafish community will attend the meeting.  Therefore, it is likely that TALKS will be limited to one per laboratory.  However, everyone is invited to present a POSTER.<br>
   <br>
The current idea about the format of the meeting is the following:  Themeeting will include 10 sessions of talks and at least 2 sessions of posters. Each session of scientific talks will be led by a chairperson who will, in a 20-25 minute introductory talk, provide an overview of the contemporary questions in the field and will present work from her/his own laboratory.  The other talks in the session will be about 15 minutes.  In addition, there will be a keynote address and a summary address.  There will also be a community organizational session to discuss issues of importance to the entire community.  WE INVITE BRIEF ABSTRACTS OR SUGGESTIONS FOR THE COMMUNITY ORGANIZATIONAL SESSION TO BE SUBMITTED DIRECTLY TO:Judith S. Eisen, Institute of Neuroscience, University of Oregon, Eugene, Oregon 97403 USA; FAX:503-346-4548; e-mail:eisen@uoneuro.uoregon.edu.  ALSO: If you anticipate requiring CHILD CARE AT THE MEETING, please contact David Grunwald FAX:801-581-5374; e-mail: grunwald@gene1.utah.med.edu<br>
<b><br>
LOW COST DOMITORY ACCOMMODATIONS FOR CSH MEETING</b><br>
<br>
Cold Spring Harbor may be able to provide a limited number (50 to 100) of meeting participants with low cost accommodation in a dormitory at a local University. Transportation to and from the University would be provided by CSH. The costs for the meeting participants staying at this location will be reduced by 100 to 150 Dollar as compared to "regular" participants, some of which will stay at CSH, but many at a Hotel about as far away from CSH as the University.We need to know as soon as possible how many are interested in this option.<br>
<br>
Please:<br>
	<br>
1. e-mail (driever@helix.MGH.harvard.edu) or fax (USA-617-726-5806) a short note to Wolfgang Driever,and	<p>
2. indicate your interest in low cost accommodation on the registration form.<br>
<br>
We will then try to negotiate a sufficient number of low cost accommodations.<br>
<b><br>
Please note:</b><br>
	<br>
1. We can not guarantee low cost accommodation to everyone who applies -we will offer this opportunity preferentially to graduate students or postdocs (who might otherwise have no chance to come to the meeting) and use the first come first serve basis.	<p>
2. The low cost accommodation does not change the deposit -anyone who wants housing has to include the 200 Dollar deposit with the registration as indicated on the form.<br>
<br>
Please reply by January 20!<br>
<br>
<b><br>
<H3><A NAME="6">SEGMENT AND CELL TYPE LINEAGE RESTRICTIONS DURING PHARYNGEAL ARCH DEVELOPMENT IN THE ZEBRAFISH EMBRYO</A></H3></b><br>
(In press, <i>Development</i>)<br>
<br>
T.F. Schilling and C.B. Kimmel.<br>
Institute of Neuroscience, University of Oregon, Eugene, OR 97403-1254<br>
	<br>
In zebrafish, the segmental series of pharyngeal arches is formed predominantly by two migratory cell types, neural crest and paraxial mesoderm, which arise in the early embryo. Neural crest cells migrate ventrally out of the neuroepithelium and into the arches to form cartilage, neurons, glia and pigment cells.  Surrounding mesoderm generates muscles and endothelia.  We labeled individual pharyngeal precursor cells with fluorescent dyes and found that their clonal progeny were confined to single segments and generated single cell types. When a neural crest or mesodermal cell was marked before migration into the pharynx, its progeny dispersed but generally remained confined to a single arch primordium.  Such segmental restrictions arose first in the most rostral arches, mandibular and hyoid, and progressed caudally. The phenotypes of progeny generated by single cells were examined in the mandibular arch. Clones derived from premigratory neural crest cells generally did not contribute to more than one cell type.  Further, the progenitors of some cell types were spatially separated in the premigratory crest.  In particular, neurogenic crest cells were situated further laterally than cells that generate cartilage and connective tissues, while pigment and glial cell progenitors were more evenly distributed.  Based on these results, we suggest that arch precursors may be specified as to their eventual fates before the major morphogenetic movements that form the arch primordia. Further, cell movements are restricted during segmentation establishing a group of arch precursors as a unit of developmental patterning, as in the fashion of vertebrate rhombomeres or segmental lineage compartments in <i>Drosophila</i>.<br>
<br>
<b><br>
<H3><A NAME="7">RESTRICTION OF NEURAL CREST CELL FATE IN THE TRUNK OF THE EMBRYONIC ZEBRAFISH</A></H3></b><br>
(In press, <i>Development</i>)<br>
<br>
D.W. Raible and J.S. Eisen.  <br>
Institute of Neuroscience, University of Oregon, Eugene, Oregon 97403<br>
<br>
To learn when cell fate differences first arise in the zebrafish trunk neural crest, individual premigratory crest cells were labeled intracellularly with fluorescent vital dyes, followed in living embryos, and complete lineages recorded.  Although some of the earliest cells to migrate produced derivatives of multiple phenotypes, most zebrafish trunk neural crest cells appear to be lineage-restricted, generating type-restricted precursors that produce single kinds of derivatives.  Further, cells that produce derivatives of multiple phenotypes appear to do so by first generating type-restricted precursors.  Among the various types of derivatives, sensory and sympathetic cells arise only from early migrating crest cells.  Some type-restricted precursors display cell-type specific characteristics while still migrating.  Taken together, these observations suggest that some trunk neural crest cells are specified before reaching their final locations. <br>
<br>
<b><br>
<H3><A NAME="8">NON-ISOTOPIC <i>IN SITU</i> HYBRIDIZATION PROCEDURE FOR SECTIONED MATERIAL</A></H3></b><br>
(In press, <i>Trends in Genetics</i>)<br>
<br>
U. Str&auml;hle, P. Blader, J. Adam and P. W. Ingham<br>
Imperial Cancer Research Fund, Developmental Biology Unit, Molecular Embryology Laboratory, South Parks Road, Oxford OX1 3PS, UK<br>
<b><br>
Fixation and Sectioning</b>	<p>
1. Fix zebrafish embryos in BT-Fix  at 4&#176;C overnight.	<p>
2. Wash embryos twice in BT-Fix minus paraformaldehyde.<p>
Remove chorions with sharp watchmaker forceps.	<p>
3. Although tissue preservation is best when material is sectioned immediately after fixation, embryos can be stored at this point.  For long term storage dehydrate embryos (30% ethanol in PBS, 50% ethanol in PBS, 70% ethanol in H2O, 3-5 min each step) and store in freezer (-20&#176;C). 	<p>
4. Embed embryos in 1.5% agar (Gibco BRL), 5% sucrose.	<p>
5. After trimming, transfer agar blocks to 30% sucrose, 0.1% azide at 4&#176;C until blocks have sunk (usually overnight).	<p>
6. Cut 15&#181;m cryostat sections.  Transfer sections to 'TESPA'-coated slides.  Allow sections to air-dry for 2 hrs at room temperature, then store at -20&#176;C in an air-tight box over silica gel.  We have stored sections this way for up to one year.<br>
<b><br>
Hybridization</b>	<br>
1. Dilute antisense RNA probe (see Solutions and Materials) in hybridization buffer (usually 100-fold but the optimal dilution can vary with different probes).  In double-labeling experiments, mix the digoxygenin-labeled and the fluorescein-labeled probes at the appropriate dilutions.  Denature the probe mix at 70&#176;C for 5 min immediately before applying to the sections.	<p>
2. Add 30-50&#181;l diluted probe to each slide and coverslip.  Initially the probe might not cover all sections completely.  This does not present a problem as it spreads out during hybridization.	<p>
3. Put slides on filter paper soaked with 50% formamide, 1x Salt in a sealed box. Hybridize at 55&#176;C for 8 hrs to overnight.	<p>
4. Transfer slides into a slide rack and submerge in wash buffer 1 (prewarmed to 65&#176;C) for 15 min to allow coverslips to fall off.  Repeat wash in wash buffer 1 at 65&#176;C for 30 min.  Wash twice in wash buffer 2 at 65&#176;C for 30 min.  Transfer to PBS at room temperature for 5 min and then block for 30 min in PBT.<p>
<b><br>
Antibody staining</b>	<br>
1. Add 30 &#181;l antibody (see Solutions and Material) to each slide and coverslip.  Incubate at room temperature for 30 min or in a box on filter paper soaked with PBS at 4&#176;C overnight.	<p>
2. Wash four times for 10 to 20 min in PTw.	<p>
3. Transfer slides into staining jars with NBT/BCIP or Vector Red staining solution and allow color to develop in the dark at room temperature for several hours to overnight.  Slides can be removed, checked for staining and returned to the staining jar.	<p>
4. Wash in PTw for 10 min.  Apply second antibody (repeating steps 1 to 4) or dehydrate (30, 50, 75, 95 100% ethanol, 1 min each step), clear in histoclear and mount. <b><p>
Solutions and Material</b><u><p>
BT -Fix :</u> 4% (w/v) paraformaldehyde, 4% (w/v) sucrose, 0.12mM CaCl2, 0.1M Na-phosphate pH 7.4.  Can be stored at 4&#176;C for up to a week.<u><p>
10xPBS</u>: 2.5M NaCl, 0.2M Na-phosphate pH 7.4<p>
'<u>TESPA'-coated slides</u>: Wash slides overnight in 1% (v/v) HCl, 70% (v/v) ethanol.  Rinse in H2O and dry at 70&#176;C.  Submerge slides in 2% (v/v) TESPA (3'aminopropyl-triethoxy silane, Sigma) in acetone for about 10 sec; rinse in acetone for 10 sec; then wash in H2O for 10 sec and bake at 160&#176;C.<u><p>
Digoxygenin- and fluorescein-labeled antisense RNA probes</u>: Synthesize probes following the protocol recommended by the supplier of the modified digoxygenein and flourescein nucleotides (Boehringer, Mannheim).  It is better to use transcripts longer than 1 kb, although shorter transcripts will work but with some loss in sensitivity.  Probes do not need to be hydrolyzed and are stored in 50% formamide at -20&#176;C.<u><p>
100x Denhardt's: </u>2% (w/v) bovine serum albumin, 2% (w/v) Ficoll, 2%  (w/v) polyvinyl pyrrolidone<u><p>
10xSalt</u>: 3M NaCl, 100mM Na phosphate, 100mM EDTA, 100mM Tris/HCl pH 7.5 (modified)<u><p>
Hybridization buffer: </u>1xSalt, 50% (v/v) formamide (Fluka p.a.), 10% (w/v) dextran sulphate (Pharmacia), 1mg/ml rRNA (Sigma R7125), 1x Denhardt's<u><p>
20xSSC </u>: 3M NaCl, O.3M Na-citrate<u><p>
Wash-buffer 1</u>: 50% (v/v) formamide, 2x SSC<u><p>
Wash-buffer 2:</u> 25% (v/v) formamide, 1xSSC, 0.5xPBS<u><p>
PBT</u>: 1xPBS, 0.2% (w/v) BSA, 0.1% (v/v) Tween 20<u><p>
Acetone powder</u>:<b> </b>Sacrifice adult<b> </b>zebrafish by treatment in 3-amino benzoic acidethyl ester (40 mg/ml, pH 7).  Freeze fish in liquid nitrogen and grind under nitrogen in a mortar to a fine powder.  Transfer the powder into a centrifuge tube, add cold acetone and keep on ice for 30 min.  Occasionally, shake vigorously.  Spin in cooled Sorvall centrifuge (HB4 rotor), 10 krpm for 10 min and resuspend pellet in ice-cold acetone, keep on ice for 10 min, shake occasionally. Repeat centrifugation and resuspend in acetone.  Dry on a filter paper and store at 4&#176;C  in a sealed vial.<u><p>
Preabsorption of anti-digoxygenin and anti-fluorescein antibodies</u>: Dilute anti-digoxygenin (DIG) and anti-fluorescein alkaline phosphatase-conjugated antibodies (Boehringer Mannheim) 1/400 and 1/100 in PBT, respectively.  Incubate antibodies with 6mg/ml acetone powder at 4&#176;C overnight.  Centrifuge to remove tissue debris and dilute preabsorbed anti-DIG and anti-fluorescein antibody in PBT to give 1/2000 and 1/500 final dilutions, respectively.<u><p>
PTw</u>: 1xPBS, 0.1% Tween 20<u><p>
NBT/BCIP staining solution </u>(Prepare fresh): 100 mM NaCl, 50 mM MgCl2,100 mM Tris pH 9.5, 0.1% Tween 20, 5 mM levamisole, 0.34 mg/ml<b> </b>nitroblue tetrazolium salt (NBT, Gibco BRL) and 0.175 mg/ml 5-bromo-4-chloro-3-indolyl-phosphate (BCIP, Gibco BRL)<u><p>
Vector Red Staining solution</u> Prepare according to manufacturer's instructions (Vector Labs). The red precipitate obtained with the Vector Red Alkaline phosphatase substrates fluoresces strongly using a rhodamine filter set.<br>
<br>
<b><br>
<H3><A NAME="9">ZF ADDRESS UPDATE</A></H3> <BR>
Since 10/25/93</b><br>
<b><br>
ADDITIONS</b><br>
<br>
Pia Aanstad<br>
Department of Physiology<br>
University College London<br>
Gower Street<br>
London WC1E 6BT, UK<br>
44-71-836-8851/ FAX 44-71-497-9078<br>
<br>
Sharon Amacher<br>
Institute of Neuroscience<br>
1254 University of Oregon<br>
Eugene, OR 97403-1254<br>
(503) 346-4506/ FAX (503) 346-4548<br>
amacher@uoneuro.uoregon.edu<br>
<br>
Kimiko Amanuma<br>
Laboratory of Gene Technology &amp; Safety<br>
Inst/Physical &amp; Chemical Res<br>
3-1-1 Koyadai, Tsukuba, Ibaraki 305<br>
JAPAN<br>
<br>
C. Ramana Bhasker<br>
Department of Biochemistry<br>
University of Adelaide<br>
Adelaide, AUSTRALIA 5005<br>
61-8-336-2355 (Phone &amp; FAX)<br>
<br>
Johannes Beckers<br>
Lab/d'Embryolgie moleculaire <br>
  et Morphogenese, Sciences III<br>
30, quai Ernest-Ansermet<br>
Ch-1211 Gene<LI>ve 4, SWITZERLAND<br>
beckers@sc2a.unige.ch<br>
<br>
Lucy Byrnes<br>
National Diagnostics Centre<br>
University College<br>
Galway, IRELAND<br>
353-91-586559/FAX 353-91-586570<br>
0066497s@bodkin.ucg.ie<br>
<br>
Gi-Iou Chen<br>
Department of Zoology<br>
Academia Sinica<br>
Non-Gun, Taipei, Taiwan<br>
ROC<br>
<br>
Jan-Nian Chen<br>
Cardiovascular Research Center<br>
Massachusetts General Hospital-East<br>
149 13th St., 4th Floor<br>
Charlestown, MA 02129<br>
<br>
Matthias Conrad<br>
Institut f&uuml;r Zoologie<br>
University of Mainz<br>
Postfach 3980<br>
6500 Mainz, GERMANY<br>
49-6131-393500/FAX 49-6131-394652<br>
<br>
Geoffrey Doerre<br>
309 Goddard Labs<br>
University of Pennsylvania<br>
Philadelphia, PA 19104<br>
(215) 898-2640/FAX (215) 898-8780<br>
gdoerre@sas.upenn.edu<br>
<br>
Bernd Fritzsch<br>
Creighton University<br>
Dept/Biomedical Sciences<br>
Division of Anatomy<br>
Omaha, NE 68178<br>
(402) 280-2915/ FAX (402) 280-5556<br>
berndfri@bif.creighton.edu<br>
<br>
Ruowen Ge<br>
Roche Institute/Molecular Biology<br>
340 Kingsland St.<br>
Nutley, NJ 07110<br>
(201) 235-2685/ FAX (201) 235-2318<br>
<br>
Michel Janowski<br>
Vlaamse Instelling voor <br>
  Technologisch Onderzoek<br>
Boeretang 200<br>
B-2400 Mol, BELGIUM<br>
32-14-333-111/FAX 32-14-320-372<br>
<br>
Robert Karp<br>
Division of Basic Research<br>
Nat Inst/Alcohol Abuse &amp; Alcoholism<br>
6000 Executive Blvd., Room 402<br>
Rockville, MD 20892<br>
(301) 443-4223/ FAX (301) 594-0673<br>
rkarp@aoaa1.ssw.dhhs.gov<br>
<br>
Robert Kelsh<br>
Max-Planck-Institut/Entwicklungs<br>
Spemannstrabe 35/III<br>
D-72076 Tubingen 1, GERMANY<br>
7071-601936 / FAX 7071-601300<br>
kelsh@zappa.mpib-tuebingen.mpg.de<br>
<br>
Chris Langdon<br>
Hatfield Marine Science Center<br>
Newport, OR 97365<br>
(503) 867-0231/ FAX (503) 867-0105<br>
langdonc@ccmail.orst.edu<br>
<br>
Gilles Leclerc<br>
Dept/Biological Sciences<br>
University of South Carolina<br>
700 Sumter St. (CLS 515)<br>
Columbia, SC 29208<br>
(803) 777-1017/ FAX (803) 777-4002<br>
fishtec@biol.scarolina.edu<br>
<br>
Manuel Mari-Beffa<br>
Depto. Biologia Celular y Genetica<br>
Facultad de Ciencias/Campus de Teatinos<br>
Universidad de Malaga<br>
29071 Malaga, SPAIN<br>
345-2131954/FAX 345-2132000<br>
beffa@ccuma.uma.es<br>
<br>
Sharyn Marks<br>
Department of Biological Sciences<br>
Humboldt State University<br>
Arcata, CA 95521-8299<br>
(707) 826-3245/FAX (707) 826-3201<br>
sharynm@aol.com<br>
<br>
Rita Meyer<br>
Biology Department<br>
Rutgers University<br>
3rd &amp; Penn Sts.<br>
Camden, NJ 08102<br>
(609) 225-6336/ FAX (609) 225-6312<br>
rimeyer@crab.rutgers.edu<br>
<br>
Naomasa Miki<br>
Department of Pharmacology I<br>
Osaka University School of Medicine<br>
2-2 Yamadaoka, Suita 565<br>
Osaka 565, JAPAN<br>
81-6-879-3520/FAX 81-6-879-3529<br>
<br>
Cecilia Moens<br>
Institute of Neuroscience<br>
1254 University of Oregon<br>
Eugene, OR 97403-1254<br>
(503) 346-4506/ FAX (503) 346-4548<br>
moens@uoneuro.uoregon.edu<br>
<br>
Carey Phillips<br>
Department of Biology<br>
Bowdoin College<br>
Brunswick, ME 04011<br>
(207) 725-3573<br>
cphillip@polar.bowdoin.edu<br>
<br>
Renate Reimschuessel<br>
UMAB Aquatic Pathobiology Center<br>
Dept/Pathology, University/Maryland<br>
10 S. Pine St.<br>
Baltimore, MD 21201-1192<br>
(410) 706-7230/ FAX (410) 706-8414<br>
<br>
Michael C. Schneider<br>
Renal Division<br>
Brigham &amp; Women's Hospital<br>
MRB 318, 75 Francis St.<br>
Boston, MA 02115<br>
(617) 732-5884/ FAX (617) 732-6392<br>
genemi@harvarda.harvard.edu<br>
<br>
Minoru Tanaka<br>
Dept/Reproductive Biology<br>
National Institute/Basic Biology<br>
Okazaki, 444 JAPAN<br>
81-564-55-7556/FAX 81-564-53-7400<br>
<br>
Heine Trepte<br>
Max Planck Inst/Biophysikalische Chemie<br>
Kar-Friedrich-Bonhoeffer-Institut<br>
Am Fa&szlig;berg, Postfach 968<br>
D-3400 G&ouml;ttingen-Nikolausberg, GERMANY<br>
<br>
Huey-Jen Tsay<br>
National Yang-ming Medical College<br>
Institute of Neuroscience<br>
Shih-Pai, Taipei, TAIWAN<br>
ROC<br>
<br>
Klaus Unsicker<br>
Dept/Anatomy &amp; Cell Biology III<br>
University of Heidelberg<br>
Im Neuenheimer Feld 307<br>
D-69120 Heidelberg, GERMANY<br>
49-6221-569227/FAX 49-6221-565604<p>
James F. Welsh<br>
Department of Zoology<br>
Humboldt State University<br>
Arcata, CA 95521<br>
<br>
Kate Whitlock<br>
Institute of Neuroscience<br>
1254 University of Oregon<br>
Eugene, OR 97403-1254<br>
(503) 346-4596/ FAX (503) 346-4548<br>
whitlock@uoneuro.uoregon.edu<br>
<br>
Jen-Leih Wu<br>
Institute of Zoology<br>
Academia Sinica<br>
Nankang, Taipei 11529<br>
Taiwan, ROC<br>
FAX 886-2-7899503<br>
<br>
<b><br>
CORRECTIONS</b><br>
<br>
Angel Amores<br>
FAX 34-5-2132000<br>
a_amores@ccuma.uma.es<br>
<br>
Werner Hoffmann<br>
Institute f&uuml;r Biochemie (Med. Fak.)<br>
Otto-von-Guericke-Universit&auml;t<br>
Leipziger Stra&szlig;e 44<br>
D-39120 Magdeburg, GERMANY<br>
49-391-67-2895/ FAX 49-391-67-2898<br>
<br>
De-Yu Lu<br>
357 East 14th Avenue  Apt 1B<br>
Columbus, OH 43201-1926<br>
<br>
Paul Z. Myers<br>
Department of Biology<br>
Temple University<br>
Philadelphia, PA 19122<br>
<br>
Paolo Sordino<br>
Dep. Zoologie et Biologie Animale<br>
Embryol Molec/ Morphogen, Sci III<br>
Univ/Geneve, Quai E.Ansermet 30<br>
CH-1211, Geneve 4, SWITZERLAND<br>
22-7026779/ FAX 22-7811747<br>
sordino@sc2a.unige.ch<br>
<br>
Hiroyuki Takeda<br>
Dept/Molecular Biology<br>
School of Science<br>
Nagoya University<br>
Chikusa-ku, Nagoya, Japan 464-01<br>
81-52-782-0943/FAX 81-52-78-0954<br>
<br>
<b><br>
DELETIONS</b><br>
<br>
Adolf Maas<br>
Department of Biochemistry<br>
University of Washington<br>
Mailstop SJ-70<br>
Seattle, WA 98195<br>
("Addressee Unknown")<br>
<br>
<b><br>
E-MAIL ADDRESSES WHICH NO LONGER WORK.....</b>(Please correct)<br>
<br>
Sue Barnes<br>
brehm@usmepg.bitnet<br>
<br>
Richard Behringer<br>
richard_behringer.molec_genetic@darcqm.mda.uth.tmc.edu<br>
<br>
Ira Blitz<br>
lamins@biovax.uchicago.edu<br>
<br>
Norman Maclean<br>
n.maclean@southampton.ac.uk<br>
<br>
Michael Mote<br>
mim35@temple.edu<br>
<br>
Mike Rust<br>
miker@max.u.washington.edu<br>
<br>
Henri Stroband<br>
henri_stroband@ontw.edc.wau.nl<br>
<br>
Gary Stuart<br>
lgstuart@indst.indstate.edu<br>
<br>
Filip Volckaert<br>
fgdcr01@cc1.kuleuven.ac.be<br>
<br>
Ellen Wilson<br>
wilson@genetics.med.utah.edu<br>
<br>
Peter Wolbert<br>
wolbert@vax.r2.uni-wuerzburg.dbp.de<b><p>
<H3><A NAME="10">ZEBRAFISH REFERENCES</A></H3></b><br>
(Since 10/25/93)<br>
<br>
<br>
Alestrom, P., G. Kisen, H. Klungland, and O. Andersen (1992) Fish gonadotropin-releasing hormone gene and molecular approaches for control of sexual maturation: development of a transgenic fish model. <i>Mol. Mar. Biol. Biotechnol.</i> <b>1</b>:376-379.<p>
Bierkamp, C. and J.A. Campos-Ortega (1993) A zebrafish homologue of the <i>Drosophila</i> neurogenic gene <i>Notch</i> and its pattern of transcription during early embryogenesis.  <i>Mech. Dev.</i> <b>43</b>:87-100.<p>
Braunbeck, T., P. Burkhardt-Holm, G. Gorge, and R. Nagel (1992) Rainbow trout and zebrafish, two models for continuous toxicity tests: relative sensitivity species and organ specificity in cytopathologic reaction of liver and intestines to atrazine.  <i>Schriftenr-Ver-Wasser-Boden-Lufthyg.</i> <b>89</b>:109-145.<p>
Buchmann, A., R. Wannemacher, E. Kulzer, D.R. Buhler, and K.W. Bock (1993) Immunohistochemical localization of the cytochrome P450 isozymes LMC2 and LM4B (P4501A1) in 2,3,7,8-tetrachlorodibenzo-<i>p</i>-dioxin-treated zebrafish (<i>Brachydanio rerio</i>).  <i>Toxicol. Appl. Pharmacol.</i> <b>123</b>:160-169.<p>
Buono, R.J. and P.J. Linser (1992) Transient expression of RSVCAT in transgenic zebrafish made by electroporation.  <i>Mol. Mar. Biol. Biotechnol</i>. <b>1</b>:271-275.<p>
Collodi, P., Y. Kamei, A. Sharps, D. Weber, and D. Barnes (1992) Fish embryo cell cultures for derivation of stem cells and transgenic chimeras.  <i>Mol. Mar. Biol. Biotechnol</i>. <b>1</b>:257-265.<p>
Dorit, R.L., O. O'Hara, and W. Gilbert (1993) One-sided anchored polymerase chain reaction for amplification and sequencing of complementary DNA.  <i>Meth. Enzymol.</i> <b>218</b>:36-47.<p>
Driever, W. and Z. Rangini (1993) Characterization of a cell line derived from zebrafish (<i>Brachydanio rerio</i>).  <i>In Vitro Cell Dev. Biol.</i> <b>29A</b>:749-754.<p>
Ekker, M., M.-A. Akimenko, R. BreMiller, and M. Westerfield (1992) Regional expression of three homeobox transcripts in the inner ear of zebrafish embryos.  <i>Neuron</i> <b>9</b>:27-35.<p>
Halpern, M.E., R.K. Ho, C. Walker, and C.B. Kimmel (1993) Induction of muscle pioneers and floor plate is distinguished by the zebrafish <i>no tail</i> mutation.  <i>Cell</i> <b>75</b>:99-111.<p>
He, L., Z. Zhu, A.J. Faras, K.S. Guise, P.B. Hackett, and A.R. Kapuscinski (1992) Characterization of AluI repeats of zebrafish (<i>Brachydanio rerio</i>).  <i>Mol. Mar. Biol. Biotechnol.</i>) <b>1</b>:125-135.<p>
Helde, K.A. and D.J. Grunwald (1993) The DVR-1 (Vg1) transcript of zebrafish is maternally supplied and distributed throughout the embryo.  <i>Dev. Biol.</i> <b>159</b>:418-426.<p>
Ivics, Z., Z. Izsv'ak, and P.B. Hackett (1993) Enhanced incorporation of transgenic DNA into zebrafish chromosomes by a retroviral integration protein.  <i>Mol. Mar. Biol. Biotechnol</i>. <b>2</b>:162-173.<p>
Joly, J.S., M. Maury, C. Joly, P. Duprey, H. Boulekbache, and H. Condamine (1992) Expression of a zebrafish caudal homeobox gene correlates with the establishment of posterior cell lineages at gastrulation.  <i>Differentiation</i> <b>50</b>:75-87.<p>
Kane, D.A. and C.B. Kimmel (1993) The zebrafish midblastula transition.  <i>Development</i> <b>119</b>:447-456.<p>
Kispert, A. and B.G. Herrmann (1993) The Brachyury gene encodes a novel DNA binding protein.  <i>EMBO J.</i> <b>12</b>:3211-3220.<p>
Lee, R.K., R.C. Eaton, and S.J. Zottoli (1993) Segmental arrangement of reticulospinal neurons in the goldfish hindbrain.  <i>J. Comp. Neurol.</i> <b>329</b>:539-556.<p>
Miklos, G.L. (1993) Molecules and cognition: the latterday lessons of levels, language, and lac. Evolutionary overview of brain structure and function in some vertebrates and invertebrates.  <i>J. Neurobiol.</i> <b>24</b>:842-890.<p>
Moav, B., Z. Liu, L.D. Caldovic, M.L. Gross, A.J. Faras, and P.B. Hackett (1993) Regulation of expression of transgenes in developing fish.  <i>Transgen. Res.</i> <b>2</b>:153-161.<p>
Molven, A., I. Hordvik, P.R. Njolstad, M. van Ghelue, and A. Fjose (1992) The zebrafish homeobox gene <i>hox[zf-114]</i>: primary structure, expression pattern and evolutionary aspects.  <i>Int. J. Dev. Biol.</i> <b>36</b>:229-237.<p>
Muller, F., Z. Lele, L. V'ardi, L. Menczel, and L. Orb'an (1993) Efficient transient expression system based on square pulse electroporation and <i>in vivo</i> luciferase assay of fertilized fish eggs.  <i>FEBS Lett</i>.<b>324</b>:27-32.<p>
Nirenberg, S. and C. Cepko (1993) Targeted ablation of diverse cell classes in the nervous system <i>in vivo</i>. <i>J. Neurosci.</i> <b>13</b>:3238-3251.<p>
Powers, D.A., L. Hereford, T. Cole, T.T. Chen, C.M. Lin, K. Kight, K. Creech, and R. Dunham (1992) Electroporation: a method for transferring genes into the gametes of zebrafish (<i>Brachydanio rerio</i>), channel catfish (<i>Ictalurus punctatus</i>), and common carp (<i>Cyprinus carpio</i>).  <i>Mol. Mar. Biol. Biotechnol</i>. <b>1</b>:301-308.<p>
Selman, K., R.A. Wallace, A. Sarka, and X. Qi (1993) Stages of oocyte development in the zebrafish, <i>Brachydanio rerio</i>.  <i>J. Morphol.</i> <b>218</b>:203-224.<p>
Stainier, D.Y.R., R.K. Lee, and M.C. Fishman (1993) Cardiovascular development in the zebrafish.  I. Myocardial fate map and heart tube formation.  <i>Development</i> <b>119</b>:31-40.<p>
S&uuml;ltmann, H., W.E. Mayer, F. Figueroa, C. O'hUigin, and J. Klein (1993) Zebrafish <i>Mhc</i> class II a chain-encoding genes: polymorphism, expression, and function.  <i>Immunogenetics</i> <b>38</b>:408-420.<p>
Van Asselt, E., F. De Graaf, and W. Van Raamsdonk (1993) Ultrastructural characteristics of zebrafish spinal motoneurons innervating glycolytic white, and oxidative red and intermediate muscle fibers. <i>Acta Histochem.</i> <b>95</b>:31-44.<p>
Wilson, E.T., K.A. Helde, and D.J. Grunwald (1993) Something's fishy here--Rethinking cell  movements and cell fate in the zebrafish embryo.  <i>Trends Genet.</i> <b>9</b>:348-352.

</body>

<script language="JavaScript" src="/javascript/footer.js"></script>
