<TITLE>The Zebrafish Database Class:  MAP_MARKER</TITLE>
</HEAD>
<BODY>
<CENTER><H4>Data Descriptions for The Zebrafish Database (ZDB)</H4>
</CENTER>
<H2>SECTION II:  Experimental Data</H2>
<H3>Abstract Class:  DATA_ITEM</H3>
<H4>Class:  MAP_MARKER</H4>
<hr>
Return to the <A HREF="data_model.html"> Table of Contents</a>
<hr>

<P>	<B>Role: </B> The purpose of this class is to capture information that is common to all kinds of map markers (with the exception of mapped mutations which are represented by the CHROMOSOME and ALTERATION classes). Thus, this class represents data attributes that are shared by cloned genes, ESTs, RAPDs, SSCPs, RFLPs, and STSs. The CLONED_GENE and EST classes extend this class (as subclasses) to include information specific to, respectively, cloned genes and ESTs. The remaining markers (i.e. RAPDs, STSs, RFLPs, and SSCPs) are represented by directly instantiating the MAP_MARKER class, because all their are included in this class.
	
<P>	<B>How/Who enters:</B>  Initially, the DB Admin enters all known map markers. Authorized submitters are also allowed to create and enter new MAP MARKER records into the database.
	
<P>	<B>Security:</B>  Only authorized submitters are permitted to enter new MAP MARKERS and update existing ones. Exceptions are the primary_pub, record_status, and name fields which, once specified, can be changed only via the DB Admin.
<BR>
<BR>
<BR>
<TABLE>
<TR><TD><B><BIG>Superclasses:</B></TD> <TD> <B>GENERIC_OBJECT</B></BIG></TD></TR>
<TR><TD></TD><TD><B>DATA_ITEM</B></TD><TD>comments, primary_pub, related_pubs, record_status</TD></TR>
</TABLE>
			
<H4>ATTRIBUTES:</H4>
<P>	<B>name (TEXT, REQUIRED).</B> Contains the full text name of the map marker. 
<P>	<B>marker_type (closed set:  cloned gene, EST, SSCP, RAPD, RFLP, STS).</B> Selected from the closed set listed here, this field identifies the type of the marker. This attribute is useful both to users, to constrain searches, and to the interface, to determine how to display a given marker (i.e., may vary by marker type).
<P>	<B>map_location (TEXT, OPTIONAL). </B>Describes the genetic map location of the map marker, if known. As with other map locations, users may enter the actual data from mapping crosses, with the location computed from that data to periodically update this field, or they may enter the location derived from their results.
<P>	<B>discoverer (TEXT, OPTIONAL). </B> The ZDB_ID of the person who discovered the marker.
<P>	<B>amp_product (AMP_PRODUCT). </B>THIS FIELD DOES NOT PHYSICALLY EXIST IN THIS RECORD; it is generated by joining with AMP_PRODUCT. With the exception of cloned genes and ESTs, which may<I> sometimes</I> have amplification products associated with them, all map markers must have an AMP_PRODUCT record associated with them, specifying the primer(s) used to segregate the marker.

<H3>Open Issues/Questions:</H3>  
<P>			•  Still not clear how to handle map location! Are mapping crosses entered or not? Is “MAPPING_CROSS” another data category? Where would one expect to enter such data in the interface? Can we compute the “average” location based on such cross data? How often should we recompute?

<hr>
<B>Data Descriptions for The Zebrafish Database (ZDB)</B><br>
For questions or comments, email: 
<a href=mailto:edwards@uoneuro.uoregon.edu> edwards@uoneuro.uoregon.edu</a><br>
Go to <A HREF="sectionii_4.html"> next section</a><br>
Return to the <A HREF="data_model.html"> Table of Contents</a>
<HR>
<A HREF="/index.html">ZFIN  <img src="Fish_Net.GIF"></A>
</BODY>
</HTML>
