<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html>
	<head>
		<title>
			The Zebrafish Database Class: ALTERATION
		</title>
	</head>

<script language="JavaScript" src="/javascript/header.js"></script>

<CENTER><H4>Data Descriptions for The Zebrafish Database (ZDB)</H4>
</CENTER>
<H2>SECTION II:  Experimental Data</H2>
<H3>Abstract Class:  DATA_ITEM</H3>
<H4>Class:  ALTERATION</H4>
<hr>
Return to the <A HREF="data_model.html"> Table of Contents</a>
<hr>

<P>	<B>Role:</B>  The ALTERATION class represents specific alterations to a chromosome. One or more ALTERATION records are associated with a given CHROMOSOME record to describe the mutations to a given chromosome carried by a mutant strain of fish. An ALTERATION record describes a specific chromosomal change, e.g., a single insertion, translocation, or point mutation. Initially, much of the information in an ALTERATION may be left blank, to be filled in as further research is done. The phenotype field is particularly difficult to fill because it is difficult to determine the phenotypic contribution of a specific alteration to a chromosome, if there are other alterations on that chromosome as well. Thus, we expect the most important and heavily used phenotypic description to be found in the CHROMOSOME record, with the phenotype of ALTERATIONs filled in only when a chromosome has sustained a single alteration, or in those rare instances in which the phenotypic contribution of an alteration has been established.
	
<P>	<B>How/Who enters: </B> Initially, the DB Admin enters known ALTERATIONS in the course of entering information on currently known mutant chromosomes. Authorized submitters are allowed to create new ALTERATION records in the course of creating or updating CHROMOSOME records.
	
<P>	<B>Security:</B>  Only authorized submitters are permitted to enter new ALTERATION records or update them. All fields may be updated by an authorized submitter, with the exception of the primary_pub, which is fixed at the time the ALTERATION record is created. Of course, changes to this field may be requested via the DB Admin. 
<BR><BR><BR>			
<TABLE>
<TR><TD><B><BIG>Superclasses:</B></TD> <TD> <B>GENERIC_OBJECT</B></BIG></TD></TR>
<TR><TD></TD><TD><B>DATA_ITEM</B></TD><TD>comments, primary_pub, related_pubs, record_status</TD></TR>
</TABLE>
	
<H4>ATTRIBUTES:</H4>
<P>	<B>allele (TEXT, REQUIRED). </B> The official allele designation of this alteration. If an allele consists of more than one alteration (e.g., allele &lsquo;b345&rsquo; consists of an insertion AND a translocation), then both alterations have the same allele name.
<P>	<B>mutagen (TEXT, REQUIRED). </B>The mutagenizing agent. Must be one of {ENU, EMS, g-rays, DNA, spontaneous}.
<P>	<B>protocol (TEXT, REQUIRED). </B> The mutagenizing protocol. Describes what was treated with the mutagenizing agent. One of {adult males, adult females, sperm, embryos}.
<P>	<B>chrom_change (Closed Set, REQUIRED).</B> Describes the nature of the mutation. Valid values come from the set {Dp, Df, T, Tp, In, UNKNOWN} standing for Duplication, Deficiency, Translocation, Transposition, Insertion.
<P>	<B>locus (text, OPTIONAL). </B> The name of the locus affected by this mutation.
<P>	<B>genetic_loc (text, OPTIONAL). </B>Gives location in genetic terms.
<P>	<B>cyto_loc (text, OPTIONAL). </B>Gives location of mutation in cytological terms.
<P>	<B>breakpoints (text, OPTIONAL).</B> Two genetic and cytological locations.
<P>	<B>affected_genes (GENE, OPTIONAL).</B> Points to all genes that are known to be affected by this genetic alteration. THIS FIELD DOES NOT EXIST PHYSICALLY; it is an intersection table. Whenever a GENE-SEQ is found to be affected by the mutation described by this CHROMOSOME record, the person that discovers it can update this field, adding the GENE-SEQ to the list of known mutations. 
<P>	<B>potential_genes (GENE, OPTIONAL). </B>Points to all <I>possible</I> genes that could be affected by this alteration. The difference between this attribute and the affected_genes is that the genes in this list are <I>suspected</I> to be affected, but this has not been formally established. THIS FIELD DOES NOT PHYSICALLY EXIST. Rather the information is computed dynamically by comparing the breakpoint information in this record with the genetic locations of GENE-SEQ records. The idea is that if a GENE-SEQ falls on or between your breakpoints, then it may be affected by the mutation. The interface should do this calculation and present the resulting list of GENE-SEQs as &ldquo;genes potentially affected by mutation&rdquo;.
<P>	<B>chrom_id (TEXT, REQUIRED). </B> For system use. Contains the ZDB_ID of the CHROMOSOME record with which this alteration is associated.

<H3>Open Issues/Questions:  </H3>
<P>			&bull;  (Eck, 7/19/96) Given the brief discussion in ROLE above, should we even have a phenotype attribute at the ALTERATION level? If I understand it correctly, the only way to pick apart the phenotypic contributions of individual alterations is to isolate them on a chromosome by somehow knocking out all other alterations. But then wouldn&rsquo;t this create a completely new chromosome, with its own CHROMOSOME record --- which would have its own phenotype. 
<P>			&bull;  Format for genetic location needs to be determined.  Probably something like &ldquo;{u|d}real#&rdquo;, where u/d refers to the long/short (up/down) arm of the chromosome, followed by a recombination (centimorgan) distance.
<P>			&bull;  Format for cytological location needs to be determined.  Something like (integer, letter, integer) to locate the exact region of the mutation.
<BR>
<BR>
<BR>
<hr>
<B>Data Descriptions for The Zebrafish Database (ZDB)</B><br>

Go to <A HREF="sectionii_14.html"> next section</a><br>
Return to the <A HREF="data_model.html"> Table of Contents</a>

</BODY>

<script language="JavaScript" src="/javascript/footer.js"></script>
