<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html>
	<head>
		<title>
			The Zebrafish Database Class:  CLONED_GENE<
		</title>
	</head>

<script language="JavaScript" src="/javascript/header.js"></script>

<CENTER><H4>Data Descriptions for The Zebrafish Database (ZDB)</H4>
</CENTER>
<H2>SECTION II:  Experimental Data</H2>
<H3>Abstract Class:  DATA_ITEM</H3>
<H4>Class:  CLONED_GENE</H4>
<hr>
Return to the <A HREF="data_model.html"> Table of Contents</a>
<hr>

<P>	<B>Role:</B>   The CLONED_GENE class extends the MAP_MARKER class to represent additional information specifically relevant to cloned genes. 
	
<P>	<B>How/Who enters:</B> Currently known cloned genes are entered initially by the DB Admin. In the future, it is expected that authorized submitters will add new cloned genes and update or flesh out information in existing records.  
	
<P>	<B>Security: </B> Only authorized submitters are allowed to add new CLONED_GENE records to the database; as always, submissions are tagged with the time and the identity of the submitter. Changes to abbreviation and allele constitute nomenclature changes and must be handled via requests to the DB Admin. All other fields may be updated by authorized submitters.
<BR>
<BR>
<TABLE>
<TR><TD><B><BIG>Superclasses:</B></TD> <TD> <B>GENERIC_OBJECT</B></BIG></TD></TR>
<TR><TD></TD><TD><B>DATA_ITEM</B></TD><TD>comments, primary_pub, related_pubs, record_status</TD></TR>
<TR><TD></TD><TD><B>MAP_MARKER</B></TD><TD>name, marker_type, map_location, amp_product</TD></TR>
</TABLE>
			
<H4>ATTRIBUTES:</H4>
<P>	<B>abbreviation (TEXT, REQUIRED). </B>This field holds the abbreviated name of the cloned gene. It is up to the submitter to provide an abbreviation that conforms to established conventions; the system checks to be sure the abbreviation is unique.
<P>	<B>allele (TEXT, REQUIRED).</B> This field holds the unique designation for the allele cloned. Is null if the sequence in question is the wild-type.
<P>	<B>source (SOURCE, OPTIONAL). </B>This field holds the ZDB_ID# of a SOURCE (i.e. person, lab, or company) from which the cloned gene is currently available.
<P>	<B>cDNA sequence, protein sequence, genomic sequence (SEQUENCE, OPTIONAL). </B>THIS FIELD DOES NOT EXIST PHYSICALLY; the association is established by joining with the SEQUENCE class. Each cloned gene may have one each of genomic, protein, and cDNA sequence records associated with it.
<P>	<B>orthologs (ORTHOLOG, OPTIONAL).</B> THIS FIELD DOES NOT EXIST PHYSICALLY IN THIS RECORD; the association is established by joining with the ORTHOLOG table. A cloned gene may have any number of ORTHOLOG records associated with it, one for each species with which an orthologous relationship has been established.
<P>	<B>expr_pattern (text, REQUIRED).</B> A rigid and compact summary of the expression pattern, to be used to support searches. The format is (PART, STAGE-start, STAGE-end), where the values for this set are determined by presenting the user with the following interface: &ldquo;Please summarize the expression pattern of this gene: Expressed in &lt;PART&gt; from stage &lt;STAGE-start&gt; to &lt;STAGE-end&gt;.&rdquo;
			
<H3>Open Issues/Questions:  </H3>
<P>			&bull;  Should we try to enforce conventions for naming alleles and coming up with abbreviations? Currently, people are allowed to enter any text string.
<BR>
<BR>
<BR>
<hr>
<B>Data Descriptions for The Zebrafish Database (ZDB)</B><br>

Go to <A HREF="sectionii_5.html"> next section</a><br>
Return to the <A HREF="data_model.html"> Table of Contents</a>

</BODY>

<script language="JavaScript" src="/javascript/footer.js"></script>
