<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html>
	<head>
		<title>
			The Zebrafish Database Class: SEQUENCE
		</title>
	</head>

<script language="JavaScript" src="/javascript/header.js"></script>

<CENTER><H4>Data Descriptions for The Zebrafish Database (ZDB)</H4>
</CENTER>
<H2>SECTION II:  Experimental Data</H2>
<H3>Abstract Class:  DATA_ITEM</H3>
<H4>Class:  SEQUENCE</H4>
<hr>
Return to the <A HREF="data_model.html"> Table of Contents</a>

<P>	<B>Role: </B>  The purpose of the SEQUENCE class is to represent all types of sequence information that may be associated with various map markers, with the exception of primer sequences which have their own (simplified) data type. That is, rather than making the actual text sequences physically part of each such record, all sequences are stored in a separate table and associated with the appropriate map marker through a join operation. Only one SEQUENCE record of a given seq_type can be associated with a given map marker. Other restrictions may apply as well, e.g., you would not expect to associate a sequence with a RAPD.
	
<P>	<B>How/Who enters:</B>  Currently known SEQUENCES are entered initially by the DB Admin, as part of entering the information for known map markers. In the future, it is expected that authorized submitters will add/update SEQUENCES in the course of describing various map markers.  
	
<P>	<B>Security:</B>  Only authorized submitters are allowed to add new SEQUENCE records to the database; as always, submissions are tagged with the time and the identity of the submitter. Additions to related_pubs are allowed, as usual, but all other fields can be changed only by DB_Admin once the submission has been made.
<BR>
<BR>
<TABLE>
<TR><TD><B><BIG>Superclasses:</B></TD> <TD> <B>GENERIC_OBJECT</B></BIG></TD></TR>
<TR><TD></TD><TD><B>DATA_ITEM</B></TD><TD>comments, primary_pub, related_pubs, record_status</TD></TR>
</TABLE>
			
<H4>ATTRIBUTES:</H4>
<P>	<B>marker_id (CLONED_GENE, REQUIRED).</B> Contains the zdb_id of the CLONED_GENE which this SEQUENCE record describes. 
<P>	<B>seq_type ( closed list{protein, genomic, cDNA, EST}, REQUIRED). </B>Describes the type of sequence that is described in this SEQUENCE record. 
<P>	<B>sequence ( TEXT, REQUIRED).</B> The actual sequence, i.e., a list of base pairs.
<P>	<B>genbank_num (TEXT, REQUIRED).</B> Contains the GenBank accession number for the sequence. The fact that this field is required implies that we accept only sequences that have been submitted to GenBank.
<P>	<B>start_codon (integer, OPTIONAL). </B>Integer that locates the start codon within the sequence
<P>	<B>stop_codon (integer, OPTIONAL). </B> Integer that locates the stop codon within the sequence.
<P>	<B>introns (TEXT, OPTIONAL). </B>Contains a comma-separated list of integers that describe the intron locations in the sequence.
	
<H3>Open Issues/Questions:  </H3>
<P>			&bull;  How long are the strings of base pairs that we want to put in sequence attribute? Need to make sure it isn&rsquo;t higher than limit for TEXT data type.
<P>			&bull; Are we going to check the validity of the GenBank number they give us dynamically?
<P>			&bull; How to handle updates to the sequence stored in GenBank --> if we are caching the actual sequence, we have an updating/consistency problem. TED: Don&rsquo;t cache it, get it dynamically each time. ECK: let&rsquo;s cache and find a way to make updates automatically.
<BR>
<BR>
<BR>
<hr>
<B>Data Descriptions for The Zebrafish Database (ZDB)</B><br>

Go to <A HREF="sectionii_8.html"> next section</a><br>
Return to the <A HREF="data_model.html"> Table of Contents</a>

</BODY>

<script language="JavaScript" src="/javascript/footer.js"></script>
